// @strapi/content-manager/dist/server/mcp/schemas/output-schemas.js
'use strict';

var strapiUtils = require('@strapi/utils');
var shapeRelations = require('../sanitizers/shape-relations.js');

const relationIdentity = strapiUtils.z.object({
    documentId: strapiUtils.z.string(),
    locale: strapiUtils.z.string().optional(),
    __type: strapiUtils.z.string().optional(),
    // Computed publish status preserved on `localizations` entries (calculate, then strip).
    status: strapiUtils.z.string().optional()
});
const buildAttributeSchema = (key, attributes)=>{
    const attribute = attributes[key];
    if (attribute?.type !== 'relation') {
        return strapiUtils.z.unknown().optional();
    }
    // admin::user relations are out of scope — left as unknown
    if (attribute.target === 'admin::user') {
        return strapiUtils.z.unknown().optional();
    }
    // Must match the runtime shaping cardinality (shape-relations.ts) — the MCP SDK
    // validates structuredContent against this schema, so a mismatch fails the tool call.
    if (shapeRelations.isManyRelationForMcp(attribute)) {
        return strapiUtils.z.array(relationIdentity).optional();
    }
    return relationIdentity.nullable().optional();
};
const buildOutputDataSchema = (attributes, permittedFields)=>{
    const readableKeys = Object.keys(attributes).filter((key)=>permittedFields === null || permittedFields.has(key));
    if (readableKeys.length === 0) {
        return strapiUtils.z.record(strapiUtils.z.string(), strapiUtils.z.unknown());
    }
    const shape = Object.fromEntries(readableKeys.map((key)=>[
            key,
            buildAttributeSchema(key, attributes)
        ]));
    return strapiUtils.z.object(shape).loose();
};
/**
 * Builds the MCP output schema for a single-document response (`{ data, meta }`).
 * Field shape is constrained to `readFields` when non-null (RBAC field filtering).
 */ const buildDocumentOutputSchema = (attributes, readFields)=>strapiUtils.z.object({
        data: buildOutputDataSchema(attributes, readFields).nullable(),
        meta: strapiUtils.z.object({
            availableLocales: strapiUtils.z.array(strapiUtils.z.record(strapiUtils.z.string(), strapiUtils.z.unknown())).optional(),
            availableStatus: strapiUtils.z.array(strapiUtils.z.record(strapiUtils.z.string(), strapiUtils.z.unknown())).optional()
        }).optional()
    }).loose();
/**
 * Builds the MCP output schema for a paginated list response (`{ results, pagination }`).
 * Field shape is constrained to `readFields` when non-null (RBAC field filtering).
 */ const buildListOutputSchema = (attributes, readFields)=>strapiUtils.z.object({
        results: strapiUtils.z.array(buildOutputDataSchema(attributes, readFields)),
        pagination: strapiUtils.z.object({
            page: strapiUtils.z.number(),
            pageSize: strapiUtils.z.number(),
            pageCount: strapiUtils.z.number(),
            total: strapiUtils.z.number()
        })
    }).loose();
/**
 * Builds the MCP output schema for a delete response (`{ data }`).
 * Delete handlers return an empty data object, so do not require document relation fields.
 */ const buildDeleteOutputSchema = (// eslint-disable-next-line @typescript-eslint/no-unused-vars
_attributes, // eslint-disable-next-line @typescript-eslint/no-unused-vars
_readFields)=>strapiUtils.z.object({
        data: strapiUtils.z.object({}).loose().nullable()
    }).loose();

exports.buildDeleteOutputSchema = buildDeleteOutputSchema;
exports.buildDocumentOutputSchema = buildDocumentOutputSchema;
exports.buildListOutputSchema = buildListOutputSchema;
