// @strapi/admin/dist/server/server/src/controllers/authenticated-user.js
'use strict';

var index = require('../utils/index.js');
var normalizeEmail = require('../utils/normalize-email.js');
var user = require('../validation/user.js');
var sessionAuth = require('../../../shared/utils/session-auth.js');

var authenticatedUser = {
    async getMe (ctx) {
        const userInfo = index.getService('user').sanitizeUser(ctx.state.user);
        ctx.body = {
            data: userInfo
        };
    },
    async updateMe (ctx) {
        const data = normalizeEmail.normalizeEmail(ctx.request.body);
        await user.validateProfileUpdateInput(data);
        const userService = index.getService('user');
        const authServer = index.getService('auth');
        const { currentPassword, ...userInfo } = data;
        const isChangingPassword = Boolean(currentPassword && userInfo.password);
        if (currentPassword && userInfo.password) {
            const isValid = await authServer.validatePassword(currentPassword, ctx.state.user.password);
            if (!isValid) {
                return ctx.badRequest('ValidationError', {
                    currentPassword: [
                        'Invalid credentials'
                    ]
                });
            }
        }
        if (userInfo.email !== undefined) {
            const emailAlreadyTaken = await userService.exists({
                id: {
                    $ne: ctx.state.user.id
                },
                email: userInfo.email
            });
            if (emailAlreadyTaken === true) {
                return ctx.badRequest('ValidationError', {
                    email: [
                        'Email already taken'
                    ]
                });
            }
        }
        // Invalidate all sessions when password changes for security. This must run only once the
        // update is going to be persisted, so a rejected request (e.g. duplicate email) does not log
        // the user out without applying any change.
        if (isChangingPassword) {
            const sessionManager = sessionAuth.getSessionManager();
            if (sessionManager && sessionManager.hasOrigin('admin')) {
                await sessionManager('admin').invalidateRefreshToken(String(ctx.state.user.id));
            }
        }
        const updatedUser = await userService.updateById(ctx.state.user.id, userInfo);
        ctx.body = {
            data: userService.sanitizeUser(updatedUser)
        };
    },
    async getOwnPermissions (ctx) {
        const { findUserPermissions, sanitizePermission } = index.getService('permission');
        const { user } = ctx.state;
        const userPermissions = await findUserPermissions(user);
        ctx.body = {
            // @ts-expect-error - transform response type to sanitized permission
            data: userPermissions.map(sanitizePermission)
        };
    }
};

module.exports = authenticatedUser;
