// @strapi/upload/dist/admin/future/pages/Assets/components/AssetDetails/AssetDetailsDrawer.mjs
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import * as React from 'react';
import { useQueryParams, useNotification, Form, Blocker, getDisplayName, useForm, useField, useClipboard } from '@strapi/admin/strapi-admin';
import { VisuallyHidden, Flex, Loader, Alert, Typography, Box, Button, IconButton, Field, TextInput, Tooltip, SingleSelect, SingleSelectOption, Dialog } from '@strapi/design-system';
import { FileError, ArrowLineRight, Crop, Trash, WarningCircle, Link, Download, ArrowsCounterClockwise } from '@strapi/icons';
import { useIntl } from 'react-intl';
import { styled } from 'styled-components';
import { ASSET_TYPES } from '../../../../../enums.mjs';
import { Drawer } from '../../../../components/Drawer.mjs';
import { useUploadFileSilentlyMutation } from '../../../../services/api.mjs';
import { useGetAssetQuery, useUpdateAssetMutation, useReplaceAssetMutation, useDeleteAssetMutation } from '../../../../services/assets.mjs';
import { useGetAllFoldersQuery } from '../../../../services/folders.mjs';
import { useGetSettingsQuery } from '../../../../services/settings.mjs';
import { downloadFile } from '../../../../utils/downloadFile.mjs';
import { formatBytes, getFileExtension, prefixFileUrlWithBackendUrl } from '../../../../utils/files.mjs';
import { getAssetIcon } from '../../../../utils/getAssetIcon.mjs';
import { getTranslationKey } from '../../../../utils/translations.mjs';
import { useFolderInfo } from '../../hooks/useFolderInfo.mjs';
import { AssetCropEditor } from './AssetCropEditor.mjs';
import { AssetPreview } from './AssetPreview.mjs';

// Name of the parameter to look for in the URL to open the drawer
const URL_PARAM = 'assetId';
const DrawerNotifyContext = /*#__PURE__*/ React.createContext(null);
const useDrawerNotify = ()=>{
    const ctx = React.useContext(DrawerNotifyContext);
    if (!ctx) {
        throw new Error('useDrawerNotify must be used within AssetDetails');
    }
    return ctx;
};
const AssetOperationsContext = /*#__PURE__*/ React.createContext(null);
const useAssetOperation = ()=>{
    const ctx = React.useContext(AssetOperationsContext);
    if (!ctx) {
        throw new Error('useAssetOperation must be used within AssetDetails');
    }
    return ctx;
};
/* -------------------------------------------------------------------------------------------------
 * useAssetDetailsParam - sync drawer visibility with URL ?{URL_PARAM}={id}
 * -----------------------------------------------------------------------------------------------*/ const useAssetDetailsParam = ()=>{
    const [{ query }, setQuery] = useQueryParams();
    const detailsId = query?.[URL_PARAM];
    const assetId = detailsId ? parseInt(detailsId, 10) : null;
    const hasValidId = assetId !== null && !Number.isNaN(assetId);
    // Closing is driven by removing the URL param (a navigation), so navigation
    // guards like <Blocker> can intercept it. `isMounted` keeps the drawer in the
    // tree through the slide-out: it stays true once opened and only flips false
    // when the close animation actually ends (see onCloseAnimationEnd), so the
    // close duration lives entirely in CSS — no JS timer.
    const [isMounted, setIsMounted] = React.useState(hasValidId);
    const displayAssetId = React.useRef(null);
    React.useEffect(()=>{
        if (hasValidId) {
            displayAssetId.current = assetId;
            setIsMounted(true);
        }
    }, [
        hasValidId,
        assetId
    ]);
    const onCloseAnimationEnd = React.useCallback((event)=>{
        // Ignore animations bubbling up from descendants, and the slide-in.
        if (event.target === event.currentTarget && !hasValidId) {
            setIsMounted(false);
        }
    }, [
        hasValidId
    ]);
    const openDetails = React.useCallback((id)=>{
        setQuery({
            [URL_PARAM]: String(id)
        }, 'push', true);
    }, [
        setQuery
    ]);
    const closeDetails = React.useCallback(()=>{
        setQuery({
            [URL_PARAM]: undefined
        }, 'remove', true);
    }, [
        setQuery
    ]);
    return {
        assetId: hasValidId ? assetId : displayAssetId.current,
        isVisible: hasValidId,
        shouldRenderDrawer: isMounted,
        onCloseAnimationEnd,
        openDetails,
        closeDetails
    };
};
const DetailItemContainer = styled(Flex)`
  flex: 0 0 calc(50% - ${({ theme })=>theme.spaces[2]});
`;
const DetailItem = ({ label, value })=>/*#__PURE__*/ jsxs(DetailItemContainer, {
        direction: "column",
        justifyContent: "flex-start",
        alignItems: "flex-start",
        gap: 1,
        children: [
            /*#__PURE__*/ jsx(Typography, {
                variant: "sigma",
                textColor: "neutral600",
                fontWeight: "semiBold",
                textTransform: "uppercase",
                children: label
            }),
            /*#__PURE__*/ jsx(Typography, {
                variant: "pi",
                textColor: "neutral700",
                children: value ?? '-'
            })
        ]
    });
/* -------------------------------------------------------------------------------------------------
 * DetailField
 * -----------------------------------------------------------------------------------------------*/ /**
 * Make the asset details Form behave as a flex column inside Drawer.Body so
 * the scrollable area can grow while the footer stays pinned at the bottom.
 * The Form component from `@strapi/admin/strapi-admin` only forwards `width`
 * + `height` to its Box, so we target the rendered `<form>` element via a
 * styled-components descendant rule.
 */ const FormShell = styled(Box)`
  display: flex;
  flex-direction: column;
  flex: 1;
  min-height: 0;

  > form {
    display: flex;
    flex-direction: column;
    flex: 1;
    min-height: 0;
    position: relative;
  }
`;
/**
 * In-drawer toast container
 */ const DrawerToastSlot = styled(Box)`
  position: absolute;
  top: ${({ theme })=>theme.spaces[2]};
  left: 50%;
  transform: translateX(-50%);
  z-index: 10;
  width: calc(100% - ${({ theme })=>theme.spaces[2]});
`;
/**
 * Full-form overlay rendered during long-running drawer-scoped mutations
 * (e.g. replacing the binary). Sits above the toast slot (z-index 10) and
 * the in-drawer Alert so the user can't interact with the form mid-flight.
 */ const DrawerBusyOverlay = styled(Flex)`
  position: absolute;
  inset: 0;
  z-index: 20;
  align-items: center;
  justify-content: center;
  background: ${({ theme })=>theme.colors.neutral0};
  opacity: 0.7;
`;
/**
 * Map the drawer-scoped mutation flags to a single i18n message for the busy
 * overlay loader. Returns `null` when nothing is in flight.
 */ const getBusyMessage = (state)=>{
    if (state.isDeleting) {
        return {
            id: getTranslationKey('asset-details.delete.loading'),
            defaultMessage: 'Deleting the file…'
        };
    }
    if (state.isCropCopying) {
        return {
            id: getTranslationKey('asset-details.crop.loading'),
            defaultMessage: 'Saving the cropped copy…'
        };
    }
    if (state.isReplacing) {
        return {
            id: getTranslationKey('asset-details.replace.loading'),
            defaultMessage: 'Replacing the file…'
        };
    }
    return null;
};
const StyledWarning = styled(WarningCircle)`
  width: 1.6rem;
  height: 1.6rem;

  path {
    fill: ${({ theme })=>theme.colors.warning500};
  }
`;
const DetailField = ({ name, label, required })=>{
    const { formatMessage } = useIntl();
    const field = useField(name);
    const isSubmitting = useForm('DetailField', (state)=>state.isSubmitting);
    const value = field.value ?? '';
    const emptyTooltipLabel = formatMessage({
        id: getTranslationKey('asset-details.field.empty'),
        defaultMessage: '{label} is currently empty.'
    }, {
        label
    });
    return /*#__PURE__*/ jsxs(Field.Root, {
        name: name,
        required: required,
        children: [
            /*#__PURE__*/ jsx(Field.Label, {
                children: label
            }),
            /*#__PURE__*/ jsx(TextInput, {
                value: value,
                onChange: (event)=>field.onChange(name, event.target.value),
                endAction: !value ? /*#__PURE__*/ jsx(Tooltip, {
                    label: emptyTooltipLabel,
                    children: /*#__PURE__*/ jsx(StyledWarning, {
                        "aria-label": emptyTooltipLabel,
                        role: "img"
                    })
                }) : undefined,
                type: "text",
                disabled: isSubmitting
            })
        ]
    });
};
const LocationField = ({ label, rootLabel, folders })=>{
    const field = useField('folder');
    const isSubmitting = useForm('LocationField', (state)=>state.isSubmitting);
    return /*#__PURE__*/ jsxs(Field.Root, {
        name: "folder",
        required: true,
        children: [
            /*#__PURE__*/ jsx(Field.Label, {
                children: label
            }),
            /*#__PURE__*/ jsxs(SingleSelect, {
                value: field.value == null ? '' : String(field.value),
                onChange: (value)=>{
                    const next = value === '' ? null : Number(value);
                    field.onChange('folder', next);
                },
                disabled: isSubmitting,
                children: [
                    /*#__PURE__*/ jsx(SingleSelectOption, {
                        value: "",
                        children: rootLabel
                    }),
                    folders.map((folder)=>/*#__PURE__*/ jsx(SingleSelectOption, {
                            value: String(folder.id),
                            children: folder.name
                        }, folder.id))
                ]
            })
        ]
    });
};
/* -------------------------------------------------------------------------------------------------
 * DeleteAssetButton
 * -----------------------------------------------------------------------------------------------*/ const DeleteAssetButton = ()=>{
    const { formatMessage } = useIntl();
    const { deleteAsset, isDeleting } = useAssetOperation();
    const [isOpen, setIsOpen] = React.useState(false);
    const handleConfirm = async ()=>{
        await deleteAsset();
        setIsOpen(false);
    };
    const triggerLabel = formatMessage({
        id: getTranslationKey('asset-details.delete.trigger'),
        defaultMessage: 'Delete this file'
    });
    return /*#__PURE__*/ jsxs(Dialog.Root, {
        open: isOpen,
        onOpenChange: setIsOpen,
        children: [
            /*#__PURE__*/ jsx(Dialog.Trigger, {
                children: /*#__PURE__*/ jsx(IconButton, {
                    withTooltip: false,
                    label: triggerLabel,
                    variant: "danger-light",
                    children: /*#__PURE__*/ jsx(Trash, {})
                })
            }),
            /*#__PURE__*/ jsxs(Dialog.Content, {
                children: [
                    /*#__PURE__*/ jsx(Dialog.Header, {
                        children: formatMessage({
                            id: getTranslationKey('asset-details.delete.title'),
                            defaultMessage: 'Delete this media file?'
                        })
                    }),
                    /*#__PURE__*/ jsx(Dialog.Body, {
                        icon: /*#__PURE__*/ jsx(WarningCircle, {
                            width: "24px",
                            height: "24px",
                            fill: "danger600"
                        }),
                        textAlign: "center",
                        children: formatMessage({
                            id: getTranslationKey('asset-details.delete.description'),
                            defaultMessage: 'This file cannot be recovered once deleted. If it is currently in use, linked content will break and image containers will be empty.'
                        })
                    }),
                    /*#__PURE__*/ jsxs(Dialog.Footer, {
                        children: [
                            /*#__PURE__*/ jsx(Dialog.Cancel, {
                                children: /*#__PURE__*/ jsx(Button, {
                                    variant: "tertiary",
                                    disabled: isDeleting,
                                    fullWidth: true,
                                    children: formatMessage({
                                        id: 'app.components.Button.cancel',
                                        defaultMessage: 'Cancel'
                                    })
                                })
                            }),
                            /*#__PURE__*/ jsx(Dialog.Action, {
                                children: /*#__PURE__*/ jsx(Button, {
                                    variant: "danger-light",
                                    loading: isDeleting,
                                    onClick: handleConfirm,
                                    fullWidth: true,
                                    children: formatMessage({
                                        id: 'app.components.Button.confirm',
                                        defaultMessage: 'Confirm'
                                    })
                                })
                            })
                        ]
                    })
                ]
            })
        ]
    });
};
const CopyLinkButton = ({ asset })=>{
    const { formatMessage } = useIntl();
    const { copy } = useClipboard();
    const notify = useDrawerNotify();
    const handleCopy = async ()=>{
        const url = prefixFileUrlWithBackendUrl(asset.url);
        if (!url) return;
        const didCopy = await copy(url);
        notify({
            type: didCopy ? 'success' : 'danger',
            message: didCopy ? formatMessage({
                id: getTranslationKey('asset-details.copy-link.success'),
                defaultMessage: 'Link copied.'
            }) : formatMessage({
                id: getTranslationKey('asset-details.copy-link.error'),
                defaultMessage: 'Failed to copy the link.'
            })
        });
    };
    return /*#__PURE__*/ jsx(IconButton, {
        withTooltip: false,
        label: formatMessage({
            id: getTranslationKey('asset-details.copy-link.trigger'),
            defaultMessage: 'Copy link'
        }),
        variant: "tertiary",
        onClick: handleCopy,
        children: /*#__PURE__*/ jsx(Link, {})
    });
};
const DownloadAssetButton = ({ asset })=>{
    const { formatMessage } = useIntl();
    const notify = useDrawerNotify();
    const [isDownloading, setIsDownloading] = React.useState(false);
    const handleDownload = async ()=>{
        const url = prefixFileUrlWithBackendUrl(asset.url);
        if (!url) return;
        setIsDownloading(true);
        try {
            await downloadFile(url, asset.name);
        } catch  {
            notify({
                type: 'danger',
                message: formatMessage({
                    id: getTranslationKey('asset-details.download.error'),
                    defaultMessage: 'Failed to download the file.'
                })
            });
        } finally{
            setIsDownloading(false);
        }
    };
    return /*#__PURE__*/ jsx(IconButton, {
        withTooltip: false,
        label: formatMessage({
            id: getTranslationKey('asset-details.download.trigger'),
            defaultMessage: 'Download'
        }),
        variant: "tertiary",
        onClick: handleDownload,
        disabled: isDownloading,
        children: /*#__PURE__*/ jsx(Download, {})
    });
};
/* -------------------------------------------------------------------------------------------------
 * ReplaceAssetButton
 * -----------------------------------------------------------------------------------------------*/ const ReplaceAssetButton = ()=>{
    const { formatMessage } = useIntl();
    const { replaceAsset, isReplacing } = useAssetOperation();
    const fileInputRef = React.useRef(null);
    const [isDialogOpen, setIsDialogOpen] = React.useState(false);
    const { data: settings } = useGetSettingsQuery();
    const aiEnabled = settings?.data?.aiMetadata ?? false;
    const handleTriggerClick = ()=>{
        setIsDialogOpen(true);
    };
    const handleContinue = ()=>{
        // Confirm first, then open the native picker so the user only commits to
        // replacing after acknowledging the warning. The actual POST is delegated
        // to the parent (which owns the mutation + loading state).
        setIsDialogOpen(false);
        fileInputRef.current?.click();
    };
    const handleFileChange = async (event)=>{
        const file = event.target.files?.[0];
        // Reset the native input so the same file can be picked again later.
        event.target.value = '';
        if (!file) {
            return;
        }
        await replaceAsset(file);
    };
    return /*#__PURE__*/ jsxs(Fragment, {
        children: [
            /*#__PURE__*/ jsx(VisuallyHidden, {
                children: /*#__PURE__*/ jsx("input", {
                    ref: fileInputRef,
                    type: "file",
                    multiple: false,
                    onChange: handleFileChange,
                    "aria-hidden": true,
                    tabIndex: -1
                })
            }),
            /*#__PURE__*/ jsx(IconButton, {
                withTooltip: false,
                label: formatMessage({
                    id: getTranslationKey('asset-details.replace.trigger'),
                    defaultMessage: 'Replace this file'
                }),
                variant: "tertiary",
                onClick: handleTriggerClick,
                disabled: isReplacing,
                children: /*#__PURE__*/ jsx(ArrowsCounterClockwise, {})
            }),
            /*#__PURE__*/ jsx(Dialog.Root, {
                open: isDialogOpen,
                onOpenChange: setIsDialogOpen,
                children: /*#__PURE__*/ jsxs(Dialog.Content, {
                    children: [
                        /*#__PURE__*/ jsx(Dialog.Header, {
                            children: formatMessage({
                                id: getTranslationKey('asset-details.replace.title'),
                                defaultMessage: 'Replace this media file?'
                            })
                        }),
                        /*#__PURE__*/ jsx(Dialog.Body, {
                            textAlign: "center",
                            children: /*#__PURE__*/ jsxs(Flex, {
                                direction: "column",
                                textAlign: "center",
                                children: [
                                    /*#__PURE__*/ jsx(Typography, {
                                        variant: "omega",
                                        children: formatMessage({
                                            id: getTranslationKey('asset-details.replace.description'),
                                            defaultMessage: 'Current content will be permanently replaced.'
                                        })
                                    }),
                                    aiEnabled ? /*#__PURE__*/ jsx(Typography, {
                                        variant: "omega",
                                        children: formatMessage({
                                            id: getTranslationKey('asset-details.replace.description.ai'),
                                            defaultMessage: 'AI will generate new metadata after upload.'
                                        })
                                    }) : null
                                ]
                            })
                        }),
                        /*#__PURE__*/ jsxs(Dialog.Footer, {
                            children: [
                                /*#__PURE__*/ jsx(Dialog.Cancel, {
                                    children: /*#__PURE__*/ jsx(Button, {
                                        variant: "tertiary",
                                        fullWidth: true,
                                        children: formatMessage({
                                            id: 'app.components.Button.cancel',
                                            defaultMessage: 'Cancel'
                                        })
                                    })
                                }),
                                /*#__PURE__*/ jsx(Dialog.Action, {
                                    children: /*#__PURE__*/ jsx(Button, {
                                        variant: "secondary",
                                        onClick: handleContinue,
                                        fullWidth: true,
                                        children: formatMessage({
                                            id: getTranslationKey('asset-details.replace.continue'),
                                            defaultMessage: 'Continue'
                                        })
                                    })
                                })
                            ]
                        })
                    ]
                })
            })
        ]
    });
};
const AssetImageActions = ({ onCrop })=>{
    const { formatMessage } = useIntl();
    const isSubmitting = useForm('AssetImageActions', (state)=>state.isSubmitting);
    return /*#__PURE__*/ jsxs(Flex, {
        direction: "column",
        gap: 2,
        children: [
            /*#__PURE__*/ jsx(IconButton, {
                withTooltip: false,
                label: formatMessage({
                    id: getTranslationKey('asset-details.crop.trigger'),
                    defaultMessage: 'Crop'
                }),
                variant: "tertiary",
                onClick: onCrop,
                disabled: isSubmitting || !onCrop,
                children: /*#__PURE__*/ jsx(Crop, {})
            }),
            /*#__PURE__*/ jsx(ReplaceAssetButton, {})
        ]
    });
};
const AssetDetails = ({ asset, closeDetails })=>{
    const { formatMessage, formatDate } = useIntl();
    const { data: folders = [] } = useGetAllFoldersQuery();
    const { toggleNotification } = useNotification();
    const [updateAsset] = useUpdateAssetMutation();
    const [replaceMutation, { isLoading: isReplacing }] = useReplaceAssetMutation();
    const [deleteMutation, { isLoading: isDeleting }] = useDeleteAssetMutation();
    const [uploadCroppedCopy, { isLoading: isCropCopying }] = useUploadFileSilentlyMutation();
    const [isCropOpen, setIsCropOpen] = React.useState(false);
    // In-drawer toast slot
    const [drawerToast, setDrawerToast] = React.useState(null);
    React.useEffect(()=>{
        if (!drawerToast) return;
        const timer = window.setTimeout(()=>setDrawerToast(null), 5000);
        return ()=>window.clearTimeout(timer);
    }, [
        drawerToast
    ]);
    // Local alias matching the DrawerNotifyContext signature, so the drawer's
    // own handlers (replace, update) read like the consumers do.
    const notify = React.useCallback((toast)=>setDrawerToast(toast), []);
    const isImage = asset.mime?.includes(ASSET_TYPES.Image);
    const initialValues = {
        name: asset.name ?? '',
        caption: asset.caption ?? '',
        alternativeText: asset.alternativeText ?? '',
        folder: typeof asset.folder === 'object' && asset.folder !== null ? asset.folder.id ?? null : asset.folder ?? null
    };
    const handleSubmit = async (values)=>{
        const fileInfo = {
            name: values.name,
            caption: values.caption,
            alternativeText: values.alternativeText,
            folder: values.folder
        };
        const res = await updateAsset({
            id: asset.id,
            fileInfo
        });
        if ('error' in res) {
            notify({
                type: 'danger',
                message: formatMessage({
                    id: getTranslationKey('asset-details.update.error'),
                    defaultMessage: 'Failed to update the file.'
                })
            });
            return;
        }
        notify({
            type: 'success',
            message: formatMessage({
                id: getTranslationKey('asset-details.update.success'),
                defaultMessage: 'File updated'
            })
        });
    };
    const { title: folderName } = useFolderInfo(typeof asset.folder === 'object' && asset.folder !== null ? asset.folder.id ?? null : asset.folder ?? null);
    // Owns the replace upload so isReplacing can drive the busy overlay.
    const handleReplace = async (file)=>{
        const res = await replaceMutation({
            id: asset.id,
            file
        });
        if ('error' in res) {
            const error = res.error;
            const message = error?.data?.error?.message ?? error?.data?.message ?? formatMessage({
                id: getTranslationKey('asset-details.replace.error'),
                defaultMessage: 'Failed to replace the file.'
            });
            notify({
                type: 'danger',
                message
            });
            return;
        }
        notify({
            type: 'success',
            message: formatMessage({
                id: getTranslationKey('asset-details.replace.success'),
                defaultMessage: 'File replaced.'
            })
        });
    };
    // Owns the delete: on error notify in-drawer (drawer stays), on success fire
    // a persistent global notification then close the drawer.
    const handleDelete = async ()=>{
        const res = await deleteMutation(asset.id);
        if ('error' in res) {
            const error = res.error;
            const message = error?.data?.error?.message ?? error?.data?.message ?? formatMessage({
                id: getTranslationKey('asset-details.delete.error'),
                defaultMessage: 'Failed to delete the asset.'
            });
            notify({
                type: 'danger',
                message
            });
            return;
        }
        toggleNotification({
            type: 'success',
            message: formatMessage({
                id: getTranslationKey('asset-details.delete.success'),
                defaultMessage: '1 element have been deleted from {folderName}'
            }, {
                folderName
            })
        });
        closeDetails();
    };
    const notifyCropError = ()=>{
        notify({
            type: 'danger',
            message: formatMessage({
                id: getTranslationKey('asset-details.crop.error'),
                defaultMessage: 'Failed to crop the file.'
            })
        });
    };
    // Apply: replace the original binary with the cropped file + focal point.
    // Close the editor first so the drawer's busy overlay is visible while the
    // mutation runs (the editor renders in a Portal above the drawer).
    const handleCropApply = async (file, focalPoint)=>{
        setIsCropOpen(false);
        const res = await replaceMutation({
            id: asset.id,
            file,
            fileInfo: {
                focalPoint
            }
        });
        if ('error' in res) {
            notifyCropError();
            return;
        }
        notify({
            type: 'success',
            message: formatMessage({
                id: getTranslationKey('asset-details.crop.success'),
                defaultMessage: 'File cropped.'
            })
        });
    };
    // Save as copy: upload the cropped file as a new asset in the same folder.
    const handleCropSaveAsCopy = async (file, focalPoint)=>{
        setIsCropOpen(false);
        // Silent single-file upload (no global progress dialog). Carry over the
        // source asset's caption/alt so the copy keeps its metadata.
        const res = await uploadCroppedCopy({
            file,
            fileInfo: {
                name: asset.name,
                caption: asset.caption ?? '',
                alternativeText: asset.alternativeText ?? '',
                folder: initialValues.folder,
                focalPoint
            }
        });
        if ('error' in res) {
            notifyCropError();
            return;
        }
        notify({
            type: 'success',
            message: formatMessage({
                id: getTranslationKey('asset-details.crop.copy-success'),
                defaultMessage: 'Copy created.'
            })
        });
    };
    const operations = React.useMemo(()=>({
            replaceAsset: handleReplace,
            deleteAsset: handleDelete,
            isReplacing,
            isDeleting
        }), // handleReplace / handleDelete close over asset+mutations and don't need a
    // stable identity here; the consumers re-render with the new context value.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [
        isReplacing,
        isDeleting
    ]);
    return(// `key={asset.id}` resets the form when the drawer switches to a different
    // asset so cached values from the previous asset don't bleed in.
    /*#__PURE__*/ jsx(DrawerNotifyContext.Provider, {
        value: notify,
        children: /*#__PURE__*/ jsx(AssetOperationsContext.Provider, {
            value: operations,
            children: /*#__PURE__*/ jsx(FormShell, {
                children: /*#__PURE__*/ jsx(Form, {
                    method: "POST",
                    initialValues: initialValues,
                    onSubmit: handleSubmit,
                    children: ({ modified, isSubmitting, values, resetForm })=>{
                        const nameIsEmpty = (values.name ?? '').trim() === '';
                        const busyMessage = getBusyMessage({
                            isDeleting,
                            isReplacing,
                            isCropCopying
                        });
                        return /*#__PURE__*/ jsxs(Fragment, {
                            children: [
                                /*#__PURE__*/ jsx(Blocker, {
                                    onProceed: resetForm
                                }),
                                isCropOpen && isImage ? /*#__PURE__*/ jsx(AssetCropEditor, {
                                    asset: asset,
                                    onClose: ()=>setIsCropOpen(false),
                                    onApply: handleCropApply,
                                    onSaveAsCopy: handleCropSaveAsCopy
                                }) : null,
                                busyMessage ? /*#__PURE__*/ jsx(DrawerBusyOverlay, {
                                    children: /*#__PURE__*/ jsx(Loader, {
                                        children: formatMessage(busyMessage)
                                    })
                                }) : null,
                                drawerToast ? /*#__PURE__*/ jsx(DrawerToastSlot, {
                                    children: /*#__PURE__*/ jsx(Alert, {
                                        variant: drawerToast.type === 'success' ? 'success' : 'danger',
                                        closeLabel: formatMessage({
                                            id: 'global.close',
                                            defaultMessage: 'Close'
                                        }),
                                        onClose: ()=>setDrawerToast(null),
                                        children: drawerToast.message
                                    })
                                }) : null,
                                /*#__PURE__*/ jsxs(Drawer.ScrollableContent, {
                                    children: [
                                        /*#__PURE__*/ jsx(AssetPreview, {
                                            asset: asset,
                                            actions: isImage ? /*#__PURE__*/ jsx(AssetImageActions, {
                                                onCrop: ()=>setIsCropOpen(true)
                                            }) : null
                                        }),
                                        /*#__PURE__*/ jsxs(Flex, {
                                            direction: "column",
                                            alignItems: "stretch",
                                            gap: 4,
                                            paddingTop: 4,
                                            paddingBottom: 4,
                                            paddingLeft: 5,
                                            paddingRight: 5,
                                            children: [
                                                /*#__PURE__*/ jsx(Typography, {
                                                    variant: "beta",
                                                    fontWeight: "semiBold",
                                                    tag: "h3",
                                                    children: formatMessage({
                                                        id: getTranslationKey('asset-details.fileInfo'),
                                                        defaultMessage: 'File info'
                                                    })
                                                }),
                                                /*#__PURE__*/ jsxs(Flex, {
                                                    wrap: "wrap",
                                                    gap: 4,
                                                    background: "neutral100",
                                                    paddingTop: 4,
                                                    paddingBottom: 4,
                                                    paddingLeft: 6,
                                                    paddingRight: 6,
                                                    alignItems: "flex-start",
                                                    children: [
                                                        /*#__PURE__*/ jsx(DetailItem, {
                                                            label: formatMessage({
                                                                id: getTranslationKey('asset-details.creationDate'),
                                                                defaultMessage: 'Creation date'
                                                            }),
                                                            value: asset.createdAt ? formatDate(new Date(asset.createdAt), {
                                                                dateStyle: 'long',
                                                                timeStyle: 'short'
                                                            }) : null
                                                        }),
                                                        /*#__PURE__*/ jsx(DetailItem, {
                                                            label: formatMessage({
                                                                id: getTranslationKey('asset-details.lastUpdated'),
                                                                defaultMessage: 'Last updated'
                                                            }),
                                                            value: asset.updatedAt ? formatDate(new Date(asset.updatedAt), {
                                                                dateStyle: 'long',
                                                                timeStyle: 'short'
                                                            }) : null
                                                        }),
                                                        /*#__PURE__*/ jsx(DetailItem, {
                                                            label: formatMessage({
                                                                id: getTranslationKey('asset-details.createdBy'),
                                                                defaultMessage: 'Created by'
                                                            }),
                                                            value: asset.createdBy ? getDisplayName({
                                                                firstname: asset.createdBy.firstname ?? undefined,
                                                                lastname: asset.createdBy.lastname ?? undefined,
                                                                username: asset.createdBy.username ?? undefined,
                                                                email: asset.createdBy.email ?? undefined
                                                            }) ?? '-' : null
                                                        }),
                                                        /*#__PURE__*/ jsx(DetailItem, {
                                                            label: formatMessage({
                                                                id: getTranslationKey('asset-details.size'),
                                                                defaultMessage: 'Size'
                                                            }),
                                                            value: asset.size ? formatBytes(asset.size, 1) : null
                                                        }),
                                                        isImage && (asset.width != null || asset.height != null) && /*#__PURE__*/ jsx(DetailItem, {
                                                            label: formatMessage({
                                                                id: getTranslationKey('asset-details.dimensions'),
                                                                defaultMessage: 'Dimensions'
                                                            }),
                                                            value: asset.width != null && asset.height != null ? `${asset.width} × ${asset.height}` : null
                                                        }),
                                                        /*#__PURE__*/ jsx(DetailItem, {
                                                            label: formatMessage({
                                                                id: getTranslationKey('asset-details.extension'),
                                                                defaultMessage: 'Extension'
                                                            }),
                                                            value: getFileExtension(asset.ext)
                                                        }),
                                                        /*#__PURE__*/ jsx(DetailItem, {
                                                            label: formatMessage({
                                                                id: getTranslationKey('asset-details.assetId'),
                                                                defaultMessage: 'Asset ID'
                                                            }),
                                                            value: String(asset.id)
                                                        })
                                                    ]
                                                }),
                                                /*#__PURE__*/ jsx(DetailField, {
                                                    name: "name",
                                                    label: formatMessage({
                                                        id: getTranslationKey('asset-details.fileName'),
                                                        defaultMessage: 'File name'
                                                    }),
                                                    required: true
                                                }),
                                                /*#__PURE__*/ jsx(LocationField, {
                                                    label: formatMessage({
                                                        id: getTranslationKey('asset-details.location'),
                                                        defaultMessage: 'Location'
                                                    }),
                                                    rootLabel: formatMessage({
                                                        id: getTranslationKey('plugin.home'),
                                                        defaultMessage: 'Home'
                                                    }),
                                                    folders: folders
                                                }),
                                                isImage && /*#__PURE__*/ jsxs(Fragment, {
                                                    children: [
                                                        /*#__PURE__*/ jsx(DetailField, {
                                                            name: "caption",
                                                            label: formatMessage({
                                                                id: getTranslationKey('asset-details.caption'),
                                                                defaultMessage: 'Caption'
                                                            })
                                                        }),
                                                        /*#__PURE__*/ jsx(DetailField, {
                                                            name: "alternativeText",
                                                            label: formatMessage({
                                                                id: getTranslationKey('asset-details.alternativeText'),
                                                                defaultMessage: 'Alternative text'
                                                            })
                                                        })
                                                    ]
                                                })
                                            ]
                                        })
                                    ]
                                }),
                                /*#__PURE__*/ jsxs(Flex, {
                                    justifyContent: "space-between",
                                    alignItems: "center",
                                    gap: 2,
                                    padding: 3,
                                    borderColor: "neutral150",
                                    borderStyle: "solid",
                                    borderWidth: "1px 0 0 0",
                                    background: "neutral0",
                                    children: [
                                        /*#__PURE__*/ jsxs(Flex, {
                                            gap: 2,
                                            children: [
                                                /*#__PURE__*/ jsx(DeleteAssetButton, {}),
                                                /*#__PURE__*/ jsx(CopyLinkButton, {
                                                    asset: asset
                                                }),
                                                /*#__PURE__*/ jsx(DownloadAssetButton, {
                                                    asset: asset
                                                })
                                            ]
                                        }),
                                        /*#__PURE__*/ jsx(Button, {
                                            type: "submit",
                                            variant: "default",
                                            loading: isSubmitting,
                                            // File name is required; block submit when it's empty or whitespace so the API can't 400 on a blank value.
                                            disabled: !modified || isSubmitting || nameIsEmpty,
                                            children: formatMessage({
                                                id: getTranslationKey('asset-details.save'),
                                                defaultMessage: 'Save changes'
                                            })
                                        })
                                    ]
                                })
                            ]
                        });
                    }
                }, asset.id)
            })
        })
    }));
};
const DrawerHeader = ({ asset, closeDetails })=>{
    const DocIcon = asset ? getAssetIcon(asset.mime, asset.ext) : FileError;
    return /*#__PURE__*/ jsxs(Flex, {
        gap: 2,
        paddingLeft: 5,
        paddingTop: 3,
        paddingBottom: 3,
        paddingRight: 3,
        borderColor: "neutral150",
        borderStyle: "solid",
        borderWidth: "0 0 1px 0",
        children: [
            /*#__PURE__*/ jsx(DocIcon, {
                width: 20,
                height: 20
            }),
            /*#__PURE__*/ jsx(Drawer.Title, {
                asChild: true,
                children: /*#__PURE__*/ jsx(Typography, {
                    variant: "omega",
                    fontWeight: "semiBold",
                    overflow: "hidden",
                    ellipsis: true,
                    tag: "h2",
                    children: asset.name
                })
            }),
            /*#__PURE__*/ jsx(Box, {
                marginLeft: "auto",
                children: /*#__PURE__*/ jsx(Drawer.CloseButton, {
                    onClose: closeDetails,
                    children: /*#__PURE__*/ jsx(ArrowLineRight, {})
                })
            })
        ]
    });
};
const DrawerContent = ({ assetId, closeDetails })=>{
    const { formatMessage } = useIntl();
    const { data: asset, isLoading, error } = useGetAssetQuery(assetId, {
        refetchOnMountOrArgChange: false,
        refetchOnReconnect: false,
        refetchOnFocus: false
    });
    if (isLoading) {
        return /*#__PURE__*/ jsx(Flex, {
            justifyContent: "center",
            padding: 8,
            children: /*#__PURE__*/ jsx(Loader, {
                children: formatMessage({
                    id: 'app.loading',
                    defaultMessage: 'Loading...'
                })
            })
        });
    }
    if (error || !asset) {
        return /*#__PURE__*/ jsx(Flex, {
            direction: "column",
            alignItems: "stretch",
            gap: 4,
            padding: 4,
            children: /*#__PURE__*/ jsx(Alert, {
                variant: "danger",
                closeLabel: formatMessage({
                    id: 'global.close',
                    defaultMessage: 'Close'
                }),
                onClose: closeDetails,
                children: formatMessage({
                    id: getTranslationKey('asset-details.error'),
                    defaultMessage: 'Failed to load file details.'
                })
            })
        });
    }
    return /*#__PURE__*/ jsxs(Fragment, {
        children: [
            /*#__PURE__*/ jsx(DrawerHeader, {
                asset: asset,
                closeDetails: closeDetails
            }),
            /*#__PURE__*/ jsx(AssetDetails, {
                asset: asset,
                closeDetails: closeDetails
            })
        ]
    });
};
/* -------------------------------------------------------------------------------------------------
 * AssetDetailsDrawer
 * -----------------------------------------------------------------------------------------------*/ const AssetDetailsDrawer = ()=>{
    const { formatMessage } = useIntl();
    const { assetId, isVisible, shouldRenderDrawer, onCloseAnimationEnd, closeDetails } = useAssetDetailsParam();
    if (!shouldRenderDrawer || assetId === null) {
        return null;
    }
    return /*#__PURE__*/ jsxs(Drawer.Root, {
        isVisible: isVisible,
        onClose: closeDetails,
        children: [
            /*#__PURE__*/ jsx("div", {
                children: /*#__PURE__*/ jsxs(VisuallyHidden, {
                    children: [
                        /*#__PURE__*/ jsx(Drawer.Title, {
                            children: formatMessage({
                                id: getTranslationKey('asset-details.title'),
                                defaultMessage: 'File details'
                            })
                        }),
                        /*#__PURE__*/ jsx(Drawer.Description, {
                            children: formatMessage({
                                id: getTranslationKey('asset-details.description'),
                                defaultMessage: 'Displays file information and metadata'
                            })
                        })
                    ]
                })
            }),
            /*#__PURE__*/ jsx(Drawer.Body, {
                animationDirection: "left",
                width: "41.6rem",
                height: "100vh",
                onAnimationEnd: onCloseAnimationEnd,
                children: /*#__PURE__*/ jsx(DrawerContent, {
                    assetId: assetId,
                    closeDetails: closeDetails
                })
            })
        ]
    });
};

export { AssetDetails, AssetDetailsDrawer, AssetOperationsContext, DrawerNotifyContext, useAssetDetailsParam };
