// @strapi/upload/dist/admin/future/pages/Assets/components/AssetDetails/AssetPreview.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var designSystem = require('@strapi/design-system');
var reactIntl = require('react-intl');
var styledComponents = require('styled-components');
var enums = require('../../../../../enums.js');
var files = require('../../../../utils/files.js');
var getAssetIcon = require('../../../../utils/getAssetIcon.js');
var translations = require('../../../../utils/translations.js');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

/* -------------------------------------------------------------------------------------------------
 * Styled components
 * -----------------------------------------------------------------------------------------------*/ const PreviewContainer = styledComponents.styled(designSystem.Box)`
  position: relative;
  width: 100%;
  aspect-ratio: 16 / 9;
  max-height: 24rem;
  overflow: hidden;
  border-radius: ${({ theme })=>theme.borderRadius};
  padding: ${({ theme })=>theme.spaces[3]};
  background: repeating-conic-gradient(
      ${({ theme })=>theme.colors.neutral100} 0% 25%,
      transparent 0% 50%
    )
    50% / 20px 20px;
`;
const AssetContainer = styledComponents.styled(designSystem.Flex)`
  justify-content: center;
  position: relative;
  z-index: 2;
  width: 100%;
  height: 100%;
`;
const StyledImage = styledComponents.styled.img`
  max-width: 100%;
  max-height: 100%;
  object-fit: contain;
`;
/**
 * Top-right overlay slot for image actions (crop). Sits above the
 * image (z-index 3 > AssetContainer 2) so the buttons stay clickable.
 */ const ActionsOverlay = styledComponents.styled(designSystem.Flex)`
  position: absolute;
  top: ${({ theme })=>theme.spaces[3]};
  right: ${({ theme })=>theme.spaces[3]};
  z-index: 3;
`;
const StyledVideo = styledComponents.styled.video`
  max-width: 100%;
  max-height: 100%;
  object-fit: contain;
`;
const StyledAudio = styledComponents.styled.audio`
  width: 100%;
`;
const StyledPdfIframe = styledComponents.styled.iframe`
  width: 100%;
  height: 100%;
  min-height: 200px;
  border: none;
`;
const IconFallback = styledComponents.styled(designSystem.Flex)`
  height: 100%;
  aspect-ratio: 1;
  width: auto;
  max-width: 100%;
  margin: 0 auto;
  color: ${({ theme })=>theme.colors.neutral500};
  background: ${({ theme })=>theme.colors.neutral150};
`;
const LoaderOverlay = styledComponents.styled(designSystem.Flex)`
  position: absolute;
  inset: 0;
  z-index: 1;
`;
/* -------------------------------------------------------------------------------------------------
 * AssetLoader
 * -----------------------------------------------------------------------------------------------*/ const AssetLoader = ()=>{
    const { formatMessage } = reactIntl.useIntl();
    return /*#__PURE__*/ jsxRuntime.jsx(LoaderOverlay, {
        justifyContent: "center",
        alignItems: "center",
        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Loader, {
            children: formatMessage({
                id: 'app.loading',
                defaultMessage: 'Loading...'
            })
        })
    });
};
const AssetPreview = ({ asset, actions, isLoading = false })=>{
    const { formatMessage } = reactIntl.useIntl();
    const { alternativeText, ext, mime, url, updatedAt, isUrlSigned, isLocal } = asset;
    // Append the asset's `updatedAt` as a cache-buster so a freshly replaced
    // file (often served at the same URL) shows the new content instead of the
    // browser-cached old version. Signed URLs are excluded: their query string
    // is part of the SigV4 signature, so appending a param invalidates the
    // signature and the request fails with 403. See #26581.
    const cacheKey = updatedAt && !isUrlSigned ? new Date(updatedAt).getTime() : undefined;
    const appendCacheBuster = (raw)=>{
        if (!raw || cacheKey === undefined) return raw;
        return raw.includes('?') ? `${raw}&v=${cacheKey}` : `${raw}?v=${cacheKey}`;
    };
    const mediaUrl = appendCacheBuster(files.prefixFileUrlWithBackendUrl(url));
    const [isMediaLoaded, setIsMediaLoaded] = React__namespace.useState(false);
    React__namespace.useEffect(()=>{
        setIsMediaLoaded(false);
    }, [
        mediaUrl
    ]);
    const imageRef = React__namespace.useRef(null);
    React__namespace.useEffect(()=>{
        const image = imageRef.current;
        if (!image) return;
        const recompute = ()=>{
            const parent = image.parentElement;
            if (!parent) return;
            const c = parent.getBoundingClientRect();
            // offsetWidth/Height = pre-transform layout box (transforms ignored).
            const w = image.offsetWidth;
            const h = image.offsetHeight;
            if (!w || !h || !c.width || !c.height) return;
        };
        recompute();
        const ro = new ResizeObserver(recompute);
        ro.observe(image);
        if (image.parentElement) ro.observe(image.parentElement);
        return ()=>ro.disconnect();
    }, [
        isMediaLoaded
    ]);
    if (mime?.includes(enums.ASSET_TYPES.Image)) {
        const imageUrl = appendCacheBuster(files.prefixFileUrlWithBackendUrl(url));
        if (imageUrl) {
            return /*#__PURE__*/ jsxRuntime.jsxs(PreviewContainer, {
                children: [
                    (!isMediaLoaded || isLoading) && /*#__PURE__*/ jsxRuntime.jsx(AssetLoader, {}),
                    actions ? /*#__PURE__*/ jsxRuntime.jsx(ActionsOverlay, {
                        children: actions
                    }) : null,
                    /*#__PURE__*/ jsxRuntime.jsx(AssetContainer, {
                        children: /*#__PURE__*/ jsxRuntime.jsx(StyledImage, {
                            ref: imageRef,
                            src: imageUrl,
                            alt: alternativeText || asset.name || '',
                            crossOrigin: !isLocal && isUrlSigned ? 'anonymous' : undefined,
                            onLoad: ()=>setIsMediaLoaded(true),
                            onError: ()=>setIsMediaLoaded(true)
                        })
                    })
                ]
            });
        }
    }
    if (mime?.includes(enums.ASSET_TYPES.Video) && mediaUrl) {
        return /*#__PURE__*/ jsxRuntime.jsxs(PreviewContainer, {
            children: [
                !isMediaLoaded && /*#__PURE__*/ jsxRuntime.jsx(AssetLoader, {}),
                /*#__PURE__*/ jsxRuntime.jsx(AssetContainer, {
                    children: /*#__PURE__*/ jsxRuntime.jsx(StyledVideo, {
                        src: mediaUrl,
                        controls: true,
                        title: asset.name,
                        onLoadedData: ()=>setIsMediaLoaded(true),
                        onError: ()=>setIsMediaLoaded(true),
                        children: formatMessage({
                            id: translations.getTranslationKey('asset-details.videoNotSupported'),
                            defaultMessage: 'Your browser does not support the video tag.'
                        })
                    })
                })
            ]
        });
    }
    if (mime?.includes(enums.ASSET_TYPES.Audio) && mediaUrl) {
        return /*#__PURE__*/ jsxRuntime.jsxs(PreviewContainer, {
            children: [
                !isMediaLoaded && /*#__PURE__*/ jsxRuntime.jsx(AssetLoader, {}),
                /*#__PURE__*/ jsxRuntime.jsx(AssetContainer, {
                    children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                        width: "100%",
                        padding: 4,
                        justifyContent: "center",
                        alignItems: "center",
                        height: "100%",
                        minHeight: "12rem",
                        children: /*#__PURE__*/ jsxRuntime.jsx(StyledAudio, {
                            src: mediaUrl,
                            controls: true,
                            onLoadedData: ()=>setIsMediaLoaded(true),
                            onError: ()=>setIsMediaLoaded(true)
                        })
                    })
                })
            ]
        });
    }
    const isPdf = ext?.toLowerCase() === 'pdf' || ext?.toLowerCase() === '.pdf' || mime === 'application/pdf';
    if (isPdf && mediaUrl) {
        return /*#__PURE__*/ jsxRuntime.jsxs(PreviewContainer, {
            children: [
                !isMediaLoaded && /*#__PURE__*/ jsxRuntime.jsx(AssetLoader, {}),
                /*#__PURE__*/ jsxRuntime.jsx(AssetContainer, {
                    children: /*#__PURE__*/ jsxRuntime.jsx(StyledPdfIframe, {
                        src: `${mediaUrl}#toolbar=0`,
                        title: asset.name,
                        onLoad: ()=>setIsMediaLoaded(true)
                    })
                })
            ]
        });
    }
    const DocIcon = getAssetIcon.getAssetIcon(mime, ext);
    return /*#__PURE__*/ jsxRuntime.jsx(PreviewContainer, {
        children: /*#__PURE__*/ jsxRuntime.jsxs(IconFallback, {
            justifyContent: "center",
            alignItems: "center",
            gap: 1,
            direction: "column",
            hasRadius: true,
            children: [
                /*#__PURE__*/ jsxRuntime.jsx(DocIcon, {
                    width: 24,
                    height: 24
                }),
                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                    variant: "pi",
                    children: formatMessage({
                        id: translations.getTranslationKey('asset-details.noPreview'),
                        defaultMessage: 'No preview available'
                    })
                })
            ]
        })
    });
};

exports.AssetPreview = AssetPreview;
