// @strapi/content-manager/dist/admin/pages/ListView/components/TableCells/CellValue.js
'use strict';

var parseISO = require('date-fns/parseISO');
var toString = require('lodash/toString');
var reactIntl = require('react-intl');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var parseISO__default = /*#__PURE__*/_interopDefault(parseISO);
var toString__default = /*#__PURE__*/_interopDefault(toString);

const CellValue = ({ isIdColumn = false, type, value })=>{
    const { formatDate, formatTime, formatNumber, formatMessage } = reactIntl.useIntl();
    let formattedValue = value;
    if (type === 'date' && typeof value === 'string') {
        formattedValue = formatDate(parseISO__default.default(value), {
            dateStyle: 'full'
        });
    }
    if (type === 'datetime') {
        formattedValue = formatDate(value, {
            dateStyle: 'full',
            timeStyle: 'short'
        });
    }
    if (type === 'time' && typeof value === 'string') {
        const [hour, minute, second] = value.split(':');
        const date = new Date();
        date.setHours(Number(hour));
        date.setMinutes(Number(minute));
        date.setSeconds(Number(second));
        formattedValue = formatTime(date, {
            timeStyle: 'short'
        });
    }
    if ([
        'float',
        'decimal'
    ].includes(type)) {
        formattedValue = formatNumber(value, {
            // Should be kept in sync with the corresponding value
            // in the design-system/NumberInput: https://github.com/strapi/design-system/blob/main/packages/strapi-design-system/src/NumberInput/NumberInput.js#L53
            maximumFractionDigits: 20
        });
    }
    if (type === 'boolean') {
        const booleanValue = Boolean(value);
        formattedValue = formatMessage({
            id: booleanValue ? 'app.components.ToggleCheckbox.on-label' : 'app.components.ToggleCheckbox.off-label',
            defaultMessage: booleanValue ? 'true' : 'false'
        });
    }
    if ([
        'integer',
        'biginteger'
    ].includes(type)) {
        formattedValue = formatNumber(value, {
            maximumFractionDigits: 0,
            useGrouping: !isIdColumn
        });
    }
    return toString__default.default(formattedValue);
};

exports.CellValue = CellValue;
