// @strapi/upload/dist/admin/future/pages/Assets/components/Dnd/AssetsDndProvider.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import { useState, useCallback, useMemo, createContext, useContext } from 'react';
import { useSensors, useSensor, PointerSensor, DndContext, pointerWithin, DragOverlay } from '@dnd-kit/core';
import { useNotification } from '@strapi/admin/strapi-admin';
import { VisuallyHidden, Box } from '@strapi/design-system';
import { useIntl } from 'react-intl';
import { useGetFolderStructureQuery, useBulkMoveMutation } from '../../../../services/folders.mjs';
import { buildBulkMovePayload } from '../../../../utils/buildBulkMovePayload.mjs';
import { canDropItemOnFolder } from '../../../../utils/canDropItemOnFolder.mjs';
import { getBulkMoveErrorMessage } from '../../../../utils/getBulkMoveErrorMessage.mjs';
import { getTranslationKey } from '../../../../utils/translations.mjs';
import { parseFolderTargetId } from './dndIds.mjs';
import { DragOverlayChip } from './DragOverlayChip.mjs';

const AssetsDndContext = /*#__PURE__*/ createContext(null);
const useAssetsDndOptional = ()=>useContext(AssetsDndContext);
const getDragItems = (activeData)=>{
    if (!activeData) {
        return [];
    }
    // TODO(CMS-433): include selectedItems from active data for multi-drag payload members.
    return [
        activeData
    ];
};
const AssetsDndProvider = ({ children })=>{
    const { formatMessage } = useIntl();
    const { toggleNotification } = useNotification();
    const { data: folderStructure = [] } = useGetFolderStructureQuery();
    const [bulkMove, { isLoading: isMovePending }] = useBulkMoveMutation();
    const [activeItem, setActiveItem] = useState(null);
    const [liveAnnouncement, setLiveAnnouncement] = useState('');
    const announceToLiveRegion = useCallback((message)=>{
        setLiveAnnouncement('');
        requestAnimationFrame(()=>setLiveAnnouncement(message));
    }, []);
    const sensors = useSensors(useSensor(PointerSensor, {
        activationConstraint: {
            distance: 8
        }
    }));
    const contextValue = useMemo(()=>({
            isInternalDragActive: activeItem !== null,
            isMovePending
        }), [
        activeItem,
        isMovePending
    ]);
    const handleDragStart = useCallback((event)=>{
        const data = event.active.data.current;
        setActiveItem(data ?? null);
    }, []);
    const handleDragEnd = useCallback(async (event)=>{
        const { active, over } = event;
        setActiveItem(null);
        if (isMovePending || !over) {
            return;
        }
        const targetFolderId = parseFolderTargetId(over.id);
        const activeData = active.data.current;
        if (targetFolderId == null || !activeData) {
            return;
        }
        const items = getDragItems(activeData);
        if (!canDropItemOnFolder({
            items,
            targetFolderId,
            folderStructure
        })) {
            return;
        }
        const payload = buildBulkMovePayload(items);
        const successMessage = formatMessage({
            id: getTranslationKey('modal.move.success-label'),
            defaultMessage: 'Elements have been moved successfully.'
        });
        const errorFallback = formatMessage({
            id: getTranslationKey('modal.move.error-label'),
            defaultMessage: 'An error occurred while moving the elements.'
        });
        try {
            await bulkMove({
                ...payload,
                destinationFolderId: targetFolderId
            }).unwrap();
            announceToLiveRegion(successMessage);
            toggleNotification({
                type: 'success',
                message: successMessage
            });
        } catch (error) {
            const errorMessage = getBulkMoveErrorMessage(error, errorFallback);
            announceToLiveRegion(formatMessage({
                id: getTranslationKey('dnd.announce.move-failure'),
                defaultMessage: 'Move failed. {message}'
            }, {
                message: errorMessage
            }));
            toggleNotification({
                type: 'danger',
                message: errorMessage
            });
        }
    }, [
        announceToLiveRegion,
        bulkMove,
        folderStructure,
        formatMessage,
        isMovePending,
        toggleNotification
    ]);
    const handleDragCancel = useCallback(()=>{
        setActiveItem(null);
    }, []);
    const announcements = useMemo(()=>({
            onDragStart: ({ active })=>{
                const data = active.data.current;
                return data ? formatMessage({
                    id: getTranslationKey('dnd.announce.drag-start'),
                    defaultMessage: 'Picked up {name}. Drop on a folder to move.'
                }, {
                    name: data.name
                }) : '';
            },
            onDragOver: ()=>'',
            onDragEnd: ({ active, over })=>{
                if (!over) {
                    return formatMessage({
                        id: getTranslationKey('dnd.announce.cancel'),
                        defaultMessage: 'Drag cancelled.'
                    });
                }
                const targetFolderId = parseFolderTargetId(over.id);
                const activeData = active.data.current;
                if (targetFolderId == null || !activeData) {
                    return '';
                }
                const items = getDragItems(activeData);
                if (!canDropItemOnFolder({
                    items,
                    targetFolderId,
                    folderStructure
                })) {
                    return formatMessage({
                        id: getTranslationKey('dnd.announce.invalid-drop'),
                        defaultMessage: 'Cannot move item to this folder.'
                    });
                }
                return '';
            },
            onDragCancel: ()=>formatMessage({
                    id: getTranslationKey('dnd.announce.cancel'),
                    defaultMessage: 'Drag cancelled.'
                })
        }), [
        folderStructure,
        formatMessage
    ]);
    return /*#__PURE__*/ jsx(AssetsDndContext.Provider, {
        value: contextValue,
        children: /*#__PURE__*/ jsxs(DndContext, {
            sensors: sensors,
            collisionDetection: pointerWithin,
            onDragStart: handleDragStart,
            onDragEnd: handleDragEnd,
            onDragCancel: handleDragCancel,
            accessibility: {
                announcements
            },
            children: [
                /*#__PURE__*/ jsx(VisuallyHidden, {
                    "aria-live": "polite",
                    "aria-atomic": "true",
                    children: liveAnnouncement
                }),
                /*#__PURE__*/ jsx(Box, {
                    position: "relative",
                    children: children
                }),
                /*#__PURE__*/ jsx(DragOverlay, {
                    dropAnimation: null,
                    children: activeItem ? /*#__PURE__*/ jsx(DragOverlayChip, {
                        item: activeItem
                    }) : null
                })
            ]
        })
    });
}; // TODO(CMS-229): FolderTree droppables + keyboard move path.

export { AssetsDndProvider, useAssetsDndOptional };
