// @strapi/upload/dist/admin/future/pages/Assets/components/Dnd/AssetsDndProvider.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var core = require('@dnd-kit/core');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var reactIntl = require('react-intl');
var folders = require('../../../../services/folders.js');
var buildBulkMovePayload = require('../../../../utils/buildBulkMovePayload.js');
var canDropItemOnFolder = require('../../../../utils/canDropItemOnFolder.js');
var getBulkMoveErrorMessage = require('../../../../utils/getBulkMoveErrorMessage.js');
var translations = require('../../../../utils/translations.js');
var dndIds = require('./dndIds.js');
var DragOverlayChip = require('./DragOverlayChip.js');

const AssetsDndContext = /*#__PURE__*/ React.createContext(null);
const useAssetsDndOptional = ()=>React.useContext(AssetsDndContext);
const getDragItems = (activeData)=>{
    if (!activeData) {
        return [];
    }
    // TODO(CMS-433): include selectedItems from active data for multi-drag payload members.
    return [
        activeData
    ];
};
const AssetsDndProvider = ({ children })=>{
    const { formatMessage } = reactIntl.useIntl();
    const { toggleNotification } = strapiAdmin.useNotification();
    const { data: folderStructure = [] } = folders.useGetFolderStructureQuery();
    const [bulkMove, { isLoading: isMovePending }] = folders.useBulkMoveMutation();
    const [activeItem, setActiveItem] = React.useState(null);
    const [liveAnnouncement, setLiveAnnouncement] = React.useState('');
    const announceToLiveRegion = React.useCallback((message)=>{
        setLiveAnnouncement('');
        requestAnimationFrame(()=>setLiveAnnouncement(message));
    }, []);
    const sensors = core.useSensors(core.useSensor(core.PointerSensor, {
        activationConstraint: {
            distance: 8
        }
    }));
    const contextValue = React.useMemo(()=>({
            isInternalDragActive: activeItem !== null,
            isMovePending
        }), [
        activeItem,
        isMovePending
    ]);
    const handleDragStart = React.useCallback((event)=>{
        const data = event.active.data.current;
        setActiveItem(data ?? null);
    }, []);
    const handleDragEnd = React.useCallback(async (event)=>{
        const { active, over } = event;
        setActiveItem(null);
        if (isMovePending || !over) {
            return;
        }
        const targetFolderId = dndIds.parseFolderTargetId(over.id);
        const activeData = active.data.current;
        if (targetFolderId == null || !activeData) {
            return;
        }
        const items = getDragItems(activeData);
        if (!canDropItemOnFolder.canDropItemOnFolder({
            items,
            targetFolderId,
            folderStructure
        })) {
            return;
        }
        const payload = buildBulkMovePayload.buildBulkMovePayload(items);
        const successMessage = formatMessage({
            id: translations.getTranslationKey('modal.move.success-label'),
            defaultMessage: 'Elements have been moved successfully.'
        });
        const errorFallback = formatMessage({
            id: translations.getTranslationKey('modal.move.error-label'),
            defaultMessage: 'An error occurred while moving the elements.'
        });
        try {
            await bulkMove({
                ...payload,
                destinationFolderId: targetFolderId
            }).unwrap();
            announceToLiveRegion(successMessage);
            toggleNotification({
                type: 'success',
                message: successMessage
            });
        } catch (error) {
            const errorMessage = getBulkMoveErrorMessage.getBulkMoveErrorMessage(error, errorFallback);
            announceToLiveRegion(formatMessage({
                id: translations.getTranslationKey('dnd.announce.move-failure'),
                defaultMessage: 'Move failed. {message}'
            }, {
                message: errorMessage
            }));
            toggleNotification({
                type: 'danger',
                message: errorMessage
            });
        }
    }, [
        announceToLiveRegion,
        bulkMove,
        folderStructure,
        formatMessage,
        isMovePending,
        toggleNotification
    ]);
    const handleDragCancel = React.useCallback(()=>{
        setActiveItem(null);
    }, []);
    const announcements = React.useMemo(()=>({
            onDragStart: ({ active })=>{
                const data = active.data.current;
                return data ? formatMessage({
                    id: translations.getTranslationKey('dnd.announce.drag-start'),
                    defaultMessage: 'Picked up {name}. Drop on a folder to move.'
                }, {
                    name: data.name
                }) : '';
            },
            onDragOver: ()=>'',
            onDragEnd: ({ active, over })=>{
                if (!over) {
                    return formatMessage({
                        id: translations.getTranslationKey('dnd.announce.cancel'),
                        defaultMessage: 'Drag cancelled.'
                    });
                }
                const targetFolderId = dndIds.parseFolderTargetId(over.id);
                const activeData = active.data.current;
                if (targetFolderId == null || !activeData) {
                    return '';
                }
                const items = getDragItems(activeData);
                if (!canDropItemOnFolder.canDropItemOnFolder({
                    items,
                    targetFolderId,
                    folderStructure
                })) {
                    return formatMessage({
                        id: translations.getTranslationKey('dnd.announce.invalid-drop'),
                        defaultMessage: 'Cannot move item to this folder.'
                    });
                }
                return '';
            },
            onDragCancel: ()=>formatMessage({
                    id: translations.getTranslationKey('dnd.announce.cancel'),
                    defaultMessage: 'Drag cancelled.'
                })
        }), [
        folderStructure,
        formatMessage
    ]);
    return /*#__PURE__*/ jsxRuntime.jsx(AssetsDndContext.Provider, {
        value: contextValue,
        children: /*#__PURE__*/ jsxRuntime.jsxs(core.DndContext, {
            sensors: sensors,
            collisionDetection: core.pointerWithin,
            onDragStart: handleDragStart,
            onDragEnd: handleDragEnd,
            onDragCancel: handleDragCancel,
            accessibility: {
                announcements
            },
            children: [
                /*#__PURE__*/ jsxRuntime.jsx(designSystem.VisuallyHidden, {
                    "aria-live": "polite",
                    "aria-atomic": "true",
                    children: liveAnnouncement
                }),
                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                    position: "relative",
                    children: children
                }),
                /*#__PURE__*/ jsxRuntime.jsx(core.DragOverlay, {
                    dropAnimation: null,
                    children: activeItem ? /*#__PURE__*/ jsxRuntime.jsx(DragOverlayChip.DragOverlayChip, {
                        item: activeItem
                    }) : null
                })
            ]
        })
    });
}; // TODO(CMS-229): FolderTree droppables + keyboard move path.

exports.AssetsDndProvider = AssetsDndProvider;
exports.useAssetsDndOptional = useAssetsDndOptional;
