// @strapi/upload/dist/admin/future/pages/Assets/components/Dnd/DragOverlayChip.mjs
import { jsxs, jsx } from 'react/jsx-runtime';
import { Typography, Flex } from '@strapi/design-system';
import { Folder, File } from '@strapi/icons';
import { styled } from 'styled-components';

const Chip = styled(Flex)`
  align-items: center;
  gap: ${({ theme })=>theme.spaces[2]};
  padding: ${({ theme })=>`${theme.spaces[2]} ${theme.spaces[3]}`};
  border-radius: ${({ theme })=>theme.borderRadius};
  background: ${({ theme })=>theme.colors.primary100};
  box-shadow: ${({ theme })=>theme.shadows.tableShadow};
  cursor: grabbing;
  max-width: 24rem;
`;
const DragOverlayChip = ({ item })=>{
    const Icon = item.kind === 'folder' ? Folder : File;
    return /*#__PURE__*/ jsxs(Chip, {
        children: [
            /*#__PURE__*/ jsx(Icon, {
                width: 20,
                height: 20
            }),
            /*#__PURE__*/ jsx(Typography, {
                textColor: "neutral800",
                fontWeight: "semiBold",
                ellipsis: true,
                children: item.name
            })
        ]
    });
}; // TODO composite overlay with folder/file counts + total badge.

export { DragOverlayChip };
