// @strapi/upload/dist/admin/future/pages/Assets/components/Dnd/DragOverlayChip.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var styledComponents = require('styled-components');

const Chip = styledComponents.styled(designSystem.Flex)`
  align-items: center;
  gap: ${({ theme })=>theme.spaces[2]};
  padding: ${({ theme })=>`${theme.spaces[2]} ${theme.spaces[3]}`};
  border-radius: ${({ theme })=>theme.borderRadius};
  background: ${({ theme })=>theme.colors.primary100};
  box-shadow: ${({ theme })=>theme.shadows.tableShadow};
  cursor: grabbing;
  max-width: 24rem;
`;
const DragOverlayChip = ({ item })=>{
    const Icon = item.kind === 'folder' ? icons.Folder : icons.File;
    return /*#__PURE__*/ jsxRuntime.jsxs(Chip, {
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(Icon, {
                width: 20,
                height: 20
            }),
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                textColor: "neutral800",
                fontWeight: "semiBold",
                ellipsis: true,
                children: item.name
            })
        ]
    });
}; // TODO composite overlay with folder/file counts + total badge.

exports.DragOverlayChip = DragOverlayChip;
