// @strapi/upload/dist/admin/future/pages/Assets/components/Dnd/useAssetDnd.js
'use strict';

var React = require('react');
var core = require('@dnd-kit/core');
var folders = require('../../../../services/folders.js');
var canDropItemOnFolder = require('../../../../utils/canDropItemOnFolder.js');
var AssetsDndProvider = require('./AssetsDndProvider.js');
var dndIds = require('./dndIds.js');

const useFileDraggable = (asset)=>{
    const { isMovePending } = AssetsDndProvider.useAssetsDndOptional() ?? {
        isMovePending: false
    };
    const data = React.useMemo(()=>({
            kind: 'file',
            id: asset.id,
            name: asset.name,
            folderId: asset.folder == null ? null : typeof asset.folder === 'number' ? asset.folder : Number(asset.folder) || null
        }), [
        asset.folder,
        asset.id,
        asset.name
    ]);
    return core.useDraggable({
        id: dndIds.toFileDraggableId(asset.id),
        data,
        disabled: isMovePending
    });
};
const useFolderDraggableDroppable = (folder)=>{
    const { isMovePending } = AssetsDndProvider.useAssetsDndOptional() ?? {
        isMovePending: false
    };
    const { active } = core.useDndContext();
    const { data: folderStructure = [] } = folders.useGetFolderStructureQuery();
    const parentId = typeof folder.parent === 'object' && folder.parent != null ? folder.parent.id ?? null : folder.parent ?? null;
    const dragData = React.useMemo(()=>({
            kind: 'folder',
            id: folder.id,
            name: folder.name,
            parentId
        }), [
        folder.id,
        folder.name,
        parentId
    ]);
    const dropData = React.useMemo(()=>({
            kind: 'folder-target',
            id: folder.id,
            name: folder.name
        }), [
        folder.id,
        folder.name
    ]);
    const draggable = core.useDraggable({
        id: dndIds.toFolderDraggableId(folder.id),
        data: dragData,
        disabled: isMovePending
    });
    const droppable = core.useDroppable({
        id: dndIds.toFolderTargetId(folder.id),
        data: dropData,
        disabled: isMovePending
    });
    const activeData = active?.data.current;
    const isValidDropTarget = React.useMemo(()=>{
        if (!activeData || activeData.id === folder.id && activeData.kind === 'folder') {
            return false;
        }
        return canDropItemOnFolder.canDropItemOnFolder({
            items: [
                activeData
            ],
            targetFolderId: folder.id,
            folderStructure
        });
    }, [
        activeData,
        folder.id,
        folderStructure
    ]);
    const isOver = droppable.isOver;
    const showValidDropHighlight = isOver && isValidDropTarget;
    const showInvalidDropCursor = isOver && !isValidDropTarget && active != null;
    return {
        dragData,
        draggable,
        droppable,
        isDragging: draggable.isDragging,
        showValidDropHighlight,
        showInvalidDropCursor
    };
};

exports.useFileDraggable = useFileDraggable;
exports.useFolderDraggableDroppable = useFolderDraggableDroppable;
