// @strapi/core/dist/configuration/index.js
'use strict';

var os = require('os');
var path = require('path');
var _ = require('lodash');
var fp = require('lodash/fp');
var dotenv = require('dotenv');
var strapiUtils = require('@strapi/utils');
var urls = require('./urls.js');
var configLoader = require('./config-loader.js');
var getDirs = require('./get-dirs.js');
var _package = require('../package.json.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var os__default = /*#__PURE__*/_interopDefault(os);
var path__default = /*#__PURE__*/_interopDefault(path);
var ___default = /*#__PURE__*/_interopDefault(_);
var dotenv__default = /*#__PURE__*/_interopDefault(dotenv);

dotenv__default.default.config({
    path: process.env.ENV_PATH
});
process.env.NODE_ENV = process.env.NODE_ENV || 'development';
const defaultServerConfig = {
    host: process.env.HOST || os__default.default.hostname() || 'localhost',
    port: Number(process.env.PORT) || 1337,
    proxy: false,
    cron: {
        enabled: false
    },
    dirs: {
        public: './public'
    },
    transfer: {
        remote: {
            enabled: true
        }
    },
    logger: {
        updates: {
            enabled: true
        },
        startup: {
            enabled: true
        }
    },
    openapi: {
        'content-api': {
            access: 'disabled',
            route: {
                path: '/openapi.json'
            },
            cache: {
                enabled: true,
                maxAgeMs: 60000,
                filePath: '.strapi/openapi/content-api.json'
            }
        },
        admin: {
            access: 'disabled',
            route: {
                path: '/openapi.json'
            },
            cache: {
                enabled: true,
                maxAgeMs: 60000,
                filePath: '.strapi/openapi/admin.json'
            }
        }
    }
};
const defaultConfig = {
    server: defaultServerConfig,
    admin: {},
    api: {
        rest: {
            prefix: '/api'
        }
    }
};
const loadConfiguration = (opts)=>{
    const { appDir, distDir, autoReload = false, serveAdminPanel = true } = opts;
    const pkgJSON = require(path__default.default.resolve(appDir, 'package.json'));
    const configDir = path__default.default.resolve(distDir || process.cwd(), 'config');
    const rootConfig = {
        launchedAt: Date.now(),
        autoReload,
        environment: process.env.NODE_ENV,
        uuid: ___default.default.get(pkgJSON, 'strapi.uuid'),
        installId: ___default.default.get(pkgJSON, 'strapi.installId'),
        packageJsonStrapi: ___default.default.omit(___default.default.get(pkgJSON, 'strapi', {}), 'uuid'),
        info: {
            ...pkgJSON,
            strapi: _package.version
        },
        admin: {
            serveAdminPanel
        }
    };
    // See packages/core/core/src/domain/module/index.ts for plugin config loading
    const baseConfig = fp.omit('plugins', configLoader(configDir)); // plugin config will be loaded later
    const envDir = path__default.default.resolve(configDir, 'env', process.env.NODE_ENV);
    const envConfig = configLoader(envDir);
    const config = ___default.default.merge(rootConfig, defaultConfig, baseConfig, envConfig);
    const { serverUrl, adminUrl } = urls.getConfigUrls(config);
    const serverAbsoluteUrl = urls.getAbsoluteServerUrl(config);
    const adminAbsoluteUrl = urls.getAbsoluteAdminUrl(config);
    const sameOrigin = new URL(adminAbsoluteUrl).origin === new URL(serverAbsoluteUrl).origin;
    const adminPath = sameOrigin ? adminUrl.replace(strapiUtils.strings.getCommonPath(serverUrl, adminUrl), '') : new URL(adminUrl).pathname;
    ___default.default.set(config, 'server.url', serverUrl);
    ___default.default.set(config, 'server.absoluteUrl', serverAbsoluteUrl);
    ___default.default.set(config, 'admin.url', adminUrl);
    ___default.default.set(config, 'admin.path', adminPath);
    ___default.default.set(config, 'admin.absoluteUrl', adminAbsoluteUrl);
    ___default.default.set(config, 'dirs', getDirs.getDirs(opts, config));
    return config;
};

exports.loadConfiguration = loadConfiguration;
