// @strapi/typescript-utils/dist/generators/common/models/mappers.js
'use strict';

var ts = require('typescript');
var _ = require('lodash/fp');
var utils = require('./utils.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var ts__namespace = /*#__PURE__*/_interopNamespace(ts);
var ___default = /*#__PURE__*/_interopDefault(_);

const { factory } = ts__namespace;
function string() {
    return [
        utils.withAttributeNamespace('String')
    ];
}
function text() {
    return [
        utils.withAttributeNamespace('Text')
    ];
}
function richtext() {
    return [
        utils.withAttributeNamespace('RichText')
    ];
}
function password() {
    return [
        utils.withAttributeNamespace('Password')
    ];
}
function email() {
    return [
        utils.withAttributeNamespace('Email')
    ];
}
function date() {
    return [
        utils.withAttributeNamespace('Date')
    ];
}
function time() {
    return [
        utils.withAttributeNamespace('Time')
    ];
}
function datetime() {
    return [
        utils.withAttributeNamespace('DateTime')
    ];
}
function timestamp() {
    return [
        utils.withAttributeNamespace('Timestamp')
    ];
}
function integer() {
    return [
        utils.withAttributeNamespace('Integer')
    ];
}
function biginteger() {
    return [
        utils.withAttributeNamespace('BigInteger')
    ];
}
function float() {
    return [
        utils.withAttributeNamespace('Float')
    ];
}
function decimal() {
    return [
        utils.withAttributeNamespace('Decimal')
    ];
}
function uid({ attribute }) {
    const { targetField, options } = attribute;
    // If there are no params to compute, then return the attribute type alone
    if (targetField === undefined && options === undefined) {
        return [
            utils.withAttributeNamespace('UID')
        ];
    }
    const params = [];
    // If the targetField property is defined, then reference it,
    // otherwise, put `undefined` keyword type node as placeholder
    const targetFieldParam = ___default.default.isUndefined(targetField) ? factory.createKeywordTypeNode(ts__namespace.SyntaxKind.UndefinedKeyword) : factory.createStringLiteral(targetField);
    params.push(targetFieldParam);
    // If the options property is defined, transform it to
    // a type literal node and add it to the params list
    if (___default.default.isObject(options)) {
        params.push(utils.toTypeLiteral(options));
    }
    return [
        utils.withAttributeNamespace('UID'),
        params
    ];
}
function enumeration({ attribute }) {
    const { enum: enumValues } = attribute;
    return [
        utils.withAttributeNamespace('Enumeration'),
        [
            utils.toTypeLiteral(enumValues)
        ]
    ];
}
function boolean() {
    return [
        utils.withAttributeNamespace('Boolean')
    ];
}
function json() {
    return [
        utils.withAttributeNamespace('JSON')
    ];
}
function blocks() {
    return [
        utils.withAttributeNamespace('Blocks')
    ];
}
function media({ attribute }) {
    const { allowedTypes, multiple } = attribute;
    const params = [];
    const typesParam = allowedTypes ? factory.createUnionTypeNode(allowedTypes.map((allowedType)=>factory.createStringLiteral(allowedType))) : factory.createKeywordTypeNode(ts__namespace.SyntaxKind.UndefinedKeyword);
    if (allowedTypes || multiple) {
        params.push(typesParam);
    }
    if (multiple) {
        params.push(factory.createTrue());
    }
    return [
        utils.withAttributeNamespace('Media'),
        params
    ];
}
function relation({ attribute }) {
    const { relation: relationType, target } = attribute;
    const isMorphRelation = relationType.toLowerCase().includes('morph');
    if (isMorphRelation) {
        return [
            utils.withAttributeNamespace('Relation'),
            [
                factory.createStringLiteral(relationType, true)
            ]
        ];
    }
    return [
        utils.withAttributeNamespace('Relation'),
        [
            factory.createStringLiteral(relationType, true),
            factory.createStringLiteral(target, true)
        ]
    ];
}
function component({ attribute }) {
    const target = attribute.component;
    const params = [
        factory.createStringLiteral(target, true)
    ];
    if (attribute.repeatable) {
        params.push(factory.createTrue());
    } else {
        params.push(factory.createFalse());
    }
    return [
        utils.withAttributeNamespace('Component'),
        params
    ];
}
function dynamiczone({ attribute }) {
    const componentsParam = factory.createTupleTypeNode(attribute.components.map((c)=>factory.createStringLiteral(c)));
    return [
        utils.withAttributeNamespace('DynamicZone'),
        [
            componentsParam
        ]
    ];
}
const mappers = {
    string,
    text,
    richtext,
    password,
    email,
    date,
    time,
    datetime,
    timestamp,
    integer,
    biginteger,
    float,
    decimal,
    uid,
    enumeration,
    boolean,
    json,
    blocks,
    media,
    relation,
    component,
    dynamiczone
};

exports.mappers = mappers;
