// @strapi/typescript-utils/dist/generators/common/models/utils.js
'use strict';

var ts = require('typescript');
var _ = require('lodash/fp');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var ts__namespace = /*#__PURE__*/_interopNamespace(ts);

const { factory } = ts__namespace;
const NAMESPACES = {
    Struct: 'Struct',
    Schema: 'Schema'
};
/**
 * Extract a valid interface name from a schema uid
 */ const getSchemaInterfaceName = _.pipe(_.replace(/(:.)/, ' '), _.camelCase, _.upperFirst);
const getSchemaModelType = (schema)=>{
    const { modelType, kind } = schema;
    // Components
    if (modelType === 'component') {
        return 'component';
    }
    // Content-Types
    if (modelType === 'contentType') {
        return kind;
    }
    return null;
};
/**
 * Get the parent type name to extend based on the schema's nature
 */ const getSchemaExtendsTypeName = (schema)=>{
    const base = getSchemaModelType(schema);
    if (base === null) {
        return null;
    }
    return `${NAMESPACES.Struct}.${_.upperFirst(base)}Schema`;
};
/**
 * Get a type node based on a type and its params
 */ const getTypeNode = (typeName, params = [])=>{
    return factory.createTypeReferenceNode(factory.createIdentifier(typeName), params);
};
/**
 * Transform a regular JavaScript object or scalar value into a literal expression
 */ const toTypeLiteral = (data)=>{
    if (_.isUndefined(data)) {
        return factory.createLiteralTypeNode(ts__namespace.SyntaxKind.UndefinedKeyword);
    }
    if (_.isNull(data)) {
        return factory.createLiteralTypeNode(ts__namespace.SyntaxKind.NullKeyword);
    }
    if (_.isString(data)) {
        return factory.createStringLiteral(data, true);
    }
    if (_.isNumber(data)) {
        return data < 0 ? factory.createPrefixUnaryExpression(ts__namespace.SyntaxKind.MinusToken, factory.createNumericLiteral(-data)) : factory.createNumericLiteral(data);
    }
    if (_.isBoolean(data)) {
        return data ? factory.createTrue() : factory.createFalse();
    }
    if (_.isArray(data)) {
        return factory.createTupleTypeNode(data.map((item)=>toTypeLiteral(item)));
    }
    if (_.isDate(data)) {
        return factory.createStringLiteral(data.toISOString());
    }
    if (typeof data !== 'object') {
        throw new Error(`Cannot convert to object literal. Unknown type "${typeof data}"`);
    }
    const entries = Object.entries(data).sort((a, b)=>a[0].localeCompare(b[0]));
    const props = entries.reduce((acc, [key, value])=>{
        // Handle keys such as content-type-builder & co.
        const identifier = key.includes('-') ? factory.createStringLiteral(key, true) : factory.createIdentifier(key);
        return [
            ...acc,
            factory.createPropertyDeclaration(undefined, identifier, undefined, toTypeLiteral(value), undefined)
        ];
    }, []);
    return factory.createTypeLiteralNode(props);
};
/**
 * Add the Schema.Attribute namespace before the typename
 */ const withAttributeNamespace = (typeName)=>`${NAMESPACES.Schema}.Attribute.${typeName}`;

exports.NAMESPACES = NAMESPACES;
exports.getSchemaExtendsTypeName = getSchemaExtendsTypeName;
exports.getSchemaInterfaceName = getSchemaInterfaceName;
exports.getSchemaModelType = getSchemaModelType;
exports.getTypeNode = getTypeNode;
exports.toTypeLiteral = toTypeLiteral;
exports.withAttributeNamespace = withAttributeNamespace;
