// @strapi/typescript-utils/dist/generators/common/models/utils.mjs
import * as ts from 'typescript';
import { pipe, replace, camelCase, upperFirst, isUndefined, isNull, isString, isNumber, isBoolean, isArray, isDate } from 'lodash/fp';

const { factory } = ts;
const NAMESPACES = {
    Struct: 'Struct',
    Schema: 'Schema'
};
/**
 * Extract a valid interface name from a schema uid
 */ const getSchemaInterfaceName = pipe(replace(/(:.)/, ' '), camelCase, upperFirst);
const getSchemaModelType = (schema)=>{
    const { modelType, kind } = schema;
    // Components
    if (modelType === 'component') {
        return 'component';
    }
    // Content-Types
    if (modelType === 'contentType') {
        return kind;
    }
    return null;
};
/**
 * Get the parent type name to extend based on the schema's nature
 */ const getSchemaExtendsTypeName = (schema)=>{
    const base = getSchemaModelType(schema);
    if (base === null) {
        return null;
    }
    return `${NAMESPACES.Struct}.${upperFirst(base)}Schema`;
};
/**
 * Get a type node based on a type and its params
 */ const getTypeNode = (typeName, params = [])=>{
    return factory.createTypeReferenceNode(factory.createIdentifier(typeName), params);
};
/**
 * Transform a regular JavaScript object or scalar value into a literal expression
 */ const toTypeLiteral = (data)=>{
    if (isUndefined(data)) {
        return factory.createLiteralTypeNode(ts.SyntaxKind.UndefinedKeyword);
    }
    if (isNull(data)) {
        return factory.createLiteralTypeNode(ts.SyntaxKind.NullKeyword);
    }
    if (isString(data)) {
        return factory.createStringLiteral(data, true);
    }
    if (isNumber(data)) {
        return data < 0 ? factory.createPrefixUnaryExpression(ts.SyntaxKind.MinusToken, factory.createNumericLiteral(-data)) : factory.createNumericLiteral(data);
    }
    if (isBoolean(data)) {
        return data ? factory.createTrue() : factory.createFalse();
    }
    if (isArray(data)) {
        return factory.createTupleTypeNode(data.map((item)=>toTypeLiteral(item)));
    }
    if (isDate(data)) {
        return factory.createStringLiteral(data.toISOString());
    }
    if (typeof data !== 'object') {
        throw new Error(`Cannot convert to object literal. Unknown type "${typeof data}"`);
    }
    const entries = Object.entries(data).sort((a, b)=>a[0].localeCompare(b[0]));
    const props = entries.reduce((acc, [key, value])=>{
        // Handle keys such as content-type-builder & co.
        const identifier = key.includes('-') ? factory.createStringLiteral(key, true) : factory.createIdentifier(key);
        return [
            ...acc,
            factory.createPropertyDeclaration(undefined, identifier, undefined, toTypeLiteral(value), undefined)
        ];
    }, []);
    return factory.createTypeLiteralNode(props);
};
/**
 * Add the Schema.Attribute namespace before the typename
 */ const withAttributeNamespace = (typeName)=>`${NAMESPACES.Schema}.Attribute.${typeName}`;

export { NAMESPACES, getSchemaExtendsTypeName, getSchemaInterfaceName, getSchemaModelType, getTypeNode, toTypeLiteral, withAttributeNamespace };
