// @strapi/upload/dist/server/register.js
'use strict';

var _ = require('lodash');
var sharp = require('sharp');
var utils = require('@strapi/utils');
var upload = require('./middlewares/upload.js');
var contentApi = require('./documentation/content-api.json.js');
var aiMetadataJob = require('./models/ai-metadata-job.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var ___default = /*#__PURE__*/_interopDefault(_);
var sharp__default = /*#__PURE__*/_interopDefault(sharp);

const { PayloadTooLargeError } = utils.errors;
const { bytesToHumanReadable, kbytesToBytes } = utils.file;
/**
 * Register upload plugin
 */ async function register({ strapi }) {
    // Register AI metadata job model
    strapi.get('models').add(aiMetadataJob.aiMetadataJob);
    const raw = strapi.config.get('plugin::upload') ?? {};
    // createProvider needs a provider; empty get() (e.g. in tests) still has defaults.
    const uploadConfig = {
        provider: 'local',
        providerOptions: {},
        actionOptions: {},
        ...raw
    };
    // Configure sharp memory management
    const { cache = false, concurrency = 1 } = uploadConfig.sharp ?? {};
    sharp__default.default.cache(cache);
    sharp__default.default.concurrency(concurrency);
    strapi.plugin('upload').provider = createProvider(uploadConfig);
    await upload({
        strapi
    });
    if (strapi.plugin('graphql')) {
        const { installGraphqlExtension } = await Promise.resolve().then(function () { return require('./graphql.js'); });
        installGraphqlExtension({
            strapi
        });
    }
    if (strapi.plugin('documentation')) {
        strapi.plugin('documentation').service('override').registerOverride(contentApi.default, {
            pluginOrigin: 'upload',
            excludeFromGeneration: [
                'upload'
            ]
        });
    }
}
const createProvider = (config)=>{
    const { providerOptions, actionOptions = {} } = config;
    const providerName = ___default.default.toLower(config.provider);
    let provider;
    let modulePath;
    try {
        modulePath = require.resolve(`@strapi/provider-upload-${providerName}`);
    } catch (error) {
        if (typeof error === 'object' && error !== null && 'code' in error && error.code === 'MODULE_NOT_FOUND') {
            modulePath = providerName;
        } else {
            throw error;
        }
    }
    try {
        provider = require(modulePath);
    } catch (err) {
        const newError = new Error(`Could not load upload provider "${providerName}".`);
        if (err instanceof Error) {
            newError.stack = err.stack;
        }
        throw newError;
    }
    const providerInstance = provider.init(providerOptions);
    if (!providerInstance.delete) {
        throw new Error(`The upload provider "${providerName}" doesn't implement the delete method.`);
    }
    if (!providerInstance.upload && !providerInstance.uploadStream) {
        throw new Error(`The upload provider "${providerName}" doesn't implement the uploadStream nor the upload method.`);
    }
    if (!providerInstance.uploadStream) {
        process.emitWarning(`The upload provider "${providerName}" doesn't implement the uploadStream function. Strapi will fallback on the upload method. Some performance issues may occur.`);
    }
    const wrappedProvider = ___default.default.mapValues(providerInstance, (method, methodName)=>{
        return async (file, options = actionOptions[methodName])=>providerInstance[methodName](file, options);
    });
    return Object.assign(Object.create(baseProvider), wrappedProvider);
};
const baseProvider = {
    extend (obj) {
        Object.assign(this, obj);
    },
    checkFileSize (file, { sizeLimit }) {
        if (sizeLimit && kbytesToBytes(file.size) > sizeLimit) {
            throw new PayloadTooLargeError(`${file.originalFilename} exceeds size limit of ${bytesToHumanReadable(sizeLimit)}.`);
        }
    },
    getSignedUrl (file) {
        return file;
    },
    isPrivate () {
        return false;
    }
};

exports.register = register;
