// @strapi/cloud-cli/dist/services/logger.js
'use strict';

var chalk = require('chalk');
var stringify = require('fast-safe-stringify');
var ora = require('ora');
var cliProgress = require('cli-progress');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var chalk__default = /*#__PURE__*/_interopDefault(chalk);
var stringify__default = /*#__PURE__*/_interopDefault(stringify);
var ora__default = /*#__PURE__*/_interopDefault(ora);
var cliProgress__namespace = /*#__PURE__*/_interopNamespace(cliProgress);

const stringifyArg = (arg)=>{
    return typeof arg === 'object' ? stringify__default.default(arg) : arg;
};
const createLogger = (options = {})=>{
    const { silent = false, debug = false, timestamp = true } = options;
    const state = {
        errors: 0,
        warning: 0
    };
    return {
        get warnings () {
            return state.warning;
        },
        get errors () {
            return state.errors;
        },
        async debug (...args) {
            if (silent || !debug) {
                return;
            }
            console.log(chalk__default.default.cyan(`[DEBUG]${timestamp ? `\t[${new Date().toISOString()}]` : ''}`), ...args.map(stringifyArg));
        },
        info (...args) {
            if (silent) {
                return;
            }
            console.info(chalk__default.default.blue(`[INFO]${timestamp ? `\t[${new Date().toISOString()}]` : ''}`), ...args.map(stringifyArg));
        },
        log (...args) {
            if (silent) {
                return;
            }
            console.info(chalk__default.default.blue(`${timestamp ? `\t[${new Date().toISOString()}]` : ''}`), ...args.map(stringifyArg));
        },
        success (...args) {
            if (silent) {
                return;
            }
            console.info(chalk__default.default.green(`[SUCCESS]${timestamp ? `\t[${new Date().toISOString()}]` : ''}`), ...args.map(stringifyArg));
        },
        warn (...args) {
            state.warning += 1;
            if (silent) {
                return;
            }
            console.warn(chalk__default.default.yellow(`[WARN]${timestamp ? `\t[${new Date().toISOString()}]` : ''}`), ...args.map(stringifyArg));
        },
        error (...args) {
            state.errors += 1;
            if (silent) {
                return;
            }
            console.error(chalk__default.default.red(`[ERROR]${timestamp ? `\t[${new Date().toISOString()}]` : ''}`), ...args.map(stringifyArg));
        },
        // @ts-expect-error – returning a subpart of ora is fine because the types tell us what is what.
        spinner (text) {
            if (silent) {
                return {
                    succeed () {
                        return this;
                    },
                    fail () {
                        return this;
                    },
                    start () {
                        return this;
                    },
                    text: '',
                    isSpinning: false
                };
            }
            return ora__default.default(text);
        },
        progressBar (totalSize, text) {
            if (silent) {
                return {
                    start () {
                        return this;
                    },
                    stop () {
                        return this;
                    },
                    update () {
                        return this;
                    }
                };
            }
            const progressBar = new cliProgress__namespace.SingleBar({
                format: `${text ? `${text} |` : ''}${chalk__default.default.green('{bar}')}| {percentage}%`,
                barCompleteChar: '\u2588',
                barIncompleteChar: '\u2591',
                hideCursor: true,
                forceRedraw: true
            });
            progressBar.start(totalSize, 0);
            return progressBar;
        }
    };
};

exports.createLogger = createLogger;
