// @strapi/core/dist/ee/license.js
'use strict';

var fs = require('fs');
var path = require('path');
var crypto = require('crypto');
var z = require('zod/v4');
var strapiUtils = require('@strapi/utils');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var fs__default = /*#__PURE__*/_interopDefault(fs);
var crypto__default = /*#__PURE__*/_interopDefault(crypto);
var z__namespace = /*#__PURE__*/_interopNamespace(z);

const DEFAULT_FEATURES = {
    bronze: [],
    silver: [],
    gold: [
        {
            name: 'sso'
        },
        // Set a null retention duration to allow the user to override it
        // The default of 90 days is set in the audit logs service
        {
            name: 'audit-logs',
            options: {
                retentionDays: null
            }
        },
        {
            name: 'review-workflows'
        },
        {
            name: 'cms-content-releases'
        },
        {
            name: 'cms-content-history',
            options: {
                retentionDays: 99999
            }
        },
        {
            name: 'cms-advanced-preview'
        }
    ]
};
const LICENSE_REGISTRY_URI = 'https://license.strapi.io';
// Shape of a successful (200) response from the license registry.
const licenseSuccessSchema = z__namespace.object({
    data: z__namespace.object({
        license: z__namespace.string()
    })
});
// Shape of a bad-request (400) response from the license registry.
const licenseErrorSchema = z__namespace.object({
    error: z__namespace.object({
        message: z__namespace.string()
    })
});
const publicKey = fs__default.default.readFileSync(path.resolve(__dirname, '../../resources/key.pub'));
class LicenseCheckError extends Error {
    constructor(message, shouldFallback = false){
        super(message), this.shouldFallback = false;
        this.shouldFallback = shouldFallback;
    }
}
const readLicense = (directory)=>{
    try {
        const path$1 = path.join(directory, 'license.txt');
        return fs__default.default.readFileSync(path$1).toString();
    } catch (error) {
        if (typeof error === 'object' && error !== null && 'code' in error && error.code !== 'ENOENT') {
            throw Error('License file not readable, review its format and access rules.');
        }
    }
};
const verifyLicense = (license)=>{
    const [signature, base64Content] = Buffer.from(license, 'base64').toString().split('\n');
    if (!signature || !base64Content) {
        throw new Error('Invalid license.');
    }
    const stringifiedContent = Buffer.from(base64Content, 'base64').toString();
    const verify = crypto__default.default.createVerify('RSA-SHA256');
    verify.update(stringifiedContent);
    verify.end();
    const verified = verify.verify(publicKey, signature, 'base64');
    if (!verified) {
        throw new Error('Invalid license.');
    }
    const licenseInfo = JSON.parse(stringifiedContent);
    if (!licenseInfo.features) {
        licenseInfo.features = DEFAULT_FEATURES[licenseInfo.type];
    }
    if (!licenseInfo.isTrial) {
        licenseInfo.isTrial = false;
    }
    Object.freeze(licenseInfo.features);
    return licenseInfo;
};
const createShouldFallbackError = ()=>{
    return new LicenseCheckError('Could not proceed to the online validation of your license.', true);
};
const throwShouldFallbackError = ()=>{
    throw createShouldFallbackError();
};
const fetchLicense = async ({ strapi }, key, projectId)=>{
    const { installId: installIdFromPackageJson } = strapi.config;
    const response = await strapi.fetch(`${LICENSE_REGISTRY_URI}/api/licenses/validate`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            key,
            projectId,
            deviceId: strapiUtils.generateInstallId(projectId, installIdFromPackageJson)
        })
    }).catch(throwShouldFallbackError);
    const contentType = response.headers.get('Content-Type');
    if (contentType?.includes('application/json')) {
        const body = await response.json();
        switch(response.status){
            case 200:
                {
                    const result = licenseSuccessSchema.safeParse(body);
                    if (!result.success) throw createShouldFallbackError();
                    return result.data.data.license;
                }
            case 400:
                {
                    const result = licenseErrorSchema.safeParse(body);
                    if (!result.success) throw createShouldFallbackError();
                    throw new LicenseCheckError(result.data.error.message);
                }
            case 404:
                throw new LicenseCheckError('The license used does not exists.');
            default:
                throw createShouldFallbackError();
        }
    } else {
        throw createShouldFallbackError();
    }
};

exports.LICENSE_REGISTRY_URI = LICENSE_REGISTRY_URI;
exports.LicenseCheckError = LicenseCheckError;
exports.fetchLicense = fetchLicense;
exports.readLicense = readLicense;
exports.verifyLicense = verifyLicense;
