// @strapi/admin/dist/admin/admin/src/components/LazyOutlet.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var designSystem = require('@strapi/design-system');
var reactRouterDom = require('react-router-dom');
var PageHelpers = require('./PageHelpers.js');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

const NestedLoading = ()=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
        alignItems: "center",
        justifyContent: "center",
        minHeight: "50vh",
        padding: 8,
        "aria-busy": true,
        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Loader, {
            children: "Loading content."
        })
    });
/**
 * Renders child routes with a visible loading state while lazy chunks load.
 * Suspense fallbacks are suppressed during React Router transitions, so we also
 * check `useNavigation().state` to replace stale outlet content while navigating.
 */ const LazyOutlet = ({ nested = false, suspenseOnly = false })=>{
    const navigation = reactRouterDom.useNavigation();
    const isNavigating = !suspenseOnly && navigation.state === 'loading';
    const fallback = nested ? /*#__PURE__*/ jsxRuntime.jsx(NestedLoading, {}) : /*#__PURE__*/ jsxRuntime.jsx(PageHelpers.Page.Loading, {});
    if (isNavigating) {
        return fallback;
    }
    return /*#__PURE__*/ jsxRuntime.jsx(React__namespace.Suspense, {
        fallback: fallback,
        children: /*#__PURE__*/ jsxRuntime.jsx(reactRouterDom.Outlet, {})
    });
};

exports.LazyOutlet = LazyOutlet;
