// @strapi/admin/dist/admin/admin/src/components/LazyOutlet.mjs
import { jsx } from 'react/jsx-runtime';
import * as React from 'react';
import { Flex, Loader } from '@strapi/design-system';
import { useNavigation, Outlet } from 'react-router-dom';
import { Page } from './PageHelpers.mjs';

const NestedLoading = ()=>/*#__PURE__*/ jsx(Flex, {
        alignItems: "center",
        justifyContent: "center",
        minHeight: "50vh",
        padding: 8,
        "aria-busy": true,
        children: /*#__PURE__*/ jsx(Loader, {
            children: "Loading content."
        })
    });
/**
 * Renders child routes with a visible loading state while lazy chunks load.
 * Suspense fallbacks are suppressed during React Router transitions, so we also
 * check `useNavigation().state` to replace stale outlet content while navigating.
 */ const LazyOutlet = ({ nested = false, suspenseOnly = false })=>{
    const navigation = useNavigation();
    const isNavigating = !suspenseOnly && navigation.state === 'loading';
    const fallback = nested ? /*#__PURE__*/ jsx(NestedLoading, {}) : /*#__PURE__*/ jsx(Page.Loading, {});
    if (isNavigating) {
        return fallback;
    }
    return /*#__PURE__*/ jsx(React.Suspense, {
        fallback: fallback,
        children: /*#__PURE__*/ jsx(Outlet, {})
    });
};

export { LazyOutlet };
