// @strapi/content-manager/dist/admin/utils/layouts/normalizeContentManagerLayout.mjs
const DEFAULT_RESPONSE_SETTINGS = {
    bulkable: true,
    filterable: true,
    searchable: true,
    pageSize: 10,
    mainField: 'id',
    defaultSortBy: '',
    defaultSortOrder: 'ASC'
};
const isRecord = (value)=>{
    return Boolean(value) && typeof value === 'object' && !Array.isArray(value);
};
const warnOnceInDevelopment = (()=>{
    const warnings = new Set();
    return (message)=>{
        if (process.env.NODE_ENV !== 'development' || warnings.has(message)) {
            return;
        }
        warnings.add(message);
        console.warn(`[Content Manager] ${message}`);
    };
})();
const normalizeResponseSettings = (settings, path)=>{
    if (!isRecord(settings)) {
        warnOnceInDevelopment(`Received malformed settings at "${path}". Falling back to defaults.`);
    }
    return {
        ...DEFAULT_RESPONSE_SETTINGS,
        ...isRecord(settings) ? settings : {}
    };
};
const normalizeResponseMetadatas = (metadatas, path)=>{
    if (!isRecord(metadatas)) {
        warnOnceInDevelopment(`Received malformed metadatas at "${path}". Falling back to empty.`);
        return {};
    }
    return Object.entries(metadatas).reduce((acc, [name, metadata])=>{
        const metadataRecord = isRecord(metadata) ? metadata : {};
        if (!isRecord(metadata)) {
            warnOnceInDevelopment(`Received malformed metadata for "${path}.${name}". Repairing entry.`);
        }
        if (!isRecord(metadataRecord.edit)) {
            warnOnceInDevelopment(`Received malformed edit metadata for "${path}.${name}". Falling back to empty.`);
        }
        if (!isRecord(metadataRecord.list)) {
            warnOnceInDevelopment(`Received malformed list metadata for "${path}.${name}". Falling back to empty.`);
        }
        acc[name] = {
            edit: isRecord(metadataRecord.edit) ? metadataRecord.edit : {},
            list: isRecord(metadataRecord.list) ? metadataRecord.list : {}
        };
        return acc;
    }, {});
};
const normalizeResponseLayouts = (layouts, path)=>{
    if (!isRecord(layouts)) {
        warnOnceInDevelopment(`Received malformed layouts at "${path}". Falling back to empty.`);
        return {
            edit: [],
            list: []
        };
    }
    if (!Array.isArray(layouts.edit)) {
        warnOnceInDevelopment(`Received malformed edit layout at "${path}". Falling back to empty.`);
    }
    if (!Array.isArray(layouts.list)) {
        warnOnceInDevelopment(`Received malformed list layout at "${path}". Falling back to empty.`);
    }
    return {
        edit: Array.isArray(layouts.edit) ? layouts.edit.map((row, rowIndex)=>{
            if (!Array.isArray(row)) {
                warnOnceInDevelopment(`Dropped malformed edit layout row "${path}.edit.${rowIndex}".`);
                return [];
            }
            return row.filter((field, fieldIndex)=>{
                const isValidField = isRecord(field) && typeof field.name === 'string' && typeof field.size === 'number';
                if (!isValidField) {
                    warnOnceInDevelopment(`Dropped malformed edit layout field "${path}.edit.${rowIndex}.${fieldIndex}".`);
                }
                return isValidField;
            });
        }) : [],
        list: Array.isArray(layouts.list) ? layouts.list.filter((field, index)=>{
            const isValidField = typeof field === 'string';
            if (!isValidField) {
                warnOnceInDevelopment(`Dropped malformed list layout field "${path}.list.${index}".`);
            }
            return isValidField;
        }) : []
    };
};
const normalizeConfigurationResponse = (configuration, uid, path = uid ?? 'configuration')=>{
    if (!isRecord(configuration)) {
        warnOnceInDevelopment(`Received malformed configuration at "${path}". Repairing entry.`);
    }
    const configurationRecord = isRecord(configuration) ? configuration : {};
    return {
        ...configurationRecord,
        ...uid ? {
            uid
        } : {},
        settings: normalizeResponseSettings(configurationRecord.settings, `${path}.settings`),
        metadatas: normalizeResponseMetadatas(configurationRecord.metadatas, `${path}.metadatas`),
        layouts: normalizeResponseLayouts(configurationRecord.layouts, `${path}.layouts`)
    };
};
const normalizeComponentsRecord = (components)=>{
    if (!isRecord(components)) {
        warnOnceInDevelopment('Received malformed component configuration record. Falling back to empty.');
        return {};
    }
    return Object.entries(components).reduce((acc, [uid, configuration])=>{
        acc[uid] = normalizeConfigurationResponse(configuration, uid, `components.${uid}`);
        return acc;
    }, {});
};
const normalizeContentTypeConfigurationResponse = (data, uid)=>{
    if (!isRecord(data)) {
        warnOnceInDevelopment('Received malformed content-type configuration response. Repairing data.');
    }
    const dataRecord = isRecord(data) ? data : {};
    return {
        contentType: normalizeConfigurationResponse(dataRecord.contentType, uid, 'contentType'),
        components: normalizeComponentsRecord(dataRecord.components)
    };
};
const normalizeContentTypeConfigurationUpdateResponse = (data)=>{
    if (!isRecord(data)) {
        warnOnceInDevelopment('Received malformed content-type configuration update response. Repairing data.');
    }
    const dataRecord = isRecord(data) ? data : {};
    return {
        contentType: normalizeConfigurationResponse(dataRecord.contentType, undefined, 'contentType'),
        components: normalizeComponentsRecord(dataRecord.components)
    };
};
const normalizeComponentConfigurationResponse = (data, uid)=>{
    if (!isRecord(data)) {
        warnOnceInDevelopment('Received malformed component configuration response. Repairing data.');
    }
    const dataRecord = isRecord(data) ? data : {};
    return {
        component: normalizeConfigurationResponse(dataRecord.component, uid, 'component'),
        components: normalizeComponentsRecord(dataRecord.components)
    };
};
const normalizeComponentConfigurationUpdateResponse = (data)=>{
    return normalizeConfigurationResponse(data, undefined, 'component');
};
const normalizeContentTypeSettingsResponse = (data)=>{
    if (!Array.isArray(data)) {
        warnOnceInDevelopment('Received malformed content-type settings response. Falling back to empty.');
        return [];
    }
    return data.filter((entry, index)=>{
        const isValidEntry = isRecord(entry);
        if (!isValidEntry) {
            warnOnceInDevelopment(`Dropped malformed content-type settings entry "${index}".`);
        }
        return isValidEntry;
    }).filter((entry)=>{
        const hasUid = typeof entry.uid === 'string';
        if (!hasUid) {
            warnOnceInDevelopment('Dropped content-type settings entry without a uid.');
        }
        return hasUid;
    }).map((entry)=>({
            uid: entry.uid,
            settings: normalizeResponseSettings(entry.settings, `contentTypeSettings.${entry.uid}`)
        }));
};
const hasAttribute = (attributes, name)=>{
    return Boolean(name && attributes?.[name]);
};
const getTargetAttributes = (attribute, { components, schemas })=>{
    if (attribute.type === 'component') {
        return components[attribute.component]?.attributes;
    }
    if (attribute.type === 'relation') {
        const target = 'targetModel' in attribute ? attribute.targetModel : 'target' in attribute ? attribute.target : undefined;
        return schemas.find((schema)=>schema.uid === target)?.attributes;
    }
    return undefined;
};
const hasTargetSchema = (attribute, { schemas })=>{
    if (attribute.type !== 'relation') {
        return true;
    }
    const target = 'targetModel' in attribute ? attribute.targetModel : 'target' in attribute ? attribute.target : undefined;
    return schemas.some((schema)=>schema.uid === target);
};
const normalizeMainField = (attribute, mainField, context, path)=>{
    if (!mainField) {
        return mainField;
    }
    const targetAttributes = getTargetAttributes(attribute, context);
    if (!targetAttributes) {
        if (attribute.type === 'relation' && !hasTargetSchema(attribute, context)) {
            return mainField;
        }
        if (attribute.type !== 'component' && attribute.type !== 'relation') {
            return mainField;
        }
        warnOnceInDevelopment(`Reset invalid mainField "${mainField}" at "${path}" to "id".`);
        return 'id';
    }
    if (hasAttribute(targetAttributes, mainField)) {
        return mainField;
    }
    warnOnceInDevelopment(`Reset invalid mainField "${mainField}" at "${path}" to "id".`);
    return 'id';
};
const normalizeSettings = (settings, attributes, path)=>{
    if (!settings.mainField || hasAttribute(attributes, settings.mainField) || settings.mainField === 'id') {
        return settings;
    }
    warnOnceInDevelopment(`Reset invalid mainField "${settings.mainField}" at "${path}" to "id".`);
    return {
        ...settings,
        mainField: 'id'
    };
};
const hasResolvableComponent = (attribute, context)=>{
    if (!attribute || !('component' in attribute) || typeof attribute.component !== 'string') {
        return true;
    }
    return Boolean(context.components[attribute.component]);
};
const normalizeMetadatas = (metadatas, attributes, context, path)=>{
    return Object.entries(metadatas).reduce((acc, [name, metadata])=>{
        const attribute = attributes?.[name];
        if (!attribute) {
            warnOnceInDevelopment(`Dropped metadata for unknown field "${path}.metadatas.${name}".`);
            return acc;
        }
        if (!hasResolvableComponent(attribute, context)) {
            warnOnceInDevelopment(`Dropped metadata for field "${path}.metadatas.${name}" because its component schema is unavailable.`);
            return acc;
        }
        acc[name] = {
            edit: {
                ...metadata.edit,
                // `mainField` is present at runtime for component/relation edit metadata.
                ...'mainField' in metadata.edit ? {
                    mainField: normalizeMainField(attribute, metadata.edit.mainField, context, `${path}.metadatas.${name}.edit.mainField`)
                } : {}
            },
            list: {
                ...metadata.list,
                mainField: normalizeMainField(attribute, metadata.list.mainField, context, `${path}.metadatas.${name}.list.mainField`)
            }
        };
        return acc;
    }, {});
};
const normalizeLayouts = (layouts, attributes, metadatas, context, path)=>{
    const hasRenderableField = (name)=>{
        const attribute = attributes?.[name];
        return Boolean(attribute && metadatas[name] && hasResolvableComponent(attribute, context));
    };
    return {
        edit: (layouts.edit ?? []).map((row, rowIndex)=>row.filter((field)=>{
                const isRenderable = hasRenderableField(field.name);
                if (!isRenderable) {
                    warnOnceInDevelopment(`Dropped non-renderable edit layout field "${path}.layouts.edit.${rowIndex}.${field.name}".`);
                }
                return isRenderable;
            })).filter((row)=>row.length > 0),
        list: (layouts.list ?? []).filter((name, index)=>{
            const isRenderable = hasRenderableField(name);
            if (!isRenderable) {
                warnOnceInDevelopment(`Dropped non-renderable list layout field "${path}.layouts.list.${index}.${name}".`);
            }
            return isRenderable;
        })
    };
};
const normalizeConfiguration = (configuration, attributes, context)=>{
    const path = configuration.uid ?? 'configuration';
    const metadatas = normalizeMetadatas(configuration.metadatas, attributes, context, path);
    const layouts = normalizeLayouts(configuration.layouts, attributes, metadatas, context, path);
    return {
        ...configuration,
        settings: normalizeSettings(configuration.settings, attributes, `${path}.settings.mainField`),
        metadatas,
        layouts
    };
};
const normalizeComponents = (componentConfigurations, context)=>{
    return Object.entries(componentConfigurations).reduce((acc, [uid, configuration])=>{
        const componentSchema = context.components[uid];
        if (!componentSchema) {
            return acc;
        }
        acc[uid] = normalizeConfiguration(configuration, componentSchema.attributes, context);
        return acc;
    }, {});
};
const normalizeContentManagerLayout = (data, context)=>{
    const normalizedComponents = normalizeComponents(data.components, context);
    return {
        contentType: normalizeConfiguration(data.contentType, context.schema?.attributes, context),
        components: normalizedComponents
    };
};
const normalizeComponentConfigurationLayout = (data, context)=>{
    const normalizedComponents = normalizeComponents(data.components, {
        components: context.components,
        schemas: context.schemas
    });
    return {
        component: normalizeConfiguration(data.component, context.schema?.attributes, {
            components: context.components,
            schemas: context.schemas
        }),
        components: normalizedComponents
    };
};

export { normalizeComponentConfigurationLayout, normalizeComponentConfigurationResponse, normalizeComponentConfigurationUpdateResponse, normalizeContentManagerLayout, normalizeContentTypeConfigurationResponse, normalizeContentTypeConfigurationUpdateResponse, normalizeContentTypeSettingsResponse };
