// @strapi/typescript-utils/dist/index.js
'use strict';

var compile = require('./compile.js');
var index = require('./compilers/index.js');
var index$1 = require('./generators/index.js');
var isUsingTypescript = require('./utils/is-using-typescript.js');
var isUsingTypescriptSync = require('./utils/is-using-typescript-sync.js');
var getConfigPath = require('./utils/get-config-path.js');
var reportDiagnostics = require('./utils/report-diagnostics.js');
var resolveConfigOptions = require('./utils/resolve-config-options.js');
var formatHost = require('./utils/format-host.js');
var resolveOutdir = require('./utils/resolve-outdir.js');
var resolveOutdirSync = require('./utils/resolve-outdir-sync.js');



exports.compile = compile.compile;
exports.compilers = index;
exports.generators = index$1;
exports.isUsingTypeScript = isUsingTypescript.isUsingTypeScript;
exports.isUsingTypeScriptSync = isUsingTypescriptSync.isUsingTypeScriptSync;
exports.getConfigPath = getConfigPath.getConfigPath;
exports.reportDiagnostics = reportDiagnostics.reportDiagnostics;
exports.resolveConfigOptions = resolveConfigOptions.resolveConfigOptions;
exports.formatHost = formatHost.formatHost;
exports.resolveOutDir = resolveOutdir.resolveOutDir;
exports.resolveOutDirSync = resolveOutdirSync.resolveOutDirSync;
