// @strapi/content-type-builder/dist/server/services/builder.js
'use strict';

var utils = require('@strapi/utils');

/**
 * Reserved names are centralised in `@strapi/utils` so the bootstrap validator,
 * the schema-to-model transform and this CTB service all share a single source.
 *
 * The arrays exported here are kept for backwards compatibility with consumers
 * that import them directly. New call sites should import from `@strapi/utils`.
 */ const reservedAttributes = utils.contentTypes.getReservedAttributeNames();
const reservedModels = utils.contentTypes.getReservedModelNames();
const getReservedNames = ()=>{
    return {
        models: reservedModels,
        attributes: reservedAttributes
    };
};
const isReservedModelName = (name)=>utils.contentTypes.isReservedModelName(name);
const isReservedAttributeName = (name)=>utils.contentTypes.isReservedAttributeName(name);

exports.getReservedNames = getReservedNames;
exports.isReservedAttributeName = isReservedAttributeName;
exports.isReservedModelName = isReservedModelName;
exports.reservedAttributes = reservedAttributes;
exports.reservedModels = reservedModels;
