// @strapi/admin/dist/server/ee/server/src/controllers/user.js
'use strict';

var _ = require('lodash');
var fp = require('lodash/fp');
var utils = require('@strapi/utils');
var user$2 = require('../validation/user.js');
var user$1 = require('../../../../server/src/validation/user.js');
var normalizeEmail = require('../../../../server/src/utils/normalize-email.js');
var index = require('../utils/index.js');
var ssoLock = require('../utils/sso-lock.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var ___default = /*#__PURE__*/_interopDefault(_);

const { ApplicationError, ForbiddenError } = utils.errors;
const pickUserCreationAttributes = fp.pick([
    'firstname',
    'lastname',
    'email',
    'roles'
]);
const hasAdminSeatsAvaialble = async ()=>{
    if (!strapi.EE) {
        return true;
    }
    const permittedSeats = strapi.ee.seats;
    if (fp.isNil(permittedSeats)) {
        return true;
    }
    const userCount = await strapi.service('admin::user').getCurrentActiveUserCount();
    if (userCount < permittedSeats) {
        return true;
    }
};
var user = {
    async create (ctx) {
        if (!await hasAdminSeatsAvaialble()) {
            throw new ForbiddenError('License seat limit reached. You cannot create a new user');
        }
        const { body } = ctx.request;
        const cleanData = {
            ...body,
            email: ___default.default.get(body, `email`, ``).toLowerCase()
        };
        await user$2.validateUserCreationInput(cleanData);
        const attributes = pickUserCreationAttributes(cleanData);
        const { useSSORegistration } = cleanData;
        const userAlreadyExists = await index.getService('user').exists({
            email: attributes.email
        });
        if (userAlreadyExists) {
            throw new ApplicationError('Email already taken');
        }
        if (useSSORegistration) {
            Object.assign(attributes, {
                registrationToken: null,
                isActive: true
            });
        }
        const createdUser = await index.getService('user').create(attributes);
        const userInfo = index.getService('user').sanitizeUser(createdUser);
        // Note: We need to assign manually the registrationToken to the
        // final user payload so that it's not removed in the sanitation process.
        Object.assign(userInfo, {
            registrationToken: createdUser.registrationToken
        });
        ctx.created({
            data: userInfo
        });
    },
    async update (ctx) {
        const { id } = ctx.params;
        const data = normalizeEmail.normalizeEmail(ctx.request.body);
        await user$1.validateUserUpdateInput(data);
        if (___default.default.has(data, 'email')) {
            const uniqueEmailCheck = await index.getService('user').exists({
                id: {
                    $ne: id
                },
                email: data.email
            });
            if (uniqueEmailCheck) {
                throw new ApplicationError('A user with this email address already exists');
            }
        }
        const user = await index.getService('user').findOne(id, null);
        if (!await hasAdminSeatsAvaialble() && !user.isActive && data.isActive) {
            throw new ForbiddenError('License seat limit reached. You cannot active this user');
        }
        const updatedUser = await index.getService('user').updateById(id, data);
        if (!updatedUser) {
            return ctx.notFound('User does not exist');
        }
        ctx.body = {
            data: index.getService('user').sanitizeUser(updatedUser)
        };
    },
    async isSSOLocked (ctx) {
        const { user } = ctx.state;
        const isSSOLocked = await ssoLock.isSsoLocked(user);
        ctx.body = {
            data: {
                isSSOLocked
            }
        };
    }
};

module.exports = user;
