// @strapi/strapi/dist/src/node/core/admin-vite-optimize-exclude.js
'use strict';

var path = require('node:path');
var fs = require('node:fs/promises');
var readPkgUp = require('read-pkg-up');
var adminViteAliasModules = require('./admin-vite-alias-modules.js');
var dependencies = require('./dependencies.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var path__default = /*#__PURE__*/_interopDefault(path);
var fs__default = /*#__PURE__*/_interopDefault(fs);
var readPkgUp__default = /*#__PURE__*/_interopDefault(readPkgUp);

const REACT_PEER_DEPENDENCIES = new Set([
    'react',
    'react-dom'
]);
/**
 * Packages explicitly pre-bundled or aliased for the admin singleton contract.
 * Never auto-exclude these — they must stay on the optimizeDeps.include / dedupe path.
 *
 * The admin entry host (@strapi/strapi) must never land in optimizeDeps.exclude (#26944, #27014).
 * CJS-only deps imported by @strapi/admin (e.g. invariant, lodash) belong in optimizeDeps.include
 * (see vite/config.ts — #26964, #26944, #27014).
 */ const PINNED_OPTIMIZE_MODULES = new Set([
    ...adminViteAliasModules.ADMIN_VITE_ALIAS_MODULES,
    '@strapi/strapi'
]);
const isOfficialStrapiPackage = (name)=>name.startsWith('@strapi/');
const getRootPackageExport = (pkg)=>{
    const exportsField = pkg.exports;
    if (!exportsField || typeof exportsField === 'string') {
        return undefined;
    }
    return exportsField['.'];
};
/**
 * @internal exported for tests
 */ const isEsmPackage = (pkg)=>{
    if (pkg.type === 'module') {
        return true;
    }
    const rootExport = getRootPackageExport(pkg);
    return typeof rootExport === 'object' && rootExport !== null && 'import' in rootExport && typeof rootExport.import === 'string';
};
/**
 * @internal exported for tests
 */ const hasReactPeerDependency = (pkg)=>Object.keys(pkg.peerDependencies ?? {}).some((name)=>REACT_PEER_DEPENDENCIES.has(name));
/**
 * Targets libraries that ship a pre-built dist bundle (e.g. plugin UI kits) rather than
 * source packages like react-intl that still benefit from Vite's dep optimizer.
 *
 * @internal exported for tests
 */ const shipsPreBuiltDist = (pkg)=>{
    const rootExport = getRootPackageExport(pkg);
    const exportImport = typeof rootExport === 'object' && rootExport !== null && 'import' in rootExport ? rootExport.import : undefined;
    const entryPaths = [
        pkg.module,
        pkg.main,
        exportImport
    ].filter((entry)=>typeof entry === 'string');
    if (entryPaths.some((entry)=>entry.includes('/dist/') || entry.startsWith('./dist/') || entry.startsWith('dist/'))) {
        return true;
    }
    return Array.isArray(pkg.files) && pkg.files.some((file)=>file === 'dist' || file.startsWith('dist/'));
};
/**
 * @internal exported for tests
 */ const shouldExcludeFromOptimizeDeps = (pkg)=>hasReactPeerDependency(pkg) && isEsmPackage(pkg) && shipsPreBuiltDist(pkg);
/**
 * @internal exported for tests
 */ const getPluginPackageName = (modulePath)=>{
    if (modulePath.startsWith('@')) {
        const [scope, name] = modulePath.split('/');
        return `${scope}/${name}`;
    }
    return modulePath.split('/')[0] ?? modulePath;
};
const collectDependencyNames = (pkg)=>{
    const names = new Set();
    for (const section of [
        pkg.dependencies,
        pkg.devDependencies
    ]){
        if (section) {
            for (const name of Object.keys(section)){
                names.add(name);
            }
        }
    }
    return [
        ...names
    ];
};
const getPluginPackageJson = async (plugin, cwd)=>{
    if (plugin.type === 'local' && plugin.path) {
        try {
            const content = await fs__default.default.readFile(path__default.default.join(plugin.path, 'package.json'), 'utf8');
            return JSON.parse(content);
        } catch  {
            return null;
        }
    }
    return dependencies.getModule(getPluginPackageName(plugin.modulePath), cwd);
};
const loadAppPackageJson = async (cwd)=>{
    const result = await readPkgUp__default.default({
        cwd
    });
    return result?.packageJson ?? null;
};
/**
 * Pre-built ESM libraries with React peers (shared plugin UI kits) are incompatible with
 * Strapi's React/design-system pre-bundling. Skip dep optimization so they resolve through
 * the admin resolve aliases instead of being re-bundled by Vite.
 *
 * Scans app and plugin dependency trees (#26944). Official @strapi/* packages and pinned
 * singletons are never auto-excluded — @strapi/strapi matches the heuristic but must stay on
 * the optimizeDeps.include path (#26944, #27014).
 *
 * @internal
 */ const collectAdminOptimizeDepsExclude = async (cwd, plugins)=>{
    const candidateNames = new Set();
    const appPkg = await loadAppPackageJson(cwd);
    if (appPkg) {
        for (const name of collectDependencyNames(appPkg)){
            candidateNames.add(name);
        }
    }
    for (const plugin of plugins){
        const pluginPkg = await getPluginPackageJson(plugin, cwd);
        if (pluginPkg) {
            for (const name of collectDependencyNames(pluginPkg)){
                candidateNames.add(name);
            }
        }
    }
    const exclude = [];
    for (const name of candidateNames){
        if (PINNED_OPTIMIZE_MODULES.has(name) || isOfficialStrapiPackage(name)) {
            continue;
        }
        const pkg = await dependencies.getModule(name, cwd);
        if (pkg && shouldExcludeFromOptimizeDeps(pkg)) {
            exclude.push(name);
        }
    }
    return exclude.sort();
};

exports.collectAdminOptimizeDepsExclude = collectAdminOptimizeDepsExclude;
exports.getPluginPackageName = getPluginPackageName;
exports.hasReactPeerDependency = hasReactPeerDependency;
exports.isEsmPackage = isEsmPackage;
exports.shipsPreBuiltDist = shipsPreBuiltDist;
exports.shouldExcludeFromOptimizeDeps = shouldExcludeFromOptimizeDeps;
