// @strapi/strapi/dist/src/node/core/plugins.js
'use strict';

var os = require('node:os');
var path = require('node:path');
var fs = require('node:fs');
var camelCase = require('lodash/camelCase');
var utils = require('@strapi/utils');
var dependencies = require('./dependencies.js');
var files = require('./files.js');
var errors = require('./errors.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var os__default = /*#__PURE__*/_interopDefault(os);
var path__default = /*#__PURE__*/_interopDefault(path);
var fs__default = /*#__PURE__*/_interopDefault(fs);
var camelCase__default = /*#__PURE__*/_interopDefault(camelCase);

const validatePackageHasStrapi = (pkg)=>'strapi' in pkg && typeof pkg.strapi === 'object' && !Array.isArray(pkg.strapi) && pkg.strapi !== null;
const validatePackageIsPlugin = (pkg)=>validatePackageHasStrapi(pkg) && pkg.strapi.kind === 'plugin';
const isPluginConfigEnabled = (config)=>{
    if (typeof config === 'boolean') {
        return config;
    }
    return config.enabled !== false;
};
const getEnabledPlugins = async ({ cwd, logger, runtimeDir, strapi })=>{
    const plugins = {};
    /**
   * This is the list of dependencies that are installed in the user's project.
   * It will include libraries like "react", so we need to collect the ones that
   * are plugins.
   */ const deps = strapi.config.get('info.dependencies', {});
    logger.debug("Dependencies from user's project", os__default.default.EOL, deps);
    const userPluginsFile = await loadUserPluginsFile(strapi.dirs.app.config);
    logger.debug("User's plugins file", os__default.default.EOL, userPluginsFile);
    for (const dep of Object.keys(deps)){
        const pkg = await dependencies.getModule(dep, cwd);
        if (pkg && validatePackageIsPlugin(pkg)) {
            const name = pkg.strapi.name || pkg.name;
            if (!name) {
                /**
         * Unlikely to happen, but you never know.
         */ throw Error("You're trying to import a plugin that doesn't have a name – check the package.json of that plugin!");
            }
            const userPluginConfig = userPluginsFile[name];
            if (userPluginConfig !== undefined && !isPluginConfigEnabled(userPluginConfig)) {
                continue;
            }
            plugins[name] = {
                name,
                importName: camelCase__default.default(name),
                type: 'module',
                modulePath: dep
            };
        }
    }
    for (const [userPluginName, userPluginConfig] of Object.entries(userPluginsFile)){
        /**
     * Local plugins must be explicitly enabled to be registered, matching the
     * server-side loader (`@strapi/core` get-enabled-plugins), which drops a
     * `{ resolve }` declaration without a truthy `enabled`. Treating an omitted
     * `enabled` as enabled here would bundle the plugin into the admin while the
     * server leaves it unloaded.
     */ if (typeof userPluginConfig === 'object' && userPluginConfig.enabled && userPluginConfig.resolve) {
            const sysPath = files.convertModulePathToSystemPath(userPluginConfig.resolve);
            plugins[userPluginName] = {
                name: userPluginName,
                importName: camelCase__default.default(userPluginName),
                type: 'local',
                /**
         * User plugin paths are resolved from the entry point
         * of the app, because that's how you import them.
         */ modulePath: files.convertSystemPathToModulePath(path__default.default.relative(runtimeDir, sysPath)),
                path: sysPath
            };
        }
    }
    return plugins;
};
const PLUGIN_CONFIGS = [
    'plugins.js',
    'plugins.mjs',
    'plugins.ts'
];
const loadUserPluginsFile = async (root)=>{
    for (const file of PLUGIN_CONFIGS){
        const filePath = path__default.default.join(root, file);
        const configFile = await files.loadFile(filePath);
        if (configFile) {
            /**
       * Configs can be a function or they can be just an object!
       */ return typeof configFile === 'function' ? configFile({
                env: utils.env
            }) : configFile;
        }
    }
    return {};
};
const getMapOfPluginsWithAdmin = (plugins)=>{
    /**
   * This variable stores the import paths for plugins.
   * The keys are the module paths of the plugins, and the values are the paths
   * to the admin part of the plugins, which is either loaded from the
   * package.json exports or from the legacy strapi-admin.js file.
   */ const pluginImportPaths = {};
    return Object.values(plugins).filter((plugin)=>{
        if (!plugin) {
            return false;
        }
        /**
       * There are two ways a plugin should be imported, either it's local to the strapi app,
       * or it's an actual npm module that's installed and resolved via node_modules.
       *
       * We first check if the plugin is local to the strapi app, using a regular `fs.existsSync` because
       * the pathToPlugin will be relative i.e. `/Users/my-name/strapi-app/src/plugins/my-plugin`.
       *
       * If the file doesn't exist well then it's probably a node_module, so instead we use `require.resolve`
       * which will resolve the path to the module in node_modules. If it fails with the specific code `MODULE_NOT_FOUND`
       * then it doesn't have an admin part to the package.
       */ try {
            const localPluginPath = plugin.path;
            if (localPluginPath) {
                // Here we are loading a locally installed plugin
                const packageJsonPath = path__default.default.join(localPluginPath, 'package.json');
                if (fs__default.default.existsSync(packageJsonPath)) {
                    const packageJson = JSON.parse(fs__default.default.readFileSync(packageJsonPath, 'utf-8'));
                    const localAdminPath = packageJson?.exports?.['./strapi-admin']?.import;
                    if (localAdminPath) {
                        pluginImportPaths[plugin.modulePath] = localAdminPath;
                        return true;
                    }
                }
                // Check if legacy admin file exists in local plugin
                if (fs__default.default.existsSync(path__default.default.join(localPluginPath, 'strapi-admin.js'))) {
                    pluginImportPaths[plugin.modulePath] = 'strapi-admin';
                    return true;
                }
            }
            // This plugin is a module, so we need to check if it has a strapi-admin export
            if (require.resolve(`${plugin.modulePath}/strapi-admin`)) {
                pluginImportPaths[plugin.modulePath] = 'strapi-admin';
                return true;
            }
            return false;
        } catch (err) {
            if (errors.isError(err) && 'code' in err && (err.code === 'MODULE_NOT_FOUND' || err.code === 'ERR_PACKAGE_PATH_NOT_EXPORTED')) {
                /**
           * the plugin does not contain FE code, so we
           * don't want to import it anyway
           */ return false;
            }
            throw err;
        }
    }).map((plugin)=>({
            ...plugin,
            modulePath: `${plugin.modulePath}/${pluginImportPaths[plugin.modulePath]}`
        }));
};

exports.getEnabledPlugins = getEnabledPlugins;
exports.getMapOfPluginsWithAdmin = getMapOfPluginsWithAdmin;
