// @strapi/content-manager/dist/admin/pages/EditView/utils/draftRelationCounts.mjs
const EMPTY_DRAFT_RELATION_COUNTS = {
    unpublishedRelations: 0,
    draftM2mLinks: 0
};
const isBidirectionalManyToMany = (attribute)=>{
    if (attribute.type !== 'relation' || attribute.relation !== 'manyToMany') {
        return false;
    }
    const relationAttribute = attribute;
    return Boolean(relationAttribute.inversedBy || relationAttribute.mappedBy);
};
const mergeDraftRelationCounts = (left, right)=>({
        unpublishedRelations: left.unpublishedRelations + right.unpublishedRelations,
        draftM2mLinks: left.draftM2mLinks + right.draftM2mLinks
    });
const normalizeDraftRelationCounts = (payload)=>{
    if (payload && typeof payload === 'object' && 'unpublishedRelations' in payload && 'draftM2mLinks' in payload) {
        return payload;
    }
    if (payload && typeof payload === 'object' && 'data' in payload) {
        return normalizeDraftRelationCounts(payload.data);
    }
    return EMPTY_DRAFT_RELATION_COUNTS;
};
const resolveDraftRelationCounts = (documentId, modified, localCounts, serverCounts)=>{
    if (!documentId) {
        return localCounts;
    }
    const hasUnsavedDraftRelations = localCounts.unpublishedRelations > 0 || localCounts.draftM2mLinks > 0;
    if (modified || hasUnsavedDraftRelations) {
        return mergeDraftRelationCounts(localCounts, serverCounts);
    }
    return serverCounts;
};
const getDraftRelationsPublishState = (counts)=>{
    const hasUnpublishedRelations = counts.unpublishedRelations > 0;
    const hasDraftM2mLinks = counts.draftM2mLinks > 0;
    const hasDraftRelations = hasUnpublishedRelations || hasDraftM2mLinks;
    const isM2mOnlyDraftRelations = hasDraftM2mLinks && !hasUnpublishedRelations;
    return {
        hasUnpublishedRelations,
        hasDraftM2mLinks,
        hasDraftRelations,
        isM2mOnlyDraftRelations,
        dialogVariant: isM2mOnlyDraftRelations ? 'default' : 'danger',
        bodyIcon: isM2mOnlyDraftRelations ? 'default' : 'danger',
        confirmLabel: isM2mOnlyDraftRelations ? 'publish' : 'publish-without-relations'
    };
};
/**
 * Counts draft relations in unsaved form values, excluding self-referential relations
 * (preserved on publish via document-service self-referential-relations).
 */ const countLocalDraftRelations = (data, schema, components, contentTypeUid)=>{
    if (!schema?.attributes) {
        return EMPTY_DRAFT_RELATION_COUNTS;
    }
    return Object.keys(schema.attributes).reduce((counts, attributeName)=>{
        const attribute = schema.attributes[attributeName];
        const value = data[attributeName];
        if (!value) {
            return counts;
        }
        switch(attribute.type){
            case 'relation':
                {
                    if (!('target' in attribute) || attribute.target === contentTypeUid) {
                        return counts;
                    }
                    if (typeof value === 'object' && value !== null && 'connect' in value) {
                        const draftConnectCount = value.connect?.filter((relation)=>relation.status === 'draft').length ?? 0;
                        if (draftConnectCount === 0) {
                            return counts;
                        }
                        if (isBidirectionalManyToMany(attribute)) {
                            return {
                                ...counts,
                                draftM2mLinks: counts.draftM2mLinks + draftConnectCount
                            };
                        }
                        return {
                            ...counts,
                            unpublishedRelations: counts.unpublishedRelations + draftConnectCount
                        };
                    }
                    return counts;
                }
            case 'component':
                {
                    const componentItems = Array.isArray(value) ? value : [
                        value
                    ];
                    const componentSchema = components[attribute.component];
                    return componentItems.reduce((componentCounts, componentValue)=>mergeDraftRelationCounts(componentCounts, countLocalDraftRelations(componentValue, componentSchema, components, contentTypeUid)), counts);
                }
            case 'dynamiczone':
                {
                    return value.reduce((zoneCounts, componentValue)=>{
                        const componentUid = componentValue.__component;
                        return mergeDraftRelationCounts(zoneCounts, countLocalDraftRelations(componentValue, components[componentUid], components, contentTypeUid));
                    }, counts);
                }
            default:
                return counts;
        }
    }, EMPTY_DRAFT_RELATION_COUNTS);
};

export { EMPTY_DRAFT_RELATION_COUNTS, countLocalDraftRelations, getDraftRelationsPublishState, isBidirectionalManyToMany, mergeDraftRelationCounts, normalizeDraftRelationCounts, resolveDraftRelationCounts };
