// @strapi/upload/dist/admin/future/services/assets.js
'use strict';

var api = require('./api.js');

const assetsApi = api.uploadApi.injectEndpoints({
    endpoints: (builder)=>({
            getAssets: builder.query({
                query: (params = {})=>{
                    const { folder, ...rest } = params;
                    const queryParams = {
                        ...rest
                    };
                    if (folder != null) {
                        queryParams['filters'] = {
                            $and: [
                                {
                                    folder: {
                                        id: folder
                                    }
                                }
                            ]
                        };
                    } else {
                        queryParams['filters'] = {
                            $and: [
                                {
                                    folder: {
                                        id: {
                                            $null: true
                                        }
                                    }
                                }
                            ]
                        };
                    }
                    return {
                        url: '/upload/files',
                        method: 'GET',
                        config: {
                            params: queryParams
                        }
                    };
                },
                transformResponse: (response)=>response,
                providesTags: (result)=>result ? [
                        ...result.results.map(({ id })=>({
                                type: 'Asset',
                                id
                            })),
                        {
                            type: 'Asset',
                            id: 'LIST'
                        }
                    ] : [
                        {
                            type: 'Asset',
                            id: 'LIST'
                        }
                    ]
            }),
            getAsset: builder.query({
                query: (id)=>({
                        url: `/upload/files/${id}`,
                        method: 'GET'
                    }),
                providesTags: (_result, _error, id)=>[
                        {
                            type: 'Asset',
                            id
                        }
                    ]
            }),
            /**
     * Update the editable metadata of an existing asset.
     * Hits the legacy `POST /upload?id=<id>` endpoint which dispatches to
     * `admin-upload.updateFileInfo`.
     */ updateAsset: builder.mutation({
                query: ({ id, fileInfo })=>{
                    const formData = new FormData();
                    formData.append('fileInfo', JSON.stringify(fileInfo));
                    return {
                        url: '/upload',
                        method: 'POST',
                        data: formData,
                        config: {
                            params: {
                                id
                            }
                        }
                    };
                },
                invalidatesTags: (_result, _error, { id })=>[
                        {
                            type: 'Asset',
                            id
                        },
                        {
                            type: 'Asset',
                            id: 'LIST'
                        }
                    ]
            }),
            /**
     * Replace the binary content of an existing asset.
     * Hits `POST /upload?id=<id>` with a multipart body — the controller
     * dispatches to `admin-upload.replaceFile` when a `files` part is present.
     * Uses the standard axios baseQuery (no streaming) since we only ever
     * replace one file at a time and don't need per-byte progress here.
     */ replaceAsset: builder.mutation({
                query: ({ id, file, fileInfo })=>{
                    const formData = new FormData();
                    formData.append('files', file);
                    if (fileInfo) {
                        formData.append('fileInfo', JSON.stringify(fileInfo));
                    }
                    return {
                        url: '/upload',
                        method: 'POST',
                        data: formData,
                        config: {
                            params: {
                                id
                            }
                        }
                    };
                },
                invalidatesTags: (_result, _error, { id })=>[
                        {
                            type: 'Asset',
                            id
                        },
                        {
                            type: 'Asset',
                            id: 'LIST'
                        }
                    ]
            }),
            /**
     * Permanently delete an asset by id. Hits the same endpoint as the legacy
     * `useRemoveAsset` hook so server behaviour is unchanged.
     */ deleteAsset: builder.mutation({
                query: (id)=>({
                        url: `/upload/files/${id}`,
                        method: 'DELETE'
                    }),
                invalidatesTags: (_result, _error, id)=>[
                        {
                            type: 'Asset',
                            id
                        },
                        {
                            type: 'Asset',
                            id: 'LIST'
                        }
                    ]
            })
        })
});
const { useGetAssetsQuery, useGetAssetQuery, useUpdateAssetMutation, useReplaceAssetMutation, useDeleteAssetMutation } = assetsApi;

exports.useDeleteAssetMutation = useDeleteAssetMutation;
exports.useGetAssetQuery = useGetAssetQuery;
exports.useGetAssetsQuery = useGetAssetsQuery;
exports.useReplaceAssetMutation = useReplaceAssetMutation;
exports.useUpdateAssetMutation = useUpdateAssetMutation;
