// @strapi/upload/dist/admin/future/services/folders.js
'use strict';

var api = require('./api.js');

const isDataEnvelope = (response)=>typeof response === 'object' && response !== null && 'data' in response;
const unwrapData = (response)=>isDataEnvelope(response) ? response.data : response;
const foldersApi = api.uploadApi.injectEndpoints({
    endpoints: (builder)=>({
            getFolders: builder.query({
                query: (params = {})=>{
                    const { parentId } = params;
                    const queryParams = {
                        // Match sidebar FolderTree order (server getStructure uses sortBy('name')).
                        sort: 'name:ASC'
                    };
                    if (parentId != null) {
                        queryParams['filters'] = {
                            $and: [
                                {
                                    parent: {
                                        id: parentId
                                    }
                                }
                            ]
                        };
                    } else {
                        queryParams['filters'] = {
                            $and: [
                                {
                                    parent: {
                                        id: {
                                            $null: true
                                        }
                                    }
                                }
                            ]
                        };
                    }
                    return {
                        url: '/upload/folders',
                        method: 'GET',
                        config: {
                            params: queryParams
                        }
                    };
                },
                transformResponse: (response)=>unwrapData(response),
                providesTags: (results)=>{
                    if (results) {
                        return [
                            ...results.map(({ id })=>({
                                    type: 'Folder',
                                    id
                                })),
                            {
                                type: 'Folder',
                                id: 'LIST'
                            }
                        ];
                    }
                    return [
                        {
                            type: 'Folder',
                            id: 'LIST'
                        }
                    ];
                }
            }),
            createFolder: builder.mutation({
                query: (body)=>({
                        url: '/upload/folders',
                        method: 'POST',
                        data: body
                    }),
                transformResponse: (response)=>response.data,
                invalidatesTags: [
                    {
                        type: 'Folder',
                        id: 'LIST'
                    },
                    {
                        type: 'Folder',
                        id: 'STRUCTURE'
                    }
                ]
            }),
            /**
     * Hierarchical folder tree used by the FolderTree sidebar. Returned as a
     * single nested array — the server walks the folders table once and
     * assembles parent/child relationships.
     *
     * TODO: filter results like `admin-folder.find` for folder-scoped roles
     * (v1 returns all folders; relies on global `plugin::upload.read`).
     */ getFolderStructure: builder.query({
                query: ()=>({
                        url: '/upload/folder-structure',
                        method: 'GET'
                    }),
                // TODO: align GetFolderStructure contract with the real /upload/folder-structure
                // envelope and drop this defensive unwrap (data[] vs { data: [] }).
                transformResponse: (response)=>response?.data ?? response ?? [],
                providesTags: [
                    {
                        type: 'Folder',
                        id: 'STRUCTURE'
                    }
                ]
            }),
            /**
     * Flat list of every folder, used to populate the "Location" select in the
     * asset details drawer. No parent filter — we want the entire tree in one
     * query so the select can render any destination folder.
     */ getAllFolders: builder.query({
                query: ()=>({
                        url: '/upload/folders',
                        method: 'GET'
                    }),
                transformResponse: (response)=>unwrapData(response ?? []),
                providesTags: (results)=>{
                    if (results) {
                        return [
                            ...results.map(({ id })=>({
                                    type: 'Folder',
                                    id
                                })),
                            {
                                type: 'Folder',
                                id: 'LIST'
                            }
                        ];
                    }
                    return [
                        {
                            type: 'Folder',
                            id: 'LIST'
                        }
                    ];
                }
            }),
            getFolder: builder.query({
                query: ({ id })=>({
                        url: `/upload/folders/${id}`,
                        method: 'GET',
                        config: {
                            params: {
                                populate: {
                                    parent: {
                                        populate: {
                                            parent: '*'
                                        }
                                    },
                                    children: {
                                        count: true
                                    },
                                    files: {
                                        count: true
                                    }
                                }
                            }
                        }
                    }),
                transformResponse: (response)=>response.data,
                providesTags: (_result, _error, { id })=>[
                        {
                            type: 'Folder',
                            id
                        }
                    ]
            }),
            bulkMove: builder.mutation({
                query: ({ fileIds = [], folderIds = [], destinationFolderId })=>({
                        url: '/upload/actions/bulk-move',
                        method: 'POST',
                        data: {
                            fileIds,
                            folderIds,
                            destinationFolderId
                        }
                    }),
                transformResponse: (response)=>response.data,
                invalidatesTags: [
                    {
                        type: 'Asset',
                        id: 'LIST'
                    },
                    {
                        type: 'Folder',
                        id: 'LIST'
                    },
                    {
                        type: 'Folder',
                        id: 'STRUCTURE'
                    }
                ]
            })
        })
});
const { useCreateFolderMutation, useGetFoldersQuery, useGetFolderQuery, useGetAllFoldersQuery, useGetFolderStructureQuery, useBulkMoveMutation } = foldersApi;

exports.useBulkMoveMutation = useBulkMoveMutation;
exports.useCreateFolderMutation = useCreateFolderMutation;
exports.useGetAllFoldersQuery = useGetAllFoldersQuery;
exports.useGetFolderQuery = useGetFolderQuery;
exports.useGetFolderStructureQuery = useGetFolderStructureQuery;
exports.useGetFoldersQuery = useGetFoldersQuery;
