// @strapi/upload/dist/admin/components/AssetCard/AssetCard.mjs
import { jsx } from 'react/jsx-runtime';
import { ASSET_TYPES } from '../../enums.mjs';
import { createAssetUrl } from '../../utils/createAssetUrl.mjs';
import 'byte-size';
import 'date-fns';
import 'qs';
import { getFileExtension } from '../../utils/getFileExtension.mjs';
import { prefixFileUrlWithBackendUrl } from '../../utils/prefixFileUrlWithBackendUrl.mjs';
import '../../utils/urlYupSchema.mjs';
import { AudioAssetCard } from './AudioAssetCard.mjs';
import { DocAssetCard } from './DocAssetCard.mjs';
import { ImageAssetCard } from './ImageAssetCard.mjs';
import { VideoAssetCard } from './VideoAssetCard.mjs';

const AssetCard = ({ asset, isSelected = false, onSelect, onEdit, onRemove, size = 'M', local = false, className })=>{
    const handleSelect = onSelect ? ()=>onSelect(asset) : undefined;
    const commonAssetCardProps = {
        id: asset.id,
        isSelectable: asset.isSelectable,
        extension: getFileExtension(asset.ext),
        name: asset.name,
        url: local ? asset.url : createAssetUrl(asset, true),
        mime: asset.mime,
        onEdit: onEdit ? ()=>onEdit(asset) : undefined,
        onSelect: handleSelect,
        onRemove: onRemove ? ()=>onRemove(asset) : undefined,
        selected: isSelected,
        size,
        className
    };
    if (asset.mime?.includes(ASSET_TYPES.Video)) {
        return /*#__PURE__*/ jsx(VideoAssetCard, {
            ...commonAssetCardProps
        });
    }
    if (asset.mime?.includes(ASSET_TYPES.Image)) {
        return /*#__PURE__*/ jsx(ImageAssetCard, {
            alt: asset.alternativeText || asset.name,
            height: asset.height,
            thumbnail: prefixFileUrlWithBackendUrl(asset?.formats?.thumbnail?.url || asset.url),
            width: asset.width,
            updatedAt: asset.updatedAt,
            isUrlSigned: asset?.isUrlSigned || false,
            // Only signed remote URLs need crossOrigin: they are loaded without a
            // cache-buster, so thumbnail and preview share a cache entry and must
            // both opt into CORS to avoid a cache collision. Public/unsigned remote
            // URLs are cache-busted, so they must not require a bucket CORS rule to
            // render. See #26581.
            crossOrigin: !local && asset?.isUrlSigned ? 'anonymous' : undefined,
            ...commonAssetCardProps
        });
    }
    if (asset.mime?.includes(ASSET_TYPES.Audio)) {
        return /*#__PURE__*/ jsx(AudioAssetCard, {
            ...commonAssetCardProps
        });
    }
    return /*#__PURE__*/ jsx(DocAssetCard, {
        ...commonAssetCardProps
    });
};

export { AssetCard };
