// @strapi/content-type-builder/dist/admin/components/ComponentList.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var get = require('lodash/get');
var ComponentRow = require('./ComponentRow.js');
var useDataManager = require('./DataManager/useDataManager.js');
var List = require('./List.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var get__default = /*#__PURE__*/_interopDefault(get);

const ComponentList = ({ component, isFromDynamicZone = false, firstLoopComponentUid })=>{
    const { components } = useDataManager.useDataManager();
    const type = get__default.default(components, component);
    if (!type) return;
    return /*#__PURE__*/ jsxRuntime.jsx(ComponentRow.ComponentRow, {
        $isChildOfDynamicZone: isFromDynamicZone,
        className: "component-row",
        children: /*#__PURE__*/ jsxRuntime.jsx(List.List, {
            type: type,
            firstLoopComponentUid: firstLoopComponentUid || component,
            isFromDynamicZone: isFromDynamicZone,
            isSub: true,
            secondLoopComponentUid: firstLoopComponentUid ? component : null
        })
    });
};

exports.ComponentList = ComponentList;
