// @strapi/openapi/dist/assemblers/document/security.js
'use strict';

const DEFAULT_BEARER_AUTH = {
    type: 'http',
    scheme: 'bearer',
    bearerFormat: 'JWT',
    description: 'JWT Bearer token for authentication'
};
class DocumentSecurityAssembler {
    assemble(context) {
        const { strapi } = context;
        const securitySchemes = this._getSecuritySchemes(strapi);
        const security = this._getGlobalSecurity(strapi);
        if (context.output.data.components === undefined) {
            context.output.data.components = {};
        }
        context.output.data.components.securitySchemes = securitySchemes;
        context.output.data.security = security;
    }
    _getSecuritySchemes(strapi) {
        const securityConfig = strapi.config.get('openapi.security') ?? {};
        const schemes = {};
        if (!securityConfig.bearerAuth) {
            schemes.bearerAuth = DEFAULT_BEARER_AUTH;
        }
        for (const [name, config] of Object.entries(securityConfig)){
            schemes[name] = config;
        }
        return schemes;
    }
    _getGlobalSecurity(strapi) {
        const globalSecurity = strapi.config.get('openapi.globalSecurity');
        if (globalSecurity) {
            return globalSecurity;
        }
        return [
            {
                bearerAuth: []
            }
        ];
    }
}

exports.DocumentSecurityAssembler = DocumentSecurityAssembler;
