// @strapi/cloud-cli/dist/config/local.js
'use strict';

var path = require('path');
var os = require('os');
var fse = require('fs-extra');
var XDGAppPaths = require('xdg-app-paths');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var path__default = /*#__PURE__*/_interopDefault(path);
var os__default = /*#__PURE__*/_interopDefault(os);
var fse__default = /*#__PURE__*/_interopDefault(fse);
var XDGAppPaths__default = /*#__PURE__*/_interopDefault(XDGAppPaths);

const APP_FOLDER_NAME = 'com.strapi.cli';
const CONFIG_FILENAME = 'config.json';
async function checkDirectoryExists(directoryPath) {
    try {
        const fsStat = await fse__default.default.lstat(directoryPath);
        return fsStat.isDirectory();
    } catch (e) {
        return false;
    }
}
// Determine storage path based on the operating system
async function getTmpStoragePath() {
    const storagePath = path__default.default.join(os__default.default.tmpdir(), APP_FOLDER_NAME);
    await fse__default.default.ensureDir(storagePath);
    return storagePath;
}
async function getConfigPath() {
    const configDirs = XDGAppPaths__default.default(APP_FOLDER_NAME).configDirs();
    const configPath = configDirs.find(checkDirectoryExists);
    if (!configPath) {
        await fse__default.default.ensureDir(configDirs[0]);
        return configDirs[0];
    }
    return configPath;
}
async function getLocalConfig() {
    const configPath = await getConfigPath();
    const configFilePath = path__default.default.join(configPath, CONFIG_FILENAME);
    await fse__default.default.ensureFile(configFilePath);
    try {
        return await fse__default.default.readJSON(configFilePath, {
            encoding: 'utf8',
            throws: true
        });
    } catch (e) {
        return {};
    }
}
async function saveLocalConfig(data) {
    const configPath = await getConfigPath();
    const configFilePath = path__default.default.join(configPath, CONFIG_FILENAME);
    await fse__default.default.writeJson(configFilePath, data, {
        encoding: 'utf8',
        spaces: 2,
        mode: 0o600
    });
}

exports.CONFIG_FILENAME = CONFIG_FILENAME;
exports.getLocalConfig = getLocalConfig;
exports.getTmpStoragePath = getTmpStoragePath;
exports.saveLocalConfig = saveLocalConfig;
