// @strapi/content-manager/dist/admin/services/components.mjs
import { normalizeComponentConfigurationUpdateResponse, normalizeComponentConfigurationResponse } from '../utils/layouts/normalizeContentManagerLayout.mjs';
import { contentManagerApi } from './api.mjs';

const componentsApi = contentManagerApi.injectEndpoints({
    endpoints: (builder)=>({
            getComponentConfiguration: builder.query({
                query: (uid)=>`/content-manager/components/${uid}/configuration`,
                transformResponse: (response, _meta, uid)=>normalizeComponentConfigurationResponse(response.data, uid),
                providesTags: (_result, _error, uid)=>[
                        {
                            type: 'ComponentConfiguration',
                            id: uid
                        }
                    ]
            }),
            updateComponentConfiguration: builder.mutation({
                query: ({ uid, ...body })=>({
                        url: `/content-manager/components/${uid}/configuration`,
                        method: 'PUT',
                        data: body
                    }),
                transformResponse: (response)=>normalizeComponentConfigurationUpdateResponse(response.data),
                invalidatesTags: (_result, _error, { uid })=>[
                        {
                            type: 'ComponentConfiguration',
                            id: uid
                        },
                        // otherwise layouts already fetched will have stale component configuration data.
                        {
                            type: 'ContentTypeSettings',
                            id: 'LIST'
                        }
                    ]
            })
        })
});
const { useGetComponentConfigurationQuery, useUpdateComponentConfigurationMutation } = componentsApi;

export { useGetComponentConfigurationQuery, useUpdateComponentConfigurationMutation };
