// @strapi/upload/dist/server/controllers/admin-upload.js
'use strict';

var os = require('os');
var path = require('path');
var fse = require('fs-extra');
var _ = require('lodash');
var utils = require('@strapi/utils');
var index = require('../utils/index.js');
var constants = require('../constants.js');
var upload = require('./validation/admin/upload.js');
var findEntityAndCheckPermissions = require('./utils/find-entity-and-check-permissions.js');
var mimeValidation = require('../utils/mime-validation.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var os__default = /*#__PURE__*/_interopDefault(os);
var path__default = /*#__PURE__*/_interopDefault(path);
var fse__default = /*#__PURE__*/_interopDefault(fse);
var ___default = /*#__PURE__*/_interopDefault(_);

var adminUpload = {
    async bulkUpdateFileInfo (ctx) {
        const { state: { userAbility, user }, request: { body } } = ctx;
        const { updates } = await upload.validateBulkUpdateBody(body);
        const uploadService = index.getService('upload');
        const results = await utils.async.map(updates, async ({ id, fileInfo })=>{
            const { pm } = await findEntityAndCheckPermissions.findEntityAndCheckPermissions(userAbility, constants.ACTIONS.update, constants.FILE_MODEL_UID, id);
            const updated = await uploadService.updateFileInfo(id, fileInfo, {
                user
            });
            // Sign file urls for private providers
            const signedFile = await index.getService('file').signFileUrls(updated);
            return pm.sanitizeOutput(signedFile, {
                action: constants.ACTIONS.read
            });
        });
        ctx.body = results;
    },
    async updateFileInfo (ctx) {
        const { state: { userAbility, user }, query: { id }, request: { body } } = ctx;
        if (typeof id !== 'string') {
            throw new utils.errors.ValidationError('File id is required');
        }
        const uploadService = index.getService('upload');
        const { pm } = await findEntityAndCheckPermissions.findEntityAndCheckPermissions(userAbility, constants.ACTIONS.update, constants.FILE_MODEL_UID, id);
        const data = await upload.validateUploadBody(body);
        const file = await uploadService.updateFileInfo(id, data.fileInfo, {
            user
        });
        // Sign file urls for private providers
        const signedFile = await index.getService('file').signFileUrls(file);
        ctx.body = await pm.sanitizeOutput(signedFile, {
            action: constants.ACTIONS.read
        });
    },
    async replaceFile (ctx) {
        const { state: { userAbility, user }, query: { id }, request: { body, files: { files: filesInput } = {} } } = ctx;
        if (typeof id !== 'string') {
            throw new utils.errors.ValidationError('File id is required');
        }
        const uploadService = index.getService('upload');
        const { pm } = await findEntityAndCheckPermissions.findEntityAndCheckPermissions(userAbility, constants.ACTIONS.update, constants.FILE_MODEL_UID, id);
        if (Array.isArray(filesInput) && filesInput.length > 1) {
            throw new utils.errors.ApplicationError('Cannot replace a file with multiple ones');
        }
        const files = Array.isArray(filesInput) ? filesInput[0] : filesInput;
        const { validFiles, filteredBody, errors: validationErrors } = await mimeValidation.prepareUploadRequest(files, body, strapi);
        if (validFiles.length === 0) {
            throw new utils.errors.ValidationError(validationErrors[0].message);
        }
        const data = await upload.validateUploadBody(filteredBody);
        const replacedFile = await uploadService.replace(id, {
            data,
            file: validFiles[0]
        }, {
            user
        });
        // Regenerate AI metadata for image replacements so the alt text / caption
        // reflect the new file content. Mirrors the post-upload hook in
        // `uploadFiles`; failure is logged and swallowed to keep the replace flow
        // resilient when the AI provider is unavailable.
        const aiMetadataService = index.getService('aiMetadata');
        if (replacedFile?.mime?.startsWith('image/') && await aiMetadataService.isEnabled()) {
            try {
                const metadataResults = await aiMetadataService.processFiles([
                    replacedFile
                ]);
                await aiMetadataService.updateFilesWithAIMetadata([
                    replacedFile
                ], metadataResults, user);
            } catch (error) {
                strapi.log.warn('AI metadata generation failed on replace, proceeding without it', {
                    error: error instanceof Error ? error.message : String(error)
                });
            }
        }
        // Sign file urls for private providers
        const signedFile = await index.getService('file').signFileUrls(replacedFile);
        ctx.body = await pm.sanitizeOutput(signedFile, {
            action: constants.ACTIONS.read
        });
    },
    async uploadFiles (ctx) {
        const { state: { userAbility, user }, request: { body, files: { files } = {} } } = ctx;
        const uploadService = index.getService('upload');
        const pm = strapi.service('admin::permission').createPermissionsManager({
            ability: userAbility,
            action: constants.ACTIONS.create,
            model: constants.FILE_MODEL_UID
        });
        if (!pm.isAllowed) {
            return ctx.forbidden();
        }
        const { validFiles, filteredBody, errors: validationErrors } = await mimeValidation.prepareUploadRequest(files, body, strapi);
        if (validFiles.length === 0) {
            throw new utils.errors.ValidationError(validationErrors[0].message);
        }
        const isMultipleFiles = validFiles.length > 1;
        const data = await upload.validateUploadBody(filteredBody, isMultipleFiles);
        let filesArray = validFiles;
        if (data.fileInfo && Array.isArray(data.fileInfo) && filesArray.length === data.fileInfo.length) {
            // Reorder filesArray to match data.fileInfo order
            const alignedFilesArray = data.fileInfo.map((info)=>{
                return filesArray.find((file)=>file.originalFilename === info.name);
            }).filter(Boolean);
            filesArray = alignedFilesArray;
        }
        // Upload files first to get thumbnails
        const uploadedFiles = await uploadService.upload({
            data,
            files: filesArray
        }, {
            user
        });
        if (uploadedFiles.some((file)=>file.mime?.startsWith('image/'))) {
            await index.getService('metrics').trackUsage('didUploadImage');
        }
        const aiMetadataService = index.getService('aiMetadata');
        // AFTER upload - generate AI metadata for images
        if (await aiMetadataService.isEnabled()) {
            try {
                const metadataResults = await aiMetadataService.processFiles(uploadedFiles);
                // Update the uploaded files with AI metadata
                await aiMetadataService.updateFilesWithAIMetadata(uploadedFiles, metadataResults, user);
            } catch (error) {
                strapi.log.warn('AI metadata generation failed, proceeding without AI enhancements', {
                    error: error instanceof Error ? error.message : String(error)
                });
            }
        }
        // Sign file urls for private providers
        const signedFiles = await utils.async.map(uploadedFiles, index.getService('file').signFileUrls);
        ctx.body = await pm.sanitizeOutput(signedFiles, {
            action: constants.ACTIONS.read
        });
        ctx.status = 201;
    },
    /**
   * @experimental
   * Upload a single file and return the created File.
   *
   * Accepts one file per request (multipart `files` + `fileInfo`) and returns a
   * single `File` object. Unlike `uploadFiles`, it does **not** run AI metadata
   * generation inline — that responsibility is decoupled and will be handled by a
   * background job. Auth and permission checks mirror `POST /upload`.
   */ async unstable_uploadFile (ctx) {
        const { state: { userAbility, user }, request: { body, files: { files } = {} } } = ctx;
        const uploadService = index.getService('upload');
        const pm = strapi.service('admin::permission').createPermissionsManager({
            ability: userAbility,
            action: constants.ACTIONS.create,
            model: constants.FILE_MODEL_UID
        });
        if (!pm.isAllowed) {
            return ctx.forbidden();
        }
        if (___default.default.isEmpty(files) || !Array.isArray(files) && files.size === 0) {
            throw new utils.errors.ApplicationError('Files are empty');
        }
        // Accept a single file per request; ignore any extras defensively.
        const file = Array.isArray(files) ? files[0] : files;
        const fileName = file.originalFilename || 'unknown';
        // Parse the single fileInfo object from the multipart body.
        let fileInfo = {
            name: fileName,
            caption: null,
            alternativeText: null,
            folder: null
        };
        if (body?.fileInfo) {
            const raw = body.fileInfo;
            if (typeof raw === 'string') {
                fileInfo = JSON.parse(raw);
            } else if (Array.isArray(raw)) {
                fileInfo = typeof raw[0] === 'string' ? JSON.parse(raw[0]) : raw[0];
            } else {
                fileInfo = raw;
            }
        }
        // Validate this single file using security checks.
        const { validFiles, errors: validationErrors } = await mimeValidation.prepareUploadRequest(file, {
            fileInfo: JSON.stringify(fileInfo)
        }, strapi);
        if (validFiles.length === 0) {
            throw new utils.errors.ValidationError(validationErrors[0]?.message || 'Validation failed');
        }
        const data = await upload.validateUploadBody({
            fileInfo
        }, false);
        const [uploadedFile] = await uploadService.upload({
            data,
            files: [
                validFiles[0]
            ]
        }, {
            user
        });
        if (uploadedFile.mime?.startsWith('image/')) {
            await index.getService('metrics').trackUsage('didUploadImage');
        }
        // No inline AI metadata generation — intentionally decoupled for the async job.
        // Sign file url for private providers.
        const signedFile = await index.getService('file').signFileUrls(uploadedFile);
        ctx.body = await pm.sanitizeOutput(signedFile, {
            action: constants.ACTIONS.read
        });
        ctx.status = 201;
    },
    /**
   * @experimental
   * Upload files from URLs with SSE streaming for per-file progress
   *
   * Accepts JSON body with URLs and fetches them server-side.
   * Streams Server-Sent Events as each URL is fetched and uploaded:
   * - file:fetching  — when starting to fetch a URL
   * - file:uploading — when upload starts for a fetched file
   * - file:complete  — when a file is successfully uploaded
   * - file:error     — when a URL fetch or upload fails
   * - stream:complete — final summary with all results
   */ async unstable_uploadFromUrls (ctx) {
        const { state: { userAbility, user }, request: { body } } = ctx;
        const uploadService = index.getService('upload');
        const fileService = index.getService('file');
        const pm = strapi.service('admin::permission').createPermissionsManager({
            ability: userAbility,
            action: constants.ACTIONS.create,
            model: constants.FILE_MODEL_UID
        });
        if (!pm.isAllowed) {
            return ctx.forbidden();
        }
        // Parse and validate request body
        const { urls, folderId } = body;
        if (!urls || !Array.isArray(urls) || urls.length === 0) {
            throw new utils.errors.ApplicationError('URLs are required');
        }
        if (urls.length > 20) {
            throw new utils.errors.ApplicationError('Maximum 20 URLs allowed per request');
        }
        // Take manual control of the response for SSE streaming
        ctx.respond = false;
        const res = ctx.res;
        res.writeHead(200, {
            'Content-Type': 'text/event-stream',
            'Cache-Control': 'no-cache',
            Connection: 'keep-alive'
        });
        const writeSSE = (event, data)=>{
            res.write(`event: ${event}\ndata: ${JSON.stringify(data)}\n\n`);
        };
        const total = urls.length;
        const uploadErrors = [];
        const successfulFiles = [];
        // Create temp directory for fetched files
        const tmpWorkingDirectory = await fse__default.default.mkdtemp(path__default.default.join(os__default.default.tmpdir(), 'strapi-url-upload-'));
        const { sizeLimit } = strapi.config.get('plugin::upload');
        try {
            // Process each URL sequentially
            for(let i = 0; i < urls.length; i += 1){
                const url = urls[i];
                writeSSE('file:fetching', {
                    url,
                    index: i,
                    total
                });
                try {
                    // Fetch URL to temp file
                    const { file } = await fileService.fetchUrlToInputFile(url, tmpWorkingDirectory, sizeLimit);
                    const fileName = file.originalFilename;
                    writeSSE('file:uploading', {
                        name: fileName,
                        index: i,
                        total,
                        size: file.size
                    });
                    // Validate using security checks
                    const fileInfo = {
                        name: fileName,
                        caption: null,
                        alternativeText: null,
                        folder: folderId ?? null
                    };
                    const { validFiles, errors: validationErrors } = await mimeValidation.prepareUploadRequest(file, {
                        fileInfo: JSON.stringify(fileInfo)
                    }, strapi);
                    if (validFiles.length === 0) {
                        const errorMessage = validationErrors[0]?.message || 'Validation failed';
                        uploadErrors.push({
                            name: fileName,
                            message: errorMessage
                        });
                        writeSSE('file:error', {
                            name: fileName,
                            url,
                            index: i,
                            message: errorMessage
                        });
                    } else {
                        // Upload the file
                        const data = await upload.validateUploadBody({
                            fileInfo
                        }, false);
                        const [uploadedFile] = await uploadService.upload({
                            data,
                            files: [
                                validFiles[0]
                            ]
                        }, {
                            user
                        });
                        // Sign file url
                        const signedFile = await fileService.signFileUrls(uploadedFile);
                        successfulFiles.push(signedFile);
                        writeSSE('file:complete', {
                            name: fileName,
                            index: i,
                            file: signedFile
                        });
                    }
                } catch (error) {
                    const errorMessage = error instanceof Error ? error.message : String(error);
                    uploadErrors.push({
                        name: url,
                        message: errorMessage
                    });
                    writeSSE('file:error', {
                        url,
                        index: i,
                        message: errorMessage
                    });
                }
            }
            // Track image upload metric once if any images were uploaded
            if (successfulFiles.some((file)=>file.mime?.startsWith('image/'))) {
                await index.getService('metrics').trackUsage('didUploadImage');
            }
            // Send final stream summary
            writeSSE('stream:complete', {
                data: await pm.sanitizeOutput(successfulFiles, {
                    action: constants.ACTIONS.read
                }),
                errors: uploadErrors
            });
        } finally{
            // Clean up temp directory
            await fse__default.default.remove(tmpWorkingDirectory);
        }
        res.end();
    },
    // TODO: split into multiple endpoints
    async upload (ctx) {
        const { query: { id }, request: { files: { files } = {} } } = ctx;
        if (___default.default.isEmpty(files) || !Array.isArray(files) && files.size === 0) {
            if (id) {
                return this.updateFileInfo(ctx);
            }
            throw new utils.errors.ApplicationError('Files are empty');
        }
        await (id ? this.replaceFile : this.uploadFiles)(ctx);
    }
};

module.exports = adminUpload;
