// @strapi/strapi/dist/admin.js
'use strict';

var strapiAdmin = require('@strapi/admin/strapi-admin');
var contentTypeBuilder = require('@strapi/content-type-builder/strapi-admin');
var contentManager = require('@strapi/content-manager/strapi-admin');
var email = require('@strapi/email/strapi-admin');
var upload = require('@strapi/upload/strapi-admin');
var i18n = require('@strapi/i18n/strapi-admin');
var contentReleases = require('@strapi/content-releases/strapi-admin');
var reviewWorkflows = require('@strapi/review-workflows/strapi-admin');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var contentTypeBuilder__default = /*#__PURE__*/_interopDefault(contentTypeBuilder);
var contentManager__default = /*#__PURE__*/_interopDefault(contentManager);
var email__default = /*#__PURE__*/_interopDefault(email);
var upload__default = /*#__PURE__*/_interopDefault(upload);
var i18n__default = /*#__PURE__*/_interopDefault(i18n);
var contentReleases__default = /*#__PURE__*/_interopDefault(contentReleases);
var reviewWorkflows__default = /*#__PURE__*/_interopDefault(reviewWorkflows);

const render = (mountNode, { plugins, ...restArgs })=>{
    return strapiAdmin.renderAdmin(mountNode, {
        ...restArgs,
        plugins: {
            'content-manager': contentManager__default.default,
            'content-type-builder': contentTypeBuilder__default.default,
            email: email__default.default,
            upload: upload__default.default,
            contentReleases: contentReleases__default.default,
            i18n: i18n__default.default,
            reviewWorkflows: reviewWorkflows__default.default,
            ...plugins
        }
    });
};

Object.defineProperty(exports, "private_AutoReloadOverlayBlockerProvider", {
  enumerable: true,
  get: function () { return contentTypeBuilder.private_AutoReloadOverlayBlockerProvider; }
});
Object.defineProperty(exports, "private_useAutoReloadOverlayBlocker", {
  enumerable: true,
  get: function () { return contentTypeBuilder.private_useAutoReloadOverlayBlocker; }
});
Object.defineProperty(exports, "unstable_useContentManagerContext", {
  enumerable: true,
  get: function () { return contentManager.unstable_useContentManagerContext; }
});
Object.defineProperty(exports, "unstable_useDocument", {
  enumerable: true,
  get: function () { return contentManager.unstable_useDocument; }
});
Object.defineProperty(exports, "unstable_useDocumentActions", {
  enumerable: true,
  get: function () { return contentManager.unstable_useDocumentActions; }
});
Object.defineProperty(exports, "unstable_useDocumentLayout", {
  enumerable: true,
  get: function () { return contentManager.unstable_useDocumentLayout; }
});
Object.defineProperty(exports, "useDocumentRBAC", {
  enumerable: true,
  get: function () { return contentManager.useDocumentRBAC; }
});
exports.renderAdmin = render;
Object.keys(strapiAdmin).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return strapiAdmin[k]; }
  });
});
