// @strapi/plugin-users-permissions/dist/admin/components/UsersPermissions/reducer.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var immer = require('immer');
var get = require('lodash/get');
var set = require('lodash/set');
var take = require('lodash/take');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var get__default = /*#__PURE__*/_interopDefault(get);
var set__default = /*#__PURE__*/_interopDefault(set);
var take__default = /*#__PURE__*/_interopDefault(take);

const initialState = {
    initialData: {},
    modifiedData: {},
    routes: {},
    selectedAction: '',
    policies: []
};
const reducer = (state, action)=>immer.produce(state, (draftState)=>{
        switch(action.type){
            case 'ON_CHANGE':
                {
                    const keysLength = action.keys.length;
                    const isChangingCheckbox = action.keys[keysLength - 1] === 'enabled';
                    if (action.value && isChangingCheckbox) {
                        const selectedAction = take__default.default(action.keys, keysLength - 1).join('.');
                        draftState.selectedAction = selectedAction;
                    }
                    set__default.default(draftState, [
                        'modifiedData',
                        ...action.keys
                    ], action.value);
                    break;
                }
            case 'ON_CHANGE_SELECT_ALL':
                {
                    const pathToValue = [
                        'modifiedData',
                        ...action.keys
                    ];
                    const oldValues = get__default.default(state, pathToValue, {});
                    const updatedValues = Object.keys(oldValues).reduce((acc, current)=>{
                        acc[current] = {
                            ...oldValues[current],
                            enabled: action.value
                        };
                        return acc;
                    }, {});
                    set__default.default(draftState, pathToValue, updatedValues);
                    break;
                }
            case 'ON_RESET':
                {
                    draftState.modifiedData = state.initialData;
                    break;
                }
            case 'ON_SUBMIT_SUCCEEDED':
                {
                    draftState.initialData = state.modifiedData;
                    break;
                }
            case 'SELECT_ACTION':
                {
                    const { actionToSelect } = action;
                    draftState.selectedAction = actionToSelect === state.selectedAction ? '' : actionToSelect;
                    break;
                }
            default:
                return draftState;
        }
    });

exports.default = reducer;
exports.initialState = initialState;
