// @strapi/admin/dist/admin/admin/src/components/FormInputs/Renderer.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var designSystem = require('@strapi/design-system');
var useFocusInputField = require('../../hooks/useFocusInputField.js');
var Form = require('../Form.js');
var Boolean = require('./Boolean.js');
var Checkbox = require('./Checkbox.js');
var Date = require('./Date.js');
var DateTime = require('./DateTime.js');
var Email = require('./Email.js');
var Enumeration = require('./Enumeration.js');
var Json = require('./Json.js');
var Number = require('./Number.js');
var Password = require('./Password.js');
var String = require('./String.js');
var Textarea = require('./Textarea.js');
var Time = require('./Time.js');

/* -------------------------------------------------------------------------------------------------
 * InputRenderer
 * -----------------------------------------------------------------------------------------------*/ /**
 * @internal
 * @experimental
 * @description A generic form renderer for Strapi forms. Similar to GenericInputs but with a different API. The entire component is memoized to avoid re-renders in large forms.
 */ const InputRenderer = /*#__PURE__*/ React.memo((props)=>{
    switch(props.type){
        case 'biginteger':
        case 'timestamp':
        case 'string':
        case 'uid':
            return /*#__PURE__*/ jsxRuntime.jsx(String.StringInput, {
                ...props
            });
        case 'boolean':
            return /*#__PURE__*/ jsxRuntime.jsx(Boolean.BooleanInput, {
                ...props
            });
        case 'checkbox':
            return /*#__PURE__*/ jsxRuntime.jsx(Checkbox.CheckboxInput, {
                ...props
            });
        case 'datetime':
            return /*#__PURE__*/ jsxRuntime.jsx(DateTime.DateTimeInput, {
                ...props
            });
        case 'date':
            return /*#__PURE__*/ jsxRuntime.jsx(Date.DateInput, {
                ...props
            });
        case 'decimal':
        case 'float':
        case 'integer':
            return /*#__PURE__*/ jsxRuntime.jsx(Number.NumberInput, {
                ...props
            });
        case 'json':
            return /*#__PURE__*/ jsxRuntime.jsx(Json.JsonInput, {
                ...props
            });
        case 'email':
            return /*#__PURE__*/ jsxRuntime.jsx(Email.EmailInput, {
                ...props
            });
        case 'enumeration':
            return /*#__PURE__*/ jsxRuntime.jsx(Enumeration.EnumerationInput, {
                ...props
            });
        case 'password':
            return /*#__PURE__*/ jsxRuntime.jsx(Password.PasswordInput, {
                ...props
            });
        case 'text':
            return /*#__PURE__*/ jsxRuntime.jsx(Textarea.TextareaInput, {
                ...props
            });
        case 'time':
            return /*#__PURE__*/ jsxRuntime.jsx(Time.TimeInput, {
                ...props
            });
        default:
            {
                const notSupportedProps = props;
                return /*#__PURE__*/ jsxRuntime.jsx(NotSupportedField, {
                    label: notSupportedProps.label,
                    hint: notSupportedProps.hint,
                    name: notSupportedProps.name,
                    required: notSupportedProps.required,
                    type: notSupportedProps.type,
                    labelAction: notSupportedProps.labelAction
                });
            }
    }
});
const NotSupportedField = ({ label, hint, name, required, type, labelAction })=>{
    const { error } = Form.useField(name);
    const fieldRef = useFocusInputField.useFocusInputField(name);
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Field.Root, {
        error: error,
        name: name,
        hint: hint,
        required: required,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Label, {
                action: labelAction,
                children: label
            }),
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.TextInput, {
                ref: fieldRef,
                disabled: true,
                placeholder: `Unsupported field type: ${type}`,
                required: required,
                type: "text",
                value: ""
            }),
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Hint, {}),
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Error, {})
        ]
    });
};
const MemoizedInputRenderer = /*#__PURE__*/ React.memo(InputRenderer);

exports.InputRenderer = MemoizedInputRenderer;
