// @strapi/admin/dist/admin/admin/src/features/Auth.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var reactRouterDom = require('react-router-dom');
var Context = require('../components/Context.js');
var hooks = require('../core/store/hooks.js');
var StrapiApp = require('./StrapiApp.js');
var useIdleSessionLogout = require('../hooks/useIdleSessionLogout.js');
var useQueryParams = require('../hooks/useQueryParams.js');
var reducer = require('../reducer.js');
var api = require('../services/api.js');
var auth = require('../services/auth.js');
var normalizeAdminLocale = require('../translations/normalizeAdminLocale.js');
var deviceId = require('../utils/deviceId.js');
var getFetchClient = require('../utils/getFetchClient.js');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

/**
 * ensure the Auth context never exposes a non-function for checkUserHasPermissions.
 * When this is undefined (e.g. context timing in production builds), consumers would throw
 * "p is not a function" / "checkUserHasPermissions is not a function". By always passing
 * a function here, all current and future consumers are protected without per-call-site guards.
 *
 * When would the fallback run? Only if the real checkUserHasPermissions were ever undefined
 * when we pass to the Provider (e.g. a rare timing/build edge case). In normal runs it is
 * always defined (useCallback), so the real function is passed and behavior is unchanged.
 *
 * If the fallback ever did run: it returns [] so consumers (which use .length > 0) treat it
 * as "no permission" for that render—under-permissive. On the next AuthProvider re-render we
 * pass the real function again, so the context updates and the view corrects quickly.
 * @see https://github.com/strapi/strapi/issues/24384
 */ const NOOP_CHECK_USER_HAS_PERMISSIONS = async ()=>[];
const [Provider, useAuth] = Context.createContext('Auth');
const STORAGE_KEYS = {
    STATUS: 'isLoggedIn'
};
const AuthProvider = ({ children, _defaultPermissions = [], _disableRenewToken = false })=>{
    const dispatch = hooks.useTypedDispatch();
    const runRbacMiddleware = StrapiApp.useStrapiApp('AuthProvider', (state)=>state.rbac.run);
    const location = reactRouterDom.useLocation();
    const [{ rawQuery }] = useQueryParams.useQueryParams();
    const locationRef = React__namespace.useRef(location);
    // Update ref without causing re-render
    React__namespace.useEffect(()=>{
        locationRef.current = location;
    }, [
        location
    ]);
    const token = hooks.useTypedSelector((state)=>state.admin_app.token ?? null);
    const { data: user, isLoading: isLoadingUser } = auth.useGetMeQuery(undefined, {
        /**
     * If there's no token, we don't try to fetch
     * the user data because it will fail.
     */ skip: !token
    });
    const { data: userPermissions = _defaultPermissions, refetch, isUninitialized, isLoading: isLoadingPermissions } = auth.useGetMyPermissionsQuery(undefined, {
        skip: !token
    });
    const navigate = reactRouterDom.useNavigate();
    const [loginMutation] = auth.useLoginMutation();
    const [logoutMutation] = auth.useLogoutMutation();
    const clearStateAndLogout = React__namespace.useCallback(()=>{
        dispatch(api.adminApi.util.resetApiState());
        dispatch(reducer.logout());
        navigate('/auth/login');
    }, [
        dispatch,
        navigate
    ]);
    /**
   * Clear *only this tab's* session and redirect to login, without removing the
   * shared `isLoggedIn`/`jwtToken` storage keys. Unlike `clearStateAndLogout`,
   * this does not dispatch `logoutAction`, so it does not fire the cross-tab
   * `storage` event that logs every tab out. Used by the speculative idle timer,
   * which fires per-tab at the access token's `exp`: an idle tab must never tear
   * down a session that another tab is still actively using. A genuine logout
   * (user-initiated, or a server-confirmed 401 via `setOnSessionExpired`) still
   * goes through `clearStateAndLogout` and broadcasts to all tabs.
   */ const clearLocalSessionAndRedirect = React__namespace.useCallback(()=>{
        dispatch(api.adminApi.util.resetApiState());
        dispatch(reducer.setToken(null));
        navigate('/auth/login');
    }, [
        dispatch,
        navigate
    ]);
    const resyncToken = React__namespace.useCallback((newToken)=>{
        dispatch(reducer.setToken(newToken));
    }, [
        dispatch
    ]);
    /**
   * Track the timestamp of the user's last interaction. This is the activity
   * signal `useIdleSessionLogout` uses to tell an active user (silently renew
   * their session) apart from a genuinely idle one (log out). Successful API
   * calls don't refresh the short-lived access token on their own, so without
   * this an actively-working user is logged out the moment the token expires.
   */ const lastActivityRef = React__namespace.useRef(Date.now());
    React__namespace.useEffect(()=>{
        const markActivity = ()=>{
            lastActivityRef.current = Date.now();
        };
        const events = [
            'pointerdown',
            'keydown',
            'click',
            'scroll',
            'touchstart'
        ];
        events.forEach((event)=>window.addEventListener(event, markActivity, {
                passive: true,
                capture: true
            }));
        return ()=>{
            events.forEach((event)=>window.removeEventListener(event, markActivity, {
                    capture: true
                }));
        };
    }, []);
    const getLastActivityAt = React__namespace.useCallback(()=>lastActivityRef.current, []);
    /**
   * Silently rotate the refresh token and mint a new access token. Returns
   * `true` on success (the resulting `setToken` re-arms the idle timer) and
   * `false` when the server rejects it (session genuinely over).
   */ const renewSession = React__namespace.useCallback(async ()=>{
        try {
            await getFetchClient.attemptTokenRefresh();
            return true;
        } catch  {
            return false;
        }
    }, []);
    React__namespace.useEffect(()=>{
        if (user) {
            if (user.preferedLanguage) {
                dispatch(reducer.setLocale(normalizeAdminLocale.normalizeAdminLocale(user.preferedLanguage)));
            }
        }
    }, [
        dispatch,
        user
    ]);
    /**
   * Register a callback to update Redux state when the token is refreshed.
   * This ensures the app state stays in sync with the token stored in localStorage/cookies.
   */ React__namespace.useEffect(()=>{
        getFetchClient.setOnTokenUpdate((newToken)=>{
            dispatch(reducer.setToken(newToken));
        });
        return ()=>{
            getFetchClient.setOnTokenUpdate(null);
        };
    }, [
        dispatch
    ]);
    /**
   * Register the session-expired handler that the fetch layer / RTK baseQuery
   * call when the server rejects the refresh token. This is what redirects
   * the active tab to /auth/login on a 401, instead of leaving the user on
   * a stale page until they click something.
   */ React__namespace.useEffect(()=>{
        getFetchClient.setOnSessionExpired(()=>{
            clearStateAndLogout();
        });
        return ()=>{
            getFetchClient.setOnSessionExpired(null);
        };
    }, [
        clearStateAndLogout
    ]);
    /**
   * Session lifecycle at access-token expiry. The timer fires per-tab; the hook
   * then either (1) re-syncs from a token another tab refreshed, (2) silently
   * renews when the user has been active, or (3) logs out. Idle logout is local
   * so an idle tab can't force-logout an actively-used one; a server-confirmed
   * dead session (renewal rejected) broadcasts globally. See
   * `useIdleSessionLogout` for the full rationale.
   */ useIdleSessionLogout.useIdleSessionLogout({
        token,
        onExpired: clearLocalSessionAndRedirect,
        onSessionDead: clearStateAndLogout,
        getStoredToken: reducer.getStoredToken,
        onResync: resyncToken,
        getLastActivityAt,
        renewSession,
        disabled: _disableRenewToken
    });
    React__namespace.useEffect(()=>{
        /**
     * This will log a user out of all tabs if they log out in one tab.
     */ const handleUserStorageChange = (event)=>{
            if (event.key === STORAGE_KEYS.STATUS && event.newValue === null) {
                clearStateAndLogout();
            }
        };
        window.addEventListener('storage', handleUserStorageChange);
        return ()=>{
            window.removeEventListener('storage', handleUserStorageChange);
        };
    });
    const login = React__namespace.useCallback(async ({ rememberMe, ...body })=>{
        const res = await loginMutation({
            ...body,
            deviceId: deviceId.getOrCreateDeviceId(),
            rememberMe
        });
        /**
       * There will always be a `data` key in the response
       * because if something fails, it will throw an error.
       */ if ('data' in res) {
            const { token } = res.data;
            dispatch(reducer.login({
                token,
                persist: rememberMe
            }));
        }
        return res;
    }, [
        dispatch,
        loginMutation
    ]);
    const logout = React__namespace.useCallback(async ()=>{
        await logoutMutation({
            deviceId: deviceId.getOrCreateDeviceId()
        });
        clearStateAndLogout();
    }, [
        clearStateAndLogout,
        logoutMutation
    ]);
    const refetchPermissions = React__namespace.useCallback(async ()=>{
        if (!isUninitialized) {
            await refetch();
        }
    }, [
        isUninitialized,
        refetch
    ]);
    const [checkPermissions] = auth.useLazyCheckPermissionsQuery();
    const checkUserHasPermissions = React__namespace.useCallback(async (permissions, passedPermissions, // TODO:
    // Here we have parameterised checkUserHasPermissions in order to pass
    // query context from elsewhere in the application.
    // See packages/core/content-manager/admin/src/features/DocumentRBAC.tsx
    // This is in order to calculate permissions on accurate query params.
    // We should be able to rely on the query params in this provider
    // If we need to pass additional context to the RBAC middleware
    // we should define a better context type.
    rawQueryContext)=>{
        /**
       * If there's no permissions to check, then we allow it to
       * pass to preserve existing behaviours.
       *
       * TODO: should we review this? it feels more dangerous than useful.
       */ if (!permissions || permissions.length === 0) {
            return [
                {
                    action: '',
                    subject: ''
                }
            ];
        }
        /**
       * Given the provided permissions, return the permissions from either passedPermissions
       * or userPermissions as this is expected to be the full permission entity.
       */ const actualUserPermissions = passedPermissions ?? userPermissions;
        const matchingPermissions = actualUserPermissions.filter((permission)=>permissions.findIndex((perm)=>perm.action === permission.action && // Only check the subject if it's provided
                (perm.subject == undefined || perm.subject === permission.subject)) >= 0);
        const middlewaredPermissions = await runRbacMiddleware({
            user,
            permissions: userPermissions,
            pathname: locationRef.current.pathname,
            search: (rawQueryContext || rawQuery).split('?')[1] ?? ''
        }, matchingPermissions);
        const shouldCheckConditions = middlewaredPermissions.some((perm)=>Array.isArray(perm.conditions) && perm.conditions.length > 0);
        if (!shouldCheckConditions) {
            return middlewaredPermissions;
        }
        const { data, error } = await checkPermissions({
            permissions: middlewaredPermissions.map((perm)=>({
                    action: perm.action,
                    subject: perm.subject
                }))
        });
        if (error) {
            throw error;
        } else {
            return middlewaredPermissions.filter((_, index)=>data?.data[index] === true);
        }
    }, [
        checkPermissions,
        rawQuery,
        runRbacMiddleware,
        user,
        userPermissions
    ]);
    const isLoading = isLoadingUser || isLoadingPermissions;
    return /*#__PURE__*/ jsxRuntime.jsx(Provider, {
        token: token,
        user: user,
        login: login,
        logout: logout,
        permissions: userPermissions,
        checkUserHasPermissions: checkUserHasPermissions ?? NOOP_CHECK_USER_HAS_PERMISSIONS,
        refetchPermissions: refetchPermissions,
        isLoading: isLoading,
        children: children
    });
};

exports.AuthProvider = AuthProvider;
exports.STORAGE_KEYS = STORAGE_KEYS;
exports.useAuth = useAuth;
