// @strapi/core/dist/services/server/koa.js
'use strict';

var fp = require('lodash/fp');
var Koa = require('koa');
var createError = require('http-errors');
var delegate = require('delegates');
var errors = require('../errors.js');
var koaMethods = require('./koa-methods.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var Koa__default = /*#__PURE__*/_interopDefault(Koa);
var createError__default = /*#__PURE__*/_interopDefault(createError);
var delegate__default = /*#__PURE__*/_interopDefault(delegate);

const addCustomMethods = (app)=>{
    const delegator = delegate__default.default(app.context, 'response');
    /* errors */ for (const { code, statusName, methodName } of koaMethods.errorMethodEntries()){
        app.response[methodName] = function responseCode(response = statusName, details = {}) {
            const httpError = createError__default.default(code, response, {
                details
            });
            const { status, body } = errors.formatHttpError(httpError);
            this.status = status;
            this.body = body;
        };
        delegator.method(methodName);
    }
    /* send, created, deleted */ app.response.send = function send(data, status = 200) {
        this.status = status;
        this.body = data;
    };
    app.response.created = function created(data) {
        this.status = 201;
        this.body = data;
    };
    app.response.deleted = function deleted(data) {
        if (fp.isNil(data)) {
            this.status = 204;
        } else {
            this.status = 200;
            this.body = data;
        }
    };
    delegator.method('send').method('created').method('deleted');
    return app;
};
const createKoaApp = ({ proxy, keys })=>{
    const app = new Koa__default.default({
        proxy
    });
    app.keys = keys;
    addCustomMethods(app);
    return app;
};

module.exports = createKoaApp;
