// @strapi/upload/dist/admin/future/store/uploadProgress.mjs
import { createSlice, createSelector } from '@reduxjs/toolkit';

const initialState = {
    isVisible: false,
    isMinimized: false,
    totalFiles: 0,
    files: [],
    errors: [],
    uploadId: 0
};
const uploadProgressSlice = createSlice({
    name: 'uploadProgress',
    initialState,
    reducers: {
        openUploadProgress (state, action) {
            state.isVisible = true;
            state.isMinimized = false;
            // Create pending files for upload
            const pendingFiles = action.payload.fileNames.map((name, index)=>({
                    name,
                    index,
                    status: 'pending',
                    size: action.payload.fileSizes?.[index] ?? 0,
                    uploadedBytes: 0
                }));
            state.files = pendingFiles;
            state.totalFiles = action.payload.totalFiles;
            state.errors = [];
            state.uploadId += 1;
        },
        setFileUploading (state, action) {
            const { index, size } = action.payload;
            if (state.files[index]) {
                state.files[index].status = 'uploading';
                state.files[index].size = size;
            }
        },
        setFileProgress (state, action) {
            const { index, bytes } = action.payload;
            const file = state.files[index];
            if (file) {
                // Clamp to the known file size so the aggregate can never exceed 100%.
                file.uploadedBytes = Math.min(bytes, file.size);
            }
        },
        setFileComplete (state, action) {
            const { index, file } = action.payload;
            if (state.files[index]) {
                state.files[index].status = 'complete';
                state.files[index].file = file;
                // Reflect completion in the aggregate even if the final progress event was throttled.
                state.files[index].uploadedBytes = state.files[index].size;
            }
        },
        setFileError (state, action) {
            const { index, name, message } = action.payload;
            if (state.files[index]) {
                state.files[index].status = 'error';
                state.files[index].error = message;
            }
            state.errors = [
                ...state.errors,
                {
                    name,
                    message
                }
            ];
        },
        addUploadErrors (state, action) {
            state.errors = [
                ...state.errors,
                ...action.payload
            ];
        },
        closeUploadProgress (state) {
            state.isVisible = false;
            state.isMinimized = false;
            state.totalFiles = 0;
            state.files = [];
            state.errors = [];
        },
        toggleMinimize (state) {
            state.isMinimized = !state.isMinimized;
        },
        cancelUpload (state) {
            // Mark all pending and uploading files as cancelled
            state.files = state.files.map((file)=>{
                if (file.status === 'pending' || file.status === 'uploading') {
                    return {
                        ...file,
                        status: 'cancelled'
                    };
                }
                return file;
            });
        },
        setUploadFailed (state, action) {
            // Mark all pending and uploading files as errored when a catastrophic failure occurs
            state.files = state.files.map((file)=>{
                if (file.status === 'pending' || file.status === 'uploading') {
                    return {
                        ...file,
                        status: 'error',
                        error: action.payload.message
                    };
                }
                return file;
            });
            state.errors = [
                ...state.errors,
                {
                    name: 'Upload Error',
                    message: action.payload.message
                }
            ];
        },
        retryCancelledFiles (state) {
            // Reset all cancelled files back to pending for retry
            state.files = state.files.map((file)=>{
                if (file.status === 'cancelled') {
                    return {
                        ...file,
                        status: 'pending',
                        uploadedBytes: 0
                    };
                }
                return file;
            });
        }
    }
});
/**
 * Byte-weighted aggregate progress across the whole batch: `sum(uploadedBytes) / sum(size)`.
 *
 * Falls back to count-based progress (settled files / total files) when all sizes are
 * zero — e.g. URL-flow rows where the file size is unknown up front.
 */ const selectAggregateProgress = createSelector((state)=>state.uploadProgress.files, (files)=>{
    if (files.length === 0) return 0;
    const totalSize = files.reduce((sum, f)=>sum + f.size, 0);
    if (totalSize === 0) {
        const settled = files.filter((f)=>f.status === 'complete' || f.status === 'error' || f.status === 'cancelled').length;
        return Math.round(settled / files.length * 100);
    }
    const uploadedBytes = files.reduce((sum, f)=>sum + f.uploadedBytes, 0);
    return Math.round(uploadedBytes / totalSize * 100);
});
const { openUploadProgress, setFileUploading, setFileProgress, setFileComplete, setFileError, addUploadErrors, closeUploadProgress, toggleMinimize, cancelUpload, setUploadFailed, retryCancelledFiles } = uploadProgressSlice.actions;
const uploadProgressReducer = uploadProgressSlice.reducer;

export { addUploadErrors, cancelUpload, closeUploadProgress, openUploadProgress, retryCancelledFiles, selectAggregateProgress, setFileComplete, setFileError, setFileProgress, setFileUploading, setUploadFailed, toggleMinimize, uploadProgressReducer };
