// @strapi/content-manager/dist/admin/pages/EditView/components/FormInputs/Wysiwyg/utils/mdRenderer.js
'use strict';

var highlight_js = require('highlight.js');
var Markdown = require('markdown-it');
var abbr = require('markdown-it-abbr');
var container = require('markdown-it-container');
var deflist = require('markdown-it-deflist');
var emoji = require('markdown-it-emoji');
var footnote = require('markdown-it-footnote');
var ins = require('markdown-it-ins');
var mark = require('markdown-it-mark');
var sub = require('markdown-it-sub');
var sup = require('markdown-it-sup');
require('highlight.js/styles/solarized-dark.css');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var Markdown__default = /*#__PURE__*/_interopDefault(Markdown);
var abbr__default = /*#__PURE__*/_interopDefault(abbr);
var container__default = /*#__PURE__*/_interopDefault(container);
var deflist__default = /*#__PURE__*/_interopDefault(deflist);
var emoji__default = /*#__PURE__*/_interopDefault(emoji);
var footnote__default = /*#__PURE__*/_interopDefault(footnote);
var ins__default = /*#__PURE__*/_interopDefault(ins);
var mark__default = /*#__PURE__*/_interopDefault(mark);
var sub__default = /*#__PURE__*/_interopDefault(sub);
var sup__default = /*#__PURE__*/_interopDefault(sup);

const md = new Markdown__default.default({
    html: true,
    xhtmlOut: false,
    breaks: true,
    langPrefix: 'language-',
    linkify: true,
    typographer: true,
    // Code from: https://github.com/markdown-it/markdown-it/blob/master/support/demo_template/index.js#L83
    highlight (str, lang) {
        if (lang && lang !== 'auto' && highlight_js.getLanguage(lang)) {
            return '<pre class="hljs language-' + md.utils.escapeHtml(lang.toLowerCase()) + '"><code>' + highlight_js.highlight(lang, str, true).value + '</code></pre>';
        }
        if (lang === 'auto') {
            const result = highlight_js.highlightAuto(str);
            return '<pre class="hljs language-' + md.utils.escapeHtml(result.language) + '"><code>' + result.value + '</code></pre>';
        }
        return '<pre class="hljs"><code>' + md.utils.escapeHtml(str) + '</code></pre>';
    }
}).use(abbr__default.default).use(container__default.default, 'warning').use(container__default.default, 'tip').use(deflist__default.default).use(emoji__default.default).use(footnote__default.default).use(ins__default.default).use(mark__default.default).use(sub__default.default).use(sup__default.default);
// Code from: https://github.com/markdown-it/markdown-it-footnote/blob/master/index.js#L29
md.renderer.rules.footnote_ref = (tokens, idx, options, env, slf)=>{
    const caption = slf.rules.footnote_caption?.(tokens, idx, options, env, slf);
    return '<sup class="footnote-ref"><span>' + caption + '</span></sup>';
};
md.renderer.rules.footnote_anchor = ()=>{
    /* ↩ with escape code to prevent display as Apple Emoji on iOS */ return ' <span class="footnote-backref">\u21a9\uFE0E</span>';
};

exports.md = md;
