// @strapi/content-manager/dist/server/services/utils/draft.js
'use strict';

var fp = require('lodash/fp');
var strapiUtils = require('@strapi/utils');
var draftRelations = require('./draft-relations.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var strapiUtils__default = /*#__PURE__*/_interopDefault(strapiUtils);

const { isVisibleAttribute, hasDraftAndPublish } = strapiUtils__default.default.contentTypes;
const isLocalizedContentType = (model)=>model.pluginOptions?.i18n?.localized === true;
const toPublishedDocumentKey = (documentId, locale)=>`${documentId}:${locale ?? ''}`;
const mergeCollectedLinks = (left, right)=>({
        m2mLinks: [
            ...left.m2mLinks,
            ...right.m2mLinks
        ],
        xToOneLinks: [
            ...left.xToOneLinks,
            ...right.xToOneLinks
        ]
    });
const toDraftRelationLink = (entry, targetUid, targetIsLocalized, documentLocale)=>{
    if (!entry?.documentId) {
        return null;
    }
    return {
        targetUid,
        documentId: entry.documentId,
        locale: targetIsLocalized ? entry.locale ?? documentLocale : null
    };
};
const collectDraftRelationLinks = (strapi, entity, uid, documentLocale)=>{
    const model = strapi.getModel(uid);
    const locale = entity.locale ?? documentLocale;
    return Object.keys(model.attributes).reduce((collected, attributeName)=>{
        const attribute = model.attributes[attributeName];
        const value = entity[attributeName];
        if (!value) {
            return collected;
        }
        switch(attribute.type){
            case 'relation':
                {
                    if (!('target' in attribute)) {
                        return collected;
                    }
                    const targetModel = strapi.getModel(attribute.target);
                    if (!targetModel || !hasDraftAndPublish(targetModel)) {
                        return collected;
                    }
                    if (attribute.target === uid || !isVisibleAttribute(model, attributeName)) {
                        return collected;
                    }
                    const targetIsLocalized = isLocalizedContentType(targetModel);
                    const relatedEntries = fp.castArray(value);
                    const links = relatedEntries.map((entry)=>toDraftRelationLink(entry, attribute.target, targetIsLocalized, locale)).filter((link)=>link !== null);
                    if (links.length === 0) {
                        return collected;
                    }
                    if (draftRelations.isBidirectionalManyToMany(attribute)) {
                        return {
                            ...collected,
                            m2mLinks: [
                                ...collected.m2mLinks,
                                ...links
                            ]
                        };
                    }
                    return {
                        ...collected,
                        xToOneLinks: [
                            ...collected.xToOneLinks,
                            ...links
                        ]
                    };
                }
            case 'component':
                {
                    return fp.castArray(value).reduce((componentCollected, componentValue)=>mergeCollectedLinks(componentCollected, collectDraftRelationLinks(strapi, componentValue, attribute.component, locale)), collected);
                }
            case 'dynamiczone':
                {
                    return value.reduce((zoneCollected, componentValue)=>{
                        return mergeCollectedLinks(zoneCollected, collectDraftRelationLinks(strapi, componentValue, componentValue.__component, locale));
                    }, collected);
                }
            default:
                return collected;
        }
    }, {
        m2mLinks: [],
        xToOneLinks: []
    });
};
const countLinksToUnpublishedDocuments = async (strapi, links)=>{
    if (links.length === 0) {
        return 0;
    }
    const linksByTarget = links.reduce((acc, link)=>{
        const targetLinks = acc.get(link.targetUid) ?? [];
        targetLinks.push(link);
        acc.set(link.targetUid, targetLinks);
        return acc;
    }, new Map());
    const counts = await Promise.all(Array.from(linksByTarget.entries()).map(async ([targetUid, targetLinks])=>{
        const targetModel = strapi.getModel(targetUid);
        const targetIsLocalized = isLocalizedContentType(targetModel);
        const documentIds = [
            ...new Set(targetLinks.map((link)=>link.documentId))
        ];
        const publishedRows = await strapi.db.query(targetUid).findMany({
            select: [
                'documentId',
                'locale'
            ],
            where: {
                documentId: {
                    $in: documentIds
                },
                publishedAt: {
                    $notNull: true
                }
            }
        });
        if (!targetIsLocalized) {
            const publishedDocumentIds = new Set(publishedRows.map((row)=>row.documentId));
            return targetLinks.filter((link)=>!publishedDocumentIds.has(link.documentId)).length;
        }
        const publishedDocumentKeys = new Set(publishedRows.map((row)=>toPublishedDocumentKey(row.documentId, row.locale)));
        return targetLinks.filter((link)=>!publishedDocumentKeys.has(toPublishedDocumentKey(link.documentId, link.locale))).length;
    }));
    return counts.reduce((total, count)=>total + count, 0);
};
/**
 * sumDraftCounts works recursively on the attributes of a model counting draft relations
 * that matter for publish warnings.
 *
 * - unpublishedRelations: xToOne / oneToMany style links stripped from the published version
 * - draftM2mLinks: bidirectional manyToMany links to documents without a published version
 */ const sumDraftCounts = async (strapi, entity, uid)=>{
    const { m2mLinks, xToOneLinks } = collectDraftRelationLinks(strapi, entity, uid);
    const [draftM2mLinks, unpublishedRelations] = await Promise.all([
        countLinksToUnpublishedDocuments(strapi, m2mLinks),
        countLinksToUnpublishedDocuments(strapi, xToOneLinks)
    ]);
    return {
        unpublishedRelations,
        draftM2mLinks
    };
};

exports.sumDraftCounts = sumDraftCounts;
