// @strapi/plugin-cloud/dist/admin/components/CLIDeployTabs.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import { useNotification, useClipboard } from '@strapi/admin/strapi-admin';
import { Box, Tabs, Typography, Flex, Tooltip, IconButton } from '@strapi/design-system';
import { Duplicate } from '@strapi/icons';
import { useIntl } from 'react-intl';
import { getTrad } from '../utils/getTrad.mjs';

const CopyCommandButton = ({ command })=>{
    const { formatMessage } = useIntl();
    const { toggleNotification } = useNotification();
    const { copy } = useClipboard();
    const copyLabel = formatMessage({
        id: getTrad('Homepage.deploy.cli.copy'),
        defaultMessage: 'Copy'
    });
    const handleCopy = async ()=>{
        const didCopy = await copy(command);
        if (didCopy) {
            toggleNotification({
                type: 'success',
                message: formatMessage({
                    id: getTrad('Homepage.deploy.cli.copySuccess'),
                    defaultMessage: 'Command copied to clipboard'
                })
            });
        }
    };
    return /*#__PURE__*/ jsx(Tooltip, {
        label: copyLabel,
        children: /*#__PURE__*/ jsx(IconButton, {
            size: "XS",
            variant: "ghost",
            label: copyLabel,
            onClick: handleCopy,
            children: /*#__PURE__*/ jsx(Duplicate, {})
        })
    });
};
const commands = {
    yarn: {
        link: 'yarn strapi link',
        deploy: 'yarn strapi deploy'
    },
    npm: {
        link: 'npm run strapi link',
        deploy: 'npm run strapi deploy'
    }
};
const CLIDeployTabs = ()=>{
    const { formatMessage } = useIntl();
    return /*#__PURE__*/ jsx(Box, {
        minWidth: "28em",
        maxWidth: "28em",
        paddingTop: 8,
        children: /*#__PURE__*/ jsxs(Tabs.Root, {
            defaultValue: "yarn",
            children: [
                /*#__PURE__*/ jsxs(Tabs.List, {
                    "aria-label": formatMessage({
                        id: getTrad('Homepage.deploy.cli.ariaLabel'),
                        defaultMessage: 'Package manager'
                    }),
                    children: [
                        /*#__PURE__*/ jsx(Tabs.Trigger, {
                            value: "yarn",
                            children: /*#__PURE__*/ jsx(Typography, {
                                variant: "omega",
                                children: "Yarn"
                            })
                        }),
                        /*#__PURE__*/ jsx(Tabs.Trigger, {
                            value: "npm",
                            children: /*#__PURE__*/ jsx(Typography, {
                                variant: "omega",
                                children: "NPM"
                            })
                        })
                    ]
                }),
                [
                    'yarn',
                    'npm'
                ].map((packageManager)=>/*#__PURE__*/ jsx(Tabs.Content, {
                        value: packageManager,
                        children: /*#__PURE__*/ jsx(Box, {
                            background: "neutral100",
                            children: /*#__PURE__*/ jsx(Box, {
                                padding: 4,
                                children: /*#__PURE__*/ jsxs(Flex, {
                                    direction: "column",
                                    gap: 4,
                                    children: [
                                        /*#__PURE__*/ jsxs(Flex, {
                                            direction: "row",
                                            justifyContent: "space-between",
                                            alignItems: "center",
                                            gap: 2,
                                            children: [
                                                /*#__PURE__*/ jsx(Typography, {
                                                    tag: "code",
                                                    textColor: "neutral800",
                                                    children: commands[packageManager].link
                                                }),
                                                /*#__PURE__*/ jsx(CopyCommandButton, {
                                                    command: commands[packageManager].link
                                                })
                                            ]
                                        }),
                                        /*#__PURE__*/ jsxs(Flex, {
                                            direction: "row",
                                            justifyContent: "space-between",
                                            alignItems: "center",
                                            gap: 2,
                                            children: [
                                                /*#__PURE__*/ jsx(Typography, {
                                                    tag: "code",
                                                    textColor: "neutral800",
                                                    children: commands[packageManager].deploy
                                                }),
                                                /*#__PURE__*/ jsx(CopyCommandButton, {
                                                    command: commands[packageManager].deploy
                                                })
                                            ]
                                        })
                                    ]
                                })
                            })
                        })
                    }, packageManager))
            ]
        })
    });
};

export { CLIDeployTabs };
