// @strapi/plugin-cloud/dist/admin/components/CloudFeatures.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var reactIntl = require('react-intl');
var getTrad = require('../utils/getTrad.js');

const CloudFeatures = ()=>{
    const { formatMessage } = reactIntl.useIntl();
    const features = [
        {
            id: 'cdn',
            message: 'Global CDN'
        },
        {
            id: 'pushToDeploy',
            message: 'Push to deploy'
        },
        {
            id: 'customDomain',
            message: 'Custom domains'
        }
    ];
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
            paddingBottom: 10,
            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                direction: "column",
                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                    direction: "row",
                    wrap: "wrap",
                    children: features.map(({ id, message })=>/*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                            paddingRight: 5,
                            children: [
                                /*#__PURE__*/ jsxRuntime.jsx(icons.Check, {
                                    fill: "primary500"
                                }),
                                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                    variant: "omega",
                                    paddingLeft: 1,
                                    children: formatMessage({
                                        id: getTrad.getTrad(`Homepage.features.${id}`),
                                        defaultMessage: message
                                    })
                                })
                            ]
                        }, id))
                })
            })
        })
    });
};

exports.CloudFeatures = CloudFeatures;
