// @strapi/cloud-cli/dist/utils/error-message-factories.js
'use strict';

var chalk = require('chalk');
var boxen = require('boxen');
var api = require('../config/api.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var chalk__default = /*#__PURE__*/_interopDefault(chalk);
var boxen__default = /*#__PURE__*/_interopDefault(boxen);

const environmentErrorMessageFactory = ({ projectName, firstLine, secondLine })=>{
    return [
        chalk__default.default.yellow(firstLine),
        '',
        chalk__default.default.cyan(secondLine),
        chalk__default.default.blue(' →  ') + chalk__default.default.blue.underline(`${api.apiConfig.dashboardBaseUrl}/projects/${projectName}`)
    ].join('\n');
};
const environmentCreationErrorFactory = (environmentErrorMessage)=>boxen__default.default(environmentErrorMessage, {
        padding: 1,
        margin: 1,
        borderStyle: 'round',
        borderColor: 'white',
        titleAlignment: 'left'
    });

exports.environmentCreationErrorFactory = environmentCreationErrorFactory;
exports.environmentErrorMessageFactory = environmentErrorMessageFactory;
