// @strapi/plugin-cloud/dist/admin/index.mjs
import { Cloud } from '@strapi/icons';
import { Initializer } from './components/Initializer.mjs';
import { pluginId } from './pluginId.mjs';
import { prefixPluginTranslations } from './utils/prefixPluginTranslations.mjs';

function __variableDynamicImportRuntime1__(path) {
  switch (path) {
    case './translations/cs.json': return import('./translations/cs.json.mjs');
    case './translations/en.json': return import('./translations/en.json.mjs');
    case './translations/es.json': return import('./translations/es.json.mjs');
    case './translations/fr.json': return import('./translations/fr.json.mjs');
    case './translations/it.json': return import('./translations/it.json.mjs');
    case './translations/ja.json': return import('./translations/ja.json.mjs');
    case './translations/ko.json': return import('./translations/ko.json.mjs');
    case './translations/nl.json': return import('./translations/nl.json.mjs');
    case './translations/pl.json': return import('./translations/pl.json.mjs');
    case './translations/ru.json': return import('./translations/ru.json.mjs');
    case './translations/sk.json': return import('./translations/sk.json.mjs');
    case './translations/tr.json': return import('./translations/tr.json.mjs');
    case './translations/uk.json': return import('./translations/uk.json.mjs');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }
const pluginName = 'Deploy';
// eslint-disable-next-line import/no-default-export
var index = {
    register (app) {
        app.addMenuLink({
            to: `plugins/${pluginId}`,
            icon: Cloud,
            intlLabel: {
                id: `${pluginId}.Plugin.name`,
                defaultMessage: pluginName
            },
            Component: ()=>import('./pages/App.mjs').then((mod)=>({
                        default: mod.App
                    })),
            permissions: []
        });
        app.registerPlugin({
            id: pluginId,
            initializer: Initializer,
            isReady: false,
            name: pluginName
        });
    },
    async registerTrads ({ locales }) {
        const importedTrads = await Promise.all(locales.map((locale)=>{
            return __variableDynamicImportRuntime1__(`./translations/${locale}.json`).then(({ default: data })=>{
                return {
                    data: prefixPluginTranslations(data, pluginId),
                    locale
                };
            }).catch(()=>{
                return {
                    data: {},
                    locale
                };
            });
        }));
        return Promise.resolve(importedTrads);
    }
};

export { index as default };
