// @strapi/database/dist/transaction-context.mjs
import { AsyncLocalStorage } from 'node:async_hooks';

/**
 * After the user calls the Knex transactor’s `commit` or `rollback` (e.g. via
 * the callback object returned from `Database#transaction`, or after the
 * container’s promise has settled in Knex), a second finalisation is invalid
 * and can throw (e.g. "Transaction query already complete"). Knex exposes
 * `isCompleted()` for this; optional for mocks that omit it.
 */ const isTransactorComplete = (trx)=>{
    const t = trx;
    return typeof t.isCompleted === 'function' && t.isCompleted();
};
const storage = new AsyncLocalStorage();
const transactionCtx = {
    async run (trx, cb) {
        const store = storage.getStore();
        return storage.run({
            trx,
            // Fill with existing callbacks if nesting transactions
            commitCallbacks: store?.commitCallbacks || [],
            rollbackCallbacks: store?.rollbackCallbacks || []
        }, cb);
    },
    get () {
        const store = storage.getStore();
        return store?.trx;
    },
    async commit (trx) {
        const store = storage.getStore();
        if (isTransactorComplete(trx)) {
            if (store?.trx) {
                store.trx = null;
            }
            return;
        }
        // Clear transaction from store
        if (store?.trx) {
            store.trx = null;
        }
        // Commit transaction
        await trx.commit();
        if (!store?.commitCallbacks.length) {
            return;
        }
        // Run callbacks
        store.commitCallbacks.forEach((cb)=>cb());
        store.commitCallbacks = [];
    },
    async rollback (trx) {
        const store = storage.getStore();
        if (isTransactorComplete(trx)) {
            if (store?.trx) {
                store.trx = null;
            }
            return;
        }
        // Clear transaction from store
        if (store?.trx) {
            store.trx = null;
        }
        // Rollback transaction
        await trx.rollback();
        if (!store?.rollbackCallbacks.length) {
            return;
        }
        // Run callbacks
        store.rollbackCallbacks.forEach((cb)=>cb());
        store.rollbackCallbacks = [];
    },
    onCommit (cb) {
        const store = storage.getStore();
        if (store?.commitCallbacks) {
            store.commitCallbacks.push(cb);
        }
    },
    onRollback (cb) {
        const store = storage.getStore();
        if (store?.rollbackCallbacks) {
            store.rollbackCallbacks.push(cb);
        }
    }
};

export { transactionCtx };
