// @strapi/typescript-utils/dist/generators/utils.mjs
import path from 'node:path';
import assert from 'node:assert';
import * as ts from 'typescript';
import fse from 'fs-extra';
import chalk from 'chalk';

const { factory } = ts;
const MODULE_DECLARATION = '@strapi/strapi';
const PUBLIC_NAMESPACE = 'Public';
/**
 * Aggregate the given TypeScript nodes into a single string
 */ const emitDefinitions = (definitions)=>{
    const nodeArray = factory.createNodeArray(definitions);
    const sourceFile = ts.createSourceFile('placeholder.ts', '', ts.ScriptTarget.ESNext, true, ts.ScriptKind.TS);
    const printer = ts.createPrinter({
        omitTrailingSemicolon: true
    });
    return printer.printList(ts.ListFormat.MultiLine, nodeArray, sourceFile);
};
/**
 * Save the given string representation of TS nodes in a file
 * If the given directory doesn't exist, it'll be created automatically
 *
 * @return The path of the created file
 */ const saveDefinitionToFileSystem = async (dir, file, content)=>{
    const filepath = path.join(dir, file);
    fse.ensureDirSync(dir);
    await fse.writeFile(filepath, content);
    return filepath;
};
/**
 * Format the given definitions.
 * Uses the existing config if one is defined in the project.
 */ const format = async (content)=>{
    // eslint-disable-next-line node/no-unsupported-features/es-syntax
    const prettier = await import('prettier'); // ESM-only
    const configFile = await prettier.resolveConfigFile();
    const config = configFile ? await prettier.resolveConfig(configFile) : {
        singleQuote: true,
        useTabs: false,
        tabWidth: 2
    };
    Object.assign(config, {
        parser: 'typescript'
    });
    return prettier.format(content, config);
};
/**
 * Generate the extension block for a shared component from strapi/strapi
 */ const generateSharedExtensionDefinition = (registry, definitions)=>{
    const properties = definitions.map(({ uid, definition })=>factory.createPropertySignature(undefined, factory.createStringLiteral(uid, true), undefined, factory.createTypeReferenceNode(factory.createIdentifier(definition.name.escapedText))));
    return factory.createModuleDeclaration([
        factory.createModifier(ts.SyntaxKind.DeclareKeyword)
    ], factory.createStringLiteral(MODULE_DECLARATION, true), factory.createModuleBlock([
        factory.createModuleDeclaration([
            factory.createModifier(ts.SyntaxKind.ExportKeyword)
        ], factory.createIdentifier(PUBLIC_NAMESPACE), factory.createModuleBlock(properties.length > 0 ? [
            factory.createInterfaceDeclaration([
                factory.createModifier(ts.SyntaxKind.ExportKeyword)
            ], factory.createIdentifier(registry), undefined, undefined, properties)
        ] : []), ts.NodeFlags.Namespace)
    ]), ts.NodeFlags.ExportContext);
};
const createLogger = (options = {})=>{
    const { silent = false, debug = false } = options;
    const state = {
        errors: 0,
        warning: 0
    };
    return {
        get warnings () {
            return state.warning;
        },
        get errors () {
            return state.errors;
        },
        debug (...args) {
            if (silent || !debug) {
                return;
            }
            console.log(chalk.cyan(`[DEBUG]\t[${new Date().toISOString()}] (Typegen)`), ...args);
        },
        info (...args) {
            if (silent) {
                return;
            }
            console.info(chalk.blue(`[INFO]\t[${new Date().toISOString()}] (Typegen)`), ...args);
        },
        warn (...args) {
            state.warning += 1;
            if (silent) {
                return;
            }
            console.warn(chalk.yellow(`[WARN]\t[${new Date().toISOString()}] (Typegen)`), ...args);
        },
        error (...args) {
            state.errors += 1;
            if (silent) {
                return;
            }
            console.error(chalk.red(`[ERROR]\t[${new Date().toISOString()}] (Typegen)`), ...args);
        }
    };
};
const timer = ()=>{
    const state = {
        start: null,
        end: null
    };
    return {
        start () {
            assert(state.start === null, 'The timer has already been started');
            assert(state.end === null, 'The timer has already been ended');
            state.start = Date.now();
            return this;
        },
        end () {
            assert(state.start !== null, 'The timer needs to be started before ending it');
            assert(state.end === null, 'The timer has already been ended');
            state.end = Date.now();
            return this;
        },
        get duration () {
            assert(state.start !== null, 'The timer has not been started');
            return ((state.end ?? Date.now) - state.start) / 1000;
        }
    };
};

export { createLogger, emitDefinitions, format, generateSharedExtensionDefinition, saveDefinitionToFileSystem, timer };
