// @strapi/typescript-utils/dist/generators/utils.js
'use strict';

var path = require('node:path');
var assert = require('node:assert');
var ts = require('typescript');
var fse = require('fs-extra');
var chalk = require('chalk');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var path__default = /*#__PURE__*/_interopDefault(path);
var assert__default = /*#__PURE__*/_interopDefault(assert);
var ts__namespace = /*#__PURE__*/_interopNamespace(ts);
var fse__default = /*#__PURE__*/_interopDefault(fse);
var chalk__default = /*#__PURE__*/_interopDefault(chalk);

const { factory } = ts__namespace;
const MODULE_DECLARATION = '@strapi/strapi';
const PUBLIC_NAMESPACE = 'Public';
/**
 * Aggregate the given TypeScript nodes into a single string
 */ const emitDefinitions = (definitions)=>{
    const nodeArray = factory.createNodeArray(definitions);
    const sourceFile = ts__namespace.createSourceFile('placeholder.ts', '', ts__namespace.ScriptTarget.ESNext, true, ts__namespace.ScriptKind.TS);
    const printer = ts__namespace.createPrinter({
        omitTrailingSemicolon: true
    });
    return printer.printList(ts__namespace.ListFormat.MultiLine, nodeArray, sourceFile);
};
/**
 * Save the given string representation of TS nodes in a file
 * If the given directory doesn't exist, it'll be created automatically
 *
 * @return The path of the created file
 */ const saveDefinitionToFileSystem = async (dir, file, content)=>{
    const filepath = path__default.default.join(dir, file);
    fse__default.default.ensureDirSync(dir);
    await fse__default.default.writeFile(filepath, content);
    return filepath;
};
/**
 * Format the given definitions.
 * Uses the existing config if one is defined in the project.
 */ const format = async (content)=>{
    // eslint-disable-next-line node/no-unsupported-features/es-syntax
    const prettier = await import('prettier'); // ESM-only
    const configFile = await prettier.resolveConfigFile();
    const config = configFile ? await prettier.resolveConfig(configFile) : {
        singleQuote: true,
        useTabs: false,
        tabWidth: 2
    };
    Object.assign(config, {
        parser: 'typescript'
    });
    return prettier.format(content, config);
};
/**
 * Generate the extension block for a shared component from strapi/strapi
 */ const generateSharedExtensionDefinition = (registry, definitions)=>{
    const properties = definitions.map(({ uid, definition })=>factory.createPropertySignature(undefined, factory.createStringLiteral(uid, true), undefined, factory.createTypeReferenceNode(factory.createIdentifier(definition.name.escapedText))));
    return factory.createModuleDeclaration([
        factory.createModifier(ts__namespace.SyntaxKind.DeclareKeyword)
    ], factory.createStringLiteral(MODULE_DECLARATION, true), factory.createModuleBlock([
        factory.createModuleDeclaration([
            factory.createModifier(ts__namespace.SyntaxKind.ExportKeyword)
        ], factory.createIdentifier(PUBLIC_NAMESPACE), factory.createModuleBlock(properties.length > 0 ? [
            factory.createInterfaceDeclaration([
                factory.createModifier(ts__namespace.SyntaxKind.ExportKeyword)
            ], factory.createIdentifier(registry), undefined, undefined, properties)
        ] : []), ts__namespace.NodeFlags.Namespace)
    ]), ts__namespace.NodeFlags.ExportContext);
};
const createLogger = (options = {})=>{
    const { silent = false, debug = false } = options;
    const state = {
        errors: 0,
        warning: 0
    };
    return {
        get warnings () {
            return state.warning;
        },
        get errors () {
            return state.errors;
        },
        debug (...args) {
            if (silent || !debug) {
                return;
            }
            console.log(chalk__default.default.cyan(`[DEBUG]\t[${new Date().toISOString()}] (Typegen)`), ...args);
        },
        info (...args) {
            if (silent) {
                return;
            }
            console.info(chalk__default.default.blue(`[INFO]\t[${new Date().toISOString()}] (Typegen)`), ...args);
        },
        warn (...args) {
            state.warning += 1;
            if (silent) {
                return;
            }
            console.warn(chalk__default.default.yellow(`[WARN]\t[${new Date().toISOString()}] (Typegen)`), ...args);
        },
        error (...args) {
            state.errors += 1;
            if (silent) {
                return;
            }
            console.error(chalk__default.default.red(`[ERROR]\t[${new Date().toISOString()}] (Typegen)`), ...args);
        }
    };
};
const timer = ()=>{
    const state = {
        start: null,
        end: null
    };
    return {
        start () {
            assert__default.default(state.start === null, 'The timer has already been started');
            assert__default.default(state.end === null, 'The timer has already been ended');
            state.start = Date.now();
            return this;
        },
        end () {
            assert__default.default(state.start !== null, 'The timer needs to be started before ending it');
            assert__default.default(state.end === null, 'The timer has already been ended');
            state.end = Date.now();
            return this;
        },
        get duration () {
            assert__default.default(state.start !== null, 'The timer has not been started');
            return ((state.end ?? Date.now) - state.start) / 1000;
        }
    };
};

exports.createLogger = createLogger;
exports.emitDefinitions = emitDefinitions;
exports.format = format;
exports.generateSharedExtensionDefinition = generateSharedExtensionDefinition;
exports.saveDefinitionToFileSystem = saveDefinitionToFileSystem;
exports.timer = timer;
