// @strapi/openapi/dist/assemblers/document/path/path-item/operation/parameters.js
'use strict';

var debug$1 = require('../../../../../utils/debug.js');
var zod = require('../../../../../utils/zod.js');
var queryParamStyles = require('../../../query-param-styles.js');

const debug = debug$1.createDebugger('assembler:parameters');
class OperationParametersAssembler {
    assemble(context, route) {
        debug('assembling parameters for %o %o...', route.method, route.path);
        const pathParameters = this._getPathParameters(route);
        debug('found %o path parameters', pathParameters.length);
        const queryParameters = this._getQueryParameters(route);
        debug('found %o query parameters', queryParameters.length);
        const parameters = [
            ...pathParameters,
            ...queryParameters
        ];
        debug('assembled %o parameters for %o %o', parameters.length, route.method, route.path);
        context.output.data.parameters = parameters;
    }
    _getPathParameters(route) {
        const { params } = route.request ?? {};
        // TODO: Allow auto inference (from path) if enabled through configuration
        if (!params) {
            return [];
        }
        const pathParams = [];
        for (const [name, zodSchema] of Object.entries(params)){
            const required = !zodSchema.isOptional();
            const schema = zod.zodToOpenAPI(zodSchema);
            pathParams.push({
                name,
                in: 'path',
                required,
                schema
            });
        }
        return pathParams;
    }
    _getQueryParameters(route) {
        const { query } = route.request ?? {};
        if (!query) {
            return [];
        }
        const queryParams = [];
        for (const [name, zodSchema] of Object.entries(query)){
            const required = !zodSchema.isOptional();
            const schema = zod.zodToOpenAPI(zodSchema);
            const resolvedSchema = this._resolveSchema(schema);
            if (queryParamStyles.hasExpandableObjectProperties(resolvedSchema)) {
                for (const [propName, propSchema] of Object.entries(resolvedSchema.properties)){
                    const isRequired = resolvedSchema.required?.includes(propName) ?? false;
                    const param = {
                        name: `${name}[${propName}]`,
                        in: 'query',
                        required: isRequired,
                        schema: propSchema
                    };
                    Object.assign(param, {
                        'x-strapi-serialize': 'querystring'
                    });
                    queryParams.push(param);
                }
            } else if (queryParamStyles.shouldUseDeepObjectStyle(name, resolvedSchema)) {
                const param = {
                    name,
                    in: 'query',
                    required,
                    schema: resolvedSchema,
                    style: 'deepObject',
                    explode: true
                };
                Object.assign(param, {
                    'x-strapi-serialize': 'querystring'
                });
                queryParams.push(param);
            } else {
                const param = {
                    name,
                    in: 'query',
                    required,
                    schema
                };
                Object.assign(param, {
                    'x-strapi-serialize': 'querystring'
                });
                queryParams.push(param);
            }
        }
        return queryParams;
    }
    /**
   * Resolve composite OpenAPI schemas (allOf/anyOf/oneOf) into a flat object shape
   * so nested Strapi query params can be emitted as bracket notation (e.g. pagination[page]).
   */ _resolveSchema(schema) {
        if (schema.type === 'object' && schema.properties) {
            return schema;
        }
        if (schema.allOf && Array.isArray(schema.allOf)) {
            const mergedSchema = {
                type: 'object',
                properties: {},
                required: []
            };
            for (const subSchema of schema.allOf){
                if (!('$ref' in subSchema)) {
                    const hasAnyOfOneOf = Boolean(subSchema.anyOf || subSchema.oneOf);
                    const resolved = this._resolveSchema(subSchema);
                    if (resolved.type === 'object' && resolved.properties) {
                        Object.assign(mergedSchema.properties, resolved.properties);
                        if (resolved.required && !hasAnyOfOneOf) {
                            mergedSchema.required = [
                                ...mergedSchema.required ?? [],
                                ...resolved.required
                            ];
                        }
                    }
                }
            }
            return queryParamStyles.hasExpandableObjectProperties(mergedSchema) ? mergedSchema : schema;
        }
        if (schema.anyOf && Array.isArray(schema.anyOf)) {
            return this._extractCommonProperties(schema.anyOf, false);
        }
        if (schema.oneOf && Array.isArray(schema.oneOf)) {
            return this._extractCommonProperties(schema.oneOf, false);
        }
        return schema;
    }
    _extractCommonProperties(schemas, preserveRequired = true) {
        const objectSchemas = schemas.filter((s)=>!('$ref' in s)).map((s)=>this._resolveSchema(s)).filter((s)=>s.type === 'object' && s.properties);
        if (objectSchemas.length === 0) {
            return schemas[0];
        }
        const allProperties = {};
        let allRequired = [];
        for (const objSchema of objectSchemas){
            Object.assign(allProperties, objSchema.properties);
            if (preserveRequired && objSchema.required) {
                allRequired.push(...objSchema.required);
            }
        }
        if (!preserveRequired) {
            allRequired = [];
        }
        if (Object.keys(allProperties).length === 0) {
            return schemas[0];
        }
        return {
            type: 'object',
            properties: allProperties,
            required: allRequired
        };
    }
}

exports.OperationParametersAssembler = OperationParametersAssembler;
