// @strapi/core/dist/package.json.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var name = "@strapi/core";
var version = "5.50.2";
var description = "Core of Strapi";
var homepage = "https://strapi.io";
var bugs = {
    url: "https://github.com/strapi/strapi/issues"
};
var repository = {
    type: "git",
    url: "https://github.com/strapi/strapi.git",
    directory: "packages/core/core"
};
var license = "SEE LICENSE IN LICENSE";
var author = {
    name: "Strapi Solutions SAS",
    email: "hi@strapi.io",
    url: "https://strapi.io"
};
var maintainers = [
    {
        name: "Strapi Solutions SAS",
        email: "hi@strapi.io",
        url: "https://strapi.io"
    }
];
var main = "./dist/index.js";
var module$1 = "./dist/index.mjs";
var source = "./src/index.ts";
var types = "./dist/index.d.ts";
var exports$1 = {
    ".": {
        types: "./dist/index.d.ts",
        source: "./src/index.ts",
        "import": "./dist/index.mjs",
        require: "./dist/index.js",
        "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
};
var files = [
    "dist/",
    "resources/"
];
var scripts = {
    build: "run -T npm-run-all clean --parallel build:code build:types",
    "build:code": "run -T  rollup -c",
    "build:types": "run -T tsc -p tsconfig.build.json --emitDeclarationOnly",
    watch: "run -T rollup -c -w",
    clean: "run -T rimraf ./dist",
    lint: "run -T eslint . --max-warnings=0",
    prepublishOnly: "yarn clean && yarn build",
    "test:ts": "run -T tsc -p tsconfig.json",
    "test:unit": "run -T jest",
    "test:unit:watch": "run -T jest --watch",
    "test:unit:vitest": "vitest run",
    "test:unit:vitest:watch": "vitest --watch"
};
var dependencies = {
    "@casl/ability": "6.7.5",
    "@koa/cors": "5.0.0",
    "@koa/router": "12.0.2",
    "@modelcontextprotocol/sdk": "1.29.0",
    "@paralleldrive/cuid2": "2.2.2",
    "@strapi/admin": "5.50.2",
    "@strapi/database": "5.50.2",
    "@strapi/generators": "5.50.2",
    "@strapi/logger": "5.50.2",
    "@strapi/openapi": "5.50.2",
    "@strapi/permissions": "5.50.2",
    "@strapi/types": "5.50.2",
    "@strapi/typescript-utils": "5.50.2",
    "@strapi/utils": "5.50.2",
    "@vercel/stega": "0.1.2",
    bcryptjs: "2.4.3",
    boxen: "5.1.2",
    chalk: "4.1.2",
    "ci-info": "4.0.0",
    "cli-table3": "0.6.5",
    commander: "8.3.0",
    configstore: "5.0.1",
    debug: "4.3.4",
    delegates: "1.0.0",
    dotenv: "16.6.1",
    execa: "5.1.1",
    "fs-extra": "11.3.4",
    glob: "13.0.6",
    "global-agent": "4.1.3",
    helmet: "6.2.0",
    "http-errors": "2.0.0",
    inquirer: "9.3.8",
    "is-docker": "2.2.1",
    "json-logic-js": "2.0.5",
    jsonwebtoken: "9.0.3",
    koa: "2.16.4",
    "koa-body": "6.0.1",
    "koa-compose": "4.1.0",
    "koa-compress": "5.1.1",
    "koa-favicon": "2.1.0",
    "koa-helmet": "7.1.0",
    "koa-ip": "^2.1.3",
    "koa-session": "7.0.2",
    "koa-static": "5.0.0",
    lodash: "4.18.1",
    "mime-types": "2.1.35",
    "node-schedule": "2.1.1",
    open: "8.4.2",
    ora: "5.4.1",
    "package-json": "10.0.1",
    "pkg-up": "3.1.0",
    qs: "6.15.3",
    "resolve.exports": "2.0.2",
    semver: "7.7.4",
    typescript: "5.9.3",
    undici: "6.27.0",
    yup: "0.32.9",
    zod: "3.25.67"
};
var devDependencies = {
    "@swc/core": "1.13.5",
    "@types/bcryptjs": "2.4.3",
    "@types/configstore": "5.0.1",
    "@types/delegates": "1.0.3",
    "@types/fs-extra": "11.0.4",
    "@types/global-agent": "3.0.0",
    "@types/http-errors": "2.0.4",
    "@types/jest": "29.5.2",
    "@types/json-logic-js": "2.0.8",
    "@types/koa": "2.15.2",
    "@types/koa-compress": "4.0.3",
    "@types/koa-static": "4.0.2",
    "@types/koa__router": "12.0.0",
    "@types/lodash": "^4.14.191",
    "@types/mime-types": "2.1.1",
    "@types/node": "20.19.41",
    "@types/node-schedule": "2.1.7",
    "eslint-config-custom": "5.50.2",
    jest: "29.6.0",
    supertest: "7.2.2",
    tsconfig: "5.50.2",
    vitest: "catalog:",
    "vitest-config": "5.50.2"
};
var engines = {
    node: ">=20.0.0 <=26.x.x",
    npm: ">=6.0.0"
};
var pkg = {
    name: name,
    version: version,
    description: description,
    homepage: homepage,
    bugs: bugs,
    repository: repository,
    license: license,
    author: author,
    maintainers: maintainers,
    main: main,
    module: module$1,
    source: source,
    types: types,
    exports: exports$1,
    files: files,
    scripts: scripts,
    dependencies: dependencies,
    devDependencies: devDependencies,
    engines: engines
};

exports.author = author;
exports.bugs = bugs;
exports.default = pkg;
exports.dependencies = dependencies;
exports.description = description;
exports.devDependencies = devDependencies;
exports.engines = engines;
exports.exports = exports$1;
exports.files = files;
exports.homepage = homepage;
exports.license = license;
exports.main = main;
exports.maintainers = maintainers;
exports.module = module$1;
exports.name = name;
exports.repository = repository;
exports.scripts = scripts;
exports.source = source;
exports.types = types;
exports.version = version;
