// @strapi/admin/dist/admin/admin/src/core/apis/Widgets.js
'use strict';

var invariant = require('invariant');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var invariant__default = /*#__PURE__*/_interopDefault(invariant);

class Widgets {
    register(widgets) {
        if (Array.isArray(widgets)) {
            this.checkWidgets(widgets);
            this.widgets = [
                ...this.widgets,
                ...widgets
            ];
        } else if (typeof widgets === 'function') {
            const newWidgets = widgets(this.widgets);
            this.checkWidgets(newWidgets);
            this.widgets = newWidgets;
        } else if (typeof widgets === 'object') {
            this.checkWidgets([
                widgets
            ]);
            this.widgets.push(widgets);
        } else {
            throw new Error('Expected widgets to be an array or a reducer function');
        }
    }
    constructor(){
        this.generateUid = (widget)=>{
            return widget.pluginId ? `plugin::${widget.pluginId}.${widget.id}` : `global::${widget.id}`;
        };
        this.checkWidgets = (widgets)=>{
            widgets.forEach((widget)=>{
                invariant__default.default(widget.id, 'An id must be provided');
                invariant__default.default(widget.component, 'A component must be provided');
                invariant__default.default(widget.title, 'A title must be provided');
                invariant__default.default(widget.icon, 'An icon must be provided');
            });
        };
        this.getAll = ()=>{
            return this.widgets.map((widget)=>{
                const { id: _id, pluginId: _pluginId, ...widgetBase } = widget;
                return {
                    ...widgetBase,
                    uid: this.generateUid(widget)
                };
            });
        };
        this.widgets = [];
    }
}

exports.Widgets = Widgets;
