// @strapi/content-type-builder/dist/admin/components/CustomRadioGroup/CustomRadioGroup.mjs
import { jsxs, jsx } from 'react/jsx-runtime';
import { Flex, Typography, Box } from '@strapi/design-system';
import { useIntl } from 'react-intl';
import { Wrapper } from './Styles.mjs';

const CustomRadioGroup = ({ intlLabel, name, onChange, radios = [], value })=>{
    const { formatMessage } = useIntl();
    return /*#__PURE__*/ jsxs(Flex, {
        direction: "column",
        alignItems: "stretch",
        gap: 2,
        children: [
            /*#__PURE__*/ jsx(Typography, {
                variant: "pi",
                fontWeight: "bold",
                textColor: "neutral800",
                htmlFor: name,
                tag: "label",
                children: formatMessage(intlLabel)
            }),
            /*#__PURE__*/ jsx(Wrapper, {
                gap: 4,
                alignItems: "stretch",
                children: radios.map((radio)=>{
                    const radioValue = radio.value.toString();
                    return /*#__PURE__*/ jsxs("label", {
                        htmlFor: radioValue,
                        className: "container",
                        children: [
                            /*#__PURE__*/ jsx("input", {
                                id: radioValue,
                                name: name,
                                className: "option-input",
                                checked: radio.value === value,
                                value: radioValue,
                                onChange: onChange,
                                type: "radio"
                            }, radioValue),
                            /*#__PURE__*/ jsx(Box, {
                                className: "option",
                                padding: 4,
                                children: /*#__PURE__*/ jsxs(Flex, {
                                    children: [
                                        /*#__PURE__*/ jsx(Box, {
                                            paddingRight: 4,
                                            children: /*#__PURE__*/ jsx("span", {
                                                className: "checkmark"
                                            })
                                        }),
                                        /*#__PURE__*/ jsxs(Flex, {
                                            direction: "column",
                                            alignItems: "stretch",
                                            gap: 2,
                                            children: [
                                                /*#__PURE__*/ jsx(Typography, {
                                                    fontWeight: "bold",
                                                    children: formatMessage(radio.title)
                                                }),
                                                /*#__PURE__*/ jsx(Typography, {
                                                    variant: "pi",
                                                    textColor: "neutral600",
                                                    children: formatMessage(radio.description)
                                                })
                                            ]
                                        })
                                    ]
                                })
                            })
                        ]
                    }, radioValue);
                })
            })
        ]
    });
};

export { CustomRadioGroup };
