// @strapi/admin/dist/admin/admin/src/pages/SessionsPage.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var reactIntl = require('react-intl');
var reactRouterDom = require('react-router-dom');
var ConfirmDialog = require('../components/ConfirmDialog.js');
var Layout = require('../components/Layouts/Layout.js');
var PageHelpers = require('../components/PageHelpers.js');
var RelativeTime = require('../components/RelativeTime.js');
var Auth = require('../features/Auth.js');
var Notifications = require('../features/Notifications.js');
var useAPIErrorHandler = require('../hooks/useAPIErrorHandler.js');
var auth = require('../services/auth.js');
var baseQuery = require('../utils/baseQuery.js');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

/* -------------------------------------------------------------------------------------------------
 * SessionsPage
 * -----------------------------------------------------------------------------------------------*/ const SessionsPage = ()=>{
    const { formatMessage } = reactIntl.useIntl();
    const navigate = reactRouterDom.useNavigate();
    const { toggleNotification } = Notifications.useNotification();
    const { _unstableFormatAPIError: formatAPIError } = useAPIErrorHandler.useAPIErrorHandler();
    const logout = Auth.useAuth('SessionsPage', (state)=>state.logout);
    const { data: sessions = [], isLoading, error } = auth.useGetActiveSessionsQuery();
    const [revokeSession] = auth.useRevokeSessionMutation();
    const [revokeAllSessions] = auth.useRevokeAllSessionsMutation();
    const handleLogoutAndRedirect = React__namespace.useCallback(async ()=>{
        await logout();
        navigate('/auth/login');
    }, [
        logout,
        navigate
    ]);
    const notifyError = React__namespace.useCallback((error)=>{
        toggleNotification({
            type: 'danger',
            message: error && baseQuery.isBaseQueryError(error) ? formatAPIError(error) : formatMessage({
                id: 'notification.error',
                defaultMessage: 'An error occurred'
            })
        });
    }, [
        formatAPIError,
        formatMessage,
        toggleNotification
    ]);
    const runSessionMutation = React__namespace.useCallback(async (action, onSuccess)=>{
        try {
            const res = await action();
            if (res && 'error' in res && res.error) {
                notifyError(res.error);
                return;
            }
            await onSuccess();
        } catch  {
            notifyError();
        }
    }, [
        notifyError
    ]);
    const handleRevoke = async (session)=>{
        await runSessionMutation(()=>revokeSession(session.id), async ()=>{
            if (session.current) {
                await handleLogoutAndRedirect();
                return;
            }
            toggleNotification({
                type: 'success',
                message: formatMessage({
                    id: 'Settings.sessions.revoke.success',
                    defaultMessage: 'Session ended'
                })
            });
        });
    };
    const handleRevokeOthers = async ()=>{
        await runSessionMutation(()=>revokeAllSessions({
                keepCurrent: true
            }), ()=>{
            toggleNotification({
                type: 'success',
                message: formatMessage({
                    id: 'Settings.sessions.revokeOthers.success',
                    defaultMessage: 'Other sessions ended'
                })
            });
        });
    };
    const handleRevokeAll = async ()=>{
        await runSessionMutation(()=>revokeAllSessions(undefined), handleLogoutAndRedirect);
    };
    if (isLoading) {
        return /*#__PURE__*/ jsxRuntime.jsx(PageHelpers.Page.Loading, {});
    }
    if (error) {
        return /*#__PURE__*/ jsxRuntime.jsx(PageHelpers.Page.Error, {});
    }
    return /*#__PURE__*/ jsxRuntime.jsxs(PageHelpers.Page.Main, {
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(PageHelpers.Page.Title, {
                children: formatMessage({
                    id: 'Settings.sessions.title',
                    defaultMessage: 'Active Devices'
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsx(Layout.Layouts.Header, {
                title: formatMessage({
                    id: 'Settings.sessions.title',
                    defaultMessage: 'Active Devices'
                }),
                subtitle: formatMessage({
                    id: 'Settings.sessions.subtitle',
                    defaultMessage: 'These are the devices and sessions currently signed in to your account.'
                }),
                primaryAction: sessions.length > 0 ? /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                    gap: 2,
                    children: [
                        sessions.length > 1 ? /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Dialog.Root, {
                            children: [
                                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Dialog.Trigger, {
                                    children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                                        variant: "secondary",
                                        startIcon: /*#__PURE__*/ jsxRuntime.jsx(icons.SignOut, {}),
                                        children: formatMessage({
                                            id: 'Settings.sessions.revokeOthers',
                                            defaultMessage: 'Log out of other devices'
                                        })
                                    })
                                }),
                                /*#__PURE__*/ jsxRuntime.jsx(ConfirmDialog.ConfirmDialog, {
                                    onConfirm: handleRevokeOthers,
                                    children: formatMessage({
                                        id: 'Settings.sessions.revokeOthers.confirm',
                                        defaultMessage: 'Are you sure? This will end every other active session. You will stay signed in on this device.'
                                    })
                                })
                            ]
                        }) : null,
                        /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Dialog.Root, {
                            children: [
                                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Dialog.Trigger, {
                                    children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                                        variant: "danger-light",
                                        startIcon: /*#__PURE__*/ jsxRuntime.jsx(icons.SignOut, {}),
                                        children: formatMessage({
                                            id: 'Settings.sessions.revokeAll',
                                            defaultMessage: 'Log out of all devices'
                                        })
                                    })
                                }),
                                /*#__PURE__*/ jsxRuntime.jsx(ConfirmDialog.ConfirmDialog, {
                                    onConfirm: handleRevokeAll,
                                    children: formatMessage({
                                        id: 'Settings.sessions.revokeAll.confirm',
                                        defaultMessage: 'Are you sure? This will end every active session, including this one, and you will need to log in again.'
                                    })
                                })
                            ]
                        })
                    ]
                }) : undefined
            }),
            /*#__PURE__*/ jsxRuntime.jsx(Layout.Layouts.Content, {
                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                    background: "neutral0",
                    hasRadius: true,
                    shadow: "filterShadow",
                    children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Table, {
                        colCount: 4,
                        rowCount: sessions.length,
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Thead, {
                                children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Tr, {
                                    children: [
                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Th, {
                                            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                variant: "sigma",
                                                children: formatMessage({
                                                    id: 'Settings.sessions.device',
                                                    defaultMessage: 'Device'
                                                })
                                            })
                                        }),
                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Th, {
                                            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                variant: "sigma",
                                                children: formatMessage({
                                                    id: 'Settings.sessions.loginAt',
                                                    defaultMessage: 'Signed in'
                                                })
                                            })
                                        }),
                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Th, {
                                            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                variant: "sigma",
                                                children: formatMessage({
                                                    id: 'Settings.sessions.lastActiveAt',
                                                    defaultMessage: 'Last used'
                                                })
                                            })
                                        }),
                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Th, {
                                            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                variant: "sigma",
                                                children: formatMessage({
                                                    id: 'Settings.sessions.actions',
                                                    defaultMessage: 'Actions'
                                                })
                                            })
                                        })
                                    ]
                                })
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Tbody, {
                                children: sessions.map((session)=>/*#__PURE__*/ jsxRuntime.jsxs(designSystem.Tr, {
                                        children: [
                                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Td, {
                                                children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                                    gap: 2,
                                                    children: [
                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                            textColor: "neutral800",
                                                            children: session.deviceName ?? (session.deviceId ? session.deviceId.slice(0, 8) : formatMessage({
                                                                id: 'Settings.sessions.unknownDevice',
                                                                defaultMessage: 'Unknown device'
                                                            }))
                                                        }),
                                                        session.current && /*#__PURE__*/ jsxRuntime.jsx(designSystem.Badge, {
                                                            children: formatMessage({
                                                                id: 'Settings.sessions.current',
                                                                defaultMessage: 'This device'
                                                            })
                                                        })
                                                    ]
                                                })
                                            }),
                                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Td, {
                                                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                    textColor: "neutral800",
                                                    children: session.loginAt ? /*#__PURE__*/ jsxRuntime.jsx(RelativeTime.RelativeTime, {
                                                        timestamp: new Date(session.loginAt)
                                                    }) : '-'
                                                })
                                            }),
                                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Td, {
                                                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                    textColor: "neutral800",
                                                    children: session.lastActiveAt ? /*#__PURE__*/ jsxRuntime.jsx(RelativeTime.RelativeTime, {
                                                        timestamp: new Date(session.lastActiveAt)
                                                    }) : '-'
                                                })
                                            }),
                                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Td, {
                                                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                                                    justifyContent: "end",
                                                    onClick: (e)=>e.stopPropagation(),
                                                    children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Dialog.Root, {
                                                        children: [
                                                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Dialog.Trigger, {
                                                                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                                                                    label: formatMessage({
                                                                        id: 'Settings.sessions.revoke',
                                                                        defaultMessage: 'End session'
                                                                    }),
                                                                    variant: "ghost",
                                                                    children: /*#__PURE__*/ jsxRuntime.jsx(icons.SignOut, {})
                                                                })
                                                            }),
                                                            /*#__PURE__*/ jsxRuntime.jsx(ConfirmDialog.ConfirmDialog, {
                                                                onConfirm: ()=>handleRevoke(session),
                                                                children: session.current ? formatMessage({
                                                                    id: 'Settings.sessions.revoke.confirmCurrent',
                                                                    defaultMessage: 'This is the device you are currently using. Ending this session will log you out.'
                                                                }) : formatMessage({
                                                                    id: 'Settings.sessions.revoke.confirm',
                                                                    defaultMessage: 'Are you sure you want to end this session? The device will need to log in again.'
                                                                })
                                                            })
                                                        ]
                                                    })
                                                })
                                            })
                                        ]
                                    }, session.id))
                            })
                        ]
                    })
                })
            })
        ]
    });
};

exports.SessionsPage = SessionsPage;
