// @strapi/admin/dist/admin/admin/src/pages/SessionsPage.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import * as React from 'react';
import { Flex, Dialog, Button, Box, Table, Thead, Tr, Th, Typography, Tbody, Td, Badge, IconButton } from '@strapi/design-system';
import { SignOut } from '@strapi/icons';
import { useIntl } from 'react-intl';
import { useNavigate } from 'react-router-dom';
import { ConfirmDialog } from '../components/ConfirmDialog.mjs';
import { Layouts } from '../components/Layouts/Layout.mjs';
import { Page } from '../components/PageHelpers.mjs';
import { RelativeTime } from '../components/RelativeTime.mjs';
import { useAuth } from '../features/Auth.mjs';
import { useNotification } from '../features/Notifications.mjs';
import { useAPIErrorHandler } from '../hooks/useAPIErrorHandler.mjs';
import { useGetActiveSessionsQuery, useRevokeSessionMutation, useRevokeAllSessionsMutation } from '../services/auth.mjs';
import { isBaseQueryError } from '../utils/baseQuery.mjs';

/* -------------------------------------------------------------------------------------------------
 * SessionsPage
 * -----------------------------------------------------------------------------------------------*/ const SessionsPage = ()=>{
    const { formatMessage } = useIntl();
    const navigate = useNavigate();
    const { toggleNotification } = useNotification();
    const { _unstableFormatAPIError: formatAPIError } = useAPIErrorHandler();
    const logout = useAuth('SessionsPage', (state)=>state.logout);
    const { data: sessions = [], isLoading, error } = useGetActiveSessionsQuery();
    const [revokeSession] = useRevokeSessionMutation();
    const [revokeAllSessions] = useRevokeAllSessionsMutation();
    const handleLogoutAndRedirect = React.useCallback(async ()=>{
        await logout();
        navigate('/auth/login');
    }, [
        logout,
        navigate
    ]);
    const notifyError = React.useCallback((error)=>{
        toggleNotification({
            type: 'danger',
            message: error && isBaseQueryError(error) ? formatAPIError(error) : formatMessage({
                id: 'notification.error',
                defaultMessage: 'An error occurred'
            })
        });
    }, [
        formatAPIError,
        formatMessage,
        toggleNotification
    ]);
    const runSessionMutation = React.useCallback(async (action, onSuccess)=>{
        try {
            const res = await action();
            if (res && 'error' in res && res.error) {
                notifyError(res.error);
                return;
            }
            await onSuccess();
        } catch  {
            notifyError();
        }
    }, [
        notifyError
    ]);
    const handleRevoke = async (session)=>{
        await runSessionMutation(()=>revokeSession(session.id), async ()=>{
            if (session.current) {
                await handleLogoutAndRedirect();
                return;
            }
            toggleNotification({
                type: 'success',
                message: formatMessage({
                    id: 'Settings.sessions.revoke.success',
                    defaultMessage: 'Session ended'
                })
            });
        });
    };
    const handleRevokeOthers = async ()=>{
        await runSessionMutation(()=>revokeAllSessions({
                keepCurrent: true
            }), ()=>{
            toggleNotification({
                type: 'success',
                message: formatMessage({
                    id: 'Settings.sessions.revokeOthers.success',
                    defaultMessage: 'Other sessions ended'
                })
            });
        });
    };
    const handleRevokeAll = async ()=>{
        await runSessionMutation(()=>revokeAllSessions(undefined), handleLogoutAndRedirect);
    };
    if (isLoading) {
        return /*#__PURE__*/ jsx(Page.Loading, {});
    }
    if (error) {
        return /*#__PURE__*/ jsx(Page.Error, {});
    }
    return /*#__PURE__*/ jsxs(Page.Main, {
        children: [
            /*#__PURE__*/ jsx(Page.Title, {
                children: formatMessage({
                    id: 'Settings.sessions.title',
                    defaultMessage: 'Active Devices'
                })
            }),
            /*#__PURE__*/ jsx(Layouts.Header, {
                title: formatMessage({
                    id: 'Settings.sessions.title',
                    defaultMessage: 'Active Devices'
                }),
                subtitle: formatMessage({
                    id: 'Settings.sessions.subtitle',
                    defaultMessage: 'These are the devices and sessions currently signed in to your account.'
                }),
                primaryAction: sessions.length > 0 ? /*#__PURE__*/ jsxs(Flex, {
                    gap: 2,
                    children: [
                        sessions.length > 1 ? /*#__PURE__*/ jsxs(Dialog.Root, {
                            children: [
                                /*#__PURE__*/ jsx(Dialog.Trigger, {
                                    children: /*#__PURE__*/ jsx(Button, {
                                        variant: "secondary",
                                        startIcon: /*#__PURE__*/ jsx(SignOut, {}),
                                        children: formatMessage({
                                            id: 'Settings.sessions.revokeOthers',
                                            defaultMessage: 'Log out of other devices'
                                        })
                                    })
                                }),
                                /*#__PURE__*/ jsx(ConfirmDialog, {
                                    onConfirm: handleRevokeOthers,
                                    children: formatMessage({
                                        id: 'Settings.sessions.revokeOthers.confirm',
                                        defaultMessage: 'Are you sure? This will end every other active session. You will stay signed in on this device.'
                                    })
                                })
                            ]
                        }) : null,
                        /*#__PURE__*/ jsxs(Dialog.Root, {
                            children: [
                                /*#__PURE__*/ jsx(Dialog.Trigger, {
                                    children: /*#__PURE__*/ jsx(Button, {
                                        variant: "danger-light",
                                        startIcon: /*#__PURE__*/ jsx(SignOut, {}),
                                        children: formatMessage({
                                            id: 'Settings.sessions.revokeAll',
                                            defaultMessage: 'Log out of all devices'
                                        })
                                    })
                                }),
                                /*#__PURE__*/ jsx(ConfirmDialog, {
                                    onConfirm: handleRevokeAll,
                                    children: formatMessage({
                                        id: 'Settings.sessions.revokeAll.confirm',
                                        defaultMessage: 'Are you sure? This will end every active session, including this one, and you will need to log in again.'
                                    })
                                })
                            ]
                        })
                    ]
                }) : undefined
            }),
            /*#__PURE__*/ jsx(Layouts.Content, {
                children: /*#__PURE__*/ jsx(Box, {
                    background: "neutral0",
                    hasRadius: true,
                    shadow: "filterShadow",
                    children: /*#__PURE__*/ jsxs(Table, {
                        colCount: 4,
                        rowCount: sessions.length,
                        children: [
                            /*#__PURE__*/ jsx(Thead, {
                                children: /*#__PURE__*/ jsxs(Tr, {
                                    children: [
                                        /*#__PURE__*/ jsx(Th, {
                                            children: /*#__PURE__*/ jsx(Typography, {
                                                variant: "sigma",
                                                children: formatMessage({
                                                    id: 'Settings.sessions.device',
                                                    defaultMessage: 'Device'
                                                })
                                            })
                                        }),
                                        /*#__PURE__*/ jsx(Th, {
                                            children: /*#__PURE__*/ jsx(Typography, {
                                                variant: "sigma",
                                                children: formatMessage({
                                                    id: 'Settings.sessions.loginAt',
                                                    defaultMessage: 'Signed in'
                                                })
                                            })
                                        }),
                                        /*#__PURE__*/ jsx(Th, {
                                            children: /*#__PURE__*/ jsx(Typography, {
                                                variant: "sigma",
                                                children: formatMessage({
                                                    id: 'Settings.sessions.lastActiveAt',
                                                    defaultMessage: 'Last used'
                                                })
                                            })
                                        }),
                                        /*#__PURE__*/ jsx(Th, {
                                            children: /*#__PURE__*/ jsx(Typography, {
                                                variant: "sigma",
                                                children: formatMessage({
                                                    id: 'Settings.sessions.actions',
                                                    defaultMessage: 'Actions'
                                                })
                                            })
                                        })
                                    ]
                                })
                            }),
                            /*#__PURE__*/ jsx(Tbody, {
                                children: sessions.map((session)=>/*#__PURE__*/ jsxs(Tr, {
                                        children: [
                                            /*#__PURE__*/ jsx(Td, {
                                                children: /*#__PURE__*/ jsxs(Flex, {
                                                    gap: 2,
                                                    children: [
                                                        /*#__PURE__*/ jsx(Typography, {
                                                            textColor: "neutral800",
                                                            children: session.deviceName ?? (session.deviceId ? session.deviceId.slice(0, 8) : formatMessage({
                                                                id: 'Settings.sessions.unknownDevice',
                                                                defaultMessage: 'Unknown device'
                                                            }))
                                                        }),
                                                        session.current && /*#__PURE__*/ jsx(Badge, {
                                                            children: formatMessage({
                                                                id: 'Settings.sessions.current',
                                                                defaultMessage: 'This device'
                                                            })
                                                        })
                                                    ]
                                                })
                                            }),
                                            /*#__PURE__*/ jsx(Td, {
                                                children: /*#__PURE__*/ jsx(Typography, {
                                                    textColor: "neutral800",
                                                    children: session.loginAt ? /*#__PURE__*/ jsx(RelativeTime, {
                                                        timestamp: new Date(session.loginAt)
                                                    }) : '-'
                                                })
                                            }),
                                            /*#__PURE__*/ jsx(Td, {
                                                children: /*#__PURE__*/ jsx(Typography, {
                                                    textColor: "neutral800",
                                                    children: session.lastActiveAt ? /*#__PURE__*/ jsx(RelativeTime, {
                                                        timestamp: new Date(session.lastActiveAt)
                                                    }) : '-'
                                                })
                                            }),
                                            /*#__PURE__*/ jsx(Td, {
                                                children: /*#__PURE__*/ jsx(Flex, {
                                                    justifyContent: "end",
                                                    onClick: (e)=>e.stopPropagation(),
                                                    children: /*#__PURE__*/ jsxs(Dialog.Root, {
                                                        children: [
                                                            /*#__PURE__*/ jsx(Dialog.Trigger, {
                                                                children: /*#__PURE__*/ jsx(IconButton, {
                                                                    label: formatMessage({
                                                                        id: 'Settings.sessions.revoke',
                                                                        defaultMessage: 'End session'
                                                                    }),
                                                                    variant: "ghost",
                                                                    children: /*#__PURE__*/ jsx(SignOut, {})
                                                                })
                                                            }),
                                                            /*#__PURE__*/ jsx(ConfirmDialog, {
                                                                onConfirm: ()=>handleRevoke(session),
                                                                children: session.current ? formatMessage({
                                                                    id: 'Settings.sessions.revoke.confirmCurrent',
                                                                    defaultMessage: 'This is the device you are currently using. Ending this session will log you out.'
                                                                }) : formatMessage({
                                                                    id: 'Settings.sessions.revoke.confirm',
                                                                    defaultMessage: 'Are you sure you want to end this session? The device will need to log in again.'
                                                                })
                                                            })
                                                        ]
                                                    })
                                                })
                                            })
                                        ]
                                    }, session.id))
                            })
                        ]
                    })
                })
            })
        ]
    });
};

export { SessionsPage };
