// @strapi/content-manager/dist/admin/pages/EditView/components/DocumentActions.mjs
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import * as React from 'react';
import { useNotification, useAPIErrorHandler, useQueryParams, useForm, useGuidedTour, GUIDED_TOUR_REQUIRED_ACTIONS, useIsDesktop, tours } from '@strapi/admin/strapi-admin';
import { Button, Dialog, Modal, Flex, Typography, Radio, Tooltip, Menu, IconButton } from '@strapi/design-system';
import { WarningCircle, Cross, More } from '@strapi/icons';
import mapValues from 'lodash/fp/mapValues';
import get from 'lodash/get';
import merge from 'lodash/merge';
import set from 'lodash/set';
import { useIntl } from 'react-intl';
import { useNavigate, useMatch, useParams } from 'react-router-dom';
import { PUBLISHED_AT_ATTRIBUTE_NAME } from '../../../constants/attributes.mjs';
import { SINGLE_TYPES } from '../../../constants/collections.mjs';
import { useDocumentRBAC } from '../../../features/DocumentRBAC.mjs';
import { useDocument, useDoc } from '../../../hooks/useDocument.mjs';
import { useDocumentActions } from '../../../hooks/useDocumentActions.mjs';
import { useDocumentContext } from '../../../hooks/useDocumentContext.mjs';
import { usePreviewContext } from '../../../preview/pages/Preview.mjs';
import { LIST_PATH, CLONE_PATH } from '../../../router.mjs';
import { useGetDraftRelationCountQuery as useLazyGetDraftRelationCountQuery, useUpdateDocumentMutation } from '../../../services/documents.mjs';
import { buildValidParams, isBaseQueryError } from '../../../utils/api.mjs';
import { getTranslation } from '../../../utils/translations.mjs';
import { handleInvisibleAttributes } from '../utils/data.mjs';
import { EMPTY_DRAFT_RELATION_COUNTS, countLocalDraftRelations, normalizeDraftRelationCounts, resolveDraftRelationCounts, getDraftRelationsPublishState } from '../utils/draftRelationCounts.mjs';
import { getEditViewShortcut } from '../utils/keyboardShortcuts.mjs';
import { useRelationModal } from './FormInputs/Relations/RelationModal.mjs';

const publishConfirmDialogOpeners = new Map();
const openPublishConfirmDialog = (scope)=>{
    publishConfirmDialogOpeners.get(scope)?.();
};
const connectRelationToParent = (parentDataToUpdate, fieldToConnect, data, fieldToConnectUID)=>{
    /*
   * Check if the fieldToConnect is already present in the parentDataToUpdate.
   * This happens in particular when in the parentDocument you have created
   * a new component without saving.
   */ const isFieldPresent = !!get(parentDataToUpdate, fieldToConnect);
    const fieldToConnectPath = isFieldPresent ? fieldToConnect : fieldToConnect.split('.').slice(0, -1).join('.');
    const fieldToConnectValue = isFieldPresent ? {
        connect: [
            {
                id: data.documentId,
                documentId: data.documentId,
                locale: data.locale
            }
        ]
    } : {
        [fieldToConnect.split('.').pop()]: {
            connect: [
                {
                    id: data.documentId,
                    documentId: data.documentId,
                    locale: data.locale
                }
            ],
            disconnect: []
        },
        // In case the object was not present you need to pass the componentUID of the parent document
        __component: fieldToConnectUID
    };
    const objectToConnect = set({}, fieldToConnectPath, fieldToConnectValue);
    return merge(parentDataToUpdate, objectToConnect);
};
const DocumentActions = ({ actions })=>{
    const { formatMessage } = useIntl();
    const isDesktop = useIsDesktop();
    const [primaryAction, secondaryAction, ...restActions] = actions.filter((action)=>{
        if (action.position === undefined) {
            return true;
        }
        const positions = Array.isArray(action.position) ? action.position : [
            action.position
        ];
        return positions.includes('panel');
    });
    if (!primaryAction) {
        return null;
    }
    const addHintTooltip = (action, children)=>{
        if (action.disabled) {
            return children;
        }
        const hint = action.type === 'publish' ? formatMessage({
            id: 'content-manager.containers.EditView.publishHint',
            defaultMessage: 'Ctrl / Cmd + Shift + Enter to publish'
        }) : formatMessage({
            id: 'content-manager.containers.EditView.saveHint',
            defaultMessage: 'Ctrl / Cmd + Enter to save'
        });
        return /*#__PURE__*/ jsx(Tooltip, {
            label: hint,
            children: /*#__PURE__*/ jsx(Flex, {
                width: "100%",
                children: children
            })
        });
    };
    const primaryActionContent = /*#__PURE__*/ jsxs(Fragment, {
        children: [
            /*#__PURE__*/ jsx(Flex, {
                flex: 1,
                alignItems: "stretch",
                direction: "column",
                children: addHintTooltip(primaryAction, /*#__PURE__*/ jsx(DocumentActionButton, {
                    ...primaryAction,
                    variant: primaryAction.variant || 'default',
                    buttonType: primaryAction.type === 'publish' ? undefined : 'submit'
                }))
            }),
            restActions.length > 0 ? /*#__PURE__*/ jsx(DocumentActionsMenu, {
                actions: restActions,
                label: formatMessage({
                    id: 'content-manager.containers.edit.panels.default.more-actions',
                    defaultMessage: 'More document actions'
                })
            }) : null
        ]
    });
    return /*#__PURE__*/ jsxs(Flex, {
        direction: {
            initial: 'row',
            large: 'column'
        },
        gap: 2,
        alignItems: "stretch",
        width: "100%",
        children: [
            /*#__PURE__*/ jsx(tours.contentManager.Publish, {
                children: isDesktop ? /*#__PURE__*/ jsx(Flex, {
                    gap: 2,
                    children: primaryActionContent
                }) : primaryActionContent
            }),
            secondaryAction ? /*#__PURE__*/ jsx(Flex, {
                flex: 1,
                order: {
                    initial: -1,
                    large: 0
                },
                alignItems: "stretch",
                direction: "column",
                children: secondaryAction.type === 'publish' ? /*#__PURE__*/ jsx(tours.contentManager.Publish, {
                    children: addHintTooltip(secondaryAction, /*#__PURE__*/ jsx(DocumentActionButton, {
                        ...secondaryAction,
                        variant: secondaryAction.variant || 'secondary'
                    }))
                }) : addHintTooltip(secondaryAction, /*#__PURE__*/ jsx(DocumentActionButton, {
                    ...secondaryAction,
                    variant: secondaryAction.variant || 'secondary',
                    buttonType: "submit"
                }))
            }) : null
        ]
    });
};
const DocumentActionButton = ({ buttonType = 'button', ...action })=>{
    const [dialogId, setDialogId] = React.useState(null);
    const { toggleNotification } = useNotification();
    React.useEffect(()=>{
        const scope = action.publishConfirmScope;
        if (action.type !== 'publish' || !action.dialog || !scope) {
            return;
        }
        const open = ()=>setDialogId(action.id);
        publishConfirmDialogOpeners.set(scope, open);
        return ()=>{
            if (publishConfirmDialogOpeners.get(scope) === open) {
                publishConfirmDialogOpeners.delete(scope);
            }
        };
    }, [
        action.type,
        action.dialog,
        action.id,
        action.publishConfirmScope
    ]);
    const handleClick = (action)=>async (e)=>{
            const { onClick = ()=>false, dialog, id } = action;
            const muteDialog = await onClick(e);
            if (dialog && !muteDialog) {
                switch(dialog.type){
                    case 'notification':
                        toggleNotification({
                            title: dialog.title,
                            message: dialog.content,
                            type: dialog.status,
                            timeout: dialog.timeout,
                            onClose: dialog.onClose
                        });
                        break;
                    case 'dialog':
                    case 'modal':
                        e.preventDefault();
                        setDialogId(id);
                }
            }
        };
    const handleClose = ()=>{
        setDialogId(null);
    };
    return /*#__PURE__*/ jsxs(Fragment, {
        children: [
            /*#__PURE__*/ jsx(Button, {
                flex: "auto",
                startIcon: action.icon,
                disabled: action.disabled,
                onClick: handleClick(action),
                justifyContent: "center",
                variant: action.variant || 'default',
                paddingTop: "7px",
                paddingBottom: "7px",
                loading: action.loading,
                type: buttonType,
                children: action.label
            }),
            action.dialog?.type === 'dialog' ? /*#__PURE__*/ jsx(DocumentActionConfirmDialog, {
                ...action.dialog,
                variant: action.dialog?.variant ?? action.variant,
                isOpen: dialogId === action.id,
                onClose: handleClose
            }) : null,
            action.dialog?.type === 'modal' ? /*#__PURE__*/ jsx(DocumentActionModal, {
                ...action.dialog,
                onModalClose: handleClose,
                isOpen: dialogId === action.id
            }) : null
        ]
    });
};
const DocumentActionsMenu = ({ actions, children, label, variant = 'tertiary' })=>{
    const [isOpen, setIsOpen] = React.useState(false);
    const [dialogId, setDialogId] = React.useState(null);
    const { formatMessage } = useIntl();
    const { toggleNotification } = useNotification();
    const isDisabled = actions.every((action)=>action.disabled) || actions.length === 0;
    const triggerRef = React.useRef(null);
    const handleClick = (action)=>async (e)=>{
            const { onClick = ()=>false, dialog, id } = action;
            const muteDialog = await onClick(e);
            if (dialog && !muteDialog) {
                switch(dialog.type){
                    case 'notification':
                        toggleNotification({
                            title: dialog.title,
                            message: dialog.content,
                            type: dialog.status,
                            timeout: dialog.timeout,
                            onClose: dialog.onClose
                        });
                        break;
                    case 'dialog':
                    case 'modal':
                        setDialogId(id);
                }
            }
        };
    const handleClose = ()=>{
        setDialogId(null);
        setIsOpen(false);
    };
    const handleOpenChange = (open)=>{
        if (!isDisabled) {
            setIsOpen(open);
        }
    };
    return /*#__PURE__*/ jsxs(Menu.Root, {
        open: isOpen,
        onOpenChange: handleOpenChange,
        children: [
            /*#__PURE__*/ jsx(Menu.Trigger, {
                ref: triggerRef,
                disabled: isDisabled,
                label: label || formatMessage({
                    id: 'content-manager.containers.edit.panels.default.more-actions',
                    defaultMessage: 'More document actions'
                }),
                tag: IconButton,
                icon: /*#__PURE__*/ jsx(More, {}),
                variant: variant
            }),
            /*#__PURE__*/ jsxs(Menu.Content, {
                maxHeight: undefined,
                popoverPlacement: "bottom-end",
                maxWidth: "25rem",
                children: [
                    actions.map((action)=>{
                        return /*#__PURE__*/ jsx(Menu.Item, {
                            disabled: action.disabled,
                            /* @ts-expect-error – TODO: this is an error in the DS where it is most likely a synthetic event, not regular. */ onSelect: handleClick(action),
                            display: "block",
                            variant: action.variant === 'danger' ? action.variant : 'default',
                            startIcon: action.icon,
                            children: /*#__PURE__*/ jsx(Flex, {
                                justifyContent: "space-between",
                                gap: 4,
                                children: /*#__PURE__*/ jsx(Flex, {
                                    gap: 2,
                                    tag: "span",
                                    children: action.label
                                })
                            })
                        }, action.id);
                    }),
                    children
                ]
            }),
            actions.map((action)=>{
                return /*#__PURE__*/ jsxs(React.Fragment, {
                    children: [
                        action.dialog?.type === 'dialog' ? /*#__PURE__*/ jsx(DocumentActionConfirmDialog, {
                            ...action.dialog,
                            variant: action.variant,
                            isOpen: dialogId === action.id,
                            onClose: handleClose
                        }) : null,
                        action.dialog?.type === 'modal' ? /*#__PURE__*/ jsx(DocumentActionModal, {
                            ...action.dialog,
                            onModalClose: handleClose,
                            isOpen: dialogId === action.id
                        }) : null
                    ]
                }, action.id);
            })
        ]
    });
};
const DocumentActionConfirmDialog = ({ onClose, onCancel, onConfirm, title, content, bodyIcon, confirmLabel, isOpen, variant = 'secondary', loading })=>{
    const { formatMessage } = useIntl();
    const handleClose = async ()=>{
        if (onCancel) {
            await onCancel();
        }
        onClose();
    };
    const handleConfirm = async ()=>{
        if (onConfirm) {
            await onConfirm();
        }
        onClose();
    };
    const dialogBody = bodyIcon ? /*#__PURE__*/ jsxs(Flex, {
        direction: "column",
        alignItems: "stretch",
        gap: 2,
        children: [
            /*#__PURE__*/ jsx(Flex, {
                justifyContent: "center",
                children: /*#__PURE__*/ jsx(WarningCircle, {
                    width: "24px",
                    height: "24px",
                    fill: bodyIcon === 'danger' ? 'danger600' : 'primary600'
                })
            }),
            /*#__PURE__*/ jsx(Typography, {
                id: "confirm-description",
                tag: "p",
                variant: "omega",
                textAlign: "center",
                children: content
            })
        ]
    }) : content;
    return /*#__PURE__*/ jsx(Dialog.Root, {
        open: isOpen,
        onOpenChange: handleClose,
        children: /*#__PURE__*/ jsxs(Dialog.Content, {
            children: [
                /*#__PURE__*/ jsx(Dialog.Header, {
                    children: title
                }),
                /*#__PURE__*/ jsx(Dialog.Body, {
                    children: dialogBody
                }),
                /*#__PURE__*/ jsxs(Dialog.Footer, {
                    children: [
                        /*#__PURE__*/ jsx(Dialog.Cancel, {
                            children: /*#__PURE__*/ jsx(Button, {
                                variant: "tertiary",
                                fullWidth: true,
                                children: formatMessage({
                                    id: 'app.components.Button.cancel',
                                    defaultMessage: 'Cancel'
                                })
                            })
                        }),
                        /*#__PURE__*/ jsx(Button, {
                            onClick: handleConfirm,
                            variant: variant,
                            fullWidth: true,
                            loading: loading,
                            children: confirmLabel ?? formatMessage({
                                id: 'app.components.Button.confirm',
                                defaultMessage: 'Confirm'
                            })
                        })
                    ]
                })
            ]
        })
    });
};
const DocumentActionModal = ({ isOpen, title, onClose, footer: Footer, content: Content, onModalClose })=>{
    const handleClose = ()=>{
        if (onClose) {
            onClose();
        }
        onModalClose();
    };
    return /*#__PURE__*/ jsx(Modal.Root, {
        open: isOpen,
        onOpenChange: handleClose,
        children: /*#__PURE__*/ jsxs(Modal.Content, {
            children: [
                /*#__PURE__*/ jsx(Modal.Header, {
                    children: /*#__PURE__*/ jsx(Modal.Title, {
                        children: title
                    })
                }),
                typeof Content === 'function' ? /*#__PURE__*/ jsx(Content, {
                    onClose: handleClose
                }) : /*#__PURE__*/ jsx(Modal.Body, {
                    children: Content
                }),
                typeof Footer === 'function' ? /*#__PURE__*/ jsx(Footer, {
                    onClose: handleClose
                }) : Footer
            ]
        })
    });
};
const transformData = (data)=>{
    if (Array.isArray(data)) {
        return data.map((value)=>transformData(value));
    }
    if (typeof data === 'object' && data !== null) {
        if ('apiData' in data) {
            return data.apiData;
        }
        return mapValues((value)=>transformData(value))(data);
    }
    return data;
};
const transformDocumentData = (data)=>transformData(data);
/* -------------------------------------------------------------------------------------------------
 * DocumentActionComponents
 * -----------------------------------------------------------------------------------------------*/ const PublishAction = ({ activeTab, documentId, model, collectionType, meta, document })=>{
    const { currentDocument: { schema } } = useDocumentContext('PublishAction');
    const navigate = useNavigate();
    const { toggleNotification } = useNotification();
    const { _unstableFormatValidationErrors: formatValidationErrors } = useAPIErrorHandler();
    const isListView = useMatch(LIST_PATH) !== null;
    const isCloning = useMatch(CLONE_PATH) !== null;
    const { id } = useParams();
    const { formatMessage } = useIntl();
    const { canPublish, canReadFields } = useDocumentRBAC('PublishAction', ({ canPublish, canReadFields })=>({
            canPublish,
            canReadFields
        }));
    const { publish, isLoading } = useDocumentActions();
    const onPreview = usePreviewContext('PublishAction', (state)=>state.onPreview, false);
    const [countDraftRelations, { isError: isErrorDraftRelations }] = useLazyGetDraftRelationCountQuery();
    const [localDraftRelationCounts, setLocalDraftRelationCounts] = React.useState(EMPTY_DRAFT_RELATION_COUNTS);
    const [serverDraftRelationCounts, setServerDraftRelationCounts] = React.useState(EMPTY_DRAFT_RELATION_COUNTS);
    const [isFetchingDraftRelations, setIsFetchingDraftRelations] = React.useState(false);
    const [{ rawQuery }] = useQueryParams();
    const modified = useForm('PublishAction', ({ modified })=>modified);
    const setSubmitting = useForm('PublishAction', ({ setSubmitting })=>setSubmitting);
    const isSubmitting = useForm('PublishAction', ({ isSubmitting })=>isSubmitting);
    const validate = useForm('PublishAction', (state)=>state.validate);
    const setErrors = useForm('PublishAction', (state)=>state.setErrors);
    const getValues = useForm('PublishAction', (state)=>state.getValues);
    const formValues = useForm('PublishAction', ({ values })=>values);
    const resetForm = useForm('PublishAction', ({ resetForm })=>resetForm);
    const { currentDocument: { components } } = useDocumentContext('PublishAction');
    // need to discriminate if the publish is coming from a relation modal or in the edit view
    const relationContext = useRelationModal('PublishAction', ()=>true, false);
    const fromRelationModal = relationContext != undefined;
    const isRelationModalOpen = useRelationModal('PublishAction', (state)=>state.state.isModalOpen, false);
    const publishConfirmScope = fromRelationModal ? 'relation-modal' : onPreview ? 'preview' : 'panel';
    const dispatch = useRelationModal('PublishAction', (state)=>state.dispatch);
    const fieldToConnect = useRelationModal('PublishAction', (state)=>state.state.fieldToConnect, false);
    const fieldToConnectUID = useRelationModal('PublishAction', (state)=>state.state.fieldToConnectUID, false);
    const documentHistory = useRelationModal('PublishAction', (state)=>state.state.documentHistory, false);
    const rootDocumentMeta = useRelationModal('PublishAction', (state)=>state.rootDocumentMeta);
    const dispatchGuidedTour = useGuidedTour('PublishAction', (s)=>s.dispatch);
    const { currentDocumentMeta } = useDocumentContext('PublishAction');
    const [updateDocumentMutation] = useUpdateDocumentMutation();
    const { _unstableFormatAPIError: formatAPIError } = useAPIErrorHandler();
    const idToPublish = currentDocumentMeta.documentId || id;
    React.useEffect(()=>{
        if (isErrorDraftRelations) {
            toggleNotification({
                type: 'danger',
                message: formatMessage({
                    id: getTranslation('error.records.fetch-draft-relatons'),
                    defaultMessage: 'An error occurred while fetching draft relations on this document.'
                })
            });
        }
    }, [
        isErrorDraftRelations,
        toggleNotification,
        formatMessage
    ]);
    React.useEffect(()=>{
        setLocalDraftRelationCounts(countLocalDraftRelations(formValues, schema, components, model));
    }, [
        components,
        formValues,
        model,
        schema
    ]);
    const fetchDraftRelationsCount = React.useCallback(async ()=>{
        if (!document?.documentId || isListView) {
            return;
        }
        setIsFetchingDraftRelations(true);
        try {
            const { data, error } = await countDraftRelations({
                collectionType,
                model,
                documentId,
                params: currentDocumentMeta.params
            });
            if (error) {
                throw error;
            }
            if (data) {
                setServerDraftRelationCounts(normalizeDraftRelationCounts(data));
            }
        } finally{
            setIsFetchingDraftRelations(false);
        }
    }, [
        collectionType,
        countDraftRelations,
        currentDocumentMeta.params,
        document?.documentId,
        documentId,
        isListView,
        model
    ]);
    React.useEffect(()=>{
        fetchDraftRelationsCount();
    }, [
        fetchDraftRelationsCount,
        document?.updatedAt
    ]);
    React.useEffect(()=>{
        const handleWindowFocus = ()=>{
            fetchDraftRelationsCount();
        };
        window.addEventListener('focus', handleWindowFocus);
        return ()=>{
            window.removeEventListener('focus', handleWindowFocus);
        };
    }, [
        fetchDraftRelationsCount
    ]);
    const parentDocumentMetaToUpdate = documentHistory?.at(-2) ?? rootDocumentMeta;
    const parentDocumentData = useDocument({
        documentId: parentDocumentMetaToUpdate?.documentId,
        model: parentDocumentMetaToUpdate?.model,
        collectionType: parentDocumentMetaToUpdate?.collectionType,
        params: parentDocumentMetaToUpdate?.params
    }, {
        skip: !parentDocumentMetaToUpdate
    });
    const { getInitialFormValues } = useDoc();
    const isDocumentPublished = (document?.[PUBLISHED_AT_ATTRIBUTE_NAME] || meta?.availableStatus.some((doc)=>doc[PUBLISHED_AT_ATTRIBUTE_NAME] !== null)) && document?.status !== 'modified';
    // eslint-disable-next-line react-hooks/exhaustive-deps
    const performPublish = async ()=>{
        setSubmitting(true);
        try {
            /**
       * Yield one microtask so React can flush any pending field updates from the same task
       * (common with fast local runners / Playwright) before we read values and validate.
       * TODO: replace with an explicit form flush contract when we can (same concern as Save flow).
       */ await Promise.resolve();
            const { errors } = await validate(true, {
                status: 'published'
            });
            if (errors) {
                const hasUnreadableRequiredField = schema && Object.keys(schema.attributes).some((fieldName)=>{
                    const attribute = schema.attributes[fieldName];
                    // For components, check if any of the component fields are readable
                    if (attribute.type === 'component') {
                        const componentFields = (canReadFields ?? []).filter((field)=>field.startsWith(`${fieldName}.`));
                        return componentFields.length === 0;
                    }
                    // For regular fields, check if the field itself is readable
                    return attribute?.required && !(canReadFields ?? []).includes(fieldName);
                });
                if (hasUnreadableRequiredField) {
                    toggleNotification({
                        type: 'danger',
                        message: formatMessage({
                            id: 'content-manager.validation.error.unreadable-required-field',
                            defaultMessage: 'Your current permissions prevent access to certain required fields. Please request access from an administrator to proceed.'
                        })
                    });
                } else {
                    toggleNotification({
                        type: 'danger',
                        message: formatMessage({
                            id: 'content-manager.validation.error',
                            defaultMessage: 'There are validation errors in your document. Please fix them before saving.'
                        })
                    });
                }
                return;
            }
            const { data } = handleInvisibleAttributes(transformDocumentData(getValues()), {
                schema,
                components
            });
            const res = await publish({
                collectionType,
                model,
                documentId,
                params: currentDocumentMeta.params
            }, data);
            // Reset form with current values as new initial values (clears errors/submitting and sets modified to false)
            if ('data' in res) {
                resetForm(getValues());
                dispatchGuidedTour({
                    type: 'set_completed_actions',
                    payload: [
                        GUIDED_TOUR_REQUIRED_ACTIONS.contentManager.createContent
                    ]
                });
            }
            if ('data' in res && collectionType !== SINGLE_TYPES) {
                /**
         * TODO: refactor the router so we can just do `../${res.data.documentId}` instead of this.
         */ if (idToPublish === 'create' && !fromRelationModal) {
                    navigate({
                        pathname: `../${collectionType}/${model}/${res.data.documentId}`,
                        search: rawQuery
                    });
                } else if (fromRelationModal) {
                    const newRelation = {
                        documentId: res.data.documentId,
                        collectionType,
                        model,
                        params: currentDocumentMeta.params
                    };
                    /*
           * Update, if needed, the parent relation with the newly published document.
           * Check if in history we have the parent relation otherwise use the
           * rootDocument
           */ if (fieldToConnect && documentHistory && (parentDocumentMetaToUpdate.documentId || parentDocumentMetaToUpdate.collectionType === SINGLE_TYPES)) {
                        const parentDataToUpdate = parentDocumentMetaToUpdate.collectionType === SINGLE_TYPES ? getInitialFormValues() : parentDocumentData.getInitialFormValues();
                        const metaDocumentToUpdate = documentHistory.at(-2) ?? rootDocumentMeta;
                        const dataToUpdate = connectRelationToParent(parentDataToUpdate, fieldToConnect, res.data, fieldToConnectUID);
                        try {
                            const updateRes = await updateDocumentMutation({
                                collectionType: metaDocumentToUpdate.collectionType,
                                model: metaDocumentToUpdate.model,
                                documentId: metaDocumentToUpdate.collectionType !== SINGLE_TYPES ? metaDocumentToUpdate.documentId : undefined,
                                params: metaDocumentToUpdate.params,
                                data: dataToUpdate
                            });
                            if ('error' in updateRes) {
                                toggleNotification({
                                    type: 'danger',
                                    message: formatAPIError(updateRes.error)
                                });
                                return;
                            }
                        } catch (err) {
                            toggleNotification({
                                type: 'danger',
                                message: formatMessage({
                                    id: 'notification.error',
                                    defaultMessage: 'An error occurred'
                                })
                            });
                            throw err;
                        }
                    }
                    dispatch({
                        type: 'GO_TO_CREATED_RELATION',
                        payload: {
                            document: newRelation,
                            shouldBypassConfirmation: true
                        }
                    });
                }
            } else if ('error' in res && isBaseQueryError(res.error) && res.error.name === 'ValidationError') {
                setErrors(formatValidationErrors(res.error));
            }
        } finally{
            setSubmitting(false);
            if (onPreview) {
                onPreview();
            }
        }
    };
    const getFreshDraftRelationCounts = React.useCallback(()=>resolveDraftRelationCounts(documentId, modified, countLocalDraftRelations(getValues(), schema, components, model), serverDraftRelationCounts), [
        components,
        documentId,
        getValues,
        model,
        modified,
        schema,
        serverDraftRelationCounts
    ]);
    const draftRelationCounts = resolveDraftRelationCounts(documentId, modified, localDraftRelationCounts, serverDraftRelationCounts);
    const { hasUnpublishedRelations, hasDraftM2mLinks, dialogVariant, bodyIcon, confirmLabel } = getDraftRelationsPublishState(draftRelationCounts);
    const supportsDraftRelationWarning = Boolean(schema?.options?.draftAndPublish);
    /**
   * Disabled when:
   *  - currently if you're cloning a document we don't support publish & clone at the same time.
   *  - the form is submitting
   *  - the active tab is the published tab
   *  - the document is already published & not modified
   *  - the document is being created & not modified
   *  - the user doesn't have the permission to publish
   */ const isDisabled = isCloning || isSubmitting || isFetchingDraftRelations || activeTab === 'published' || !modified && isDocumentPublished || !modified && !document?.documentId || !canPublish;
    // Publish on CMD+Shift+Enter (macOS) / CTRL+Shift+Enter (Windows/Linux).
    // Saving a draft (CMD/CTRL+Enter) is handled by the UpdateAction.
    React.useEffect(()=>{
        if (!schema?.options?.draftAndPublish) {
            return;
        }
        const handleKeyDown = (e)=>{
            if (getEditViewShortcut(e) !== 'publish') {
                return;
            }
            e.preventDefault();
            if (!fromRelationModal && isRelationModalOpen) {
                return;
            }
            if (isDisabled) {
                return;
            }
            if (getDraftRelationsPublishState(getFreshDraftRelationCounts()).hasDraftRelations) {
                openPublishConfirmDialog(publishConfirmScope);
                return;
            }
            performPublish();
        };
        window.addEventListener('keydown', handleKeyDown);
        return ()=>{
            window.removeEventListener('keydown', handleKeyDown);
        };
    }, [
        fromRelationModal,
        getFreshDraftRelationCounts,
        isDisabled,
        isRelationModalOpen,
        performPublish,
        publishConfirmScope,
        schema?.options?.draftAndPublish
    ]);
    if (!schema?.options?.draftAndPublish) {
        return null;
    }
    return {
        type: 'publish',
        loading: isLoading,
        position: [
            'panel',
            'preview',
            'relation-modal'
        ],
        disabled: isDisabled,
        publishConfirmScope: supportsDraftRelationWarning ? publishConfirmScope : undefined,
        label: formatMessage({
            id: 'app.utils.publish',
            defaultMessage: 'Publish'
        }),
        onClick: async ()=>{
            if (getDraftRelationsPublishState(getFreshDraftRelationCounts()).hasDraftRelations) {
                // Let DocumentActionButton open the confirmation dialog.
                return;
            }
            await performPublish();
            // Skip the registered dialog when publishing directly.
            return true;
        },
        dialog: supportsDraftRelationWarning ? {
            type: 'dialog',
            variant: dialogVariant,
            bodyIcon,
            title: formatMessage({
                id: getTranslation('popUpWarning.warning.has-draft-relations.title'),
                defaultMessage: 'Confirmation'
            }),
            content: /*#__PURE__*/ jsxs(Fragment, {
                children: [
                    hasUnpublishedRelations ? formatMessage({
                        id: getTranslation('popUpWarning.warning.has-draft-relations.message'),
                        defaultMessage: 'This entry is related to {count, plural, one {# draft entry} other {# draft entries}}. {count, plural, one {That relation will not be included in the published version.} other {Those relations will not be included in the published version.}}'
                    }, {
                        count: draftRelationCounts.unpublishedRelations
                    }) : formatMessage({
                        id: getTranslation('popUpWarning.warning.has-draft-m2m-relations.message'),
                        defaultMessage: '{count, plural, one {# linked entry is still in draft. It will appear on the live site once that entry is published.} other {# linked entries are still in draft. They will appear on the live site once those entries are published.}}'
                    }, {
                        count: draftRelationCounts.draftM2mLinks
                    }),
                    hasUnpublishedRelations && hasDraftM2mLinks ? ` ${formatMessage({
                        id: getTranslation('popUpWarning.warning.has-draft-m2m-relations.additional'),
                        defaultMessage: '{count, plural, one {# many-to-many link points} other {# many-to-many links point}} to draft entries that will become visible once published.'
                    }, {
                        count: draftRelationCounts.draftM2mLinks
                    })}` : null,
                    ' ',
                    formatMessage({
                        id: getTranslation('popUpWarning.warning.publish-question'),
                        defaultMessage: 'Do you still want to publish?'
                    })
                ]
            }),
            confirmLabel: confirmLabel === 'publish' ? formatMessage({
                id: 'app.utils.publish',
                defaultMessage: 'Publish'
            }) : formatMessage({
                id: getTranslation('popUpwarning.warning.has-draft-relations.button-confirm'),
                defaultMessage: 'Publish without relations'
            }),
            onConfirm: async ()=>{
                await performPublish();
            }
        } : undefined
    };
};
PublishAction.type = 'publish';
PublishAction.position = [
    'panel',
    'preview',
    'relation-modal'
];
const UpdateAction = ({ activeTab, documentId, model, collectionType })=>{
    const dispatchGuidedTour = useGuidedTour('UpdateAction', (s)=>s.dispatch);
    const navigate = useNavigate();
    const { toggleNotification } = useNotification();
    const { _unstableFormatValidationErrors: formatValidationErrors } = useAPIErrorHandler();
    const cloneMatch = useMatch(CLONE_PATH);
    const isCloning = cloneMatch !== null;
    const { formatMessage } = useIntl();
    const { create, update, clone, isLoading } = useDocumentActions();
    const { currentDocument: { components } } = useDocumentContext('UpdateAction');
    const [{ rawQuery }] = useQueryParams();
    const onPreview = usePreviewContext('UpdateAction', (state)=>state.onPreview, false);
    const { getInitialFormValues } = useDoc();
    const isSubmitting = useForm('UpdateAction', ({ isSubmitting })=>isSubmitting);
    const modified = useForm('UpdateAction', ({ modified })=>modified);
    const setSubmitting = useForm('UpdateAction', ({ setSubmitting })=>setSubmitting);
    const initialValues = useForm('UpdateAction', ({ initialValues })=>initialValues);
    const getValues = useForm('UpdateAction', (state)=>state.getValues);
    const validate = useForm('UpdateAction', (state)=>state.validate);
    const setErrors = useForm('UpdateAction', (state)=>state.setErrors);
    const resetForm = useForm('UpdateAction', ({ resetForm })=>resetForm);
    const dispatch = useRelationModal('UpdateAction', (state)=>state.dispatch);
    // need to discriminate if the update is coming from a relation modal or in the edit view
    const relationContext = useRelationModal('UpdateAction', ()=>true, false);
    const relationalModalSchema = useRelationModal('UpdateAction', (state)=>state.currentDocument.schema, false);
    const fieldToConnect = useRelationModal('UpdateAction', (state)=>state.state.fieldToConnect, false);
    const fieldToConnectUID = useRelationModal('PublishAction', (state)=>state.state.fieldToConnectUID, false);
    const documentHistory = useRelationModal('UpdateAction', (state)=>state.state.documentHistory, false);
    const rootDocumentMeta = useRelationModal('UpdateAction', (state)=>state.rootDocumentMeta);
    const fromRelationModal = relationContext != undefined;
    const { currentDocumentMeta } = useDocumentContext('UpdateAction');
    const [updateDocumentMutation] = useUpdateDocumentMutation();
    const { _unstableFormatAPIError: formatAPIError } = useAPIErrorHandler();
    const parentDocumentMetaToUpdate = documentHistory?.at(-2) ?? rootDocumentMeta;
    const parentDocumentData = useDocument({
        documentId: parentDocumentMetaToUpdate?.documentId,
        model: parentDocumentMetaToUpdate?.model,
        collectionType: parentDocumentMetaToUpdate?.collectionType,
        params: parentDocumentMetaToUpdate?.params
    }, {
        skip: !parentDocumentMetaToUpdate
    });
    const { schema } = useDoc();
    const suitableSchema = fromRelationModal ? relationalModalSchema : schema;
    const hasDraftAndPublished = suitableSchema?.options?.draftAndPublish ?? false;
    /**
   * Disabled when:
   * - the form is submitting
   * - the document is not modified & we're not cloning (you can save a clone entity straight away)
   * - the active tab is the published tab
   */ const isDisabled = isSubmitting || !modified && !isCloning || activeTab === 'published';
    const handleUpdate = async ()=>{
        setSubmitting(true);
        try {
            if (!modified) {
                return;
            }
            // Blur the active element so inputs that debounce into the form (e.g. blocks editor) flush
            // before validate/getValues — on fast clients Save can otherwise read stale field state.
            // Use the global DOM document — form values are not in scope here (avoid shadowing `document`).
            globalThis.document?.activeElement?.blur();
            // Yield microtasks so batched React updates after blur/onChange can settle before validate
            // (same idea as performPublish; two ticks vs one gives a bit more room after focus/blur).
            // TODO: replace with an explicit field/form flush contract when available.
            await Promise.resolve();
            await Promise.resolve();
            const { errors } = await validate(true, {
                // enforce "published" validation if not using "draft and published"
                status: !hasDraftAndPublished ? 'published' : 'draft'
            });
            if (errors) {
                toggleNotification({
                    type: 'danger',
                    message: formatMessage({
                        id: 'content-manager.validation.error',
                        defaultMessage: 'There are validation errors in your document. Please fix them before saving.'
                    })
                });
                return;
            }
            const latestValues = getValues();
            if (isCloning) {
                const res = await clone({
                    model,
                    documentId: cloneMatch.params.origin,
                    params: currentDocumentMeta.params
                }, transformDocumentData(latestValues));
                if ('data' in res) {
                    navigate({
                        pathname: `../${res.data.documentId}`,
                        search: rawQuery
                    }, {
                        relative: 'path'
                    });
                } else if ('error' in res && isBaseQueryError(res.error) && res.error.name === 'ValidationError') {
                    setErrors(formatValidationErrors(res.error));
                }
            } else if (documentId || collectionType === SINGLE_TYPES) {
                const { data } = handleInvisibleAttributes(transformDocumentData(latestValues), {
                    schema: suitableSchema,
                    initialValues,
                    components
                });
                const res = await update({
                    collectionType,
                    model,
                    documentId,
                    params: currentDocumentMeta.params
                }, data);
                if ('error' in res && isBaseQueryError(res.error) && res.error.name === 'ValidationError') {
                    setErrors(formatValidationErrors(res.error));
                } else {
                    resetForm(latestValues);
                }
            } else {
                const { data } = handleInvisibleAttributes(transformDocumentData(latestValues), {
                    schema: suitableSchema,
                    initialValues,
                    components
                });
                const res = await create({
                    model,
                    params: currentDocumentMeta.params
                }, data);
                if ('data' in res && collectionType !== SINGLE_TYPES) {
                    if (fromRelationModal) {
                        const createdRelation = {
                            documentId: res.data.documentId,
                            collectionType,
                            model,
                            params: currentDocumentMeta.params
                        };
                        /*
             * Update, if needed, the parent relation with the newly published document.
             * Check if in history we have the parent relation otherwise use the
             * rootDocument
             */ if (fieldToConnect && documentHistory && (parentDocumentMetaToUpdate.documentId || parentDocumentMetaToUpdate.collectionType === SINGLE_TYPES)) {
                            const parentDataToUpdate = parentDocumentMetaToUpdate.collectionType === SINGLE_TYPES ? getInitialFormValues() : parentDocumentData.getInitialFormValues();
                            const dataToUpdate = connectRelationToParent(parentDataToUpdate, fieldToConnect, res.data, fieldToConnectUID);
                            try {
                                const updateRes = await updateDocumentMutation({
                                    collectionType: parentDocumentMetaToUpdate.collectionType,
                                    model: parentDocumentMetaToUpdate.model,
                                    documentId: parentDocumentMetaToUpdate.collectionType !== SINGLE_TYPES ? parentDocumentMetaToUpdate.documentId : undefined,
                                    params: parentDocumentMetaToUpdate.params,
                                    data: {
                                        ...dataToUpdate
                                    }
                                });
                                if ('error' in updateRes) {
                                    toggleNotification({
                                        type: 'danger',
                                        message: formatAPIError(updateRes.error)
                                    });
                                    return;
                                }
                            } catch (err) {
                                toggleNotification({
                                    type: 'danger',
                                    message: formatMessage({
                                        id: 'notification.error',
                                        defaultMessage: 'An error occurred'
                                    })
                                });
                                throw err;
                            }
                        }
                        dispatch({
                            type: 'GO_TO_CREATED_RELATION',
                            payload: {
                                document: createdRelation,
                                shouldBypassConfirmation: true
                            }
                        });
                    } else {
                        navigate({
                            pathname: `../${res.data.documentId}`,
                            search: rawQuery
                        }, {
                            replace: true,
                            relative: 'path'
                        });
                    }
                } else if ('error' in res && isBaseQueryError(res.error) && res.error.name === 'ValidationError') {
                    setErrors(formatValidationErrors(res.error));
                }
            }
        } finally{
            dispatchGuidedTour({
                type: 'set_completed_actions',
                payload: [
                    GUIDED_TOUR_REQUIRED_ACTIONS.contentManager.createContent
                ]
            });
            setSubmitting(false);
            if (onPreview) {
                onPreview();
            }
        }
    };
    // Save a draft on CMD+Enter (macOS) / CTRL+Enter (Windows/Linux), with CMD/CTRL+S as an alias.
    // Publishing (CMD/CTRL+Shift+Enter) is handled by the PublishAction.
    // `handleUpdate` is recreated on every render, so we read the latest version (and the latest
    // disabled state) through refs and register the listener only once.
    const handleUpdateRef = React.useRef(handleUpdate);
    handleUpdateRef.current = handleUpdate;
    const isDisabledRef = React.useRef(isDisabled);
    isDisabledRef.current = isDisabled;
    React.useEffect(()=>{
        const handleKeyDown = (e)=>{
            if (getEditViewShortcut(e) !== 'save') {
                return;
            }
            e.preventDefault();
            if (!isDisabledRef.current) {
                handleUpdateRef.current();
            }
        };
        window.addEventListener('keydown', handleKeyDown);
        return ()=>{
            window.removeEventListener('keydown', handleKeyDown);
        };
    }, []);
    return {
        loading: isLoading,
        disabled: isDisabled,
        label: formatMessage({
            id: 'global.save',
            defaultMessage: 'Save'
        }),
        onClick: handleUpdate,
        position: [
            'panel',
            'preview',
            'relation-modal'
        ]
    };
};
UpdateAction.type = 'update';
UpdateAction.position = [
    'panel',
    'preview',
    'relation-modal'
];
const UNPUBLISH_DRAFT_OPTIONS = {
    KEEP: 'keep',
    DISCARD: 'discard'
};
const UnpublishAction = ({ activeTab, documentId, model, collectionType, document })=>{
    const { formatMessage } = useIntl();
    const { schema } = useDoc();
    const canPublish = useDocumentRBAC('UnpublishAction', ({ canPublish })=>canPublish);
    const { unpublish } = useDocumentActions();
    const [{ query }] = useQueryParams();
    const params = React.useMemo(()=>buildValidParams(query), [
        query
    ]);
    const { toggleNotification } = useNotification();
    const [shouldKeepDraft, setShouldKeepDraft] = React.useState(true);
    const isDocumentModified = document?.status === 'modified';
    const handleChange = (value)=>{
        setShouldKeepDraft(value === UNPUBLISH_DRAFT_OPTIONS.KEEP);
    };
    if (!schema?.options?.draftAndPublish) {
        return null;
    }
    return {
        disabled: !canPublish || activeTab === 'published' || document?.status !== 'published' && document?.status !== 'modified',
        label: formatMessage({
            id: 'app.utils.unpublish',
            defaultMessage: 'Unpublish'
        }),
        icon: /*#__PURE__*/ jsx(Cross, {}),
        onClick: async ()=>{
            /**
       * return if there's no id & we're in a collection type, or the status modified
       * for either collection type because we use a dialog to handle the process in
       * the latter case.
       */ if (!documentId && collectionType !== SINGLE_TYPES || isDocumentModified) {
                if (!documentId) {
                    // This should never, ever, happen.
                    console.error("You're trying to unpublish a document without an id, this is likely a bug with Strapi. Please open an issue.");
                    toggleNotification({
                        message: formatMessage({
                            id: 'content-manager.actions.unpublish.error',
                            defaultMessage: 'An error occurred while trying to unpublish the document.'
                        }),
                        type: 'danger'
                    });
                }
                return;
            }
            await unpublish({
                collectionType,
                model,
                documentId,
                params
            });
        },
        dialog: isDocumentModified ? {
            type: 'dialog',
            title: formatMessage({
                id: 'app.components.ConfirmDialog.title',
                defaultMessage: 'Confirmation'
            }),
            content: /*#__PURE__*/ jsxs(Flex, {
                alignItems: "flex-start",
                direction: "column",
                gap: 6,
                children: [
                    /*#__PURE__*/ jsxs(Flex, {
                        width: "100%",
                        direction: "column",
                        gap: 2,
                        children: [
                            /*#__PURE__*/ jsx(WarningCircle, {
                                width: "24px",
                                height: "24px",
                                fill: "danger600"
                            }),
                            /*#__PURE__*/ jsx(Typography, {
                                tag: "p",
                                variant: "omega",
                                textAlign: "center",
                                children: formatMessage({
                                    id: 'content-manager.actions.unpublish.dialog.body',
                                    defaultMessage: 'Are you sure?'
                                })
                            })
                        ]
                    }),
                    /*#__PURE__*/ jsxs(Radio.Group, {
                        defaultValue: UNPUBLISH_DRAFT_OPTIONS.KEEP,
                        name: "discard-options",
                        "aria-label": formatMessage({
                            id: 'content-manager.actions.unpublish.dialog.radio-label',
                            defaultMessage: 'Choose an option to unpublish the document.'
                        }),
                        onValueChange: handleChange,
                        children: [
                            /*#__PURE__*/ jsx(Radio.Item, {
                                checked: shouldKeepDraft,
                                value: UNPUBLISH_DRAFT_OPTIONS.KEEP,
                                children: formatMessage({
                                    id: 'content-manager.actions.unpublish.dialog.option.keep-draft',
                                    defaultMessage: 'Keep draft'
                                })
                            }),
                            /*#__PURE__*/ jsx(Radio.Item, {
                                checked: !shouldKeepDraft,
                                value: UNPUBLISH_DRAFT_OPTIONS.DISCARD,
                                children: formatMessage({
                                    id: 'content-manager.actions.unpublish.dialog.option.replace-draft',
                                    defaultMessage: 'Replace draft'
                                })
                            })
                        ]
                    })
                ]
            }),
            onConfirm: async ()=>{
                if (!documentId && collectionType !== SINGLE_TYPES) {
                    // This should never, ever, happen.
                    console.error("You're trying to unpublish a document without an id, this is likely a bug with Strapi. Please open an issue.");
                    toggleNotification({
                        message: formatMessage({
                            id: 'content-manager.actions.unpublish.error',
                            defaultMessage: 'An error occurred while trying to unpublish the document.'
                        }),
                        type: 'danger'
                    });
                }
                await unpublish({
                    collectionType,
                    model,
                    documentId,
                    params
                }, !shouldKeepDraft);
            }
        } : undefined,
        variant: 'danger',
        position: [
            'panel',
            'table-row'
        ]
    };
};
UnpublishAction.type = 'unpublish';
UnpublishAction.position = 'panel';
const DiscardAction = ({ activeTab, documentId, model, collectionType, document })=>{
    const { formatMessage } = useIntl();
    const { schema } = useDoc();
    const canUpdate = useDocumentRBAC('DiscardAction', ({ canUpdate })=>canUpdate);
    const { discard, isLoading } = useDocumentActions();
    const [{ query }] = useQueryParams();
    const params = React.useMemo(()=>buildValidParams(query), [
        query
    ]);
    if (!schema?.options?.draftAndPublish) {
        return null;
    }
    return {
        disabled: !canUpdate || activeTab === 'published' || document?.status !== 'modified',
        label: formatMessage({
            id: 'content-manager.actions.discard.label',
            defaultMessage: 'Discard changes'
        }),
        icon: /*#__PURE__*/ jsx(Cross, {}),
        position: [
            'panel',
            'table-row'
        ],
        variant: 'danger',
        dialog: {
            type: 'dialog',
            title: formatMessage({
                id: 'app.components.ConfirmDialog.title',
                defaultMessage: 'Confirmation'
            }),
            content: /*#__PURE__*/ jsxs(Flex, {
                direction: "column",
                gap: 2,
                children: [
                    /*#__PURE__*/ jsx(WarningCircle, {
                        width: "24px",
                        height: "24px",
                        fill: "danger600"
                    }),
                    /*#__PURE__*/ jsx(Typography, {
                        tag: "p",
                        variant: "omega",
                        textAlign: "center",
                        children: formatMessage({
                            id: 'content-manager.actions.discard.dialog.body',
                            defaultMessage: 'Are you sure?'
                        })
                    })
                ]
            }),
            loading: isLoading,
            onConfirm: async ()=>{
                await discard({
                    collectionType,
                    model,
                    documentId,
                    params
                });
            }
        }
    };
};
DiscardAction.type = 'discard';
DiscardAction.position = 'panel';
const DEFAULT_ACTIONS = [
    PublishAction,
    UpdateAction,
    UnpublishAction,
    DiscardAction
];

export { DEFAULT_ACTIONS, DocumentActionButton, DocumentActions, DocumentActionsMenu, openPublishConfirmDialog };
