// @strapi/admin/dist/server/index.js
'use strict';

var _ = require('lodash');
var bootstrap = require('./server/src/bootstrap.js');
var register = require('./server/src/register.js');
var destroy = require('./server/src/destroy.js');
var index$6 = require('./server/src/config/index.js');
var index$5 = require('./server/src/policies/index.js');
var index$4 = require('./server/src/routes/index.js');
var index$3 = require('./server/src/services/index.js');
var index$2 = require('./server/src/controllers/index.js');
var index$1 = require('./server/src/content-types/index.js');
var index = require('./server/src/middlewares/index.js');
var index$7 = require('./ee/server/src/index.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var ___default = /*#__PURE__*/_interopDefault(_);

// eslint-disable-next-line import/no-mutable-exports
let admin = {
    bootstrap,
    register,
    destroy,
    config: index$6.default,
    policies: index$5,
    routes: index$4,
    services: index$3,
    controllers: index$2,
    contentTypes: index$1,
    middlewares: index.default
};
const mergeRoutes = (a, b, key)=>{
    return ___default.default.isArray(a) && ___default.default.isArray(b) && key === 'routes' ? a.concat(b) : undefined;
};
if (strapi.EE) {
    admin = ___default.default.mergeWith({}, admin, index$7(), mergeRoutes);
}
var admin_default = admin;

module.exports = admin_default;
