// @strapi/design-system/dist/index.js
"use strict";Object.defineProperty(exports,Symbol.toStringTag,{value:"Module"});const s=require("styled-components"),X=require("react"),xr=require("@radix-ui/react-use-callback-ref"),F=require("@internationalized/date"),r=require("react/jsx-runtime"),$r=require("@radix-ui/react-tooltip"),Zn=require("@radix-ui/react-dismissable-layer"),D=require("@strapi/ui-primitives"),Jn=require("@radix-ui/react-accordion"),O=require("@strapi/icons"),eo=require("@radix-ui/react-avatar"),to=require("@radix-ui/react-dropdown-menu"),ro=require("@radix-ui/react-checkbox"),no=require("@tanstack/react-virtual"),oo=require("@radix-ui/react-scroll-area"),ao=require("@radix-ui/react-alert-dialog"),Fe=require("@floating-ui/react-dom"),so=require("@radix-ui/react-focus-guards"),io=require("@radix-ui/react-focus-scope"),lo=require("react-remove-scroll"),co=require("react-dom"),Xt=require("@codemirror/lang-json"),uo=require("@uiw/react-codemirror"),Rt=require("@codemirror/state"),bt=require("@codemirror/view"),ho=require("@radix-ui/react-dialog"),Zt=require("@internationalized/number"),fo=require("@radix-ui/react-popover"),go=require("@radix-ui/react-progress"),po=require("@radix-ui/react-radio-group"),mo=require("@radix-ui/react-switch"),bo=require("@radix-ui/react-tabs"),Je=require("lodash");function ee(e){const t=Object.create(null,{[Symbol.toStringTag]:{value:"Module"}});if(e){for(const n in e)if(n!=="default"){const o=Object.getOwnPropertyDescriptor(e,n);Object.defineProperty(t,n,o.get?o:{enumerable:!0,get:()=>e[n]})}}return t.default=e,Object.freeze(t)}const l=ee(X),xe=ee($r),xo=ee(Zn),Ue=ee(Jn),Tt=ee(eo),ce=ee(to),yr=ee(ro),ot=ee(oo),pe=ee(ao),Te=ee(ho),Le=ee(fo),wr=ee(go),It=ee(po),Cr=ee(mo),at=ee(bo),Oe=(e,t)=>{const n=l.createContext(t),o=i=>{const{children:c,...d}=i,u=l.useMemo(()=>d,Object.values(d));return r.jsx(n.Provider,{value:u,children:c})};function a(i){const c=l.useContext(n);if(c)return c;if(t!==void 0)return t;throw new Error(`\`${i}\` must be used within \`${e}\``)}return o.displayName=`${e}Provider`,[o,a]};function $o(e,t){return typeof e=="string"?!1:t in e}function et(e){return e&&typeof e=="object"&&!Array.isArray(e)}function _e(e,t,n){return e&&t&&$o(e,t)?e[t]:n}const Jt={padding:["padding-block-start","padding-inline-end","padding-block-end","padding-inline-start"],paddingTop:"padding-block-start",paddingRight:"padding-inline-end",paddingBottom:"padding-block-end",paddingLeft:"padding-inline-start",margin:["margin-block-start","margin-inline-end","margin-block-end","margin-inline-start"],marginLeft:"margin-inline-start",marginRight:"margin-inline-end",marginTop:"margin-block-start",marginBottom:"margin-block-end",borderRadius:"border-radius",borderStyle:"border-style",borderWidth:"border-width",borderColor:"border-color",fontSize:"font-size",fontWeight:"font-weight",lineHeight:"line-height",zIndex:"z-index",boxShadow:"box-shadow",pointerEvents:"pointer-events",textAlign:"text-align",textTransform:"text-transform",textDecoration:"text-decoration",flexGrow:"flex-grow",flexShrink:"flex-shrink",flexBasis:"flex-basis",minWidth:"min-width",maxWidth:"max-width",minHeight:"min-height",maxHeight:"max-height",order:"order",flexDirection:"flex-direction",flexWrap:"flex-wrap",justifyContent:"justify-content",alignItems:"align-items"},yo=e=>{const[t,n,o,a]=e,i=n??t;return[t,i,o??t,a??i]};function wo(e,t){switch(e){case"gap":case"padding":case"margin":case"paddingTop":case"paddingLeft":case"paddingRight":case"paddingBottom":case"marginTop":case"marginLeft":case"marginRight":case"marginBottom":case"left":case"right":case"top":case"bottom":case"width":case"maxWidth":case"minWidth":case"height":case"maxHeight":case"minHeight":case"borderRadius":case"borderWidth":return t.spaces;case"color":case"background":case"borderColor":return t.colors;case"fontSize":return t.fontSizes;case"fontWeight":return t.fontWeights;case"lineHeight":return t.lineHeights;case"zIndex":return t.zIndices;case"boxShadow":return t.shadows;default:return null}}const ke=(e,t)=>{const n=Object.entries(e).reduce((o,a)=>{const[i,c]=a,d=wo(i,t),u=Object.prototype.hasOwnProperty.call(Jt,i)?Jt[i]:i;return u&&(c||c===0)&&(typeof c=="object"&&!Array.isArray(c)?Object.entries(c).forEach(([h,f])=>{o[h]={...o[h],...er(u,f,d)}}):o.initial={...o.initial,...er(u,c,d)}),o},{initial:{},small:{},medium:{},large:{}});return Object.entries(n).reduce((o,[a,i])=>{if(i&&Object.keys(i).length>0){const c=Object.entries(i).reduce((d,[u,h])=>(d.push(`${u}: ${h};`),d),[]).join(`
`);a==="initial"?o.push(c):o.push(`${t.breakpoints[a]}{ ${c} }`)}return o},[]).join(`
`)},er=(e,t,n)=>{if(Array.isArray(e)&&Array.isArray(t)){const o=yo(t);return e.reduce((a,i,c)=>(a[i]=_e(n,o[c],o[c]),a),{})}else return Array.isArray(e)&&!Array.isArray(t)?e.reduce((o,a)=>(o[a]=_e(n,t,t),o),{}):!Array.isArray(e)&&!Array.isArray(t)?{[e]:_e(n,t,t)}:(console.warn("You've passed an array of values to a property that does not support it. Please check the property and value you're passing."),{})},v=l.forwardRef(({animation:e,background:t,basis:n,borderColor:o,borderRadius:a,borderStyle:i,borderWidth:c,bottom:d,color:u,cursor:h,display:f,flex:g,fontSize:p,fontWeight:w,grow:m,hasRadius:$,height:k,left:y,lineHeight:x,margin:I,marginBottom:j,marginLeft:R,marginRight:b,marginTop:S,maxHeight:z,maxWidth:E,minHeight:A,minWidth:U,order:B,overflow:N,padding:H,paddingBottom:P,paddingLeft:K,paddingRight:L,paddingTop:M,pointerEvents:Z,position:G,right:oe,shadow:Y,shrink:J,tag:_,textAlign:Q,textTransform:de,top:$e,transform:ye,transition:fe,width:Ie,zIndex:we,...Be},Ne)=>{const Ce=_||"div",V={$animation:e,$background:t,$basis:n,$borderColor:o,$borderRadius:a,$borderStyle:i,$borderWidth:c,$bottom:d,$color:u,$cursor:h,$display:f,$flex:g,$fontSize:p,$fontWeight:w,$grow:m,$hasRadius:$,$height:k,$left:y,$lineHeight:x,$margin:I,$marginBottom:j,$marginLeft:R,$marginRight:b,$marginTop:S,$maxHeight:z,$maxWidth:E,$minHeight:A,$minWidth:U,$order:B,$overflow:N,$padding:H,$paddingBottom:P,$paddingLeft:K,$paddingRight:L,$paddingTop:M,$pointerEvents:Z,$position:G,$right:oe,$shadow:Y,$shrink:J,$textAlign:Q,$textTransform:de,$top:$e,$transform:ye,$transition:fe,$width:Ie,$zIndex:we};return r.jsx(Co,{as:Ce,ref:Ne,...V,...Be})}),Co=s.styled.div`
  ${({theme:e,...t})=>ke({animation:t.$animation,background:t.$background,borderColor:t.$borderColor,borderRadius:t.$hasRadius?e.borderRadius:t.$borderRadius,borderStyle:t.$borderColor&&!t.$borderStyle?"solid":t.$borderStyle,borderWidth:t.$borderColor&&!t.$borderWidth?"1px":t.$borderWidth,bottom:t.$bottom,boxShadow:t.$shadow,color:t.$color,cursor:t.$cursor,display:t.$display,flex:t.$flex,flexBasis:t.$basis,flexGrow:t.$grow,flexShrink:t.$shrink,fontSize:t.$fontSize,fontWeight:t.$fontWeight,height:t.$height,left:t.$left,lineHeight:t.$lineHeight,margin:t.$margin,marginBottom:t.$marginBottom,marginLeft:t.$marginLeft,marginRight:t.$marginRight,marginTop:t.$marginTop,maxHeight:t.$maxHeight,maxWidth:t.$maxWidth,minHeight:t.$minHeight,minWidth:t.$minWidth,order:t.$order,overflow:t.$overflow,padding:t.$padding,paddingBottom:t.$paddingBottom,paddingLeft:t.$paddingLeft,paddingRight:t.$paddingRight,paddingTop:t.$paddingTop,pointerEvents:t.$pointerEvents,position:t.$position,right:t.$right,textAlign:t.$textAlign,textTransform:t.$textTransform,top:t.$top,transform:t.$transform,transition:t.$transition,width:t.$width,zIndex:t.$zIndex},e)};
`,ae=l.forwardRef,C=ae((e,t)=>{const{className:n,alignItems:o,direction:a,inline:i,gap:c,justifyContent:d,wrap:u,...h}=e,f={$alignItems:o,$direction:a,$gap:c,$justifyContent:d,$wrap:u,$inline:i};return r.jsx(jo,{className:n,ref:t,...f,...h})}),jo=s.styled(v)`
  ${({theme:e,$display:t="flex",$alignItems:n="center",$direction:o="row",...a})=>ke({gap:a.$gap,alignItems:n,justifyContent:a.$justifyContent,flexWrap:a.$wrap,flexDirection:o,display:a.$inline?"inline-flex":t},e)};
`,vo="alpha",ko="beta",So="delta",Ro="epsilon",tr="omega",To="pi",Io="sigma",jr=s.css`
  display: block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`,Do=({$variant:e=tr,theme:t,$fontSize:n,$lineHeight:o})=>{const a=n&&ke({fontSize:n},t),i=o&&ke({lineHeight:o},t);switch(e){case vo:return s.css`
        font-weight: ${t.fontWeights.bold};

        /* -------------------------------------------------------------
         * Font Size
         * -------------------------------------------------------------*/
        ${a||s.css`
          /* Mobile: 2.8rem */
          font-size: ${t.fontSizes[6]};

          ${t.breakpoints.medium} {
            /* Tablet + Desktop: 3.2rem */
            font-size: ${t.fontSizes[7]};
          }
        `}

        /* -------------------------------------------------------------
         * Line Height
         * -------------------------------------------------------------*/
        ${i||s.css`
          /* Mobile: 3.2rem */
          line-height: ${t.lineHeights[0]};

          ${t.breakpoints.medium} {
            /* Tablet + Desktop: 4rem */
            line-height: ${t.lineHeights[2]};
          }
        `}
      `;case ko:return s.css`
        font-weight: ${t.fontWeights.bold};

        /* -------------------------------------------------------------
         * Font Size
         * -------------------------------------------------------------*/
        ${a||s.css`
          /* Mobile: 2rem */
          font-size: ${t.fontSizes[5]};

          ${t.breakpoints.medium} {
            /* Tablet + Desktop: 1.8rem */
            font-size: ${t.fontSizes[4]};
          }
        `}

        /* -------------------------------------------------------------
         * Line Height
         * -------------------------------------------------------------*/
        ${i||s.css`
          /* Mobile: 2.4rem */
          line-height: ${t.lineHeights[1]};
        `}
      `;case So:return s.css`
        font-weight: ${t.fontWeights.semiBold};

        /* -------------------------------------------------------------
         * Font Size
         * -------------------------------------------------------------*/
        ${a||s.css`
          /* Mobile: 1.8rem */
          font-size: ${t.fontSizes[4]};

          ${t.breakpoints.medium} {
            /* Tablet + Desktop: 1.6rem */
            font-size: ${t.fontSizes[3]};
          }
        `}

        /* -------------------------------------------------------------
         * Line Height
         * -------------------------------------------------------------*/
        ${i||s.css`
          /* Mobile: 2.4rem */
          line-height: ${t.lineHeights[3]};

          ${t.breakpoints.medium} {
            /* Tablet + Desktop: 2rem */
            line-height: ${t.lineHeights[2]};
          }
        `}
      `;case Ro:return s.css`
        /* -------------------------------------------------------------
         * Font Size
         * -------------------------------------------------------------*/
        ${a||s.css`
          /* Mobile: 1.8rem */
          font-size: ${t.fontSizes[4]};

          ${t.breakpoints.medium} {
            /* Tablet + Desktop: 1.6rem */
            font-size: ${t.fontSizes[3]};
          }
        `}

        /* -------------------------------------------------------------
         * Line Height
         * -------------------------------------------------------------*/
        ${i||s.css`
          /* Mobile: 2.4rem */
          line-height: ${t.lineHeights[3]};

          ${t.breakpoints.medium} {
            /* Tablet + Desktop: 2.4rem */
            line-height: ${t.lineHeights[6]};
          }
        `}
      `;case tr:return s.css`
        /* -------------------------------------------------------------
         * Font Size
         * -------------------------------------------------------------*/
        ${a||s.css`
          /* Mobile: 1.6rem */
          font-size: ${t.fontSizes[3]};

          ${t.breakpoints.medium} {
            /* Tablet + Desktop: 1.4rem */
            font-size: ${t.fontSizes[2]};
          }
        `}

        /* -------------------------------------------------------------
         * Line Height
         * -------------------------------------------------------------*/
        ${i||s.css`
          /* Mobile: 2.4rem */
          line-height: ${t.lineHeights[6]};

          ${t.breakpoints.medium} {
            /* Tablet + Desktop: 2.0rem */
            line-height: ${t.lineHeights[4]};
          }
        `}
      `;case To:return s.css`
        /* -------------------------------------------------------------
         * Font Size
         * -------------------------------------------------------------*/
        ${a||s.css`
          /* All: 1.2rem */
          font-size: ${t.fontSizes[1]};
        `}

        /* -------------------------------------------------------------
         * Line Height
         * -------------------------------------------------------------*/
        ${i||s.css`
          /* All: 1.6rem */
          line-height: ${t.lineHeights[3]};
        `}
      `;case Io:return s.css`
        font-weight: ${t.fontWeights.bold};
        text-transform: uppercase;

        /* -------------------------------------------------------------
         * Font Size
         * -------------------------------------------------------------*/
        ${a||s.css`
          /* All: 1.1rem */
          font-size: ${t.fontSizes[0]};
        `}

        /* -------------------------------------------------------------
         * Line Height
         * -------------------------------------------------------------*/
        ${i||s.css`
          /* All: 1.6rem */
          line-height: ${t.lineHeights[5]};
        `}
      `;default:return s.css`
        /* -------------------------------------------------------------
         * Font Size
         * -------------------------------------------------------------*/
        ${a||s.css`
          /* Mobile: 1.6rem */
          font-size: ${t.fontSizes[3]};

          ${t.breakpoints.medium} {
            /* Tablet + Desktop: 1.4rem */
            font-size: ${t.fontSizes[2]};
          }
        `}
      `}},T=ae((e,t)=>{const{ellipsis:n,textColor:o="currentcolor",textDecoration:a,textTransform:i,variant:c,lineHeight:d,fontWeight:u,fontSize:h,...f}=e,g={$ellipsis:n,$textColor:o,$textDecoration:a,$textTransform:i,$variant:c,$lineHeight:d,$fontWeight:u,$fontSize:h};return r.jsx(Eo,{ref:t,tag:"span",...g,...f})}),Eo=s.styled(v)`
  ${({$variant:e,$fontSize:t,$lineHeight:n,theme:o})=>Do({$variant:e,theme:o,$fontSize:t,$lineHeight:n})}
  ${({$ellipsis:e})=>e?jr:""}

  ${({theme:e,...t})=>ke({color:t.$textColor,textDecoration:t.$textDecoration,textTransform:t.$textTransform,lineHeight:t.$lineHeight,fontWeight:t.$fontWeight,fontSize:t.$fontSize},e)}
`,[Ao,Dt]=Oe("Accordion"),zo=l.forwardRef(({children:e,size:t="S",...n},o)=>r.jsx(Lo,{ref:o,$size:t,collapsible:!0,...n,type:"single",children:r.jsx(Ao,{size:t,children:e})})),Lo=s.styled(Ue.Root)`
  background-color: ${e=>e.theme.colors.neutral0};

  ${e=>e.$size==="S"?s.css`
        border-radius: ${t=>t.theme.borderRadius};
        border: solid 1px ${t=>t.theme.colors.neutral200};
      `:s.css`
        
      `}
`,Oo=l.forwardRef((e,t)=>{const{size:n}=Dt("Item");return r.jsx(Po,{$size:n,"data-size":n,ref:t,...e})}),Po=s.styled(Ue.Item)`
  overflow: hidden;
  margin: 1px 0;

  &:first-child {
    border-top-left-radius: 0.3rem;
    border-top-right-radius: 0.3rem;
    margin-top: 0;
  }

  &:last-child {
    border-bottom-left-radius: 0.3rem;
    border-bottom-right-radius: 0.3rem;
    margin-bottom: 0;
  }

  &[data-size='S'] {
    & + & {
      border-top: solid 1px ${e=>e.theme.colors.neutral200};
    }
  }

  &[data-state='open'] {
    box-shadow: 0 0 0 1px ${e=>e.theme.colors.primary600};
  }

  &:not([data-disabled]):hover {
    box-shadow: 0 0 0 1px ${e=>e.theme.colors.primary600};
  }

  /* This applies our desired focus effect correctly. */
  &:focus-within {
    position: relative;
    z-index: 1;
    box-shadow: 0 0 0 1px ${e=>e.theme.colors.primary600};
  }

  @media (prefers-reduced-motion: no-preference) {
    transition: box-shadow ${e=>e.theme.motion.timings[120]}
      ${e=>e.theme.motion.easings.easeOutQuad};
  }
`,Mo=l.forwardRef(({caretPosition:e="left",description:t,icon:n,iconProps:o,children:a,...i},c)=>{const{size:d}=Dt("Trigger");return r.jsxs(kr,{$caretPosition:e,$size:d,ref:c,...i,children:[e==="left"?r.jsx(tt,{$size:d,children:r.jsx(O.CaretDown,{width:d==="S"?"1.2rem":"1.6rem",height:d==="S"?"1.2rem":"1.6rem"})}):null,r.jsxs(C,{tag:"span",gap:2,overflow:"hidden",children:[n&&d==="S"?r.jsx(vr,{children:r.jsx(n,{...o})}):null,r.jsxs(C,{alignItems:"flex-start",direction:"column",tag:"span",ref:c,overflow:"hidden",children:[r.jsx(T,{fontWeight:d==="S"?"bold":void 0,ellipsis:!0,variant:d==="M"?"delta":void 0,textAlign:"left",width:"100%",children:a}),t&&d==="M"?r.jsx(T,{textAlign:"left",children:t}):null]})]}),e==="right"?r.jsx(tt,{$size:d,children:r.jsx(O.CaretDown,{width:d==="S"?"1.2rem":"1.6rem",height:d==="S"?"1.2rem":"1.6rem"})}):null]})}),vr=s.styled(v)`
  color: ${e=>e.theme.colors.neutral500};
  display: flex;

  @media (prefers-reduced-motion: no-preference) {
    transition: ${e=>e.theme.transitions.color};
  }
`,tt=s.styled(C).attrs(e=>({...e,tag:"span"}))`
  background-color: ${e=>e.theme.colors.neutral200};
  width: ${e=>e.$size==="S"?"2.4rem":"3.2rem"};
  height: ${e=>e.$size==="S"?"2.4rem":"3.2rem"};
  flex: ${e=>e.$size==="S"?"0 0 2.4rem":"0 0 3.2rem"};
  border-radius: 50%;
  justify-content: center;

  @media (prefers-reduced-motion: no-preference) {
    transition:
      transform ${e=>e.theme.motion.timings[200]} ${e=>e.theme.motion.easings.authenticMotion},
      ${e=>e.theme.transitions.backgroundColor};
  }
`,kr=s.styled(Ue.Trigger)`
  display: flex;
  align-items: center;
  justify-content: ${e=>e.$caretPosition==="left"?"flex-start":"space-between"};
  width: 100%;
  gap: ${e=>e.theme.spaces[4]};
  padding-inline: ${e=>e.$size==="S"?e.theme.spaces[4]:e.theme.spaces[6]};
  padding-block: ${e=>e.$size==="S"?e.theme.spaces[3]:e.theme.spaces[6]};
  cursor: pointer;
  color: ${e=>e.theme.colors.neutral800};
  overflow: hidden;

  &[data-disabled] {
    cursor: default;
    color: ${e=>e.theme.colors.neutral600};
  }

  &[data-state='open'] > ${tt} {
    transform: rotate(180deg);
  }

  /* we remove the default focus because the entire item should have the focus style and the default would be hidden. */
  &:focus-visible {
    outline: none;
  }

  @media (prefers-reduced-motion: no-preference) {
    transition: ${e=>e.theme.transitions.color};
  }
`,Bo=l.forwardRef((e,t)=>{const{size:n}=Dt("Trigger");return r.jsx(Sr,{$size:n,...e,ref:t})}),Sr=s.styled(C).attrs(e=>({...e,tag:"span"}))`
  padding-inline: ${e=>e.$size==="S"?e.theme.spaces[2]:e.theme.spaces[6]};
  padding-block: ${e=>e.$size==="S"?e.theme.spaces[2]:e.theme.spaces[6]};

  // Remove default IconButton styles so there are no backgrounds or borders.
  & > button {
    border: none;
    background: none;
    color: ${e=>e.theme.colors.neutral600};

    @media (prefers-reduced-motion: no-preference) {
      transition: ${e=>e.theme.transitions.color};
    }
  }
`,No=l.forwardRef(({variant:e="primary",...t},n)=>r.jsx(Wo,{$variant:e,ref:n,...t})),Wo=s.styled(Ue.Header)`
  display: flex;
  align-items: center;
  background-color: ${e=>e.$variant==="primary"?e.theme.colors.neutral0:e.theme.colors.neutral100};

  &[data-disabled] {
    background-color: ${e=>e.theme.colors.neutral150};
  }

  &:not([data-disabled]) {
    &:hover,
    &[data-state='open'] {
      background-color: ${e=>e.theme.colors.primary100};

      & > ${kr} {
        color: ${e=>e.theme.colors.primary600};

        & ${vr} {
          color: ${e=>e.theme.colors.primary600};
        }

        & ${tt} {
          background-color: ${e=>e.theme.colors.primary200};
        }
      }

      & > ${Sr} > button {
        color: ${e=>e.theme.colors.primary600};
      }
    }
  }

  @media (prefers-reduced-motion: no-preference) {
    transition: ${e=>e.theme.transitions.backgroundColor};
  }
`,Fo=l.forwardRef((e,t)=>r.jsx(Vo,{ref:t,...e})),Ho=s.keyframes`
  from {
    height: 0;
  }
  to {
    height: var(--radix-accordion-content-height);
  }
`,_o=s.keyframes`
  from {
    height: var(--radix-accordion-content-height);
  }
  to {
    height: 0;
  }
`,Vo=s.styled(Ue.Content)`
  overflow: hidden;

  @media (prefers-reduced-motion: no-preference) {
    &[data-state='open'] {
      animation: ${Ho} ${e=>e.theme.motion.timings[320]}
        ${e=>e.theme.motion.easings.authenticMotion};
    }

    &[data-state='closed'] {
      animation: ${_o} ${e=>e.theme.motion.timings[320]}
        ${e=>e.theme.motion.easings.authenticMotion};
    }
  }
`,qo=Object.freeze(Object.defineProperty({__proto__:null,Actions:Bo,Content:Fo,Header:No,Item:Oo,Root:zo,Trigger:Mo},Symbol.toStringTag,{value:"Module"})),st=s.css`
  position: relative;
  outline: none;

  &:after {
    transition-property: all;
    transition-duration: 0.2s;
    border-radius: 8px;
    content: '';
    position: absolute;
    top: -4px;
    bottom: -4px;
    left: -4px;
    right: -4px;
    border: 2px solid transparent;
  }

  &:focus-visible {
    outline: none;

    &:after {
      border-radius: 8px;
      content: '';
      position: absolute;
      top: -5px;
      bottom: -5px;
      left: -5px;
      right: -5px;
      border: 2px solid ${e=>e.theme.colors.primary600};
    }
  }
`,me=({tag:e,...t})=>{const n=e||"span";return r.jsx(Uo,{...t,as:n})},Uo=s.styled.span`
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
`,Se=({children:e,label:t})=>{const n=l.Children.only(e);return r.jsxs(r.Fragment,{children:[l.cloneElement(n,{"aria-hidden":"true",focusable:"false"}),r.jsx(me,{children:t})]})};Se.displayName="AccessibleIcon";const Rr=e=>{switch(e){case"danger":return"danger100";case"success":return"success100";case"warning":return"warning100";default:return"primary100"}},Go=e=>Rr(e).replace("100","200"),xt=({theme:e,$variant:t})=>t==="danger"?e.colors.danger700:t==="success"?e.colors.success700:t==="warning"?e.colors.warning700:e.colors.primary700,Ko=s.styled(v)`
  ${st};
`,Yo=s.styled(C)`
  svg {
    height: 100%;
    width: 100%;

    path {
      fill: ${xt};
    }
  }
`,Qo=({variant:e,...t})=>e==="success"?r.jsx(O.CheckCircle,{...t}):e==="danger"||e==="warning"?r.jsx(O.WarningCircle,{...t}):r.jsx(O.Information,{...t}),Xo=s.styled(v)`
  & a > span {
    color: ${xt};
  }

  svg path {
    fill: ${xt};
  }
`,Zo=({title:e,children:t,variant:n="default",onClose:o,closeLabel:a,titleAs:i="p",action:c,...d})=>r.jsxs(C,{alignItems:"flex-start",background:Rr(n),borderColor:Go(n),shadow:"filterShadow",gap:3,hasRadius:!0,padding:5,paddingRight:6,...d,children:[r.jsx(Yo,{height:"2rem",shrink:0,$variant:n,width:"2rem",children:r.jsx(Qo,{"aria-hidden":!0,variant:n})}),r.jsxs(C,{alignItems:"start",gap:c?2:1,wrap:"wrap",role:n==="danger"?"alert":"status",width:"100%",children:[e&&r.jsx(T,{fontWeight:"bold",textColor:"neutral800",tag:i,children:e}),r.jsx(T,{tag:"p",textColor:"neutral800",children:t}),c&&r.jsx(Xo,{$variant:n,children:c})]}),r.jsx(Ko,{tag:"button",background:"transparent",borderColor:void 0,height:"1.6rem",width:"1.6rem",marginTop:1,onClick:o,color:"neutral700",children:r.jsx(Se,{label:a,children:r.jsx(O.Cross,{})})})]});function le({prop:e,defaultProp:t,onChange:n=()=>{}}){const[o,a]=Et({defaultProp:t,onChange:n}),i=e!==void 0,c=i?e:o,d=xr.useCallbackRef(n),u=l.useCallback(h=>{if(i){const g=typeof h=="function"?h(e):h;g!==e&&d(g)}else a(h)},[i,e,a,d]);return[c,u]}function Et({defaultProp:e,onChange:t}){const n=l.useState(e),[o]=n,a=l.useRef(o),i=xr.useCallbackRef(t);return l.useEffect(()=>{a.current!==o&&(i(o),a.current=o)},[o,a,i]),n}const $t={easeInSine:"cubic-bezier(0.47, 0, 0.745, 0.715)",easeOutSine:"cubic-bezier(0.39, 0.575, 0.565, 1)",easeInOutSine:"cubic-bezier(0.39, 0.575, 0.565, 1)",easeInQuad:"cubic-bezier(0.55, 0.085, 0.68, 0.53)",easeOutQuad:"cubic-bezier(0.25, 0.46, 0.45, 0.94)",easeInOutQuad:"cubic-bezier(0.25, 0.46, 0.45, 0.94)",easeInCubic:"cubic-bezier(0.55, 0.055, 0.675, 0.19)",easeOutCubic:"cubic-bezier(0.215, 0.61, 0.355, 1)",easeInOutCubic:"cubic-bezier(0.215, 0.61, 0.355, 1)",easeInQuart:"cubic-bezier(0.895, 0.03, 0.685, 0.22)",easeOutQuart:"cubic-bezier(0.165, 0.84, 0.44, 1)",easeInOutQuart:"cubic-bezier(0.165, 0.84, 0.44, 1)",easeInQuint:"cubic-bezier(0.755, 0.05, 0.855, 0.06)",easeOutQuint:"cubic-bezier(0.23, 1, 0.32, 1)",easeInOutQuint:"cubic-bezier(0.23, 1, 0.32, 1)",easeInExpo:"cubic-bezier(0.95, 0.05, 0.795, 0.035)",easeOutExpo:"cubic-bezier(0.19, 1, 0.22, 1)",easeInOutExpo:"cubic-bezier(0.19, 1, 0.22, 1)",easeInCirc:"cubic-bezier(0.6, 0.04, 0.98, 0.335)",easeOutCirc:"cubic-bezier(0.075, 0.82, 0.165, 1)",easeInOutCirc:"cubic-bezier(0.075, 0.82, 0.165, 1)",easeInBack:"cubic-bezier(0.6, -0.28, 0.735, 0.045)",easeOutBack:"cubic-bezier(0.175, 0.885, 0.32, 1.275)",easeInOutBack:"cubic-bezier(0.68, -0.55, 0.265, 1.55)",easeInOutFast:"cubic-bezier(1,0,0,1)",authenticMotion:"cubic-bezier(.4,0,.2,1)"},yt={320:"320ms",200:"200ms",120:"120ms"},Jo={color:`color ${yt[120]} ${$t.easeOutQuad}`,backgroundColor:`background-color ${yt[120]} ${$t.easeOutQuad}`},q={overlayFadeIn:s.keyframes`
    from {
      opacity: 0;
    }
    to {
      opacity: 0.2;
    }
  `,modalPopIn:s.keyframes`
    from {
      transform:translate(-50%, -50%)  scale(0.8);
      opacity: 0;
    }
    to {
      transform: translate(-50%, -50%) scale(1);
      opacity: 1;
    }
  `,modalPopOut:s.keyframes`
    from {
      transform: translate(-50%, -50%) scale(1);
      opacity: 1;
    }
    to {
      transform:translate(-50%, -50%)  scale(0.8);
      opacity: 0;
    }
  `,popIn:s.keyframes`
    from {
      transform: scale(0.8);
      opacity: 0;
    }
    to {
      transform: scale(1);
      opacity: 1;
    }
  `,popOut:s.keyframes`
    from {
      transform: scale(1);
      opacity: 1;
    }
    to {
      transform: scale(0.8);
      opacity: 0;
    }
  `,slideDownIn:s.keyframes`
    from {
      opacity: 0;
      transform: translateY(-10px);
    }
    to {
      opacity: 1;
      transform: translateY(0);
    }
  `,slideDownOut:s.keyframes`
    from {
      opacity: 1;
      transform: translateY(0);
    }
    to {
      opacity: 0;
      transform: translateY(-10px);
    }
  `,slideUpIn:s.keyframes`
    from {
      opacity: 0;
      transform: translateY(10px);
    }
    to {
      opacity: 1;
      transform: translateY(0);
    }
  `,slideUpOut:s.keyframes`
    from {
      opacity: 1;
      transform: translateY(0);
    }
    to {
      opacity: 0;
      transform: translateY(10px);
    }
  `,fadeIn:s.keyframes`
    from {
      opacity: 0;
    }
    to {
      opacity: 1;
    }
  `,fadeOut:s.keyframes`
    from {
      opacity: 1;
    }
    to {
      opacity: 0;
    }
  `},qe=32,rr=2,ea=l.forwardRef(({onLoadingStatusChange:e,delayMs:t=600,src:n,alt:o,crossOrigin:a,fallback:i,preview:c=!1,...d},u)=>{const[h,f]=le({onChange:e}),[g,p]=l.useState(!1),w=c&&h==="loaded",m=$=>{w&&p($)};return r.jsxs(xe.Root,{onOpenChange:m,children:[r.jsx(xe.Trigger,{asChild:!0,children:r.jsxs(wt,{ref:u,...d,children:[w?r.jsx(ta,{width:"100%",height:"100%",position:"absolute",background:"neutral0",zIndex:"overlay",style:{opacity:g?.4:0}}):null,r.jsx(ra,{src:n,alt:o,crossOrigin:a,onLoadingStatusChange:f}),r.jsx(Tt.Fallback,{delayMs:t,children:r.jsx(T,{fontWeight:"bold",textTransform:"uppercase",children:i})})]})}),w?r.jsx(xe.Portal,{children:r.jsx(na,{side:"top",sideOffset:4,children:r.jsx(oa,{src:n,alt:o,crossOrigin:a})})}):null]})}),Tr=s.css`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  vertical-align: middle;
  overflow: hidden;
  user-select: none;
  overflow: hidden;
  border-radius: 50%;
`,Ir=s.css`
  width: 100%;
  height: 100%;
  object-fit: cover;
  border-radius: inherit;
`,wt=s.styled(Tt.Root)`
  position: relative;
  z-index: 0;
  ${Tr}
  width: ${qe/10}rem;
  height: ${qe/10}rem;
  /* TODO: we should get the user email & hash it to turn it into a hex-value so different emails can consistently get a different background */
  background-color: ${e=>e.theme.colors.primary600};
  color: ${e=>e.theme.colors.neutral0};
`,ta=s.styled(v)`
  @media (prefers-reduced-motion: no-preference) {
    transition: opacity ${e=>e.theme.motion.timings[200]}
      ${e=>e.theme.motion.easings.authenticMotion};
  }
`,ra=s.styled(Tt.Image)`
  ${Ir}
`,na=s.styled(xe.Content)`
  ${Tr}
  width: ${qe*rr/10}rem;
  height: ${qe*rr/10}rem;

  @media (prefers-reduced-motion: no-preference) {
    animation: ${q.fadeIn} ${e=>e.theme.motion.timings[200]}
      ${e=>e.theme.motion.easings.authenticMotion};
  }
`,oa=s.styled.img`
  ${Ir}
`,aa=l.forwardRef((e,t)=>r.jsx(sa,{...e,ref:t,tag:"div"})),sa=s.styled(C)`
  & > ${wt} + ${wt} {
    margin-left: -${qe/10/2}rem;
  }
`,ia=Object.freeze(Object.defineProperty({__proto__:null,Group:aa,Item:ea},Symbol.toStringTag,{value:"Module"})),At=({active:e=!1,size:t="M",textColor:n="neutral600",backgroundColor:o="neutral150",variant:a,children:i,minWidth:c=5,...d})=>{const u=t==="S"?1:2,h=a?{backgroundColor:`${a}200`,textColor:`${a}700`}:{backgroundColor:o,textColor:n};return r.jsx(la,{inline:!0,alignItems:"center",justifyContent:"center",minWidth:c,paddingLeft:u,paddingRight:u,background:e?"primary200":h.backgroundColor,$size:t,...d,children:r.jsx(T,{variant:"sigma",textColor:e?"primary600":h.textColor,lineHeight:"1rem",children:i})})},la=s.styled(C)`
  border-radius: ${({theme:e,$size:t})=>t==="S"?"2px":e.borderRadius};
  ${({$size:e,theme:t})=>e==="S"?s.css`
        padding-block: 0.3rem;
        padding-inline ${t.spaces[1]}
      `:s.css`
      padding-block: 0.7rem;
      padding-inline ${t.spaces[2]}
    `};
`,ca=s.styled(v)`
  text-decoration: none;

  &:visited {
    color: inherit;
  }
`,Pe=ae(({href:e,disabled:t=!1,isExternal:n=!1,...o},a)=>r.jsx(ca,{tag:"a",ref:a,target:n?"_blank":void 0,rel:n?"noreferrer noopener":void 0,href:e,tabIndex:t?-1:void 0,"aria-disabled":t,pointerEvents:t?"none":void 0,cursor:t?void 0:"pointer",...o})),Dr=()=>r.jsx(v,{"aria-hidden":!0,paddingLeft:1,paddingRight:1,children:r.jsx(T,{variant:"pi",textColor:"neutral500",children:"/"})});Dr.displayName="Divider";const da=s.styled(C)`
  // CrumbLinks do have padding-x, because they need to have a
  // interaction effect, which mis-aligns the breadcrumbs on the left.
  // This normalizes the behavior by moving the first item to left by
  // the same amount it has inner padding
  & > *:first-child {
    margin-left: ${({theme:e})=>`calc(-1*${e.spaces[2]})`};
  }
`,Er=l.forwardRef(({label:e,children:t,...n},o)=>{const a=l.Children.toArray(t);return r.jsx(v,{"aria-label":e,tag:"nav",...n,ref:o,children:r.jsx(da,{tag:"ol",children:l.Children.map(a,(i,c)=>{const d=a.length>1&&c+1<a.length;return r.jsxs(C,{inline:!0,tag:"li",children:[i,d&&r.jsx(Dr,{})]})})})})});Er.displayName="Breadcrumbs";const Ar=l.forwardRef(({children:e,isCurrent:t=!1,...n},o)=>r.jsx(v,{paddingLeft:2,paddingRight:2,paddingTop:1,paddingBottom:1,ref:o,children:r.jsx(T,{variant:"pi",textColor:"neutral800",fontWeight:t?"bold":"regular","aria-current":t,...n,children:e})}));Ar.displayName="Crumb";const ua=s.styled(Pe)`
  border-radius: ${({theme:e})=>e.borderRadius};
  color: ${({theme:e})=>e.colors.neutral600};
  font-size: ${({theme:e})=>e.fontSizes[1]};
  line-height: ${({theme:e})=>e.lineHeights[4]};
  padding: ${({theme:e})=>`${e.spaces[1]} ${e.spaces[2]}`};
  text-decoration: none;

  :hover,
  :focus {
    background-color: ${({theme:e})=>e.colors.neutral200};
    color: ${({theme:e})=>e.colors.neutral700};
  }
`,zr=l.forwardRef(({children:e,...t},n)=>r.jsx(ua,{ref:n,...t,children:e}));zr.displayName="CrumbLink";const ze=e=>e.replaceAll(":","");function ha(e,t){typeof e=="function"?e(t):e!=null&&(e.current=t)}function zt(...e){return t=>e.forEach(n=>ha(n,t))}function te(...e){return l.useCallback(zt(...e),e)}function re(e){const t=X.useId();return(e==null?void 0:e.toString())??t}const Ge=(e,t,{selectorToWatch:n,skipWhen:o=!1})=>{const a=D.useCallbackRef(t);l.useEffect(()=>{if(o||!e.current)return;const i={root:e.current,rootMargin:"0px"},c=h=>{h.forEach(f=>{f.isIntersecting&&e.current&&e.current.scrollHeight>e.current.clientHeight&&a(f)})},d=new IntersectionObserver(c,i),u=e.current.querySelector(n);return u&&d.observe(u),()=>{d.disconnect()}},[o,a,n,e])},Lt="success-light",Ot="danger-light",it="default",Ke="tertiary",Ye="secondary",Lr="danger",Or="success",lt="ghost",Pt=[Lt,Ot],fa=[it,Ke,Ye,Lr,Or,lt,...Pt],ga=["XS","S","M","L"],ie=e=>e===Lt||e===Ot?`${e.substring(0,e.lastIndexOf("-"))}`:e===Ke?"neutral":e===it||e===Ye||fa.every(t=>t!==e)?"primary":`${e}`,Pr=({theme:e})=>s.css`
    border: 1px solid ${e.colors.neutral200};
    background: ${e.colors.neutral150};
    color: ${e.colors.neutral600};
    cursor: default;
  `,Mr=({theme:e,$variant:t})=>[...Pt,Ye].includes(t)?s.css`
      background-color: ${e.colors.neutral0};
    `:t===Ke?s.css`
      background-color: ${e.colors.neutral100};
    `:t===lt?s.css`
      background-color: ${e.colors.neutral100};
    `:t===it?s.css`
      border: 1px solid ${e.colors.buttonPrimary500};
      background: ${e.colors.buttonPrimary500};
    `:s.css`
    border: 1px solid ${e.colors[`${ie(t)}500`]};
    background: ${e.colors[`${ie(t)}500`]};
  `,Br=({theme:e,$variant:t})=>[...Pt,Ye].includes(t)?s.css`
      background-color: ${e.colors.neutral0};
      border: 1px solid ${e.colors[`${ie(t)}600`]};
      color: ${e.colors[`${ie(t)}600`]};
    `:t===Ke||t===lt?s.css`
      background-color: ${e.colors.neutral150};
    `:s.css`
    border: 1px solid ${e.colors[`${ie(t)}600`]};
    background: ${e.colors[`${ie(t)}600`]};
  `,Nr=({theme:e,$variant:t})=>{switch(t){case Ot:case Lt:case Ye:return s.css`
        border: 1px solid ${e.colors[`${ie(t)}200`]};
        background: ${e.colors[`${ie(t)}100`]};
        color: ${e.colors[`${ie(t)}700`]};
      `;case Ke:return s.css`
        border: 1px solid ${e.colors.neutral200};
        background: ${e.colors.neutral0};
        color: ${e.colors.neutral800};
      `;case lt:return s.css`
        border: 1px solid transparent;
        background: transparent;
        color: ${e.colors.neutral800};

        svg {
          fill: ${e.colors.neutral500};
        }
      `;case Or:case Lr:return s.css`
        border: 1px solid ${e.colors[`${ie(t)}600`]};
        background: ${e.colors[`${ie(t)}600`]};
        color: ${e.colors.neutral0};
      `;default:return s.css`
        border: 1px solid ${e.colors.buttonPrimary600};
        background: ${e.colors.buttonPrimary600};
        color: ${e.colors.buttonNeutral0};
      `}},rt=ae(({variant:e=it,startIcon:t,endIcon:n,disabled:o=!1,children:a,onClick:i,size:c=ga[1],loading:d=!1,fullWidth:u=!1,type:h="button",...f},g)=>{const p=o||d,w=k=>{!p&&i&&i(k)},m=f.tag??"button",$=m==="button"||m==="input"?h:void 0;return r.jsxs(ba,{ref:g,"aria-disabled":p,disabled:p,$size:c,$variant:e,tag:"button",onClick:w,hasRadius:!0,gap:2,inline:!0,alignItems:"center",justifyContent:"center",width:u?"100%":void 0,paddingLeft:4,paddingRight:4,cursor:"pointer",type:$,...f,children:[(t||d)&&r.jsx(C,{tag:"span","aria-hidden":!0,children:d?r.jsx(ma,{}):t}),r.jsx(T,{variant:c==="S"?"pi":void 0,fontWeight:"bold",children:a}),n&&r.jsx(C,{tag:"span","aria-hidden":!0,children:n})]})}),pa=s.keyframes`
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(359deg);
  }
`,ma=s.styled(O.Loader)`
  animation: ${pa} 2s infinite linear;
  will-change: transform;
`,ba=s.styled(C)`
  ${({theme:e,$size:t})=>{const n=e.sizes.button[t];if(typeof n=="string")return`height: ${n};`;const o=[];return Object.entries(n).forEach(([a,i])=>{if(i){if(a==="initial")o.push(`height: ${i};`);else if(a in e.breakpoints){const c=e.breakpoints[a];o.push(`${c} { height: ${i}; }`)}}}),o.join(`
`)}}
  text-decoration: none;
  ${Nr}

  &:hover {
    ${Mr}
  }

  &:active {
    ${Br}
  }

  &[aria-disabled='true'] {
    ${Pr}
  }

  @media (prefers-reduced-motion: no-preference) {
    transition:
      ${e=>e.theme.transitions.backgroundColor},
      ${e=>e.theme.transitions.color},
      border-color ${e=>e.theme.motion.timings[200]} ${e=>e.theme.motion.easings.easeOutQuad};
  }
`,ct=l.forwardRef(({children:e,description:t,label:n,defaultOpen:o,open:a,onOpenChange:i,delayDuration:c=500,disableHoverableContent:d,...u},h)=>!n&&!t?e:r.jsxs(xe.Root,{defaultOpen:o,open:a,onOpenChange:i,delayDuration:c,disableHoverableContent:d,children:[r.jsx(xe.Trigger,{asChild:!0,children:e}),r.jsx(xe.Portal,{children:r.jsx(xa,{ref:h,sideOffset:8,...u,children:r.jsx(T,{variant:"pi",fontWeight:"bold",children:n||t})})})]})),xa=s.styled(xe.Content)`
  background-color: ${e=>e.theme.colors.neutral900};
  color: ${e=>e.theme.colors.neutral0};
  padding-inline: ${e=>e.theme.spaces[2]};
  padding-block: ${e=>e.theme.spaces[2]};
  border-radius: ${e=>e.theme.borderRadius};
  z-index: ${e=>e.theme.zIndices.tooltip};
  will-change: opacity;
  transform-origin: var(--radix-tooltip-content-transform-origin);

  @media (prefers-reduced-motion: no-preference) {
    animation: ${q.fadeIn} ${e=>e.theme.motion.timings[200]}
      ${e=>e.theme.motion.easings.authenticMotion};
  }
`,ge=ae(({label:e,background:t,children:n,disabled:o=!1,onClick:a,size:i="S",variant:c="tertiary",withTooltip:d=!0,type:u="button",...h},f)=>{const g=w=>{!o&&a&&a(w)},p=r.jsx(He,{"aria-disabled":o,background:o?"neutral150":t,tag:"button",display:"inline-flex",justifyContent:"center",hasRadius:!0,cursor:"pointer",...h,ref:f,$size:i,onClick:g,$variant:c,type:u,children:r.jsx(Se,{label:e,children:n})});return d?r.jsx(ct,{label:e,children:p}):p}),He=s.styled(C)`
  text-decoration: none;

  ${e=>{switch(e.$size){case"XS":return s.css`
          padding-block: 0.7rem;
          padding-inline: 0.7rem;

          ${({theme:t})=>t.breakpoints.medium} {
            padding-block: 0.2rem;
            padding-inline: 0.2rem;
          }
        `;case"S":return s.css`
          padding-block: 1.1rem;
          padding-inline: 1.1rem;

          ${({theme:t})=>t.breakpoints.medium} {
            padding-block: 0.7rem;
            padding-inline: 0.7rem;
          }
        `;case"M":return s.css`
          padding-block: 1.1rem;
          padding-inline: 1.1rem;

          ${({theme:t})=>t.breakpoints.medium} {
            padding-block: 0.9rem;
            padding-inline: 0.9rem;
          }
        `;case"L":return s.css`
          padding-block: 1.5rem;
          padding-inline: 1.5rem;

          ${({theme:t})=>t.breakpoints.medium} {
            padding-block: 1.1rem;
            padding-inline: 1.1rem;
          }
        `}}}
  ${Nr}
  ${e=>e.$variant==="tertiary"?s.css`
          color: ${e.theme.colors.neutral500};
        `:""}

  &:hover {
    ${Mr}
    ${e=>e.$variant==="tertiary"?s.css`
            color: ${e.theme.colors.neutral600};
          `:""}
  }

  &:active {
    ${Br}
  }

  &[aria-disabled='true'] {
    ${Pr}
  }

  @media (prefers-reduced-motion: no-preference) {
    transition:
      ${e=>e.theme.transitions.backgroundColor},
      ${e=>e.theme.transitions.color},
      border-color ${e=>e.theme.motion.timings[200]} ${e=>e.theme.motion.easings.easeOutQuad};
  }
`,$a=s.styled(C)`
  & ${He}:first-child {
    border-radius: ${({theme:e})=>`${e.borderRadius} 0 0 ${e.borderRadius}`};
  }

  & ${He}:last-child {
    border-radius: ${({theme:e})=>`0 ${e.borderRadius} ${e.borderRadius} 0`};
  }

  & ${He} {
    border-radius: 0;

    & + ${He} {
      border-left: none;
    }
  }
`,Wr=ae(({children:e,href:t,disabled:n=!1,startIcon:o,endIcon:a,isExternal:i=!1,color:c="primary600",activeColor:d="primary700",...u},h)=>r.jsxs(ya,{ref:h,href:t,disabled:n,isExternal:i,$activeColor:d,$color:c,...u,children:[o,r.jsx(T,{textColor:n?"neutral600":c,children:e}),a,t&&!a&&i&&r.jsx(O.ExternalLink,{fill:c})]})),ya=s.styled(Pe)`
  display: inline-flex;
  align-items: center;
  text-decoration: none;
  gap: ${({theme:e})=>e.spaces[2]};
  pointer-events: ${({disabled:e})=>e?"none":void 0};

  svg {
    font-size: 1rem;

    path {
      fill: ${({disabled:e,$color:t,theme:n})=>e?n.colors.neutral600:n.colors[t||"primary600"]};
    }
  }

  &:hover {
    & > span {
      color: ${({theme:e,$color:t})=>e.colors[t||"primary600"]};
    }

    svg path {
      fill: ${({theme:e,$color:t})=>e.colors[t||"primary600"]};
    }
  }

  &:active {
    color: ${({theme:e,$activeColor:t})=>e.colors[t||"primary700"]};
  }

  ${st};
`,wa=e=>{switch(e){case"danger":return"danger100";default:return"primary100"}},Ct=(e,t)=>{switch(e){case"danger":return t?"danger500":"danger700";default:return t?"neutral500":"neutral800"}},Ve=(e,t)=>{switch(e){case"danger":return t?"danger500":"danger700";default:return t?"neutral300":"neutral500"}},Ca=ce.Root,ja=l.forwardRef(({label:e,endIcon:t=r.jsx(O.CaretDown,{width:"1.2rem",height:"1.2rem","aria-hidden":!0}),tag:n=rt,icon:o,...a},i)=>{const c={...a,ref:i,type:"button"};return r.jsx(ce.Trigger,{asChild:!0,disabled:c.disabled,children:n===ge?r.jsx(ge,{label:e,variant:"tertiary",...c,children:o}):r.jsx(rt,{endIcon:t,variant:"ghost",...c})})}),va=l.forwardRef(({children:e,intersectionId:t,onCloseAutoFocus:n,popoverPlacement:o="bottom-start",...a},i)=>{const[c,d]=o.split("-");return r.jsx(ce.Portal,{children:r.jsx(ka,{align:d,side:c,loop:!0,onCloseAutoFocus:n,asChild:!0,children:r.jsxs(Fr,{ref:i,direction:"column",borderColor:"neutral150",hasRadius:!0,background:"neutral0",shadow:"filterShadow",maxHeight:"15rem",padding:1,marginTop:1,marginBottom:1,alignItems:"flex-start",position:"relative",overflow:"auto",...a,children:[e,r.jsx(v,{id:t,width:"100%",height:"1px"})]})})})}),Fr=s.styled(C)`
  scrollbar-width: none;
  -ms-overflow-style: none;
  -webkit-overflow-scrolling: touch;
  z-index: ${e=>e.theme.zIndices.popover};

  &::-webkit-scrollbar {
    display: none;
  }
`,ka=s.styled(ce.Content)`
  @media (prefers-reduced-motion: no-preference) {
    animation-duration: ${e=>e.theme.motion.timings[200]};

    &[data-state='open'] {
      animation-timing-function: ${e=>e.theme.motion.easings.authenticMotion};

      &[data-side='top'] {
        animation-name: ${q.slideUpIn};
      }

      &[data-side='bottom'] {
        animation-name: ${q.slideDownIn};
      }
    }
  }
`,Sa=({onSelect:e,disabled:t=!1,isLink:n,startIcon:o,endIcon:a,isExternal:i,variant:c="default",...d})=>r.jsx(ce.Item,{asChild:!0,onSelect:e,disabled:t,children:n||i?r.jsx(Ra,{color:Ct(c,t),startIcon:o,endIcon:a,...d,isExternal:i??!1,$variant:c,children:d.children}):r.jsxs(_r,{cursor:"pointer",color:Ct(c,t),background:"transparent",borderStyle:"none",gap:2,$variant:c,...d,children:[o&&r.jsx(C,{tag:"span",color:Ve(c,t),"aria-hidden":!0,children:o}),r.jsx(T,{grow:1,children:d.children}),a&&r.jsx(C,{tag:"span",color:Ve(c,t),"aria-hidden":!0,children:a})]})}),Hr=({theme:e,$variant:t})=>s.css`
  text-align: left;
  width: 100%;
  border-radius: ${e.borderRadius};
  padding: ${e.spaces[2]} ${e.spaces[4]};

  &[aria-disabled='true'] {
    cursor: not-allowed;
  }

  &[data-highlighted] {
    background-color: ${e.colors[wa(t)]};
  }

  &:focus-visible {
    outline: none;

    &:after {
      content: none;
    }
  }

  color: ${e.colors[Ct(t,!1)]};
`,_r=s.styled(C)`
  ${({theme:e,$variant:t})=>Hr({theme:e,$variant:t})}
`,Ra=s.styled(Wr)`
  /* We include this here again because typically when people use OptionLink they provide an as prop which cancels the Box props */
  color: ${({theme:e,color:t})=>_e(e.colors,t,void 0)};
  text-decoration: none;

  &:hover {
    color: unset;

    > svg {
      path {
        fill: ${({theme:e,$variant:t})=>e.colors[Ve(t,!1)]};
      }
    }
  }
  > svg {
    path {
      fill: ${({theme:e,$variant:t})=>e.colors[Ve(t,!1)]};
    }
  }
  &[aria-disabled='true'] {
    pointer-events: none;

    > svg {
      path {
        fill: ${({theme:e,$variant:t})=>e.colors[Ve(t,!0)]};
      }
    }
  }

  ${({theme:e,$variant:t})=>Hr({theme:e,$variant:t})}
`,Ta=s.styled(v)`
  /* Negative horizontal margin to compensate Menu.Content's padding */
  margin: ${({theme:e})=>e.spaces[1]} -${({theme:e})=>e.spaces[1]};
  width: calc(100% + ${({theme:e})=>e.spaces[2]});
  /* Hide separator if there's nothing above in the menu */
  &:first-child {
    display: none;
  }
`,Ia=l.forwardRef((e,t)=>r.jsx(ce.Separator,{...e,asChild:!0,children:r.jsx(Ta,{height:"1px",shrink:0,background:"neutral150",ref:t})})),Da=l.forwardRef((e,t)=>r.jsx(ce.Label,{asChild:!0,children:r.jsx(Ea,{ref:t,variant:"sigma",textColor:"neutral600",...e})})),Ea=s.styled(T)`
  padding: ${({theme:e})=>e.spaces[2]} ${({theme:e})=>e.spaces[4]};
`,Aa=ce.Sub,za=l.forwardRef(({disabled:e=!1,...t},n)=>r.jsx(ce.SubTrigger,{asChild:!0,disabled:e,children:r.jsxs(La,{ref:n,color:"neutral800",tag:"button",type:"button",background:"transparent",borderStyle:"none",gap:5,...t,children:[r.jsx(T,{children:t.children}),r.jsx(O.ChevronRight,{fill:"neutral500",height:"1.2rem",width:"1.2rem"})]})})),La=s.styled(_r)`
  &[data-state='open'] {
    background-color: ${({theme:e})=>e.colors.primary100};
  }
`,Oa=l.forwardRef((e,t)=>r.jsx(ce.Portal,{children:r.jsx(ce.SubContent,{sideOffset:8,asChild:!0,children:r.jsx(Fr,{ref:t,direction:"column",borderStyle:"solid",borderWidth:"1px",borderColor:"neutral150",hasRadius:!0,background:"neutral0",shadow:"filterShadow",maxHeight:"15rem",padding:1,alignItems:"flex-start",overflow:"auto",...e})})})),Vr=Ca,qr=ja,Ur=va,Gr=Sa,Pa=Ia,Ma=Da,Ba=Aa,Na=za,Wa=Oa,Fa=Object.freeze(Object.defineProperty({__proto__:null,Content:Ur,Item:Gr,Label:Ma,Root:Vr,Separator:Pa,SubContent:Wa,SubRoot:Ba,SubTrigger:Na,Trigger:qr},Symbol.toStringTag,{value:"Module"})),Kr=l.forwardRef(({children:e,onOpen:t,onClose:n,popoverPlacement:o,onReachEnd:a,...i},c)=>{const d=l.useRef(null),u=te(c,d),h=l.useRef(null),[f,g]=l.useState(!1),p=k=>{a&&a(k)},w=k=>{k&&typeof t=="function"?t():!k&&typeof n=="function"&&n(),g(k)},m=re(),$=`intersection-${ze(m)}`;return Ge(h,p,{selectorToWatch:`#${$}`,skipWhen:!f}),r.jsxs(Vr,{onOpenChange:w,children:[r.jsx(qr,{ref:u,...i,children:i.label}),r.jsx(Ur,{ref:h,intersectionId:$,popoverPlacement:o,children:e})]})}),Ha=Gr,_a=s.styled(Kr)`
  padding: ${({theme:e})=>`${e.spaces[1]} ${e.spaces[2]}`};
  height: unset;

  :hover,
  :focus {
    background-color: ${({theme:e})=>e.colors.neutral200};
  }
`,Yr=l.forwardRef(({children:e,...t},n)=>r.jsx(_a,{ref:n,endIcon:null,size:"S",...t,children:e}));Yr.displayName="CrumbSimpleMenu";const Qr=l.createContext({id:""}),Xr=()=>l.useContext(Qr),Va=l.forwardRef(({id:e,...t},n)=>{const o=re(e),a=l.useMemo(()=>({id:o}),[o]);return r.jsx(Qr.Provider,{value:a,children:r.jsx(v,{ref:n,id:e,tabIndex:0,hasRadius:!0,background:"neutral0",borderStyle:"solid",borderWidth:"1px",borderColor:"neutral150",shadow:"tableShadow",tag:"article","aria-labelledby":`${o}-title`,...t})})}),Zr=l.forwardRef(({position:e,...t},n)=>r.jsx(qa,{ref:n,$position:e,...t,direction:"row",gap:2})),qa=s.styled(C)`
  position: absolute;
  top: ${({theme:e})=>e.spaces[3]};
  right: ${({$position:e,theme:t})=>{if(e==="end")return t.spaces[3]}};
  left: ${({$position:e,theme:t})=>{if(e==="start")return t.spaces[3]}};
`,Ua=({size:e="M",children:t,...n})=>r.jsx(Ka,{$size:e,children:t||r.jsx(Ga,{...n,"aria-hidden":!0})}),Ga=s.styled.img`
  // inline flows is based on typography and displays an extra white space below the image
  // switch to block is required in order to make the img stick the bottom of the container
  // addition infos: https://stackoverflow.com/questions/5804256/image-inside-div-has-extra-space-below-the-image
  margin: 0;
  padding: 0;
  max-height: 100%;
  max-width: 100%;
  object-fit: contain;
`,Ka=s.styled.div`
  display: flex;
  justify-content: center;
  height: ${({$size:e})=>e==="S"?"8.8rem":"16.4rem"};
  width: 100%;
  background: repeating-conic-gradient(${({theme:e})=>e.colors.neutral100} 0% 25%, transparent 0% 50%) 50% / 20px
    20px;
  border-top-left-radius: ${({theme:e})=>e.borderRadius};
  border-top-right-radius: ${({theme:e})=>e.borderRadius};
`,Ya=s.styled.div`
  margin-left: auto;
  flex-shrink: 0;
`,Qa=s.styled(At)`
  margin-left: ${({theme:e})=>e.spaces[1]};
`,Xa=e=>r.jsx(Ya,{children:r.jsx(Qa,{...e})}),Za=e=>r.jsx(v,{paddingLeft:3,paddingRight:3,paddingTop:2,paddingBottom:2,children:r.jsx(C,{...e,alignItems:"flex-start"})}),Ja=({fill:e,...t})=>{const{colors:n}=s.useTheme();return r.jsx(v,{tag:"svg",viewBox:"0 0 32 32",xmlns:"http://www.w3.org/2000/svg",fill:e?n[e]:void 0,...t,children:r.jsx("path",{d:"M29.0613 10.0613L13.0613 26.0613C12.9219 26.2011 12.7563 26.3121 12.574 26.3878C12.3917 26.4635 12.1962 26.5024 11.9988 26.5024C11.8013 26.5024 11.6059 26.4635 11.4235 26.3878C11.2412 26.3121 11.0756 26.2011 10.9363 26.0613L3.93626 19.0613C3.79673 18.9217 3.68605 18.7561 3.61053 18.5738C3.53502 18.3915 3.49615 18.1961 3.49615 17.9988C3.49615 17.8014 3.53502 17.606 3.61053 17.4237C3.68605 17.2414 3.79673 17.0758 3.93626 16.9363C4.07579 16.7967 4.24143 16.686 4.42374 16.6105C4.60604 16.535 4.80143 16.4962 4.99876 16.4962C5.19608 16.4962 5.39147 16.535 5.57378 16.6105C5.75608 16.686 5.92173 16.7967 6.06126 16.9363L12 22.875L26.9388 7.93876C27.2205 7.65697 27.6027 7.49866 28.0013 7.49866C28.3998 7.49866 28.782 7.65697 29.0638 7.93876C29.3455 8.22055 29.5039 8.60274 29.5039 9.00126C29.5039 9.39977 29.3455 9.78197 29.0638 10.0638L29.0613 10.0613Z"})})},nr=l.forwardRef(({defaultChecked:e,checked:t,onCheckedChange:n,...o},a)=>{const i=l.useRef(null),[c,d]=le({defaultProp:e,prop:t,onChange:n}),u=te(i,a);return r.jsx(es,{ref:u,checked:c,onCheckedChange:d,...o,children:r.jsxs(ts,{forceMount:!0,children:[c===!0?r.jsx(Ja,{width:"1.6rem",fill:"neutral0"}):null,c==="indeterminate"?r.jsx(O.Minus,{fill:"neutral0"}):null]})})}),es=s.styled(yr.Root)`
  background: ${e=>e.theme.colors.neutral0};
  width: 2rem;
  height: 2rem;
  border-radius: ${e=>e.theme.borderRadius};
  border: 1px solid ${e=>e.theme.colors.neutral300};
  position: relative;
  z-index: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  // this ensures the checkbox is always a square even in flex-containers.
  flex: 0 0 2rem;

  &[data-state='checked']:not([data-disabled]),
  &[data-state='indeterminate']:not([data-disabled]) {
    border: 1px solid ${e=>e.theme.colors.primary600};
    background-color: ${e=>e.theme.colors.primary600};
  }

  &[data-disabled] {
    background-color: ${e=>e.theme.colors.neutral200};
  }

  /* increase target size for touch devices https://www.w3.org/WAI/WCAG21/Understanding/target-size.html */
  &::before {
    content: '';
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    width: 100%;
    height: 100%;
    z-index: -1;
    min-width: 44px;
    min-height: 44px;
  }
`,ts=s.styled(yr.Indicator)`
  display: inline-flex;
  pointer-events: auto !important;
  width: 100%;
  height: 100%;
  cursor: pointer;
  justify-content: center;
  align-items: center;

  &[data-disabled] {
    cursor: not-allowed;
  }
`,dt=l.forwardRef(({children:e,...t},n)=>{const o=re(t.id);return e?r.jsxs(C,{gap:2,children:[r.jsx(nr,{id:o,...t}),r.jsx(T,{tag:"label",textColor:"neutral800",htmlFor:o,children:e})]}):r.jsx(nr,{ref:n,...t})}),rs=l.forwardRef((e,t)=>{const{id:n}=Xr();return r.jsx(Zr,{position:"start",children:r.jsx(dt,{"aria-labelledby":`${n}-title`,...e,ref:t})})}),ns=s.styled(v)`
  word-break: break-all;
`,os=({children:e,...t})=>r.jsx(ns,{...t,children:e}),as=s.styled(C)`
  border-bottom: 1px solid ${({theme:e})=>e.colors.neutral150};
`,ss=e=>r.jsx(as,{position:"relative",justifyContent:"center",...e}),is=({children:e,...t})=>r.jsx(v,{tag:"time",background:"neutral800",bottom:1,right:1,color:"neutral0",hasRadius:!0,position:"absolute",padding:1,...t,children:r.jsx(T,{variant:"pi",textColor:"neutral0",children:e})}),ls=e=>{const{id:t}=Xr();return r.jsx(T,{variant:"pi",id:`${t}-title`,textColor:"neutral800",fontWeight:"bold",tag:"div",...e})},cs=e=>r.jsx(T,{variant:"pi",...e,textColor:"neutral600",tag:"div"}),W={DOWN:"ArrowDown",UP:"ArrowUp",RIGHT:"ArrowRight",LEFT:"ArrowLeft",ESCAPE:"Escape",ENTER:"Enter",SPACE:" ",TAB:"Tab",END:"End",HOME:"Home",DELETE:"Delete",PAGE_UP:"PageUp",PAGE_DOWN:"PageDown",BACKSPACE:"Backspace",CLEAR:"Clear"},ds=s.styled(v)`
  grid-template-columns: auto 1fr auto;
  grid-template-areas: 'startAction slides endAction';
`,us=s.styled(C)`
  grid-area: slides;
`,or=s.styled(v)`
  grid-area: ${({$area:e})=>e};

  &:focus svg path,
  &:hover svg path {
    fill: ${({theme:e})=>e.colors.neutral900};
  }
`,Jr=l.forwardRef(({actions:e,children:t,label:n,nextLabel:o,onNext:a,onPrevious:i,previousLabel:c,secondaryLabel:d,selectedSlide:u,...h},f)=>{const g=l.useRef(null),p=l.useRef(null),w=l.Children.map(t,($,k)=>l.cloneElement($,{selected:k===u})),m=$=>{switch($.key){case W.RIGHT:{$.preventDefault(),p!=null&&p.current&&p.current.focus(),a&&a();break}case W.LEFT:{$.preventDefault(),g!=null&&g.current&&g.current.focus(),i&&i();break}}};return r.jsx(v,{ref:f,...h,onKeyDown:m,children:r.jsxs(v,{padding:2,borderColor:"neutral200",hasRadius:!0,background:"neutral100",children:[r.jsxs(ds,{tag:"section","aria-roledescription":"carousel","aria-label":n,display:"grid",position:"relative",children:[w&&w.length>1&&r.jsxs(r.Fragment,{children:[r.jsx(or,{tag:"button",onClick:i,$area:"startAction",ref:g,type:"button",children:r.jsx(Se,{label:c,children:r.jsx(O.ChevronLeft,{width:"1.6rem",height:"1.6rem",fill:"neutral600"})})}),r.jsx(or,{tag:"button",onClick:a,$area:"endAction",ref:p,type:"button",children:r.jsx(Se,{label:o,children:r.jsx(O.ChevronRight,{width:"1.6rem",height:"1.6rem",fill:"neutral600"})})})]}),r.jsx(us,{"aria-live":"polite",paddingLeft:2,paddingRight:2,width:"100%",overflow:"hidden",children:w}),e]}),d&&r.jsx(v,{paddingTop:2,paddingLeft:4,paddingRight:4,children:r.jsx(ct,{label:d,children:r.jsx(C,{justifyContent:"center",children:r.jsx(T,{variant:"pi",textColor:"neutral600",ellipsis:!0,children:d})})})})]})})}),hs={accordions:{S:"4.8rem",M:"8.8rem"},button:{S:{initial:"4rem",medium:"3.2rem"},M:{initial:"4.4rem",medium:"3.6rem"},L:{initial:"4.8rem",medium:"4rem"}}},en={sizes:hs,zIndices:{navigation:100,overlay:300,modal:310,dialog:320,popover:500,notification:700,tooltip:1e3},spaces:["0px","4px","8px","12px","16px","20px","24px","32px","40px","48px","56px","64px"],borderRadius:"4px",breakpoints:{initial:"@media(min-width: 0px)",small:"@media(min-width: 520px)",medium:"@media(min-width: 768px)",large:"@media(min-width: 1080px)"},fontSizes:["1.1rem","1.2rem","1.4rem","1.6rem","1.8rem","2rem","2.8rem","3.2rem"],lineHeights:[1.14,1.22,1.25,1.33,1.43,1.45,1.5],fontWeights:{regular:400,semiBold:500,bold:600},transitions:Jo,motion:{easings:$t,timings:yt}},fs={color:{alternative100:"#f6ecfc",alternative200:"#e0c1f4",alternative500:"#ac73e6",alternative600:"#9736e8",alternative700:"#8312d1",buttonNeutral0:"#ffffff",buttonPrimary500:"#7b79ff",buttonPrimary600:"#4945ff",danger100:"#fcecea",danger200:"#f5c0b8",danger500:"#ee5e52",danger600:"#d02b20",danger700:"#b72b1a",neutral0:"#ffffff",neutral100:"#f6f6f9",neutral1000:"#181826",neutral150:"#eaeaef",neutral200:"#dcdce4",neutral300:"#c0c0cf",neutral400:"#a5a5ba",neutral500:"#8e8ea9",neutral600:"#666687",neutral700:"#4a4a6a",neutral800:"#32324d",neutral900:"#212134",primary100:"#f0f0ff",primary200:"#d9d8ff",primary500:"#7b79ff",primary600:"#4945ff",primary700:"#271fe0",secondary100:"#eaf5ff",secondary200:"#b8e1ff",secondary500:"#66b7f1",secondary600:"#0c75af",secondary700:"#006096",success100:"#eafbe7",success200:"#c6f0c2",success500:"#5cb176",success600:"#328048",success700:"#2f6846",warning100:"#fdf4dc",warning200:"#fae7b9",warning500:"#f29d41",warning600:"#d9822f",warning700:"#be5d01"}},gs={shadow:{filterShadow:"0px 1px 4px rgba(33, 33, 52, 0.1)",focus:"inset 2px 0px 0px rgb(39, 31, 224), inset 0px 2px 0px rgb(39, 31, 224), inset -2px 0px 0px rgb(39, 31, 224), inset 0px -2px 0px rgb(39, 31, 224)",focusShadow:"0px 0px 6px rgba(76, 191, 255, 0.75)",popupShadow:"0px 2px 15px rgba(33, 33, 52, 0.1)",tableShadow:"0px 1px 4px rgba(33, 33, 52, 0.1)"}},Mt={colorScheme:"light",colors:fs.color,shadows:gs.shadow,...en},ps={color:{alternative100:"#181826",alternative200:"#4a4a6a",alternative500:"#ac73e6",alternative600:"#ac73e6",alternative700:"#e0c1f4",buttonNeutral0:"#ffffff",buttonPrimary500:"#7b79ff",buttonPrimary600:"#4945ff",danger100:"#181826",danger200:"#4a4a6a",danger500:"#ee5e52",danger600:"#ee5e52",danger700:"#ee5e52",neutral0:"#212134",neutral100:"#181826",neutral1000:"#ffffff",neutral150:"#32324d",neutral200:"#4a4a6a",neutral300:"#666687",neutral400:"#a5a5ba",neutral500:"#c0c0cf",neutral600:"#a5a5ba",neutral700:"#eaeaef",neutral800:"#ffffff",neutral900:"#ffffff",primary100:"#181826",primary200:"#4a4a6a",primary500:"#4945ff",primary600:"#7b79ff",primary700:"#7b79ff",secondary100:"#181826",secondary200:"#4a4a6a",secondary500:"#66b7f1",secondary600:"#66b7f1",secondary700:"#b8e1ff",success100:"#181826",success200:"#4a4a6a",success500:"#5cb176",success600:"#5cb176",success700:"#c6f0c2",warning100:"#181826",warning200:"#4a4a6a",warning500:"#f29d41",warning600:"#f29d41",warning700:"#fae7b9"}},ms={shadow:{filterShadow:"1px 1px 10px rgba(3, 3, 5, 0.35)",focus:"inset 2px 0px 0px rgb(39, 31, 224), inset 0px 2px 0px rgb(39, 31, 224), inset -2px 0px 0px rgb(39, 31, 224), inset 0px -2px 0px rgb(39, 31, 224)",focusShadow:"0px 0px 6px rgba(76, 191, 255, 0.75)",popupShadow:"1px 1px 10px rgba(3, 3, 5, 0.35)",tableShadow:"1px 1px 10px rgba(3, 3, 5, 0.2)"}},bs={colorScheme:"dark",colors:ps.color,shadows:ms.shadow,...en},ar=e=>`
${e}

The following is an example:


import { lightTheme, extendTheme } from '@strapi/design-system';

const myCustomTheme = extendTheme(lightTheme, {
    ${Object.keys(Mt).map(t=>`${t}: /* put the overrides for the ${t} key */,`).join(`
`)}
})
`,xs=(e,t)=>{if(!et(e)){const i=ar("The first argument should be an object and corresponds to the theme you want to extend.");throw new Error(i)}if(!et(t)){const i=ar("The second argument should be an object and corresponds to the keys of the theme you want to override.");throw new Error(i)}function n(i,c){if(Array.isArray(i))return c;if(et(i)&&et(c))return Je.merge(i,c)}const o=Je.cloneDeep(e??{}),a=Je.cloneDeep(t??{});return Je.assignWith(o,a,n),o},$s=e=>({theme:t,size:n})=>t.sizes[e][n],be=(e="&")=>({theme:t,$hasError:n=!1})=>s.css`
    outline: none;
    box-shadow: none;
    transition-property: border-color, box-shadow, fill;
    transition-duration: 0.2s;

    ${e}:focus-within {
      border: 1px solid ${n?t.colors.danger600:t.colors.primary600};
      box-shadow: ${n?t.colors.danger600:t.colors.primary600} 0px 0px 0px 2px;
    }
  `,[ys,ne]=Oe("Field",{}),Re=l.forwardRef(({children:e,name:t,error:n=!1,hint:o,id:a,required:i=!1,...c},d)=>{const u=re(a),[h,f]=l.useState();return r.jsx(ys,{name:t,id:u,error:n,hint:o,required:i,labelNode:h,setLabelNode:f,children:r.jsx(C,{direction:"column",alignItems:"stretch",gap:1,ref:d,...c,children:e})})}),Bt=l.forwardRef(({children:e,action:t,...n},o)=>{const{id:a,required:i,setLabelNode:c}=ne("Label"),d=te(o,c);return e?r.jsxs(C,{alignItems:"center",children:[r.jsxs(ws,{ref:d,variant:"pi",textColor:"neutral800",fontWeight:"bold",...n,id:`${a}-label`,htmlFor:a,tag:"label",ellipsis:!0,children:[e,i&&r.jsx(T,{"aria-hidden":!0,lineHeight:"1em",textColor:"danger600",children:"*"})]}),t&&r.jsx(Cs,{marginLeft:1,children:t})]}):null}),ws=s.styled(T)`
  display: block;
`,Cs=s.styled(C)`
  line-height: 0;
  color: ${({theme:e})=>e.colors.neutral500};
`,ut=l.forwardRef(({endAction:e,startAction:t,disabled:n=!1,onChange:o,hasError:a,required:i,className:c,size:d="M",...u},h)=>{const{id:f,error:g,hint:p,name:w,required:m}=ne("Input");let $;g?$=`${f}-error`:p&&($=`${f}-hint`);const k=!!g,y=l.useRef(null),x=l.useRef(null),I=te(x,h),j=R=>{!n&&o&&o(R)};return l.useLayoutEffect(()=>{if(y.current&&x.current){const R=y.current.offsetWidth,b=x.current;if(b){const S=R+8+16;b.style.paddingRight=`${S}px`}}},[e]),r.jsxs(ks,{gap:2,justifyContent:"space-between",$hasError:k||a,$disabled:n,$size:d,$hasLeftAction:!!t,$hasRightAction:!!e,className:c,children:[t,r.jsx(js,{id:f,name:w,ref:I,$size:d,"aria-describedby":$,"aria-invalid":k||a,"aria-disabled":n,disabled:n,"data-disabled":n?"":void 0,onChange:j,"aria-required":m||i,$hasLeftAction:!!t,$hasRightAction:!!e,...u}),e&&r.jsx(vs,{ref:y,children:e})]})}),js=s.styled.input`
  border: none;
  border-radius: ${({theme:e})=>e.borderRadius};
  cursor: ${e=>e["aria-disabled"]?"not-allowed":void 0};

  color: ${({theme:e})=>e.colors.neutral800};
  font-weight: 400;
  font-size: ${e=>e.theme.fontSizes[3]};
  line-height: 2.4rem;
  display: block;
  width: 100%;
  background: inherit;

  ${({theme:e})=>e.breakpoints.medium} {
    font-size: ${({theme:e})=>e.fontSizes[2]};
    line-height: 2.2rem;
  }

  &::placeholder {
    color: ${({theme:e})=>e.colors.neutral600};
    opacity: 1;
  }

  &[aria-disabled='true'] {
    color: inherit;
  }

  //focus managed by InputWrapper
  &:focus {
    outline: none;
    box-shadow: none;
  }

  ${e=>{switch(e.$size){case"S":return s.css`
          padding-inline-start: ${e.$hasLeftAction?0:e.theme.spaces[4]};
          padding-inline-end: ${e.$hasRightAction?0:e.theme.spaces[4]};
          padding-block: ${e.theme.spaces[2]};

          ${({theme:t})=>t.breakpoints.medium} {
            padding-block: ${e.theme.spaces[1]};
          }
        `;default:return s.css`
          padding-inline-start: ${e.$hasLeftAction?0:e.theme.spaces[4]};
          padding-inline-end: ${e.$hasRightAction?0:e.theme.spaces[4]};
          padding-block: ${e.theme.spaces[3]};

          ${({theme:t})=>t.breakpoints.medium} {
            padding-block: ${e.theme.spaces[2]};
          }
        `}}}
`,vs=s.styled(C)`
  position: absolute;
  right: ${({theme:e})=>e.spaces[4]};
  top: 50%;
  transform: translateY(-50%);
`,ks=s.styled(C)`
  border: 1px solid ${({theme:e,$hasError:t})=>t?e.colors.danger600:e.colors.neutral200};
  border-radius: ${({theme:e})=>e.borderRadius};
  background: ${({theme:e})=>e.colors.neutral0};
  padding-inline-start: ${({$hasLeftAction:e,theme:t})=>e?t.spaces[4]:0};
  position: relative;

  ${be()}
  ${({theme:e,$disabled:t})=>t?s.css`
          color: ${e.colors.neutral600};
          background: ${e.colors.neutral150};
        `:void 0};
`,tn=()=>{const{id:e,hint:t,error:n}=ne("Hint");return!t||n?null:r.jsx(T,{variant:"pi",tag:"p",id:`${e}-hint`,textColor:"neutral600",children:t})},rn=()=>{const{id:e,error:t}=ne("Error");return!t||typeof t!="string"?null:r.jsx(T,{variant:"pi",tag:"p",id:`${e}-error`,textColor:"danger600","data-strapi-field-error":!0,children:t})},Ss=l.forwardRef(({label:e,children:t,...n},o)=>r.jsx(Rs,{justifyContent:"unset",background:"transparent",borderStyle:"none",...n,type:"button",tag:"button",ref:o,children:r.jsx(Se,{label:e,children:t})})),Rs=s.styled(C)`
  font-size: 1.6rem;
  padding: 0;
`,Ts=Object.freeze(Object.defineProperty({__proto__:null,Action:Ss,Error:rn,Hint:tn,Input:ut,Label:Bt,Root:Re,useField:ne},Symbol.toStringTag,{value:"Module"})),Is=l.forwardRef(({actions:e,children:t,error:n,hint:o,label:a,labelAction:i,nextLabel:c,onNext:d,onPrevious:u,previousLabel:h,required:f,secondaryLabel:g,selectedSlide:p,id:w,...m},$)=>{const k=re(w);return r.jsx(Re,{hint:o,error:n,id:k,required:f,children:r.jsxs(C,{direction:"column",alignItems:"stretch",gap:1,children:[a&&r.jsx(Bt,{action:i,children:a}),r.jsx(Jr,{ref:$,actions:e,label:a,nextLabel:c,onNext:d,onPrevious:u,previousLabel:h,secondaryLabel:g,selectedSlide:p,id:k,...m,children:t}),r.jsx(tn,{}),r.jsx(rn,{})]})})}),Ds=({horizontal:e=!0,...t})=>r.jsx(C,{justifyContent:"center",direction:e?"row":"column",alignItems:e?"center":"stretch",position:"absolute",width:"100%",bottom:1,gap:1,...t}),sr=s.styled(v)`
  ${jr}
`,Es=e=>{const[t,n]=l.useState(!1),o=()=>{n(!0)};return t?r.jsx(ct,{label:e.alt??"",children:r.jsx(sr,{tag:"img",height:"100%",maxWidth:"100%",...e})}):r.jsx(sr,{tag:"img",height:"100%",maxWidth:"100%",...e,onError:o})},As=s.styled(C)`
  display: ${({$selected:e})=>e?"flex":"none"};
`,zs=({label:e,children:t,selected:n=!1,...o})=>r.jsx(As,{$selected:n,alignItems:"center",role:"group","aria-roledescription":"slide","aria-label":e,justifyContent:"center",height:"124px",width:"100%",...o,children:t}),nn=s.css`
  font-size: 1.6rem;
  line-height: 2.4rem;

  ${({theme:e})=>e.breakpoints.medium} {
    font-size: 1.4rem;
    line-height: 2.2rem;
  }
`,Nt=({$size:e="M",$hasValue:t=!1,$hasClear:n=!1,theme:o})=>{switch(e){case"S":return s.css`
        padding-block: calc(
          ${t&&n?o.spaces[1]:o.spaces[2]} - 1px
        ); // 1px to compensate for the border

        ${o.breakpoints.medium} {
          padding-block: ${o.spaces[1]};
        }
      `;default:return s.css`
        padding-block: calc(
          ${t&&n?o.spaces[2]:o.spaces[3]} - 1px
        ); // 1px to compensate for the border

        ${o.breakpoints.medium} {
          padding-block: ${o.spaces[2]};
        }
      `}},Qe=l.forwardRef(({children:e,viewportRef:t,...n},o)=>r.jsxs(Ls,{ref:o,...n,children:[r.jsx(Os,{ref:t,children:e}),r.jsx(ir,{orientation:"vertical",children:r.jsx(lr,{})}),r.jsx(ir,{orientation:"horizontal",children:r.jsx(lr,{})})]})),Ls=s.styled(ot.Root)`
  width: 100%;
  height: 100%;
  overflow: hidden;
  display: flex;
`,Os=s.styled(ot.Viewport)`
  min-width: 100%;
`,ir=s.styled(ot.Scrollbar)`
  display: flex;
  /* ensures no selection */
  user-select: none;
  /* disable browser handling of all panning and zooming gestures on touch devices */
  touch-action: none;

  &[data-orientation='vertical'] {
    width: 0.4rem;
    margin: 0.4rem;
  }

  &[data-orientation='horizontal'] {
    flex-direction: column;
    height: 0.4rem;
    margin: 0.4rem;
  }
`,lr=s.styled(ot.Thumb)`
  position: relative;
  flex: 1;
  background-color: ${e=>e.theme.colors.neutral150};
  border-radius: 0.4rem;

  /* increase target size for touch devices https://www.w3.org/WAI/WCAG21/Understanding/target-size.html */
  @media (hover: none), (pointer: coarse) {
    &::before {
      content: '';
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      width: 100%;
      height: 100%;
      min-width: 44px;
      min-height: 44px;
    }
  }
`,Ps="data:image/svg+xml,%3csvg%20width='63'%20height='63'%20viewBox='0%200%2063%2063'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20d='M42.5563%2011.9816C39.484%2010.3071%2035.8575%209.29097%2032.3354%209.13521C28.6443%208.92888%2024.8295%209.72318%2021.3336%2011.4129C20.9123%2011.5901%2020.5376%2011.8101%2020.1722%2012.0249L20.0108%2012.1179C19.8774%2012.1951%2019.7441%2012.2724%2019.608%2012.3536C19.3253%2012.5146%2019.0492%2012.6744%2018.7544%2012.8792C18.5463%2013.0329%2018.3395%2013.1759%2018.1301%2013.323C17.5658%2013.7208%2016.9868%2014.1317%2016.4983%2014.5979C14.8476%2015.9524%2013.5571%2017.6075%2012.6071%2018.9214C10.4365%2022.1566%209.08622%2025.9567%208.80702%2029.6143L8.7764%2030.1588C8.73328%2030.9196%208.68476%2031.7057%208.75353%2032.4555C8.76648%2032.6084%208.7661%2032.7638%208.77506%2032.914C8.78895%2033.229%208.80152%2033.5373%208.846%2033.8672L9.07396%2035.4221C9.09756%2035.5764%209.1198%2035.7413%209.1633%2035.9263L9.65919%2037.9272L10.138%2039.2823C10.2729%2039.6673%2010.4158%2040.0751%2010.6%2040.43C12.0292%2043.637%2014.1425%2046.4578%2016.7063%2048.585C19.0508%2050.5296%2021.824%2052.0023%2024.7491%2052.8452L26.2371%2053.2376C26.3781%2053.2693%2026.4926%2053.2889%2026.6031%2053.3058L26.7775%2053.3311C27.0052%2053.3636%2027.2195%2053.3986%2027.4445%2053.435C27.8598%2053.5076%2028.2672%2053.5748%2028.7079%2053.6183L30.5641%2053.7229C30.9516%2053.7249%2031.3352%2053.7068%2031.7081%2053.6874C31.9039%2053.681%2032.0984%2053.6681%2032.3288%2053.662C34.5253%2053.4772%2036.5106%2053.0634%2038.0516%2052.4652C38.1769%2052.4171%2038.3008%2052.3796%2038.4234%2052.3355C38.6727%2052.2499%2038.9259%2052.167%2039.1432%2052.0599L40.8591%2051.2626L42.5702%2050.266C42.9009%2050.0682%2043.0205%2049.6414%2042.8282%2049.2984C42.632%2048.9526%2042.2034%2048.8308%2041.8634%2049.0166L40.1792%2049.9218L38.4995%2050.6224C38.3169%2050.6953%2038.121%2050.7534%2037.9224%2050.8155C37.7838%2050.8489%2037.6518%2050.8983%2037.5012%2050.9408C36.0711%2051.435%2034.2445%2051.7425%2032.244%2051.8346C32.0442%2051.8383%2031.8471%2051.8379%2031.654%2051.8403C31.3051%2051.8414%2030.9602%2051.8451%2030.6392%2051.8305L28.9177%2051.6725C28.5476%2051.619%2028.1695%2051.5427%2027.7848%2051.4678C27.5639%2051.4167%2027.3376%2051.3737%2027.1299%2051.3374L26.9529%2051.2987C26.8704%2051.2834%2026.7772%2051.2667%2026.7333%2051.2543L25.3466%2050.8322C22.7651%2049.9789%2020.33%2048.5729%2018.2942%2046.7557C16.1056%2044.7951%2014.3339%2042.2335%2013.1742%2039.3582C12.0276%2036.6013%2011.5988%2033.2792%2011.9716%2030.0076C12.3145%2027.0213%2013.3948%2024.1635%2015.1858%2021.5083C15.3034%2021.3339%2015.421%2021.1596%2015.5212%2021.0196C16.4309%2019.8688%2017.5408%2018.5589%2018.9483%2017.496C19.3367%2017.1525%2019.7862%2016.856%2020.2611%2016.5478C20.4878%2016.4009%2020.7079%2016.2553%2020.8907%2016.1306C21.0974%2016.0048%2021.3188%2015.8831%2021.5348%2015.7694C21.6761%2015.6975%2021.8162%2015.619%2021.9388%2015.5576L22.1002%2015.4646C22.4002%2015.3037%2022.6749%2015.1546%2022.9908%2015.039L24.1186%2014.5715C24.3399%2014.4844%2024.5718%2014.4159%2024.7997%2014.3447C24.953%2014.2982%2025.0982%2014.2635%2025.2635%2014.2078C25.786%2014.0182%2026.3283%2013.9112%2026.9105%2013.7965C27.117%2013.7571%2027.3302%2013.7163%2027.5608%2013.6585C27.7553%2013.611%2027.9737%2013.5969%2028.2082%2013.5762C28.364%2013.5603%2028.5172%2013.5483%2028.6318%2013.5333C28.7876%2013.5173%2028.9342%2013.5066%2029.0927%2013.4867C29.3285%2013.4555%2029.5456%2013.4347%2029.7494%2013.4337C30.0237%2013.44%2030.2994%2013.4357%2030.5777%2013.4274C31.0811%2013.421%2031.5579%2013.4197%2032.0318%2013.4914C34.9664%2013.7352%2037.7144%2014.6085%2040.2052%2016.0868C42.3489%2017.3655%2044.2716%2019.1525%2045.7607%2021.264C47.0255%2023.0628%2047.9756%2025.0528%2048.4928%2027.0393C48.572%2027.3176%2048.6299%2027.5931%2048.6839%2027.8659C48.7154%2028.0428%2048.7563%2028.2145%2048.7892%2028.3636C48.8037%2028.4541%2048.8208%2028.5406%2048.8445%2028.6258C48.8749%2028.7443%2048.8986%2028.864%2048.9116%2028.9651L48.9793%2029.6047C48.9922%2029.7748%2049.0132%2029.9331%2049.0301%2030.0887C49.0668%2030.3268%2049.0889%2030.5608%2049.0964%2030.7561L49.1083%2031.9001C49.1312%2032.3307%2049.089%2032.7116%2049.0522%2033.0673C49.0384%2033.2598%2049.0126%2033.4443%2049.0123%2033.5824C48.9961%2033.6926%2048.9918%2033.7935%2048.9836%2033.8917C48.9753%2034.0072%2048.9724%2034.1148%2048.9414%2034.2554L48.5449%2036.3059C48.3134%2037.8623%2049.3793%2039.3365%2050.9488%2039.5822C52.0417%2039.7601%2053.1536%2039.2819%2053.7711%2038.3664C54.0063%2038.0176%2054.1604%2037.6257%2054.2227%2037.2064L54.5217%2035.2574C54.5514%2035.0756%2054.572%2034.83%2054.5846%2034.5791L54.6028%2034.2338C54.6098%2034.0598%2054.6223%2033.8779%2054.6347%2033.6788C54.6734%2033.1052%2054.7163%2032.4479%2054.6619%2031.8058L54.5867%2030.4289C54.5622%2030.0952%2054.5097%2029.76%2054.4559%2029.4181C54.431%2029.2572%2054.4048%2029.0896%2054.3826%2028.9074L54.2687%2028.104C54.2332%2027.9244%2054.1804%2027.7273%2054.1329%2027.5396L54.0643%2027.2454C54.0195%2027.071%2053.9773%2026.8927%2053.9338%2026.7076C53.8455%2026.3309%2053.7479%2025.9422%2053.613%2025.5571C52.84%2023.0292%2051.5383%2020.5194%2049.8338%2018.2799C47.8544%2015.682%2045.3333%2013.5087%2042.5563%2011.9816Z'%20fill='%234945FF'/%3e%3c/svg%3e",nt=l.forwardRef(({children:e,small:t=!1,...n},o)=>r.jsxs("div",{role:"alert","aria-live":"assertive",ref:o,...n,children:[r.jsx(me,{children:e}),r.jsx(Bs,{src:Ps,"aria-hidden":!0,$small:t})]})),Ms=s.keyframes`
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(359deg);
  }
`,Bs=s.styled.img`
  animation: ${Ms} 1s infinite linear;
  will-change: transform;
  ${({$small:e,theme:t})=>e&&`width: ${t.spaces[6]}; height: ${t.spaces[6]};`}
`,Ns=ae((e,t)=>{const{gap:n=0,gridCols:o=12,...a}=e;return r.jsx(Ws,{ref:t,$gap:n,$gridCols:o,...a})}),Ws=s.styled(v)`
  display: grid;
  grid-template-columns: repeat(${({$gridCols:e})=>e}, 1fr);
  ${({theme:e,$gap:t})=>ke({gap:t},e)}
`,Fs=ae(({col:e,s:t,xs:n,m:o,...a},i)=>r.jsx(Hs,{ref:i,$col:e,$s:t,$xs:n,$m:o,...a})),Hs=s.styled(C)`
  grid-column: span ${({$xs:e,$col:t})=>e??t??12};
  max-width: 100%;

  ${({theme:e})=>e.breakpoints.small} {
    grid-column: span ${({$s:e,$xs:t,$col:n})=>e??t??n??12};
  }

  ${({theme:e})=>e.breakpoints.medium} {
    grid-column: span ${({$m:e,$s:t,$xs:n,$col:o})=>e??t??n??o??12};
  }

  ${({theme:e})=>e.breakpoints.large} {
    grid-column: span ${({$col:e,$m:t,$s:n,$xs:o})=>e??t??n??o??12};
  }
`,_s=Object.freeze(Object.defineProperty({__proto__:null,Item:Fs,Root:Ns},Symbol.toStringTag,{value:"Module"})),Vs=({children:e,estimatedItemSize:t=40,overscan:n=10,itemCount:o,renderItem:a})=>{const i=X.useRef(null),[c,d]=X.useState(!1),u=X.useRef(!0);X.useEffect(()=>(u.current=!0,typeof X.startTransition=="function"&&X.startTransition(()=>{u.current&&d(!0)}),()=>{u.current=!1}),[]);const h=X.useMemo(()=>a&&o!==void 0?[]:X.Children.toArray(e),[e,a,o]),f=o??h.length,g=no.useVirtualizer({count:f,getScrollElement:()=>i.current??null,estimateSize:X.useCallback(()=>t,[t]),overscan:n,scrollMargin:0,measureElement:void 0,lanes:1}),p=c&&u.current?g.getVirtualItems():[];return c?r.jsx(v,{ref:i,height:`${g.getTotalSize()>0?g.getTotalSize():0}px`,width:"100%",position:"relative","data-testid":"virtualized-list",style:{willChange:"transform"},children:p.map(w=>{const m=a?a(w.index):h[w.index];return r.jsx(v,{"data-index":w.index,style:{position:"absolute",top:0,left:0,width:"100%",transform:`translate3d(0, ${w.start}px, 0)`},children:m},w.key)})}):r.jsx(v,{ref:i,height:"40px",width:"100%",position:"relative"})},on=l.forwardRef(({allowCustomValue:e,autocomplete:t,children:n,className:o,clearLabel:a="Clear",creatable:i=!1,creatableDisabled:c=!1,creatableStartIcon:d,createMessage:u=J=>`Create "${J}"`,defaultFilterValue:h,defaultTextValue:f,defaultOpen:g=!1,open:p,onOpenChange:w,disabled:m=!1,hasError:$,id:k,filterValue:y,hasMoreItems:x=!1,isPrintableCharacter:I,loading:j=!1,loadingMessage:R="Loading content...",name:b,noOptionsMessage:S=()=>"No results found",onChange:z,onClear:E,onCreateOption:A,onFilterValueChange:U,onInputChange:B,onTextValueChange:N,onLoadMore:H,placeholder:P="Select or enter a value",required:K=!1,size:L="M",startIcon:M,textValue:Z,value:G,...oe},Y)=>{const[J,_]=le({prop:p,defaultProp:g,onChange:w}),[Q,de]=le({prop:Z,defaultProp:e&&!f?G:f,onChange:N}),[$e,ye]=le({prop:y,defaultProp:h,onChange:U}),fe=l.useRef(null),Ie=l.useRef(null),we=te(fe,Y),Be=l.useRef(null),Ne=se=>{E&&!m&&(de(""),ye(""),E(se),fe.current.focus())},Ce=se=>{_(se)},V=se=>{de(se)},De=se=>{ye(se)},We=se=>{B&&B(se)},qn=se=>{z&&z(se)},Un=se=>{H&&x&&!j&&H(se)},Kt=()=>{A&&Q&&i!=="visible"?A(Q):A&&i==="visible"&&(A(),_(!1))},Gn=re(),Yt=`intersection-${ze(Gn)}`;Ge(Ie,Un,{selectorToWatch:`#${Yt}`,skipWhen:!J});const{error:ht,...Ze}=ne("Combobox"),Kn=!!ht||$,ft=Ze.id??k,Yn=Ze.name??b,Qn=Ze.required||K,Qt=l.Children.toArray(n).filter(Boolean).length,Xn=!!!(Q&&Q!==""||$e&&$e!=="")&&Qt>100;let gt;return ht?gt=`${ft}-error`:Ze.hint&&(gt=`${ft}-hint`),r.jsxs(D.Combobox.Root,{autocomplete:t||(i===!0?"list":"both"),onOpenChange:Ce,open:J,onTextValueChange:V,textValue:Q,allowCustomValue:!!i||e,disabled:m,required:Qn,value:G,onValueChange:qn,filterValue:$e,onFilterValueChange:De,isPrintableCharacter:I,visible:i==="visible",children:[r.jsxs(qs,{$hasError:Kn,$size:L,$hasTextValue:!!Q,$hasClear:!!(Q&&E),className:o,children:[r.jsxs(C,{flex:"1",tag:"span",gap:3,children:[M?r.jsx(C,{flex:"0 0 1.6rem",tag:"span","aria-hidden":!0,children:M}):null,r.jsx(Us,{placeholder:P,id:ft,"aria-invalid":!!ht,onChange:We,ref:we,name:Yn,"aria-describedby":gt,...oe})]}),r.jsxs(C,{tag:"span",gap:3,children:[Q&&E?r.jsx(ge,{size:"XS",variant:"ghost",onClick:Ne,"aria-disabled":m,"aria-label":a,label:a,ref:Be,children:r.jsx(O.Cross,{})}):null,j?r.jsx(nt,{small:!0,children:"Loading"}):r.jsx(Gs,{children:r.jsx(O.CaretDown,{fill:"neutral500"})})]})]}),r.jsx(D.Combobox.Portal,{children:r.jsxs(Ks,{sideOffset:4,children:[r.jsx(Qe,{viewportRef:Ie,children:r.jsx(D.Combobox.Viewport,{children:r.jsxs(v,{padding:1,children:[Xn?r.jsx(Vs,{itemCount:Qt,children:n}):n,i!==!0&&!j?r.jsx(D.Combobox.NoValueFound,{asChild:!0,children:r.jsx(jt,{$hasHover:!1,children:r.jsx(T,{children:S(Q??"")})})}):null,j?r.jsx(C,{justifyContent:"center",alignItems:"center",paddingTop:2,paddingBottom:2,children:r.jsx(nt,{small:!0,children:R})}):null,r.jsx(v,{id:Yt,width:"100%",height:"1px"})]})})}),i?r.jsx(Ys,{onPointerUp:Kt,onClick:Kt,disabled:c,asChild:!0,children:r.jsx(jt,{children:r.jsxs(C,{gap:2,children:[d&&r.jsx(v,{tag:"span","aria-hidden":!0,display:"inline-flex",children:d}),r.jsx(T,{children:u(Q??"")})]})})}):null]})})]})}),qs=s.styled(D.Combobox.Trigger)`
  position: relative;
  border: 1px solid ${({theme:e,$hasError:t})=>t?e.colors.danger600:e.colors.neutral200};
  border-radius: ${({theme:e})=>e.borderRadius};
  background: ${({theme:e})=>e.colors.neutral0};
  overflow: hidden;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: ${({theme:e})=>e.spaces[4]};
  padding-inline-start: ${({theme:e})=>e.spaces[4]};
  padding-inline-end: ${({theme:e})=>e.spaces[3]};
  ${({$size:e,$hasTextValue:t,$hasClear:n,theme:o})=>Nt({$size:e||"M",$hasValue:t||!1,$hasClear:n||!1,theme:o})}

  &[data-disabled] {
    color: ${({theme:e})=>e.colors.neutral600};
    background: ${({theme:e})=>e.colors.neutral150};
    cursor: not-allowed;
  }

  /* Required to ensure the below inputFocusStyles are adhered too */
  &:focus-visible {
    outline: none;
  }

  ${({theme:e,$hasError:t})=>be()({theme:e,$hasError:t})};
`,Us=s.styled(D.Combobox.TextInput)`
  width: 100%;
  ${nn}
  color: ${({theme:e})=>e.colors.neutral800};
  padding: 0;
  border: none;
  background-color: transparent;
  text-overflow: ellipsis;

  &:focus-visible {
    outline: none;
  }

  &::placeholder {
    color: ${({theme:e})=>e.colors.neutral600};
    opacity: 1;
  }

  &[aria-disabled='true'] {
    cursor: inherit;
  }
`,Gs=s.styled(D.Combobox.Icon)`
  border: none;
  background: transparent;
  padding: 0;
  color: ${({theme:e})=>e.colors.neutral600};
  display: flex;

  &[aria-disabled='true'] {
    cursor: inherit;
  }
`,Ks=s.styled(D.Combobox.Content)`
  background: ${({theme:e})=>e.colors.neutral0};
  box-shadow: ${({theme:e})=>e.shadows.filterShadow};
  border: 1px solid ${({theme:e})=>e.colors.neutral150};
  border-radius: ${({theme:e})=>e.borderRadius};
  width: var(--radix-combobox-trigger-width);
  /* This is from the design-system figma file. */
  max-height: 35.5rem;
  z-index: ${({theme:e})=>e.zIndices.popover};

  &:focus-visible {
    outline: ${({theme:e})=>`2px solid ${e.colors.primary600}`};
    outline-offset: 2px;
  }

  @media (prefers-reduced-motion: no-preference) {
    animation-duration: ${e=>e.theme.motion.timings[200]};

    /* The select can't animate out yet, watch https://github.com/radix-ui/primitives/issues/1893, or take a look and solve it yourself. */
    &[data-state='open'] {
      animation-timing-function: ${e=>e.theme.motion.easings.authenticMotion};

      &[data-side='top'] {
        animation-name: ${q.slideUpIn};
      }

      &[data-side='bottom'] {
        animation-name: ${q.slideDownIn};
      }
    }
  }
`,Ys=s.styled(D.Combobox.CreateItem)`
  && {
    border-top: 1px solid ${({theme:e})=>e.colors.neutral150};
    border-top-left-radius: 0;
    border-top-right-radius: 0;
    background: ${({theme:e})=>e.colors.neutral0};
    cursor: pointer;
    padding: ${({theme:e})=>e.spaces[1]};
    position: sticky;
    bottom: 0;
    left: 0;
  }
  &&:hover,
  &&[data-highlighted] {
    background: ${({theme:e})=>e.colors.neutral0};
  }
  &&[data-disabled] {
    color: ${({theme:e})=>e.colors.neutral600};
    cursor: not-allowed;
  }
  &&[data-disabled] svg {
    fill: ${({theme:e})=>e.colors.neutral300};
  }
  && > div {
    padding: ${({theme:e})=>e.spaces[2]} ${({theme:e})=>e.spaces[4]};
  }
  && > div:hover,
  &&[data-highlighted] > div {
    background-color: ${({theme:e})=>e.colors.primary100};
    border-radius: ${({theme:e})=>e.borderRadius};
  }
  &&[data-disabled] > div {
    background-color: inherit;
  }
`,an=l.forwardRef(({children:e,value:t,disabled:n,textValue:o,...a},i)=>r.jsx(D.Combobox.ComboboxItem,{asChild:!0,value:t,disabled:n,textValue:o,children:r.jsx(jt,{ref:i,...a,children:r.jsx(D.Combobox.ItemText,{asChild:!0,children:r.jsx(T,{children:e})})})})),jt=s.styled.div`
  width: 100%;
  border: none;
  text-align: left;
  outline-offset: -3px;
  padding: ${({theme:e})=>e.spaces[2]} ${({theme:e})=>e.spaces[4]};
  background-color: ${({theme:e})=>e.colors.neutral0};
  border-radius: ${({theme:e})=>e.borderRadius};
  user-select: none;

  &[data-state='checked'] {
    background-color: ${({theme:e})=>e.colors.primary100};
    color: ${({theme:e})=>e.colors.primary600};
    font-weight: bold;
  }

  &:hover,
  &[data-highlighted] {
    outline: none;
    background-color: ${({theme:e,$hasHover:t=!0})=>t?e.colors.primary100:e.colors.neutral0};
  }

  &[data-highlighted] {
    color: ${({theme:e})=>e.colors.primary600};
    font-weight: bold;
  }
`,Wt=(e,t)=>`${e}${Math.floor(t*255).toString(16).padStart(2,"0")}`,Qs=pe.Root,Xs=l.forwardRef((e,t)=>r.jsx(pe.Trigger,{...e,asChild:!0,ref:t})),Zs=l.forwardRef((e,t)=>r.jsx(pe.Portal,{children:r.jsx(Js,{children:r.jsx(ei,{ref:t,...e})})})),Js=s.styled(pe.Overlay)`
  background: ${e=>Wt(e.theme.colors.neutral800,.2)};
  position: fixed;
  inset: 0;
  z-index: ${e=>e.theme.zIndices.overlay};
  will-change: opacity;

  @media (prefers-reduced-motion: no-preference) {
    animation: ${q.overlayFadeIn} ${e=>e.theme.motion.timings[200]}
      ${e=>e.theme.motion.easings.authenticMotion};
  }
`,ei=s.styled(pe.Content)`
  max-width: 42rem;
  height: min-content;
  width: calc(100% - ${({theme:e})=>e.spaces[8]});
  overflow: hidden;
  margin: 0 auto;
  display: flex;
  flex-direction: column;

  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);

  border-radius: ${e=>e.theme.borderRadius};
  background-color: ${e=>e.theme.colors.neutral0};
  box-shadow: ${e=>e.theme.shadows.popupShadow};
  z-index: ${e=>e.theme.zIndices.modal};

  @media (prefers-reduced-motion: no-preference) {
    &[data-state='open'] {
      animation-duration: ${e=>e.theme.motion.timings[200]};
      animation-timing-function: ${e=>e.theme.motion.easings.authenticMotion};
      animation-name: ${q.modalPopIn};
    }

    &[data-state='closed'] {
      animation-duration: ${e=>e.theme.motion.timings[120]};
      animation-timing-function: ${e=>e.theme.motion.easings.easeOutQuad};
      animation-name: ${q.modalPopOut};
    }
  }
`,ti=l.forwardRef(({children:e,...t},n)=>r.jsx(pe.Title,{asChild:!0,children:r.jsx(ri,{tag:"h2",variant:"beta",ref:n,padding:6,fontWeight:"bold",...t,children:e})})),ri=s.styled(T)`
  display: flex;
  justify-content: center;
  border-bottom: solid 1px ${e=>e.theme.colors.neutral150};
`,ni=l.forwardRef(({children:e,icon:t,...n},o)=>r.jsx(C,{ref:o,gap:2,direction:"column",paddingTop:8,paddingBottom:8,paddingLeft:6,paddingRight:6,...n,children:typeof e=="string"?r.jsxs(r.Fragment,{children:[t?l.cloneElement(t,{width:24,height:24}):null,r.jsx(sn,{children:e})]}):e})),sn=l.forwardRef((e,t)=>r.jsx(pe.Description,{asChild:!0,children:r.jsx(T,{ref:t,variant:"omega",...e,tag:"p"})})),oi=l.forwardRef((e,t)=>r.jsx(ai,{ref:t,gap:2,padding:4,justifyContent:"space-between",...e,tag:"footer"})),ai=s.styled(C)`
  border-top: solid 1px ${e=>e.theme.colors.neutral150};
  flex: 1;
`,si=l.forwardRef((e,t)=>r.jsx(pe.Cancel,{...e,asChild:!0,ref:t})),ii=l.forwardRef((e,t)=>r.jsx(pe.Action,{...e,asChild:!0,ref:t})),li=Object.freeze(Object.defineProperty({__proto__:null,Action:ii,Body:ni,Cancel:si,Content:Zs,Description:sn,Footer:oi,Header:ti,Root:Qs,Trigger:Xs},Symbol.toStringTag,{value:"Module"}));function ue(e,t){const n=l.useRef(null);return t&&n.current&&ci(t,n.current)&&(t=n.current),n.current=t??null,l.useMemo(()=>new F.DateFormatter(e,t),[e,t])}function ci(e,t){if(e===t)return!0;const n=Object.keys(e),o=Object.keys(t);if(n.length!==o.length)return!1;for(const a of n)if(t[a]!==e[a])return!1;return!0}const ln=()=>typeof window>"u"||!window.navigator||/ServerSideRendering|^Deno\//.test(window.navigator.userAgent),he=ln()?X.useEffect:X.useLayoutEffect,Ae={Log:"live-region-log",Status:"live-region-status",Alert:"live-region-alert"},cr=e=>{const t=document.querySelector(`#${Ae.Log}`);t&&(t.innerText=e)},dr=e=>{const t=document.querySelector(`#${Ae.Status}`);t&&(t.innerText=e)},ur=e=>{const t=document.querySelector(`#${Ae.Alert}`);t&&(t.innerText=e)},di=()=>(l.useEffect(()=>()=>{cr(""),ur(""),dr("")},[]),{notifyLog:cr,notifyAlert:ur,notifyStatus:dr}),cn=()=>r.jsxs(me,{children:[r.jsx("p",{role:"log","aria-live":"polite",id:Ae.Log,"aria-relevant":"all"}),r.jsx("p",{role:"status","aria-live":"polite",id:Ae.Status,"aria-relevant":"all"}),r.jsx("p",{role:"alert","aria-live":"assertive",id:Ae.Alert,"aria-relevant":"all"})]}),ui=s.createGlobalStyle`
${s.css`
  *,
  *::before,
  *::after {
    box-sizing: border-box;
  }

  * {
    margin: 0;
  }

  html {
    /* Sets 1rem === 10px */
    font-size: 62.5%;
  }

  body {
    height: 100%;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans',
      'Helvetica Neue', sans-serif;
    color: ${({theme:e})=>e.colors.neutral800};
    line-height: 1.5;
    -webkit-font-smoothing: antialiased;
  }

  img,
  picture,
  video,
  canvas {
    display: block;
    max-width: 100%;
  }

  input,
  button,
  textarea,
  select {
    padding: 0;
    font: inherit;
  }

  button {
    border: unset;
    background: unset;
    padding: unset;
    margin: unset;
  }

  p,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    overflow-wrap: break-word;
    font: unset;
  }

  #root {
    isolation: isolate;
  }

  ol,
  ul {
    list-style: none;
    padding: unset;
    margin: unset;
  }

  table {
    border-collapse: collapse;
    border-spacing: 0;
  }

  *:focus-visible {
    outline: 2px solid ${({theme:e})=>e.colors.primary600};
    outline-offset: 2px;
  }

  *:has(> :disabled:not(button)) {
    cursor: not-allowed !important;
  }

  [aria-disabled='true']:not(button) {
    cursor: not-allowed !important;
  }

  /* Focusing the button with a mouse, touch, or stylus will show a subtle drop shadow. */
  *:focus:not(:focus-visible) {
    outline: none;
  }

  .lock-body-scroll {
    height: 100vh;
    overflow-y: hidden;
  }
`}
`,hr="en-EN",dn=()=>typeof navigator>"u"?hr:navigator.language?navigator.language:hr,[hi,Xe]=Oe("StrapiDesignSystem",{locale:dn()}),fi=({children:e,locale:t=dn(),theme:n=Mt,tooltipConfig:o})=>(he(()=>{const a=document.createElement("style");a.type="text/css",a.appendChild(document.createTextNode(`
        * {
          -webkit-transition: none !important;
          -moz-transition: none !important;
          -o-transition: none !important;
          -ms-transition: none !important;
          transition: none !important;
          animation: none !important;
        }
    `)),document.head.appendChild(a),window.getComputedStyle(a).opacity,document.head.removeChild(a)},[n]),r.jsx(hi,{locale:t,children:r.jsxs(s.ThemeProvider,{theme:n,children:[r.jsx($r.Provider,{...o,children:e}),r.jsx(cn,{}),r.jsx(ui,{})]})})),un=xo.Root,Ft=l.forwardRef(({container:e=(o=>(o=globalThis==null?void 0:globalThis.document)==null?void 0:o.body)(),...t},n)=>e?co.createPortal(r.jsx(v,{ref:n,...t}),e):null);Ft.displayName="Portal";const gi=l.forwardRef(({onClear:e,clearLabel:t="Clear",startIcon:n,disabled:o,hasError:a,loading:i,children:c,id:d,size:u="M",withTags:h,...f},g)=>{const p=l.useRef(null),w=k=>{e&&!o&&(e(k),p.current.focus())},{labelNode:m}=ne("SelectTrigger"),$=te(p,g);return r.jsx(D.Select.Trigger,{asChild:!0,children:r.jsxs(pi,{"aria-disabled":o,$hasError:a,ref:$,alignItems:"center",justifyContent:"space-between",position:"relative",overflow:"hidden",hasRadius:!0,background:o?"neutral150":"neutral0",gap:4,cursor:"default","aria-labelledby":m?`${d}-label`:void 0,$size:u,$withTags:h,$hasClear:!!e,...f,children:[r.jsxs(C,{flex:"1",tag:"span",gap:3,overflow:"hidden",children:[n&&r.jsx(C,{tag:"span","aria-hidden":!0,children:n}),c]}),r.jsxs(C,{tag:"span",gap:3,children:[e?r.jsx(ge,{size:"XS",variant:"ghost",onClick:w,"aria-disabled":o,"aria-label":t,label:t,children:r.jsx(O.Cross,{})}):null,i?r.jsx(nt,{small:!0,children:"Loading"}):r.jsx(mi,{children:r.jsx(O.CaretDown,{})})]})]})})}),pi=s.styled(C)`
  border: 1px solid ${({theme:e,$hasError:t})=>t?e.colors.danger600:e.colors.neutral200};
  padding-inline-start: ${({theme:e})=>e.spaces[4]};
  padding-inline-end: ${({theme:e})=>e.spaces[3]};

  ${({$size:e,$hasClear:t,$withTags:n,theme:o})=>n?s.css`
        padding-inline-start: calc(${o.spaces[2]} - 1px);
        padding-block: calc(${o.spaces[2]} - 1px);

        ${o.breakpoints.medium} {
          padding-inline-start: 0.3rem;
          padding-block: 0.3rem;
        }
      `:Nt({$size:e||"M",$hasValue:t,$hasClear:t,theme:o})}
  cursor: pointer;

  &[aria-disabled='true'] {
    color: ${e=>e.theme.colors.neutral500};
  }

  /* Required to ensure the below inputFocusStyles are adhered too */
  &:focus-visible {
    outline: none;
  }

  ${({theme:e,$hasError:t})=>be()({theme:e,$hasError:t})};
`,mi=s.styled(D.Select.Icon)`
  display: flex;
  & > svg {
    fill: ${({theme:e})=>e.colors.neutral500};
  }
`,bi=l.forwardRef(({children:e,placeholder:t,withTags:n,...o},a)=>r.jsx(xi,{ref:a,ellipsis:!0,...o,children:r.jsx($i,{placeholder:t,$withTags:n,children:e})})),xi=s.styled(T)`
  flex: 1;
  min-height: 2.4rem;

  ${({theme:e})=>e.breakpoints.medium} {
    min-height: 2.2rem;
  }
`,$i=s.styled(D.Select.Value)`
  ${({$withTags:e})=>e?s.css`
          display: flex;
          flex-wrap: wrap;
          gap: ${({theme:t})=>t.spaces[1]};
        `:null};
`,yi=l.forwardRef((e,t)=>r.jsx(wi,{ref:t,...e,children:r.jsx(Qe,{children:e.children})})),wi=s.styled(D.Select.Content)`
  background: ${({theme:e})=>e.colors.neutral0};
  box-shadow: ${({theme:e})=>e.shadows.filterShadow};
  border: 1px solid ${({theme:e})=>e.colors.neutral150};
  border-radius: ${({theme:e})=>e.borderRadius};
  min-width: var(--radix-select-trigger-width);
  max-height: 15.6rem;
  z-index: ${({theme:e})=>e.zIndices.popover};

  @media (prefers-reduced-motion: no-preference) {
    animation-duration: ${e=>e.theme.motion.timings[200]};

    /* The select can't animate out yet, watch https://github.com/radix-ui/primitives/issues/1893, or take a look and solve it yourself. */
    &[data-state='open'] {
      animation-timing-function: ${e=>e.theme.motion.easings.authenticMotion};

      &[data-side='top'] {
        animation-name: ${q.slideUpIn};
      }

      &[data-side='bottom'] {
        animation-name: ${q.slideDownIn};
      }
    }
  }
`,Ci=s.styled(D.Select.Viewport)`
  padding: ${({theme:e})=>e.spaces[1]};
`,ji=l.forwardRef((e,t)=>r.jsx(ki,{ref:t,...e})),vi=s.css`
  width: 100%;
  border: none;
  text-align: left;
  outline-offset: -3px;
  border-radius: ${e=>e.theme.borderRadius};
  padding: ${e=>`${e.theme.spaces[2]} ${e.theme.spaces[4]}`};
  padding-left: ${({theme:e})=>e.spaces[4]};
  background-color: ${({theme:e})=>e.colors.neutral0};
  display: flex;
  align-items: center;
  gap: ${({theme:e})=>e.spaces[2]};
  white-space: nowrap;
  user-select: none;
  color: ${({theme:e})=>e.colors.neutral800};

  &:focus-visible {
    outline: none;
    background-color: ${({theme:e})=>e.colors.primary100};
    cursor: pointer;
  }
`,ki=s.styled(D.Select.Item)`
  ${vi}

  &:hover {
    background-color: ${({theme:e})=>e.colors.primary100};
    cursor: pointer;
  }
`,hn=D.Select.Root,fn=gi,gn=bi,pn=D.Select.Portal,mn=yi,bn=Ci,Ht=ji,xn=D.Select.ItemIndicator,$n=D.Select.ItemText,Si=D.Select.Group,vt=l.forwardRef(({children:e,clearLabel:t="Clear",customizeContent:n,disabled:o,hasError:a,id:i,loading:c,name:d,onChange:u,onClear:h,onCloseAutoFocus:f,onReachEnd:g,placeholder:p,required:w,size:m,startIcon:$,value:k,...y},x)=>{const[I,j]=l.useState(),[R,b]=l.useState(!1),S=Y=>{b(Y)},z=Y=>{h&&h(Y),u||j("")},E=Y=>{u?u(typeof k=="number"?Number(Y):Y):j(Y)},A=l.useRef(null),U=re(),B=`intersection-${ze(U)}`;Ge(A,Y=>{g&&g(Y)},{selectorToWatch:`#${B}`,skipWhen:!R});const{error:H,required:P,...K}=ne("SingleSelect"),L=!!H||a,M=K.id??i,Z=K.name??d;let G;H?G=`${M}-error`:K.hint&&(G=`${M}-hint`);const oe=(typeof k<"u"&&k!==null?k.toString():I)??"";return r.jsxs(hn,{onOpenChange:S,disabled:o,required:P??w,onValueChange:E,value:oe,...y,children:[r.jsx(fn,{ref:x,id:M,name:Z,startIcon:$,hasError:L,disabled:o,clearLabel:t,loading:c,onClear:oe&&h?z:void 0,"aria-label":y["aria-label"],"aria-describedby":G??y["aria-describedby"],size:m,children:r.jsx(gn,{placeholder:p,textColor:oe?"neutral800":"neutral600",children:oe&&n?n(oe):void 0})}),r.jsx(pn,{children:r.jsx(mn,{position:"popper",sideOffset:4,onCloseAutoFocus:f,children:r.jsxs(bn,{ref:A,children:[e,r.jsx(v,{id:B,width:"100%",height:"1px"})]})})})]})}),kt=l.forwardRef(({value:e,startIcon:t,children:n,...o},a)=>r.jsxs(Ht,{ref:a,value:e.toString(),...o,children:[t&&r.jsx(C,{tag:"span","aria-hidden":!0,children:t}),r.jsx(T,{lineHeight:"20px",width:"100%",children:r.jsx($n,{children:n})})]})),Ri=200,fr=15,[Ti,Me]=Oe("DatePicker"),Ii=l.forwardRef(({calendarLabel:e,className:t,initialDate:n,locale:o,maxDate:a,minDate:i,monthSelectLabel:c="Month",onChange:d,value:u,yearSelectLabel:h="Year",hasError:f,id:g,name:p,disabled:w=!1,required:m=!1,onClear:$,clearLabel:k="Clear",size:y="M",...x},I)=>{const j=F.getLocalTimeZone(),R=Xe("DatePicker"),b=o??R.locale,S=ue(b,{day:"2-digit",month:"2-digit",year:"numeric"}),[z,E]=l.useState(!1),[A,U]=l.useState(null),[B,N]=l.useState(null),[H,P]=l.useState(null),[K,L]=l.useState(),[M,Z]=le({defaultProp:n?je(n):void 0,prop:u?je(u):void 0,onChange(V){d&&d(V==null?void 0:V.toDate(j))}}),[G,oe]=l.useMemo(()=>{const V=n?je(n):F.today("UTC"),De=i?je(i):V.set({day:1,month:1,year:V.year-Ri});let We=a?je(a):V.set({day:31,month:12,year:V.year+fr});return We.compare(De)<0&&(We=De.set({day:31,month:12,year:De.year+fr})),[De,We]},[i,a,n]),[Y,J]=l.useState(Di({currentValue:M,minDate:G,maxDate:oe})),_=re(),Q=l.useRef(null),de=V=>{$&&!w&&(L(""),Z(void 0),$(V),B==null||B.focus())},$e=l.useCallback(V=>{V&&M&&J(M),E(V)},[M]);he(()=>{if(u){const V=je(u);L(S.format(V.toDate(j))),J(V)}else L("")},[u,S,j]),he(()=>{if(n&&K===void 0){const V=je(n);L(S.format(V.toDate(j)))}},[n,K,S,j]);const{error:ye,...fe}=ne("Combobox"),Ie=!!ye||f,we=fe.id??g,Be=fe.name??p,Ne=fe.required||m;let Ce;return ye?Ce=`${we}-error`:fe.hint&&(Ce=`${we}-hint`),r.jsxs(Ti,{calendarDate:Y,content:H,contentId:_,disabled:w,locale:b,minDate:G,maxDate:oe,open:z,onCalendarDateChange:J,onContentChange:P,onOpenChange:$e,onTextInputChange:N,onTextValueChange:L,onTriggerChange:U,onValueChange:Z,onClear:$,required:Ne,textInput:B,textValue:K,timeZone:j,trigger:A,value:M,children:[r.jsxs(Ai,{className:t,hasError:Ie,size:y,children:[r.jsx(O.Calendar,{fill:"neutral500","aria-hidden":!0}),r.jsx(Oi,{ref:I,"aria-describedby":Ce,id:we,name:Be,...x}),K&&$?r.jsx(ge,{size:"XS",variant:"ghost",onClick:de,"aria-disabled":w,"aria-label":k,label:k,ref:Q,children:r.jsx(O.Cross,{})}):null]}),r.jsx(Ft,{children:r.jsx(Ni,{label:e,children:r.jsx(Ui,{monthSelectLabel:c,yearSelectLabel:h})})})]})}),gr=e=>!!e.match(/^[^a-zA-Z]*$/),Di=({currentValue:e,minDate:t,maxDate:n})=>{const o=F.today("UTC");return e||(F.minDate(t,o)===t&&F.maxDate(n,o)===n?o:F.minDate(t,o)===o?t:F.maxDate(n,o)===o?n:o)},Ei="DatePickerTrigger",Ai=l.forwardRef(({hasError:e,size:t,...n},o)=>{const a=Me(Ei),i=te(o,d=>a.onTriggerChange(d)),c=()=>{a.disabled||a.onOpenChange(!0)};return r.jsx(io.FocusScope,{asChild:!0,trapped:a.open,onMountAutoFocus:d=>{d.preventDefault()},onUnmountAutoFocus:d=>{var u;(u=document.getSelection())==null||u.empty(),d.preventDefault()},children:r.jsx(zi,{ref:i,$hasError:e,$hasTextValue:!!a.textValue,$size:t,$hasOnClear:!!a.onClear,...n,hasRadius:!0,gap:3,overflow:"hidden",background:a.disabled?"neutral150":"neutral0",onClick:D.composeEventHandlers(n.onClick,()=>{var d;(d=a.textInput)==null||d.focus()}),onPointerDown:D.composeEventHandlers(n.onPointerDown,d=>{var f;const u=d.target;u.hasPointerCapture(d.pointerId)&&u.releasePointerCapture(d.pointerId),(u.closest("button")??u.closest("div"))===d.currentTarget&&d.button===0&&d.ctrlKey===!1&&(c(),(f=a.textInput)==null||f.focus())})})})}),zi=s.styled(C)`
  min-width: ${({$hasOnClear:e})=>e?"160px":"130px"};
  border: 1px solid ${({theme:e,$hasError:t})=>t?e.colors.danger600:e.colors.neutral200};
  padding-inline: ${({theme:e})=>e.spaces[3]};
  ${({$size:e,$hasTextValue:t,$hasOnClear:n,theme:o})=>Nt({$size:e||"M",$hasValue:t||!1,$hasClear:n||!1,theme:o})}

  & > svg {
    flex: 1 0 auto;
  }

  &[data-disabled] {
    color: ${({theme:e})=>e.colors.neutral600};
    background: ${({theme:e})=>e.colors.neutral150};
    cursor: not-allowed;
  }

  /* Required to ensure the below inputFocusStyles are adhered too */
  &:focus-visible {
    outline: none;
  }

  ${({theme:e,$hasError:t})=>be()({theme:e,$hasError:t})};
`,Li="DatePickerTextInput",Oi=l.forwardRef(({placeholder:e,...t},n)=>{const o=Me(Li),{onTextValueChange:a,textValue:i,onTextInputChange:c,onOpenChange:d,disabled:u,locale:h}=o,f=te(n,y=>c(y)),g=()=>{u||d(!0)},p=ue(h,{year:"numeric",month:"2-digit",day:"2-digit"}),[w,m,$]=l.useMemo(()=>{var R;const y=p.formatToParts(new Date),x=y.filter(b=>b.type==="year"||b.type==="month"||b.type==="day"),I=x.map(b=>{switch(b.type){case"day":return"DD";case"month":return"MM";case"year":return"YYYY";default:return""}}),j=((R=y.find(b=>b.type==="literal"))==null?void 0:R.value)??"";return[I,j,x]},[p]),k=w.map(y=>`\\d{${y.length}}`).join(`\\${m}`);return r.jsx(Mi,{role:"combobox",type:"text",inputMode:"numeric",ref:f,"aria-autocomplete":"none","aria-controls":o.contentId,"aria-disabled":o.disabled,"aria-expanded":o.open,"aria-required":o.required,"aria-haspopup":"dialog","data-state":o.open?"open":"closed",disabled:u,"data-disabled":u?"":void 0,pattern:k,placeholder:e??w.join(m),...t,value:i??"",onBlur:D.composeEventHandlers(t.onBlur,()=>{if(!o.textValue){o.onValueChange(void 0);return}o.onTextValueChange(p.format(o.calendarDate.toDate(o.timeZone))),o.onValueChange(o.calendarDate)}),onChange:D.composeEventHandlers(t.onChange,y=>{if(gr(y.target.value)){const x=y.target.value.split(m),[I,j,R]=$.map((N,H)=>{const P=x[H];return{...N,value:P}}).sort((N,H)=>N.type==="year"?1:H.type==="year"?-1:N.type==="month"?1:H.type==="month"?-1:0).map(N=>N.value),b=o.calendarDate.year;let S=o.calendarDate.year;if(R){const N=R.length===1?`0${R}`:R;S=R.length<3?Number(`${b}`.slice(0,4-N.length)+N):Number(N)}R&&R.length<3&&S>o.maxDate.year&&(S-=100);const z=o.calendarDate.set({year:S}),E=z.calendar.getMonthsInYear(z),A=z.set({month:j&&Number(j)<=E?Number(j):void 0}),U=A.calendar.getDaysInMonth(A),B=A.set({day:I&&Number(I)<=U?Number(I):void 0});o.onCalendarDateChange(Pi(B,o.minDate,o.maxDate)),o.onTextValueChange(y.target.value)}}),onKeyDown:D.composeEventHandlers(t.onKeyDown,y=>{if(!o.open&&(gr(y.key)||["ArrowDown","Backspace"].includes(y.key)))g();else if(["Tab"].includes(y.key)&&o.open)y.preventDefault();else if(["Escape"].includes(y.key))o.open?o.onOpenChange(!1):(o.onValueChange(void 0),o.onTextValueChange("")),y.preventDefault();else if(o.open&&["ArrowDown","ArrowUp","ArrowLeft","ArrowRight"].includes(y.key))switch(y.preventDefault(),y.key){case"ArrowDown":{const x=o.calendarDate.add({weeks:1});if(o.maxDate&&x.compare(o.maxDate)>0)return;o.onCalendarDateChange(x);return}case"ArrowRight":{const x=o.calendarDate.add({days:1});if(o.maxDate&&x.compare(o.maxDate)>0)return;o.onCalendarDateChange(x);return}case"ArrowUp":{const x=o.calendarDate.subtract({weeks:1});if(o.minDate&&x.compare(o.minDate)<0)return;o.onCalendarDateChange(x);return}case"ArrowLeft":{const x=o.calendarDate.subtract({days:1});if(o.minDate&&x.compare(o.minDate)<0)return;o.onCalendarDateChange(x)}}else o.open&&["Enter"].includes(y.key)&&(y.preventDefault(),a(p.format(o.calendarDate.toDate(o.timeZone))),o.onValueChange(o.calendarDate),o.onOpenChange(!1))})})});function Pi(e,t,n){return t&&(e=F.maxDate(e,t)),n&&(e=F.minDate(e,n)),e}const Mi=s.styled.input`
  width: 100%;
  ${nn}
  color: ${({theme:e})=>e.colors.neutral800};
  border: none;
  background-color: transparent;

  &:focus-visible {
    outline: none;
  }

  &::placeholder {
    color: ${({theme:e})=>e.colors.neutral600};
    opacity: 1;
  }

  &[aria-disabled='true'] {
    cursor: inherit;
  }
`,Bi="DatePickerContent",Ni=l.forwardRef((e,t)=>{const n=Me(Bi),[o,a]=l.useState(!1);return l.useEffect(()=>{n.open&&!o&&a(!0)},[n.open,o]),!n.open&&!o||!n.open?null:r.jsx(Fi,{...e,ref:t})}),Wi="DatePickerContent",Fi=l.forwardRef((e,t)=>{const{label:n="Choose date",...o}=e,{onOpenChange:a,...i}=Me(Wi),{x:c,y:d,refs:u,strategy:h,placement:f}=Fe.useFloating({strategy:"fixed",placement:"bottom-start",middleware:[Fe.offset({mainAxis:4}),Fe.shift(),Fe.flip()],elements:{reference:i.trigger},whileElementsMounted:Fe.autoUpdate});l.useEffect(()=>{const p=()=>{a(!1)};return window.addEventListener("blur",p),window.addEventListener("resize",p),()=>{window.removeEventListener("blur",p),window.removeEventListener("resize",p)}},[a]);const g=te(t,p=>i.onContentChange(p),u.setFloating);return so.useFocusGuards(),r.jsx(lo.RemoveScroll,{allowPinchZoom:!0,children:r.jsx(un,{asChild:!0,onFocusOutside:p=>{p.preventDefault()},onDismiss:()=>{a(!1)},children:r.jsx(Hi,{ref:g,"data-state":i.open?"open":"closed","data-side":f.includes("top")?"top":"bottom",onContextMenu:p=>p.preventDefault(),id:i.contentId,role:"dialog","aria-modal":"true","aria-label":n,style:{left:c,top:d,position:h},hasRadius:!0,background:"neutral0",padding:1,...o})})})}),Hi=s.styled(v)`
  box-shadow: ${({theme:e})=>e.shadows.filterShadow};
  z-index: ${({theme:e})=>e.zIndices.popover};
  border: 1px solid ${({theme:e})=>e.colors.neutral150};

  @media (prefers-reduced-motion: no-preference) {
    animation-duration: ${e=>e.theme.motion.timings[200]};

    &[data-state='open'] {
      animation-timing-function: ${e=>e.theme.motion.easings.authenticMotion};

      &[data-side='top'] {
        animation-name: ${q.slideUpIn};
      }

      &[data-side='bottom'] {
        animation-name: ${q.slideDownIn};
      }
    }
  }
`,_i="DatePickerCalendar",[Vi,qi]=Oe("DateFormatters"),Ui=l.forwardRef(({monthSelectLabel:e,yearSelectLabel:t,...n},o)=>{const{locale:a,timeZone:i,minDate:c,maxDate:d,calendarDate:u,onCalendarDateChange:h}=Me(_i),f=F.startOfMonth(u),g=l.useMemo(()=>{const b=c.year,S=d.year;return[...Array(S-b+1).keys()].map(z=>(b+z).toString())},[c,d]),p=ue(a,{month:"long"}),w=l.useMemo(()=>[...Array(u.calendar.getMonthsInYear(u)).keys()].map(b=>p.format(u.set({month:b+1}).toDate(i))),[u,p,i]),m=ue(a,{weekday:"short"}),$=l.useMemo(()=>{const b=F.startOfWeek(F.today(i),a);return[...new Array(7).keys()].map(S=>{const E=b.add({days:S}).toDate(i);return m.format(E)})},[i,a,m]),k=ue(a,{weekday:"long",day:"numeric",month:"long",year:"numeric"}),y=ue(a,{day:"numeric",calendar:u.calendar.identifier}),x=ue(a,{day:"2-digit",month:"2-digit",year:"numeric"}),I=b=>{if(typeof b=="number")return;const S=u.set({month:w.indexOf(b)+1});h(S)},j=b=>{if(typeof b=="number")return;const S=u.set({year:parseInt(b,10)});h(S)},R=Gi(f,a);return r.jsx(Vi,{dateFormatter:k,cellDateFormatter:y,textValueFormatter:x,children:r.jsxs(C,{ref:o,direction:"column",alignItems:"stretch",padding:4,...n,children:[r.jsxs(Ki,{justifyContent:"flex-start",paddingBottom:4,paddingLeft:2,paddingRight:2,gap:2,children:[r.jsx(Re,{children:r.jsx(vt,{"aria-label":e,value:w[u.month-1],onChange:I,children:w.map(b=>r.jsx(kt,{value:b,children:b},b))})}),r.jsx(Re,{children:r.jsx(vt,{value:u.year.toString(),"aria-label":t,onChange:j,children:g.map(b=>r.jsx(kt,{value:b,children:b},b))})})]}),r.jsxs("table",{role:"grid",children:[r.jsx("thead",{"aria-hidden":!0,children:r.jsx("tr",{"aria-rowindex":0,children:$.map((b,S)=>r.jsx(Yi,{"aria-colindex":S,children:b},b))})}),r.jsx("tbody",{children:[...new Array(6).keys()].map(b=>r.jsx("tr",{"aria-rowindex":b+2,children:R(b).map((S,z)=>S?r.jsx(Zi,{"aria-colindex":z+1,date:S,startDate:f,disabled:c.compare(S)>0||S.compare(d)>0},S.toString()):r.jsx(yn,{"aria-colindex":z+1},z+1))},b))})]})]})})}),Gi=(e,t)=>n=>{let o=e.add({weeks:n});const a=[];o=F.startOfWeek(o,t);const i=F.getDayOfWeek(o,t);for(let c=0;c<i;c++)a.push(null);for(;a.length<7;){a.push(o);const c=o.add({days:1});if(F.isSameDay(o,c))break;o=c}for(;a.length<7;)a.push(null);return a},Ki=s.styled(C)`
  div[role='combobox'] {
    border: 1px solid transparent;
    background: transparent;
    font-weight: ${e=>e.theme.fontWeights.bold};

    svg {
      fill: ${({theme:e})=>e.colors.neutral500};
    }

    &:hover {
      background-color: ${({theme:e})=>e.colors.neutral100};
    }
  }
`,Yi=l.forwardRef(({children:e,...t},n)=>r.jsx(Qi,{tag:"th",role:"gridcell",ref:n,...t,height:"2.4rem",width:"3.2rem",children:r.jsx(T,{variant:"pi",fontWeight:"bold",color:"neutral800",children:e.slice(0,2)})})),Qi=s.styled(v)`
  border-radius: ${({theme:e})=>e.borderRadius};
  text-transform: capitalize;
`,Xi="DatePickerCalendarCell",Zi=l.forwardRef(({date:e,startDate:t,disabled:n,...o},a)=>{const{timeZone:i,calendarDate:c,onValueChange:d,onOpenChange:u,onTextValueChange:h,onCalendarDateChange:f}=Me(Xi),{dateFormatter:g,cellDateFormatter:p,textValueFormatter:w}=qi("DatePickerCalendarCell"),m=F.isSameDay(c,e),$=l.useMemo(()=>g.format(e.toDate(i)),[g,e,i]),k=l.useMemo(()=>p.formatToParts(e.toDate(i)).find(j=>j.type==="day").value,[p,e,i]),y=F.endOfMonth(t),x=e.compare(t)<0||e.compare(y)>0;let I="neutral900";return m?I="primary600":x&&(I="neutral600"),r.jsx(yn,{tag:"td",role:"gridcell",ref:a,"aria-selected":m,...o,hasRadius:!0,"aria-label":$,tabIndex:m?0:-1,background:m?"primary100":"neutral0",cursor:"pointer",onPointerDown:D.composeEventHandlers(o.onPointerDown,j=>{j.preventDefault(),f(e),d(e),h(w.format(e.toDate(i))),u(!1)}),"aria-disabled":n,children:r.jsx(T,{variant:"pi",textColor:I,children:k})})}),yn=s.styled(v)`
  text-align: center;
  padding: 0.7rem;
  // Trick to prevent the outline from overflowing because of the general outline-offset
  outline-offset: -2px !important;
  &[aria-disabled='true'] {
    pointer-events: none;
    opacity: 0.5;
  }

  &[aria-disabled='false'] {
    &:hover {
      background: ${({theme:e})=>e.colors.primary100};
      color: ${({theme:e})=>e.colors.primary600};
    }
  }
`,je=e=>{const t=e.toISOString(),n=F.parseAbsolute(t,"UTC");return F.toCalendarDate(n)},wn=l.memo(Ii),pt=e=>!!e.match(/^[^a-zA-Z]*$/);function Ji(e=""){return e.replace(/[.*+?^${}()|[\]\\]/g,"\\$&")}const el=s.styled(on)`
  min-width: ${({onClear:e})=>e?"160px":"130px"};
`,Cn=l.forwardRef(({step:e=15,value:t,defaultValue:n,onChange:o,...a},i)=>{const c=Xe("TimePicker"),[d,u]=l.useState(""),[h,f]=le({prop:t,defaultProp:n,onChange:o}),g=ue(c.locale,{hour:"2-digit",minute:"2-digit",hour12:!1}),p=l.useMemo(()=>{const j=g.formatToParts(new Date),{value:R}=j.find(b=>b.type==="literal");return R},[g]),w=l.useMemo(()=>{const j=60/e;return[...Array(24).keys()].flatMap(R=>[...Array(j).keys()].map(b=>g.format(new Date(0,0,0,R,b*e))))},[e,g]),m=j=>{(!j||pt(j))&&u(j)},$=j=>{const[R,b]=j.split(p);if(!R&&!b)return;const S=Number(R??"0"),z=Number(b??"0");if(!(S>23||z>59))return g.format(new Date(0,0,0,S,z))},k=j=>{const R=$(j.target.value);R?(u(R),f(R)):u(h)},y=j=>{if(typeof j<"u"){const R=$(j);f(R)}else f(j)};l.useEffect(()=>{const j=typeof t>"u"?"":t;pt(j)&&u(j)},[t,u]);const I=`\\d{2}${Ji(p)}\\d{2}`;return r.jsx(el,{...a,ref:i,value:h,onChange:y,isPrintableCharacter:pt,allowCustomValue:!0,placeholder:`--${p}--`,autocomplete:"none",startIcon:r.jsx(O.Clock,{fill:"neutral500"}),inputMode:"numeric",pattern:I,textValue:d,onTextValueChange:m,onBlur:k,children:w.map(j=>r.jsx(an,{value:j,children:j},j))})}),tl=l.forwardRef(({clearLabel:e="clear",dateLabel:t="Choose date",timeLabel:n="Choose time",disabled:o=!1,hasError:a,onChange:i,onClear:c,required:d=!1,step:u,value:h,initialDate:f,size:g,...p},w)=>{const m=l.useRef(null),[$,k]=l.useState(!1),y=h?Ee(h,!1):void 0,[x,I]=le({defaultProp:f?Ee(f,!1):void 0,prop:$?void 0:y,onChange(L){i&&i(L==null?void 0:L.toDate(F.getLocalTimeZone()))}}),j=Xe("DateTimePicker"),R=ue(j.locale,{hour:"2-digit",minute:"2-digit",hour12:!1}),b=x?R.format(x.toDate(F.getLocalTimeZone())):"",S=l.useRef(h);he(()=>{$&&x&&I(void 0)},[$,x,I]),he(()=>{const L=S.current;h==null&&L!=null&&x&&I(void 0),$&&h!==L&&k(!1),S.current=h},[h,x,$,I,k]),he(()=>{$&&x===void 0&&h==null&&k(!1)},[$,x,h,k]);const z=L=>{let M=L?Ee(L):void 0;if(b&&M){const[Z,G]=b.split(":");M=M.set({hour:parseInt(Z,10),minute:parseInt(G,10)})}I(M)},E=L=>{if(!L)return;const[M,Z]=L.split(":"),G=x?x.set({hour:parseInt(M,10),minute:parseInt(Z,10)}):Ee(new Date).set({hour:parseInt(M,10),minute:parseInt(Z,10)});I(G)},A=L=>{L.preventDefault(),k(!0),c&&c(L)},U=()=>{const L=x?x.set({hour:0,minute:0}):Ee(new Date);I(L)},B=te(m,w),{error:N,id:H,labelNode:P}=ne("DateTimePicker"),K=!!N||a;return r.jsxs(C,{"aria-labelledby":P?`${H}-label`:void 0,role:"group",flex:"1",gap:1,wrap:"wrap",children:[r.jsx(Re,{flex:"1",children:r.jsx(wn,{...p,hasError:K,size:g,value:x==null?void 0:x.toDate("UTC"),onChange:z,required:d,onClear:c?A:void 0,clearLabel:`${e} date`,disabled:o,ref:B,"aria-label":t})}),r.jsx(Re,{flex:"1",children:r.jsx(Cn,{size:g,hasError:K,value:b,onChange:E,onClear:c&&b!==void 0&&b!=="00:00"?U:void 0,clearLabel:`${e} time`,required:d,disabled:o,step:u,"aria-label":n})})]})}),Ee=(e,t=!0)=>{const n=e.toISOString();let o=F.parseAbsoluteToLocal(n);return t&&(o=o.set({hour:0,minute:0})),F.toCalendarDateTime(o)},_t=l.forwardRef((e,t)=>r.jsx(rl,{ref:t,background:"neutral150",...e,"data-orientation":"horizontal",role:"separator",tag:"div"})),rl=s.styled(v)`
  height: 1px;
  border: none;
  /* If contained in a Flex parent we want to prevent the Divider to shink */
  flex-shrink: 0;
`,nl=s.styled(v)`
  svg {
    height: 8.8rem;
  }
`,ol=l.forwardRef(({icon:e,content:t,action:n,hasRadius:o=!0,shadow:a="tableShadow"},i)=>r.jsxs(C,{ref:i,alignItems:"center",direction:"column",padding:11,background:"neutral0",hasRadius:o,shadow:a,children:[e?r.jsx(nl,{paddingBottom:6,"aria-hidden":!0,children:e}):null,r.jsx(v,{paddingBottom:4,children:r.jsx(T,{variant:"delta",tag:"p",textAlign:"center",textColor:"neutral600",children:t})}),n]})),jn=Rt.StateEffect.define(),vn=Rt.StateEffect.define(),al=bt.Decoration.mark({attributes:{style:"background-color: yellow; color: black"}}),sl=Rt.StateField.define({create(){return bt.Decoration.none},update(e,t){return e=e.map(t.changes),t.effects.forEach(n=>{n.is(jn)?e=e.update({add:n.value,sort:!0}):n.is(vn)&&(e=e.update({filter:n.value}))}),e},provide:e=>bt.EditorView.decorations.from(e)}),il=l.forwardRef(({hasError:e,required:t,id:n,value:o="",disabled:a=!1,onChange:i=()=>null,...c},d)=>{const u=l.useRef(),h=l.useRef(),f=l.useRef(),{error:g,...p}=ne("JsonInput"),w=!!g||e,m=p.id??n,$=p.required||t;let k;g?k=`${m}-error`:p.hint&&(k=`${m}-hint`);const y=E=>{var U,B;const A=(U=h.current)==null?void 0:U.doc;if(A){const{text:N,to:H}=A.line(E),P=H-N.trimStart().length;H>P&&((B=f.current)==null||B.dispatch({effects:jn.of([al.range(P,H)])}))}},x=()=>{var A,U;const E=(A=h.current)==null?void 0:A.doc;if(E){const B=E.length||0;(U=f.current)==null||U.dispatch({effects:vn.of((N,H)=>H<=0||N>=B)})}},I=({state:E,view:A})=>{f.current=A,h.current=E,x();const B=Xt.jsonParseLinter()(A);B.length&&y(E.doc.lineAt(B[0].from).number)},j=(E,A)=>{I(A),i(E)},R=(E,A)=>{f.current=E,h.current=A,I({view:E,state:A})},{setContainer:b,view:S}=uo.useCodeMirror({value:o,onCreateEditor:R,container:u.current,editable:!a,extensions:[Xt.json(),sl],onChange:j,theme:"dark",basicSetup:{lineNumbers:!0,bracketMatching:!0,closeBrackets:!0,indentOnInput:!0,syntaxHighlighting:!0,highlightSelectionMatches:!0,tabSize:2}}),z=te(u,b);return l.useImperativeHandle(d,()=>({...S==null?void 0:S.dom,focus(){S&&S.focus()},scrollIntoView(E){S&&S.dom.scrollIntoView(E)}}),[S]),r.jsx(ll,{ref:z,$disabled:a,$hasError:w,alignItems:"stretch",hasRadius:!0,"aria-required":$,id:m,"aria-describedby":k,...c})}),ll=s.styled(C)`
  font-size: ${({theme:e})=>e.fontSizes[3]};
  line-height: ${({theme:e})=>e.lineHeights[6]};

  ${({theme:e})=>e.breakpoints.medium} {
    font-size: ${({theme:e})=>e.fontSizes[2]};
    line-height: ${({theme:e})=>e.lineHeights[2]};
  }
  .cm-editor {
    /** 
     * Hard coded since the color is the same between themes,
     * theme.colors.neutral800 changes between themes 
     */
    background-color: #32324d;
    width: 100%;
    outline: none;
    cursor: ${({$disabled:e})=>e?"not-allowed":"text"};
  }

  .cm-scroller {
    border: 1px solid ${({theme:e,$hasError:t})=>t?e.colors.danger600:e.colors.neutral200};
    /* inputFocusStyle will receive hasError prop */
    ${be()}
  }

  .cm-editor,
  .cm-scroller {
    border-radius: ${({theme:e})=>e.borderRadius};
  }

  .cm-gutters,
  .cm-activeLineGutter {
    /** 
     * Hard coded since the color is the same between themes,
     * theme.colors.neutral700 changes between themes 
     */
    background-color: #4a4a6a;
  }
`,cl=ae(({disabled:e,...t},n)=>r.jsx(rt,{ref:n,tag:Pe,tabIndex:e?-1:void 0,disabled:e,...t})),dl=s.styled(v)`
  // To prevent global outline on focus visible to force an outline when Main is focused
  &:focus-visible {
    outline: none;
  }
`,ul=({labelledBy:e="main-content-title",...t})=>r.jsx(dl,{"aria-labelledby":e,tag:"main",id:"main-content",tabIndex:-1,...t}),hl=s.styled(v)`
  text-decoration: none;

  &:focus {
    left: ${({theme:e})=>e.spaces[3]};
    top: ${({theme:e})=>e.spaces[3]};
  }
`,fl=({children:e})=>r.jsx(hl,{tag:"a",href:"#main-content",background:"primary600",color:"neutral0",left:"-100%",padding:3,position:"absolute",top:"-100%",hasRadius:!0,zIndex:9999,children:e}),gl=Te.Root,pl=l.forwardRef((e,t)=>r.jsx(Te.Trigger,{...e,asChild:!0,ref:t})),ml=l.forwardRef((e,t)=>r.jsx(Te.Portal,{children:r.jsx(bl,{children:r.jsx(xl,{ref:t,...e})})})),bl=s.styled(Te.Overlay)`
  background: ${e=>Wt(e.theme.colors.neutral800,.2)};
  position: fixed;
  inset: 0;
  z-index: ${e=>e.theme.zIndices.overlay};
  will-change: opacity;

  @media (prefers-reduced-motion: no-preference) {
    animation: ${q.overlayFadeIn} ${e=>e.theme.motion.timings[200]}
      ${e=>e.theme.motion.easings.authenticMotion};
  }
`,xl=s.styled(Te.Content)`
  max-width: 83rem;
  max-height: 90vh;
  height: auto;
  width: calc(100% - ${({theme:e})=>e.spaces[8]});
  overflow: hidden;
  margin: 0 auto;
  display: flex;
  flex-direction: column;

  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);

  border-radius: ${e=>e.theme.borderRadius};
  background-color: ${e=>e.theme.colors.neutral0};
  box-shadow: ${e=>e.theme.shadows.popupShadow};
  z-index: ${e=>e.theme.zIndices.modal};

  ${({theme:e})=>e.breakpoints.medium} {
    width: calc(100% - ${({theme:e})=>e.spaces[9]});
  }
  > form {
    display: flex;
    flex-direction: column;
    overflow: hidden;
  }
  @media (prefers-reduced-motion: no-preference) {
    &[data-state='open'] {
      animation-duration: ${e=>e.theme.motion.timings[200]};
      animation-timing-function: ${e=>e.theme.motion.easings.authenticMotion};
      animation-name: ${q.modalPopIn};
    }

    &[data-state='closed'] {
      animation-duration: ${e=>e.theme.motion.timings[120]};
      animation-timing-function: ${e=>e.theme.motion.easings.easeOutQuad};
      animation-name: ${q.modalPopOut};
    }
  }
`,kn=l.forwardRef((e,t)=>r.jsx(Te.Close,{...e,asChild:!0,ref:t})),$l=l.forwardRef(({children:e,closeLabel:t="Close modal",...n},o)=>r.jsxs(yl,{ref:o,padding:4,paddingLeft:5,paddingRight:5,background:"neutral100",justifyContent:"space-between",...n,tag:"header",children:[e,r.jsx(kn,{children:r.jsx(ge,{withTooltip:!1,label:t,children:r.jsx(O.Cross,{})})})]})),yl=s.styled(C)`
  border-bottom: solid 1px ${e=>e.theme.colors.neutral150};
`,wl=l.forwardRef((e,t)=>r.jsx(Te.Title,{asChild:!0,children:r.jsx(T,{tag:"h2",variant:"omega",fontWeight:"bold",ref:t,...e})})),Cl=l.forwardRef(({children:e,...t},n)=>r.jsx(jl,{ref:n,...t,children:e})),jl=s.styled(Qe)`
  padding-inline: ${e=>e.theme.spaces[7]};

  & > div {
    padding-block: ${e=>e.theme.spaces[8]};
    /* Add negative margin and padding to avoid cropping the box shadow when the inputs are focused */
    margin: 0 -2px 0 -2px;
    padding-left: 2px;
    padding-right: 2px;

    & > div {
      // the scroll area component applies a display: table to the child, which we don't want.
      display: block !important;
    }
  }
`,vl=l.forwardRef((e,t)=>r.jsx(kl,{ref:t,padding:4,paddingLeft:5,paddingRight:5,background:"neutral100",justifyContent:"space-between",...e,tag:"footer"})),kl=s.styled(C)`
  border-top: solid 1px ${e=>e.theme.colors.neutral150};
  flex: 1;
`,Sl=Object.freeze(Object.defineProperty({__proto__:null,Body:Cl,Close:kn,Content:ml,Footer:vl,Header:$l,Root:gl,Title:wl,Trigger:pl},Symbol.toStringTag,{value:"Module"})),Rl="",Tl=l.forwardRef(({startAction:e,locale:t,onValueChange:n,value:o,step:a=1,disabled:i=!1,...c},d)=>{const u=Xe("NumberInput"),h=t||u.locale,f=l.useRef(new Zt.NumberParser(h,{style:"decimal"})),g=l.useRef(new Zt.NumberFormatter(h,{maximumFractionDigits:20})),[p,w]=Il({prop(b){const S=String(o);return isNaN(Number(S))||S!==b&&b!==""?b:g.current.format(Number(o))},defaultProp:Rl,onChange(b){if(!n)return;const S=f.current.parse(b??"");n(isNaN(S)?void 0:S)}}),m=b=>{w(String(b))},$=({target:{value:b}})=>{f.current.isValidPartialNumber(b)&&m(b)},y=(b=>{const S=b.toString();return S.includes(".")?S.split(".")[1].length:0})(a),x=()=>{if(!p){m(a);return}const b=f.current.parse(p),S=isNaN(b)?a:b+a,z=parseFloat(S.toFixed(y));m(g.current.format(z))},I=()=>{if(!p){m(-a);return}const b=f.current.parse(p),S=isNaN(b)?-a:b-a,z=parseFloat(S.toFixed(y));m(g.current.format(z))},j=b=>{if(!i)switch(b.key){case W.DOWN:{b.preventDefault(),I();break}case W.UP:{b.preventDefault(),x();break}}},R=()=>{if(p){const b=f.current.parse(p),S=isNaN(b)?"":g.current.format(b);m(S)}};return r.jsx(ut,{ref:d,startAction:e,disabled:i,type:"text",inputMode:"decimal",onChange:$,onKeyDown:j,onBlur:R,value:p,endAction:r.jsxs(C,{direction:"column",children:[r.jsx(pr,{disabled:i,"aria-hidden":!0,$reverse:!0,onClick:x,tabIndex:-1,type:"button","data-testid":"ArrowUp",children:r.jsx(O.CaretDown,{fill:"neutral500"})}),r.jsx(pr,{disabled:i,"aria-hidden":!0,onClick:I,tabIndex:-1,type:"button","data-testid":"ArrowDown",children:r.jsx(O.CaretDown,{fill:"neutral500"})})]}),...c})}),pr=s.styled.button`
  display: flex;
  justify-content: center;
  align-items: center;
  transform: translateY(${({$reverse:e})=>e?"-2px":"2px"});
  cursor: ${({disabled:e})=>e?"not-allowed":void 0};
  height: 1.1rem;

  svg {
    width: 1.2rem;
    transform: ${({$reverse:e})=>e?"rotateX(180deg)":void 0};
  }
`;function Il({prop:e,defaultProp:t,onChange:n=()=>{}}){const[o,a]=Et({defaultProp:t,onChange:n}),i=e!==void 0,c=e instanceof Function?e(o):e,d=i?c:o,u=D.useCallbackRef(n),h=l.useCallback(f=>{if(i){const p=typeof f=="function"?f(c):f;p!==c&&(u(p),a(f))}else a(f)},[i,c,a,u]);return[d,h]}const Sn=l.createContext({activePage:1,pageCount:1}),Vt=()=>l.useContext(Sn),Dl=({children:e,label:t="Pagination",activePage:n,pageCount:o})=>{const a=l.useMemo(()=>({activePage:n,pageCount:o}),[n,o]);return r.jsx(Sn.Provider,{value:a,children:r.jsx(v,{"aria-label":t,tag:"nav",children:r.jsx(C,{tag:"ol",gap:1,children:l.Children.map(e,(i,c)=>r.jsx("li",{children:i},c))})})})},El=ae(({children:e,...t},n)=>{const{activePage:o}=Vt(),a=o===1;return r.jsxs(Tn,{ref:n,"aria-disabled":a,tabIndex:a?-1:void 0,...t,children:[r.jsx(me,{children:e}),r.jsx(O.ChevronLeft,{"aria-hidden":!0})]})}),Al=ae(({children:e,...t},n)=>{const{activePage:o,pageCount:a}=Vt(),i=o===a;return r.jsxs(Tn,{ref:n,"aria-disabled":i,tabIndex:i?-1:void 0,...t,children:[r.jsx(me,{children:e}),r.jsx(O.ChevronRight,{"aria-hidden":!0})]})}),Rn=s.styled(Pe)`
  padding: ${({theme:e})=>e.spaces[3]};
  border-radius: ${({theme:e})=>e.borderRadius};
  box-shadow: ${({$active:e,theme:t})=>e?t.shadows.filterShadow:void 0};
  text-decoration: none;
  display: flex;

  ${st}
`,Tn=s.styled(Rn)`
  font-size: 1.1rem;

  svg path {
    fill: ${e=>e["aria-disabled"]?e.theme.colors.neutral300:e.theme.colors.neutral600};
  }

  &:focus,
  &:hover {
    svg path {
      fill: ${e=>e["aria-disabled"]?e.theme.colors.neutral300:e.theme.colors.neutral700};
    }
  }

  ${e=>e["aria-disabled"]?`
  pointer-events: none;
    `:void 0}
`,zl=ae(({number:e,children:t,...n},o)=>{const{activePage:a}=Vt(),i=a===e;return r.jsxs(Ll,{ref:o,...n,"aria-current":i,$active:i,children:[r.jsx(me,{children:t}),r.jsx(T,{"aria-hidden":!0,fontWeight:i?"bold":void 0,lineHeight:"revert",variant:"pi",children:e})]})}),Ll=s.styled(Rn)`
  color: ${({theme:e,$active:t})=>t?e.colors.primary700:e.colors.neutral800};
  background: ${({theme:e,$active:t})=>t?e.colors.neutral0:void 0};

  &:hover {
    box-shadow: ${({theme:e})=>e.shadows.filterShadow};
  }
`,Ol=({children:e,...t})=>r.jsxs(v,{...t,children:[r.jsx(me,{children:e}),r.jsx(T,{"aria-hidden":!0,lineHeight:"revert",textColor:"neutral800",variant:"pi",children:"…"})]}),Pl=Le.Root,Ml=Le.Anchor,Bl=Le.Arrow,Nl=l.forwardRef((e,t)=>r.jsx(Le.Trigger,{...e,asChild:!0,ref:t})),Wl=l.forwardRef((e,t)=>r.jsx(Le.Portal,{children:r.jsx(Fl,{sideOffset:4,side:"bottom",align:"start",...e,ref:t})})),Fl=s.styled(Le.Content)`
  box-shadow: ${({theme:e})=>e.shadows.filterShadow};
  z-index: ${({theme:e})=>e.zIndices.popover};
  background-color: ${e=>e.theme.colors.neutral0};
  border: 1px solid ${({theme:e})=>e.colors.neutral150};
  border-radius: ${({theme:e})=>e.borderRadius};

  @media (prefers-reduced-motion: no-preference) {
    animation-duration: ${e=>e.theme.motion.timings[200]};

    &[data-state='open'] {
      animation-timing-function: ${e=>e.theme.motion.easings.authenticMotion};

      &[data-side='top'] {
        animation-name: ${q.slideUpIn};
      }

      &[data-side='bottom'] {
        animation-name: ${q.slideDownIn};
      }
    }

    &[data-state='closed'] {
      animation-timing-function: ${e=>e.theme.motion.easings.easeOutQuad};

      &[data-side='top'] {
        animation-name: ${q.slideUpOut};
      }

      &[data-side='bottom'] {
        animation-name: ${q.slideDownOut};
      }
    }
  }
`,Hl=l.forwardRef(({children:e,intersectionId:t,onReachEnd:n,...o},a)=>{const i=l.useRef(null),c=te(i,a),d=re();return Ge(i,n??(()=>{}),{selectorToWatch:`#${ze(d)}`,skipWhen:!t||!n}),r.jsxs(_l,{ref:c,...o,children:[e,t&&n&&r.jsx(v,{id:ze(d),width:"100%",height:"1px"})]})}),_l=s.styled(Qe)`
  height: 20rem;
`,Vl=Object.freeze(Object.defineProperty({__proto__:null,Anchor:Ml,Arrow:Bl,Content:Wl,Root:Pl,ScrollArea:Hl,Trigger:Nl},Symbol.toStringTag,{value:"Module"})),ql=l.forwardRef(({size:e="M",value:t,...n},o)=>r.jsx(Ul,{ref:o,$size:e,...n,children:r.jsx(Gl,{style:{transform:`translate3D(-${100-(t??0)}%, 0, 0)`}})})),Ul=s.styled(wr.Root)`
  position: relative;
  overflow: hidden;
  width: ${e=>e.$size==="S"?"7.8rem":"10.2rem"};
  height: ${e=>e.$size==="S"?"0.4rem":"0.8rem"};
  background-color: ${e=>e.theme.colors.neutral600};
  border-radius: ${e=>e.theme.borderRadius};

  /* Fix overflow clipping in Safari */
  /* https://gist.github.com/domske/b66047671c780a238b51c51ffde8d3a0 */
  transform: translateZ(0);
`,Gl=s.styled(wr.Indicator)`
  background-color: ${({theme:e})=>e.colors.neutral0};
  border-radius: ${({theme:e})=>e.borderRadius};
  width: 100%;
  height: 100%;

  @media (prefers-reduced-motion: no-preference) {
    transition: transform ${e=>e.theme.motion.timings[320]}
      ${e=>e.theme.motion.easings.authenticMotion};
  }
`,Kl=l.forwardRef((e,t)=>r.jsx(Yl,{ref:t,...e})),Yl=s.styled(It.Root)`
  display: flex;
  flex-direction: column;
  gap: ${e=>e.theme.spaces[3]};
`,Ql=l.forwardRef(({children:e,id:t,...n},o)=>{const a=re(t);return r.jsxs(C,{gap:2,children:[r.jsx(Xl,{id:a,ref:o,...n,children:r.jsx(Zl,{})}),r.jsx(T,{tag:"label",htmlFor:a,children:e})]})}),Xl=s.styled(It.Item)`
  background: ${e=>e.theme.colors.neutral0};
  width: 2rem;
  height: 2rem;
  flex: 0 0 2rem;
  border-radius: 50%;
  border: 1px solid ${e=>e.theme.colors.neutral300};
  position: relative;
  cursor: pointer;
  z-index: 0;

  @media (prefers-reduced-motion: no-preference) {
    transition: border-color ${e=>e.theme.motion.timings[120]}
      ${e=>e.theme.motion.easings.easeOutQuad};
  }

  &[data-state='checked'] {
    border: 1px solid ${e=>e.theme.colors.primary600};
  }

  &[data-disabled] {
    cursor: not-allowed;
    background-color: ${e=>e.theme.colors.neutral200};
  }

  /* increase target size for touch devices https://www.w3.org/WAI/WCAG21/Understanding/target-size.html */
  &::before {
    content: '';
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    width: 100%;
    height: 100%;
    min-width: 44px;
    min-height: 44px;
  }
`,Zl=s.styled(It.Indicator)`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  position: relative;

  &[data-state='checked'] {
    @media (prefers-reduced-motion: no-preference) {
      animation: ${q.popIn} ${e=>e.theme.motion.timings[200]};
    }
  }

  &::after {
    content: '';
    display: block;
    width: 1rem;
    height: 1rem;
    border-radius: 50%;
    background: ${e=>e.theme.colors.primary600};
  }
`,Jl=Object.freeze(Object.defineProperty({__proto__:null,Group:Kl,Item:Ql},Symbol.toStringTag,{value:"Module"})),ec=e=>{const t=e.querySelector('[tabindex="0"]');t&&t.focus()},In=l.createContext({rowIndex:0,colIndex:0,setTableValues(){throw new Error("setTableValues must be initialized via the RawTableContext.Provider")}}),tc=()=>l.useContext(In),Dn=l.forwardRef(({colCount:e,rowCount:t,jumpStep:n=3,initialCol:o=0,initialRow:a=0,...i},c)=>{const d=l.useRef(null),u=l.useRef(!1),h=te(d,c),[f,g]=l.useState(a),[p,w]=l.useState(o),m=l.useCallback(({colIndex:y,rowIndex:x})=>{w(y),g(x)},[]);l.useEffect(()=>{u.current&&ec(d.current),u.current||(u.current=!0)},[p,f]);const $=y=>{switch(y.key){case W.RIGHT:{y.preventDefault(),w(x=>x<e-1?x+1:x);break}case W.LEFT:{y.preventDefault(),w(x=>x>0?x-1:x);break}case W.UP:{y.preventDefault(),g(x=>x>0?x-1:x);break}case W.DOWN:{y.preventDefault(),g(x=>x<t-1?x+1:x);break}case W.HOME:{y.preventDefault(),y.ctrlKey&&g(0),w(0);break}case W.END:{y.preventDefault(),y.ctrlKey&&g(t-1),w(e-1);break}case W.PAGE_DOWN:{y.preventDefault(),g(x=>x+n<t?x+n:t-1);break}case W.PAGE_UP:{y.preventDefault(),g(x=>x-n>0?x-n:0);break}}},k=l.useMemo(()=>({rowIndex:f,colIndex:p,setTableValues:m}),[p,f,m]);return r.jsx(In.Provider,{value:k,children:r.jsx("table",{role:"grid",ref:h,"aria-rowcount":t,"aria-colcount":e,onKeyDown:$,...i})})}),ve=(e,t)=>[...e.querySelectorAll('a, button, input, textarea, select, details, [tabindex]:not([tabindex="-1"])')].filter(a=>a.hasAttribute("disabled")?!1:t?!0:a.getAttribute("tabindex")!=="-1"),mt=e=>e.filter(t=>t.tagName==="INPUT"?t.type!=="checkbox"&&t.type!=="radio":!1),qt=l.forwardRef(({coords:e={col:0,row:0},tag:t="td",...n},o)=>{const a=l.useRef(null),i=te(o,a),{rowIndex:c,colIndex:d,setTableValues:u}=tc(),[h,f]=l.useState(!1),g=m=>{var y;const $=ve(a.current,!0);if($.length===0||$.length===1&&mt($).length===0)return;if($.length>1&&!$.find(x=>x.tagName!=="BUTTON")){m.preventDefault();const x=$.findIndex(I=>I===document.activeElement);if(m.key===W.RIGHT){const I=$[x+1];I&&(m.stopPropagation(),I.focus())}else if(m.key===W.LEFT){const I=$[x-1];I&&(m.stopPropagation(),I.focus())}return}const k=m.key===W.ENTER;if(k&&!h)f(!0);else if((m.key===W.ESCAPE||k)&&h){if(k&&((y=document.activeElement)==null?void 0:y.tagName)==="A")return;f(!1),a.current.focus()}else h&&m.stopPropagation()},p=c===e.row-1&&d===e.col-1;he(()=>{const m=ve(a.current,!0);m.length===0||m.length===1&&mt(m).length!==0||m.length>1&&m.find($=>$.tagName!=="BUTTON")?(a.current.setAttribute("tabIndex",!h&&p?"0":"-1"),m.forEach(($,k)=>{$.setAttribute("tabIndex",h?"0":"-1"),h&&k===0&&$.focus()})):m.forEach($=>{$.setAttribute("tabIndex",p?"0":"-1")})},[h,p]);const w=l.useCallback(()=>{const m=ve(a.current,!0);m.length>=1&&(mt(m).length!==0||!m.find($=>$.tagName!=="BUTTON"))&&f(!0),u({rowIndex:e.row-1,colIndex:e.col-1})},[e,u]);return he(()=>{const m=a.current;return ve(m,!0).forEach(k=>{k.addEventListener("focus",w)}),()=>{ve(m,!0).forEach(y=>{y.removeEventListener("focus",w)})}},[w]),r.jsx(v,{role:"gridcell",tag:t,ref:i,onKeyDown:g,...n})}),En=e=>r.jsx(qt,{...e,tag:"th"}),An=({children:e,...t})=>{const n=l.Children.toArray(e).map(o=>l.isValidElement(o)?l.cloneElement(o,{"aria-rowindex":1}):o);return r.jsx("thead",{...t,children:n})},zn=({children:e,...t})=>{const n=l.Children.toArray(e).map((o,a)=>l.isValidElement(o)?l.cloneElement(o,{"aria-rowindex":a+2}):o);return r.jsx("tbody",{...t,children:n})},Ln=({children:e,...t})=>{const n=l.Children.toArray(e).map((o,a)=>l.isValidElement(o)?l.cloneElement(o,{"aria-colindex":a+1,coords:{col:a+1,row:t["aria-rowindex"]}}):o);return r.jsx(v,{tag:"tr",...t,children:n})},rc=s.styled(O.Cross)`
  font-size: 0.5rem;
  path {
    fill: ${({theme:e})=>e.colors.neutral500};
  }
`,On=s.styled(O.Search)`
  font-size: 1rem;
  path {
    fill: ${({theme:e})=>e.colors.neutral500};
  }
`,nc=s.styled.div`
  border-radius: ${({theme:e})=>e.borderRadius};
  border: 1px solid ${({theme:e})=>e.colors.neutral150};

  &:focus-within {
    ${On} {
      fill: ${({theme:e})=>e.colors.primary600};
    }
  }
`,oc=s.styled(ut)`
  border: 1px solid ${({theme:e})=>e.colors.neutral150};
  padding: 0 0 0 8px;
  color: ${({theme:e})=>e.colors.neutral800};

  &:hover {
    button {
      cursor: pointer;
    }
  }

  ${be()}

  > input::-ms-clear {
    display: none;
    width: 0;
    height: 0;
  }

  > input::-ms-reveal {
    display: none;
    width: 0;
    height: 0;
  }

  > input::-webkit-search-decoration,
  > input::-webkit-search-cancel-button,
  > input::-webkit-search-results-button,
  > input::-webkit-search-results-decoration {
    display: none;
  }
`,Pn=l.forwardRef(({name:e,children:t,value:n="",onClear:o,clearLabel:a="Clear",...i},c)=>{const d=l.useRef(null),u=n.length>0,h=g=>{o(g),d.current.focus()},f=zt(c,d);return r.jsx(nc,{children:r.jsxs(Re,{name:e,children:[r.jsx(me,{children:r.jsx(Bt,{children:t})}),r.jsx(oc,{size:"S",type:"search",ref:f,value:n,startAction:r.jsx(On,{"aria-hidden":!0}),onKeyDown:g=>{g.stopPropagation(),g.key==="Escape"&&u&&h(g)},endAction:u?r.jsx(ge,{onClick:h,onMouseDown:g=>{g.preventDefault()},label:a,size:"XS",variant:"ghost",type:"button",children:r.jsx(rc,{})}):void 0,...i})]})})}),Mn=e=>r.jsx("form",{...e,role:"search"}),ac=s.styled(v)`
  display: inline-flex;
  border: none;

  & > svg {
    height: 1.2rem;
    width: 1.2rem;
  }

  & > svg path {
    fill: ${({theme:e,...t})=>t["aria-disabled"]?e.colors.neutral600:e.colors.primary600};
  }

  &:hover {
    cursor: ${({$iconAction:e})=>e?"pointer":"initial"};
  }
`,Bn=({children:e,icon:t,label:n,disabled:o=!1,onClick:a,...i})=>{const c=d=>{o||!a||a(d)};return r.jsxs(C,{inline:!0,background:o?"neutral200":"primary100",color:o?"neutral700":"primary600",paddingLeft:3,paddingRight:1,borderColor:o?"neutral300":"primary200",hasRadius:!0,height:"3.2rem",...i,children:[r.jsx(sc,{$disabled:o,variant:"pi",fontWeight:"bold",children:e}),r.jsx(ac,{tag:"button",disabled:o,"aria-disabled":o,"aria-label":n,padding:2,onClick:c,$iconAction:!!a,children:t})]})},sc=s.styled(T)`
  color: inherit;
  border-right: 1px solid ${({theme:e,$disabled:t})=>t?e.colors.neutral300:e.colors.primary200};
  padding-right: ${({theme:e})=>e.spaces[2]};
`,Nn=l.forwardRef(({children:e,clearLabel:t="Clear",customizeContent:n,disabled:o,hasError:a,id:i,loading:c,name:d,onChange:u,onClear:h,onCloseAutoFocus:f,onReachEnd:g,placeholder:p,required:w,size:m,startIcon:$,value:k,withTags:y,...x},I)=>{const j=l.useRef(null),[R,b]=l.useState(),[S,z]=l.useState(!1),E=_=>{u?u(_):b(_)},A=_=>()=>{const Q=Array.isArray(k)?k.filter(de=>de!==_):(R??[]).filter(de=>de!==_);u?u(Q):b(Q)},U=_=>{z(_)},B=re(),N=`intersection-${ze(B)}`;Ge(j,_=>{g&&g(_)},{selectorToWatch:`#${N}`,skipWhen:!S});const P=typeof k<"u"&&k!==null?k:R,K=_=>_&&typeof _=="object"&&_.value?r.jsx(Bn,{tabIndex:-1,disabled:o,icon:r.jsx(O.Cross,{width:`${14/16}rem`,height:`${14/16}rem`}),onClick:A(_.value),children:_.textValue},_.value):null,{error:L,...M}=ne("MultiSelect"),Z=!!L||a,G=M.id??i,oe=M.name??d,Y=M.required??w;let J;return L?J=`${G}-error`:M.hint&&(J=`${G}-hint`),r.jsxs(hn,{onOpenChange:U,disabled:o,required:Y,onValueChange:E,value:P,...x,multi:!0,children:[r.jsx(fn,{ref:I,id:G,name:oe,"aria-label":x["aria-label"],"aria-describedby":J??x["aria-describedby"],startIcon:$,hasError:Z,disabled:o,clearLabel:t,loading:c,onClear:P!=null&&P.length?h:void 0,withTags:!!(y&&((P==null?void 0:P.length)??!1)),size:m,children:r.jsx(gn,{placeholder:p,textColor:P!=null&&P.length?"neutral800":"neutral600",withTags:y,children:P!=null&&P.length?y?K:n?n(P):void 0:void 0})}),r.jsx(pn,{children:r.jsx(mn,{position:"popper",sideOffset:4,onCloseAutoFocus:f,children:r.jsxs(bn,{ref:j,children:[e,r.jsx(v,{id:N,width:"100%",height:"1px"})]})})})]})}),Ut=l.forwardRef(({value:e,children:t,startIcon:n,...o},a)=>r.jsxs(Ht,{ref:a,value:e.toString(),...o,children:[n&&r.jsx(v,{tag:"span","aria-hidden":!0,children:n}),r.jsx(xn,{children:({isSelected:i,isIntermediate:c})=>r.jsx(dt,{checked:c?"indeterminate":i})}),r.jsx(T,{children:r.jsx($n,{children:t})})]})),Wn=l.forwardRef(({children:e,label:t,startIcon:n,values:o=[],...a},i)=>r.jsxs(Si,{ref:i,children:[r.jsxs(Ht,{value:o,...a,children:[n&&r.jsx(v,{tag:"span","aria-hidden":!0,children:n}),r.jsx(xn,{children:({isSelected:c,isIntermediate:d})=>r.jsx(dt,{checked:d?"indeterminate":c})}),r.jsx(T,{children:t})]}),e]})),ic=({options:e,...t})=>r.jsx(Nn,{...t,children:e.map(n=>"children"in n?r.jsx(Wn,{label:n.label,values:n.children.map(o=>o.value.toString()),children:n.children.map(o=>r.jsx(lc,{value:o.value,children:o.label},o.value))},n.label):r.jsx(Ut,{value:n.value,children:n.label},n.value))}),lc=s.styled(Ut)`
  padding-left: ${({theme:e})=>e.spaces[7]};
`,cc=e=>e==="XS"?{paddingX:"0.6rem",paddingY:"0.2rem"}:e==="S"?{paddingX:2,paddingY:1}:{paddingX:5,paddingY:4},dc=({variant:e="primary",size:t="M",children:n,...o})=>{const a=`${e}100`,i=`${e}200`,c=`${e}600`,{paddingX:d,paddingY:u}=cc(t);return r.jsx(v,{borderColor:i,color:c,background:a,hasRadius:!0,paddingTop:u,paddingBottom:u,paddingLeft:d,paddingRight:d,...o,children:n})},uc="23.2rem",hc=l.forwardRef(({...e},t)=>r.jsx(fc,{ref:t,...e,tag:"nav"})),fc=s.styled(v)`
  width: ${uc};
  background: ${({theme:e})=>e.colors.neutral100};
  position: sticky;
  top: 0;
  height: 100vh;
  overflow-y: auto;
  border-right: 1px solid ${({theme:e})=>e.colors.neutral200};
  z-index: 1;
`,mr=s.styled(_t)`
  width: 2.4rem;
  background-color: ${({theme:e})=>e.colors.neutral200};
`,gc=({tag:e="h2",label:t,searchLabel:n="",searchable:o=!1,onChange:a=()=>{},value:i="",onClear:c=()=>{},onSubmit:d=()=>{},id:u,placeholder:h})=>{const[f,g]=l.useState(!1),p=D.usePrev(f),w=re(u),m=l.useRef(void 0),$=l.useRef(void 0);l.useEffect(()=>{f&&m.current&&m.current.focus(),p&&!f&&$.current&&$.current.focus()},[f,p]);const k=()=>{g(j=>!j)},y=j=>{c(j),m.current.focus()},x=j=>{var R;((R=j.relatedTarget)==null?void 0:R.id)!==w&&g(!1)},I=j=>{j.key===W.ESCAPE&&g(!1)};return f?r.jsxs(v,{paddingLeft:4,paddingTop:5,paddingBottom:2,paddingRight:4,children:[r.jsx(Mn,{children:r.jsx(Pn,{name:"searchbar",value:i,onChange:a,placeholder:h,onKeyDown:I,ref:m,onBlur:x,onClear:y,onSubmit:d,clearLabel:"Clear",children:n})}),r.jsx(v,{paddingLeft:2,paddingTop:4,children:r.jsx(mr,{})})]}):r.jsxs(C,{direction:"column",alignItems:"flex-start",paddingLeft:6,paddingTop:6,paddingBottom:2,paddingRight:4,children:[r.jsxs(C,{justifyContent:"space-between",alignItems:"flex-start",width:"100%",gap:2,children:[r.jsx(T,{variant:"beta",tag:e,children:t}),o&&r.jsx(ge,{ref:$,onClick:k,label:n,children:r.jsx(O.Search,{})})]}),r.jsx(v,{paddingTop:4,children:r.jsx(mr,{})})]})},pc=ae(({active:e,children:t,icon:n=null,withBullet:o=!1,isSubSectionChild:a=!1,...i},c)=>r.jsxs(mc,{background:"neutral100",paddingLeft:a?9:7,paddingBottom:2,paddingTop:2,ref:c,...i,children:[r.jsxs(C,{children:[n?r.jsx(bc,{children:n}):r.jsx(St,{$active:e}),r.jsx(T,{paddingLeft:2,children:t})]}),o&&r.jsx(C,{paddingRight:4,children:r.jsx(St,{$active:!0})})]})),St=s.styled.span`
  width: 0.4rem;
  height: 0.4rem;
  background-color: ${({theme:e,$active:t})=>t?e.colors.primary600:e.colors.neutral600};
  border-radius: 50%;
`,mc=s.styled(Pe)`
  display: flex;
  align-items: center;
  justify-content: space-between;
  text-decoration: none;
  color: ${({theme:e})=>e.colors.neutral800};
  svg > * {
    fill: ${({theme:e})=>e.colors.neutral600};
  }

  &.active {
    ${({theme:e})=>s.css`
        background-color: ${e.colors.primary100};
        border-right: 2px solid ${e.colors.primary600};
        color: ${e.colors.primary700};
        font-weight: 500;
      `}

    ${St} {
      background-color: ${({theme:e})=>e.colors.primary600};
    }
  }

  &:focus-visible {
    outline-offset: -2px;
  }
`,bc=s.styled.div`
  svg {
    height: 1.6rem;
    width: 1.6rem;
  }
`,xc=s.styled.button`
  border: none;
  padding: 0;
  background: transparent;
  display: flex;
  align-items: center;
`,$c=({label:e,children:t,id:n})=>{const[o,a]=l.useState(!0),i=re(n),c=()=>{a(d=>!d)};return r.jsxs(v,{children:[r.jsx(v,{paddingLeft:7,paddingTop:2,paddingBottom:2,paddingRight:4,children:r.jsx(C,{justifyContent:"space-between",children:r.jsxs(xc,{onClick:c,"aria-expanded":o,"aria-controls":i,children:[r.jsx(O.CaretDown,{width:"1.2rem",height:"1.2rem","aria-hidden":!0,fill:"neutral700",style:{transform:`rotateX(${o?"0deg":"180deg"})`}}),r.jsx(v,{paddingLeft:2,children:r.jsx(T,{tag:"span",fontWeight:"semiBold",textColor:"neutral800",children:e})})]})})}),o&&r.jsx("ul",{id:i,children:l.Children.map(t,(d,u)=>r.jsx("li",{children:d},u))})]})},yc=({collapsable:e=!1,label:t,onClick:n=()=>{},ariaExpanded:o,ariaControls:a})=>e?r.jsxs(C,{tag:"button",onClick:n,"aria-expanded":o,"aria-controls":a,textAlign:"left",alignItems:"center",children:[r.jsx(v,{paddingRight:1,children:r.jsx(T,{variant:"sigma",textColor:"neutral600",children:t})}),e&&r.jsx(O.CaretDown,{width:"1.2rem","aria-hidden":!0,fill:"neutral500",style:{transform:o?"rotateX(0deg)":"rotateX(180deg)"}})]}):r.jsx(C,{children:r.jsx(v,{paddingRight:1,children:r.jsx(T,{variant:"sigma",textColor:"neutral600",children:t})})}),wc=s.styled(v)`
  & > svg {
    height: 0.4rem;
    fill: ${({theme:e})=>e.colors.neutral500};
  }
`,Cc=({collapsable:e=!1,label:t,badgeLabel:n,children:o,id:a})=>{const[i,c]=l.useState(!0),d=re(a),u=()=>{c(h=>!h)};return r.jsxs(C,{direction:"column",alignItems:"stretch",gap:1,children:[r.jsx(wc,{paddingLeft:6,paddingTop:2,paddingBottom:2,paddingRight:4,children:r.jsxs(v,{position:"relative",paddingRight:n?6:0,children:[r.jsx(yc,{onClick:u,ariaExpanded:i,ariaControls:d,collapsable:e,label:t}),n&&r.jsx(At,{backgroundColor:"neutral150",textColor:"neutral600",position:"absolute",right:0,top:"50%",transform:"translateY(-50%)",children:n})]})}),(!e||i)&&r.jsx("ol",{id:d,children:l.Children.map(o,(h,f)=>r.jsx("li",{children:h},f))})]})},jc=({children:e,spacing:t=2,horizontal:n=!1,...o})=>r.jsx(v,{paddingTop:2,paddingBottom:4,children:r.jsx(C,{tag:"ol",gap:t,direction:n?"row":"column",alignItems:n?"center":"stretch",...o,children:l.Children.map(e,(a,i)=>r.jsx("li",{children:a},i))})}),vc=l.forwardRef(({visibleLabels:e,onLabel:t="On",offLabel:n="Off",onCheckedChange:o,checked:a,defaultChecked:i,disabled:c,...d},u)=>{const[h,f]=le({prop:a,defaultProp:i}),g=p=>{f(p)};return r.jsxs(C,{gap:3,children:[r.jsx(kc,{ref:u,onCheckedChange:D.composeEventHandlers(o,g),checked:h,disabled:c,...d,children:r.jsx(Sc,{})}),e?r.jsx(Rc,{"aria-hidden":!0,"data-disabled":c,"data-state":h?"checked":"unchecked",children:h?t:n}):null]})}),kc=s.styled(Cr.Root)`
  width: 4rem;
  height: 2.4rem;
  border-radius: 1.2rem;
  background-color: ${({theme:e})=>e.colors.danger500};

  &[data-state='checked'] {
    background-color: ${({theme:e})=>e.colors.success500};
  }

  &[data-disabled] {
    background-color: ${({theme:e})=>e.colors.neutral300};
  }

  @media (prefers-reduced-motion: no-preference) {
    transition: ${e=>e.theme.transitions.backgroundColor};
  }
`,Sc=s.styled(Cr.Thumb)`
  display: block;
  height: 1.6rem;
  width: 1.6rem;
  border-radius: 50%;
  background-color: ${({theme:e})=>e.colors.neutral0};
  transform: translateX(4px);

  &[data-state='checked'] {
    transform: translateX(20px);
  }

  @media (prefers-reduced-motion: no-preference) {
    transition: transform ${e=>e.theme.motion.timings[120]}
      ${e=>e.theme.motion.easings.authenticMotion};
  }
`,Rc=s.styled(T)`
  color: ${e=>e.theme.colors.danger600};

  &[data-state='checked'] {
    color: ${e=>e.theme.colors.success600};
  }

  &[data-disabled='true'] {
    color: ${({theme:e})=>e.colors.neutral500};
  }
`,[Tc,Gt]=Oe("Tabs"),Ic=l.forwardRef(({disabled:e=!1,variant:t="regular",hasError:n,...o},a)=>r.jsx(Tc,{disabled:e,hasError:n,variant:t,children:r.jsx(Dc,{ref:a,...o})})),Dc=s.styled(at.Root)`
  width: 100%;
  position: relative;
`,Ec=l.forwardRef((e,t)=>{const{variant:n}=Gt("List");return r.jsx(Ac,{ref:t,...e,$variant:n})}),Ac=s.styled(at.List)`
  display: flex;
  align-items: ${e=>e.$variant==="regular"?"flex-end":"unset"};
  position: relative;
  z-index: 0;
`,zc=l.forwardRef(({children:e,disabled:t,...n},o)=>{const{disabled:a,variant:i,hasError:c}=Gt("Trigger"),d=a===!0||a===n.value||t,u=c===n.value;return r.jsxs(Lc,{ref:o,...n,$hasError:u,$variant:i,disabled:d,children:[r.jsx(Hn,{fontWeight:"bold",variant:i==="simple"?"sigma":void 0,children:e}),i==="simple"?r.jsx(Fn,{}):null]})}),Fn=s.styled.span`
  display: block;
  width: 100%;
  background-color: currentColor;
  position: absolute;
  bottom: 0;
  left: 0;
  opacity: 0;
  height: 0.2rem;
`,Hn=s.styled(T)``,Lc=s.styled(at.Trigger)`
  position: relative;
  color: ${e=>e.$hasError?e.theme.colors.danger600:e.theme.colors.neutral600};
  cursor: pointer;
  z-index: 0;

  ${e=>e.$variant==="simple"?s.css`
        padding-block: ${t=>t.theme.spaces[4]};
        padding-inline: ${t=>t.theme.spaces[4]};

        & > ${Hn} {
          line-height: 1.2rem;
        }

        &[data-state='active'] {
          color: ${e.$hasError?e.theme.colors.danger600:e.theme.colors.primary700};

          & > ${Fn} {
            opacity: 1;
          }
        }
      `:s.css`
        padding-block: ${t=>t.theme.spaces[3]};
        padding-inline: ${t=>t.theme.spaces[3]};
        flex: 1;
        background-color: ${t=>t.theme.colors.neutral100};
        border-bottom: solid 1px ${t=>t.theme.colors.neutral150};

        &:not([data-state='active']) + &:not([data-state='active']) {
          border-left: solid 1px ${t=>t.theme.colors.neutral150};
        }

        &[data-state='active'] {
          padding-block: ${t=>t.theme.spaces[4]};
          padding-inline: ${t=>t.theme.spaces[4]};
          color: ${e.$hasError?e.theme.colors.danger600:e.theme.colors.primary700};
          border-top-right-radius: ${t=>t.theme.borderRadius};
          border-top-left-radius: ${t=>t.theme.borderRadius};
          background-color: ${t=>t.theme.colors.neutral0};
          border-bottom: solid 1px ${t=>t.theme.colors.neutral0};
          box-shadow: ${e.theme.shadows.tableShadow};
          z-index: 1;
        }
      `}

  &[data-disabled] {
    cursor: not-allowed;
    color: ${e=>e.theme.colors.neutral400};
  }
`,Oc=l.forwardRef((e,t)=>{const{variant:n}=Gt("Content");return r.jsx(Pc,{$variant:n,ref:t,...e})}),Pc=s.styled(at.Content)`
  ${e=>e.$variant==="simple"?s.css`
        
      `:s.css`
        position: relative;
        z-index: 1;
        background-color: ${t=>t.theme.colors.neutral0};
      `}
`,Mc=Object.freeze(Object.defineProperty({__proto__:null,Content:Oc,List:Ec,Root:Ic,Trigger:zc},Symbol.toStringTag,{value:"Module"})),Bc=s.styled(v)`
  overflow: hidden;
  border: 1px solid ${({theme:e})=>e.colors.neutral150};
`,Nc=s.styled(Dn)`
  width: 100%;
  white-space: nowrap;
`,Wc=s.styled(v)`
  &:before {
    // TODO: make sure to add a token for this weird stuff
    background: linear-gradient(90deg, #c0c0cf 0%, rgba(0, 0, 0, 0) 100%);
    opacity: 0.2;
    position: absolute;
    height: 100%;
    content: ${({$overflowing:e})=>e==="both"||e==="left"?"''":void 0};
    box-shadow: ${({theme:e})=>e.shadows.tableShadow};
    width: ${({theme:e})=>e.spaces[2]};
    left: 0;
  }

  &:after {
    // TODO: make sure to add a token for this weird stuff
    background: linear-gradient(270deg, #c0c0cf 0%, rgba(0, 0, 0, 0) 100%);
    opacity: 0.2;
    position: absolute;
    height: 100%;
    content: ${({$overflowing:e})=>e==="both"||e==="right"?"''":void 0};
    box-shadow: ${({theme:e})=>e.shadows.tableShadow};
    width: ${({theme:e})=>e.spaces[2]};
    right: 0;
    top: 0;
  }
`,Fc=s.styled(v)`
  overflow-x: auto;
`,Hc=l.forwardRef(({footer:e,...t},n)=>{const o=l.useRef(null),[a,i]=l.useState(),c=d=>{const u=d.target.scrollWidth-d.target.clientWidth;if(d.target.scrollLeft===0){i("right");return}if(d.target.scrollLeft===u){i("left");return}d.target.scrollLeft>0&&i("both")};return l.useEffect(()=>{o.current.scrollWidth>o.current.clientWidth&&i("right")},[]),r.jsxs(Bc,{shadow:"tableShadow",hasRadius:!0,background:"neutral0",children:[r.jsx(Wc,{$overflowing:a,position:"relative",children:r.jsx(Fc,{ref:o,onScroll:c,paddingLeft:6,paddingRight:6,children:r.jsx(Nc,{ref:n,...t})})}),e]})}),_c=s.styled(zn)`
  & tr:last-of-type {
    border-bottom: none;
  }
`,Vc=({children:e,...t})=>r.jsx(_c,{...t,children:e}),qc=s.styled(An)`
  border-bottom: 1px solid ${({theme:e})=>e.colors.neutral150};
`,Uc=({children:e,...t})=>r.jsx(qc,{...t,children:e}),Gc=s.styled(Ln)`
  border-bottom: 1px solid ${({theme:e})=>e.colors.neutral150};

  & td,
  & th {
    padding: ${({theme:e})=>e.spaces[4]};
  }

  & td:first-of-type,
  & th:first-of-type {
    padding: 0 ${({theme:e})=>e.spaces[1]};
  }

  // Resetting padding values and fixing a height
  th {
    padding-top: 0;
    padding-bottom: 0;
    height: 5.6rem;
  }
`,Kc=e=>r.jsx(Gc,{...e}),_n=s.styled(qt)`
  vertical-align: middle;
  text-align: left;
  outline-offset: -4px;

  /**
  * Hack to make sure the checkbox looks aligned
  */
  input {
    vertical-align: sub;
  }
`,Yc=l.forwardRef(({children:e,action:t,...n},o)=>r.jsx(_n,{color:"neutral600",as:En,ref:o,...n,children:r.jsxs(C,{children:[e,t]})})),Qc=l.forwardRef(({children:e,...t},n)=>r.jsx(_n,{color:"neutral800",ref:n,...t,children:e})),Xc=s.styled(v)`
  height: 2.4rem;
  width: 2.4rem;
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;

  svg {
    height: 1rem;
    width: 1rem;
  }

  svg path {
    fill: ${({theme:e})=>e.colors.primary600};
  }
`,Zc=s.styled(v)`
  border-radius: 0 0 ${({theme:e})=>e.borderRadius} ${({theme:e})=>e.borderRadius};
  display: block;
  width: 100%;
  border: none;
`,Jc=({children:e,icon:t,...n})=>r.jsxs("div",{children:[r.jsx(_t,{}),r.jsx(Zc,{tag:"button",background:"primary100",padding:5,...n,children:r.jsxs(C,{children:[r.jsx(Xc,{"aria-hidden":!0,background:"primary200",children:t}),r.jsx(v,{paddingLeft:3,children:r.jsx(T,{variant:"pi",fontWeight:"bold",textColor:"primary600",children:e})})]})})]}),ed=ae(({children:e,startIcon:t,endIcon:n,disabled:o=!1,loading:a=!1,type:i="button",...c},d)=>{const u=o||a;return r.jsxs(nd,{ref:d,disabled:u,"aria-disabled":u,tag:"button",type:i,gap:2,...c,children:[a?r.jsx(rd,{"aria-hidden":!0,children:r.jsx(O.Loader,{})}):t,r.jsx(T,{variant:"pi",children:e}),n]})}),td=s.keyframes`
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(359deg);
  }
`,rd=s.styled.span`
  display: flex;
  animation: ${td} 2s infinite linear;
  will-change: transform;
`,nd=s.styled(C)`
  border: none;
  background-color: transparent;
  color: ${e=>e.theme.colors.primary600};
  cursor: pointer;

  &[aria-disabled='true'] {
    pointer-events: none;
    color: ${e=>e.theme.colors.neutral600};
  }

  ${st}
`,Vn=l.forwardRef((e,t)=>r.jsx(ut,{ref:t,...e}));Vn.displayName="TextInput";const od=l.forwardRef(({disabled:e,hasError:t,id:n,name:o,required:a,resizable:i=!0,...c},d)=>{const{error:u,...h}=ne("Textarea"),f=!!u||t,g=h.id??n,p=h.name??o,w=h.required||a;let m;return u?m=`${g}-error`:h.hint&&(m=`${g}-hint`),r.jsx(ad,{borderColor:f?"danger600":"neutral200",$hasError:f,hasRadius:!0,children:r.jsx(sd,{"aria-invalid":f,"aria-required":w,tag:"textarea",background:e?"neutral150":"neutral0",color:e?"neutral600":"neutral800",disabled:e,hasRadius:!0,ref:d,lineHeight:4,padding:4,width:"100%",height:"100%",id:g,name:p,"aria-describedby":m,$resizable:i,...c})})}),ad=s.styled(v)`
  ${be()}
`,sd=s.styled(v)`
  display: block;
  border: none;
  resize: ${({$resizable:e})=>e?"vertical":"none"};
  min-height: ${({minHeight:e})=>e||"10.5rem"};
  font-size: ${({theme:e})=>e.fontSizes[3]};

  ${({theme:e})=>e.breakpoints.small} {
    font-size: ${({theme:e})=>e.fontSizes[2]};
  }

  &::placeholder {
    color: ${({theme:e})=>e.colors.neutral600};
    font-size: ${({theme:e})=>e.fontSizes[3]};
    opacity: 1;

    ${({theme:e})=>e.breakpoints.small} {
      font-size: ${({theme:e})=>e.fontSizes[2]};
    }
  }

  &:focus-within {
    outline: none;
  }

  &:disabled {
    cursor: not-allowed;
  }
`,id=l.forwardRef(({offLabel:e,onLabel:t,disabled:n,hasError:o,required:a,id:i,name:c,checked:d,onChange:u,...h},f)=>{const[g=!1,p]=le({prop:d}),w=g!==null&&!g,{error:m,...$}=ne("Toggle"),k=!!m||o,y=$.id??i,x=$.name??c,I=$.required||a;let j;return m?j=`${y}-error`:$.hint&&(j=`${y}-hint`),r.jsxs(ld,{position:"relative",hasRadius:!0,padding:1,background:n?"neutral150":"neutral100",borderStyle:"solid",borderWidth:"1px",borderColor:k?"danger600":"neutral200",wrap:"wrap",cursor:n?"not-allowed":"pointer",$hasError:k,children:[r.jsx(br,{hasRadius:!0,flex:"1 1 50%",paddingTop:2,paddingBottom:2,paddingLeft:3,paddingRight:3,justifyContent:"center",background:n&&w?"neutral200":w?"neutral0":"transparent",borderColor:n&&w?"neutral300":w?"neutral200":n?"neutral150":"neutral100",children:r.jsx(T,{variant:"pi",fontWeight:"bold",textTransform:"uppercase",textColor:n?"neutral700":w?"danger700":"neutral600",children:e})}),r.jsx(br,{hasRadius:!0,flex:"1 1 50%",paddingLeft:3,paddingRight:3,justifyContent:"center",background:n&&g?"neutral200":g?"neutral0":"transparent",borderColor:n&&g?"neutral300":g?"neutral200":n?"neutral150":"neutral100",children:r.jsx(T,{variant:"pi",fontWeight:"bold",textTransform:"uppercase",textColor:n?"neutral700":g?"primary600":"neutral600",children:t})}),r.jsx(cd,{...h,id:y,name:x,ref:f,onChange:R=>{p(R.currentTarget.checked),u==null||u(R)},type:"checkbox","aria-required":I,disabled:n,"aria-disabled":n,checked:!!g,"aria-describedby":j})]})}),ld=s.styled(C)`
  ${be()}
`,br=s.styled(C)`
  padding-block: 0.6rem;
`,cd=s.styled.input`
  height: 100%;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  z-index: 0;
  width: 100%;
`,dd={x:0,y:0,width:0,height:0,top:0,left:0,bottom:0,right:0};function ud(){const[e,t]=X.useState(null),[n,o]=X.useState(dd),a=X.useMemo(()=>new ResizeObserver(i=>{if(i[0]){const{x:c,y:d,width:u,height:h,top:f,left:g,bottom:p,right:w}=i[0].contentRect;o({x:c,y:d,width:u,height:h,top:f,left:g,bottom:p,right:w})}}),[]);return he(()=>{if(e)return a.observe(e),()=>{a.disconnect()}},[e,a]),[t,n]}const hd=({onEscape:e,restoreFocus:t=!0,skipAutoFocus:n=!1,...o})=>{const a=l.useRef(null);l.useEffect(()=>{let c=null;return t&&(c=document.activeElement),()=>{c&&c.focus()}},[t]),l.useEffect(()=>{if(!a.current||n)return;const c=ve(a.current);c.length>0?c[0].focus():console.warn("[FocusTrap]: it seems there are no focusable elements in the focus trap tree. Make sure there s at least one.")},[n]);const i=c=>{if(c.key===W.ESCAPE&&e){e();return}if(c.key!==W.TAB)return;const d=ve(a.current);if(d.length>0){const u=d[0],h=d[d.length-1];c.shiftKey?u===document.activeElement&&(c.preventDefault(),h.focus()):h===document.activeElement&&(c.preventDefault(),u.focus())}};return r.jsx("div",{ref:a,onKeyDown:i,...o})},fd=({tagName:e,attributeName:t="",...n})=>{const o=()=>{const c=document.activeElement;return c?e?c.tagName.toLowerCase()===e:c.hasAttribute(t):!1},a=c=>e?c.querySelectorAll(e):c.querySelectorAll(`[${t}]`),i=c=>{switch(c.key){case W.RIGHT:case W.DOWN:{if(o()){c.preventDefault();const d=document.activeElement,u=[...a(c.currentTarget)],h=u.findIndex(g=>g===d),f=h+1<u.length?h+1:0;u[f].focus()}break}case W.LEFT:case W.UP:{if(o()){c.preventDefault();const d=document.activeElement,u=[...a(c.currentTarget)],h=u.findIndex(g=>g===d),f=h-1>-1?h-1:u.length-1;u[f].focus()}break}case W.HOME:{o()&&(c.preventDefault(),a(c.currentTarget).item(0).focus());break}case W.END:{if(o()){c.preventDefault();const d=a(c.currentTarget);d.item(d.length-1).focus()}break}}};return r.jsx(v,{onKeyDown:i,...n})};Object.defineProperty(exports,"composeEventHandlers",{enumerable:!0,get:()=>D.composeEventHandlers});Object.defineProperty(exports,"useCallbackRef",{enumerable:!0,get:()=>D.useCallbackRef});Object.defineProperty(exports,"useCollator",{enumerable:!0,get:()=>D.useCollator});Object.defineProperty(exports,"useFilter",{enumerable:!0,get:()=>D.useFilter});exports.AccessibleIcon=Se;exports.Accordion=qo;exports.Alert=Zo;exports.Avatar=ia;exports.Badge=At;exports.BaseLink=Pe;exports.Box=v;exports.Breadcrumbs=Er;exports.Button=rt;exports.Card=Va;exports.CardAction=Zr;exports.CardAsset=Ua;exports.CardBadge=Xa;exports.CardBody=Za;exports.CardCheckbox=rs;exports.CardContent=os;exports.CardHeader=ss;exports.CardSubtitle=cs;exports.CardTimer=is;exports.CardTitle=ls;exports.Carousel=Jr;exports.CarouselActions=Ds;exports.CarouselImage=Es;exports.CarouselInput=Is;exports.CarouselSlide=zs;exports.Checkbox=dt;exports.Combobox=on;exports.ComboboxOption=an;exports.Crumb=Ar;exports.CrumbLink=zr;exports.CrumbSimpleMenu=Yr;exports.DatePicker=wn;exports.DateTimePicker=tl;exports.DesignSystemProvider=fi;exports.Dialog=li;exports.DismissibleLayer=un;exports.Divider=_t;exports.Dots=Ol;exports.EmptyStateLayout=ol;exports.Field=Ts;exports.Flex=C;exports.FocusTrap=hd;exports.Grid=_s;exports.IconButton=ge;exports.IconButtonGroup=$a;exports.JSONInput=il;exports.KeyboardKeys=W;exports.KeyboardNavigable=fd;exports.Link=Wr;exports.LinkButton=cl;exports.LiveRegions=cn;exports.Loader=nt;exports.Main=ul;exports.Menu=Fa;exports.MenuItem=Ha;exports.Modal=Sl;exports.MultiSelect=Nn;exports.MultiSelectGroup=Wn;exports.MultiSelectNested=ic;exports.MultiSelectOption=Ut;exports.NextLink=Al;exports.NumberInput=Tl;exports.PageLink=zl;exports.Pagination=Dl;exports.Popover=Vl;exports.Portal=Ft;exports.PreviousLink=El;exports.ProgressBar=ql;exports.Radio=Jl;exports.RawTable=Dn;exports.RawTbody=zn;exports.RawTd=qt;exports.RawTh=En;exports.RawThead=An;exports.RawTr=Ln;exports.ScrollArea=Qe;exports.SearchForm=Mn;exports.Searchbar=Pn;exports.SimpleMenu=Kr;exports.SingleSelect=vt;exports.SingleSelectOption=kt;exports.SkipToContent=fl;exports.Status=dc;exports.SubNav=hc;exports.SubNavHeader=gc;exports.SubNavLink=pc;exports.SubNavLinkSection=$c;exports.SubNavSection=Cc;exports.SubNavSections=jc;exports.Switch=vc;exports.TFooter=Jc;exports.Table=Hc;exports.Tabs=Mc;exports.Tag=Bn;exports.Tbody=Vc;exports.Td=Qc;exports.TextButton=ed;exports.TextInput=Vn;exports.Textarea=od;exports.Th=Yc;exports.Thead=Uc;exports.TimePicker=Cn;exports.Toggle=id;exports.Tooltip=ct;exports.Tr=Kc;exports.Typography=T;exports.VisuallyHidden=me;exports._internaluseUncontrolledState=Et;exports.composeRefs=zt;exports.convertUTCDateToCalendarDateTime=Ee;exports.darkTheme=bs;exports.extendTheme=xs;exports.extractStyleFromTheme=_e;exports.getThemeSize=$s;exports.handleResponsiveValues=ke;exports.inputFocusStyle=be;exports.isSSR=ln;exports.lightTheme=Mt;exports.setOpacity=Wt;exports.useComposedRefs=te;exports.useControllableState=le;exports.useDateFormatter=ue;exports.useDesignSystem=Xe;exports.useField=ne;exports.useId=re;exports.useIsomorphicLayoutEffect=he;exports.useMeasure=ud;exports.useNotifyAT=di;
