// @strapi/core/dist/utils/startup-logger.js
'use strict';

var chalk = require('chalk');
var CLITable = require('cli-table3');
var fp = require('lodash/fp');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var chalk__default = /*#__PURE__*/_interopDefault(chalk);
var CLITable__default = /*#__PURE__*/_interopDefault(CLITable);
var fp__default = /*#__PURE__*/_interopDefault(fp);

const ROUNDED_CHARS = {
    top: '─',
    'top-mid': '┬',
    'top-left': '╭',
    'top-right': '╮',
    bottom: '─',
    'bottom-mid': '┴',
    'bottom-left': '╰',
    'bottom-right': '╯',
    left: '│',
    right: '│',
    middle: '│',
    mid: '',
    'left-mid': '',
    'mid-mid': '',
    'right-mid': ''
};
const createStartupLogger = (app)=>{
    return {
        logStats () {
            const columns = Math.min(process.stderr.columns, 80) - 2;
            console.log();
            console.log(chalk__default.default.black.bgWhite(fp__default.default.padEnd(columns, ' Project information')));
            console.log();
            const infoTable = new CLITable__default.default({
                colWidths: [
                    20,
                    50
                ],
                chars: ROUNDED_CHARS
            });
            const dbInfo = app.db?.getInfo();
            const getPlan = ()=>{
                if (!app.EE) {
                    return 'Community';
                }
                return app.ee.planPriceId?.toLowerCase().includes('growth') ? 'Growth' : 'Enterprise';
            };
            infoTable.push([
                chalk__default.default.blue('Time'),
                `${new Date()}`
            ], [
                chalk__default.default.blue('Launched in'),
                `${Date.now() - app.config.launchedAt} ms`
            ], [
                chalk__default.default.blue('Environment'),
                app.config.environment
            ], [
                chalk__default.default.blue('Process PID'),
                process.pid
            ], [
                chalk__default.default.blue('Version'),
                `${app.config.info.strapi} (node ${process.version})`
            ], [
                chalk__default.default.blue('Plan'),
                getPlan()
            ], [
                chalk__default.default.blue('Database'),
                dbInfo?.client
            ], [
                chalk__default.default.blue('Database name'),
                dbInfo?.displayName
            ]);
            if (dbInfo?.schema) {
                infoTable.push([
                    chalk__default.default.blue('Database schema'),
                    dbInfo.schema
                ]);
            }
            console.log(infoTable.toString());
            console.log();
            console.log(chalk__default.default.black.bgWhite(fp__default.default.padEnd(columns, ' Actions available')));
            console.log();
        },
        logFirstStartupMessage () {
            if (!strapi.config.get('server.logger.startup.enabled')) {
                return;
            }
            this.logStats();
            console.log(chalk__default.default.bold('One more thing...'));
            console.log(chalk__default.default.grey('Create your first administrator 💻 by going to the administration panel at:'));
            console.log();
            const addressTable = new CLITable__default.default();
            const adminUrl = strapi.config.get('admin.absoluteUrl');
            addressTable.push([
                chalk__default.default.bold(adminUrl)
            ]);
            console.log(`${addressTable.toString()}`);
            console.log();
        },
        logDefaultStartupMessage () {
            if (!strapi.config.get('server.logger.startup.enabled')) {
                return;
            }
            this.logStats();
            console.log(chalk__default.default.bold('Welcome back!'));
            if (app.config.get('admin.serveAdminPanel') === true) {
                console.log(chalk__default.default.grey('To manage your project 🚀, go to the administration panel at:'));
                const adminUrl = strapi.config.get('admin.absoluteUrl');
                console.log(chalk__default.default.bold(adminUrl));
                console.log();
            }
            console.log(chalk__default.default.grey('To access the server ⚡️, go to:'));
            const serverUrl = strapi.config.get('server.absoluteUrl');
            console.log(chalk__default.default.bold(serverUrl));
            console.log();
        },
        logStartupMessage ({ isInitialized }) {
            if (!strapi.config.get('server.logger.startup.enabled')) {
                return;
            }
            if (!isInitialized) {
                this.logFirstStartupMessage();
            } else {
                this.logDefaultStartupMessage();
            }
        }
    };
};

exports.createStartupLogger = createStartupLogger;
