// @strapi/content-type-builder/dist/admin/utils/conditions.js
'use strict';

const isRecord = (value)=>typeof value === 'object' && value !== null && !Array.isArray(value);
const isFieldVar = (value)=>isRecord(value) && typeof value.var === 'string';
const isAttributeConditionValue = (value)=>typeof value === 'string' || typeof value === 'number' || typeof value === 'boolean';
const getVisibleConditionEntries = (condition)=>{
    const visible = condition?.visible;
    if (!isRecord(visible)) {
        return [];
    }
    return Object.entries(visible).reduce((entries, [operator, rule])=>{
        if (!Array.isArray(rule)) {
            return entries;
        }
        const [fieldVar, value] = rule;
        if (!isFieldVar(fieldVar) || !isAttributeConditionValue(value)) {
            return entries;
        }
        entries.push({
            operator,
            fieldVar,
            value
        });
        return entries;
    }, []);
};
const getFirstVisibleConditionEntry = (condition)=>getVisibleConditionEntries(condition)[0] ?? null;
const checkDependentRows = (contentTypes, fieldName)=>{
    const dependentRows = [];
    Object.entries(contentTypes).forEach(([contentTypeUid, contentType])=>{
        if (contentType.attributes !== undefined) {
            // Handle both array and object formats of attributes
            const attributes = Array.isArray(contentType.attributes) ? contentType.attributes.reduce((acc, attr, index)=>{
                acc[index.toString()] = attr;
                return acc;
            }, {}) : contentType.attributes;
            Object.entries(attributes).forEach(([attrName, attr])=>{
                getVisibleConditionEntries(attr.conditions).forEach(({ fieldVar })=>{
                    // Check if this condition references our field
                    if (fieldVar.var === fieldName) {
                        dependentRows.push({
                            contentTypeUid,
                            contentType: contentType.info.displayName,
                            attribute: attr.name || attrName
                        });
                    }
                });
            });
        }
    });
    return dependentRows;
};
const getAvailableConditionFields = (attributes, currentFieldName)=>{
    return attributes.filter((attr)=>{
        // Only include boolean and enum fields
        const isCorrectType = attr.type === 'boolean' || attr.type === 'enumeration';
        // Exclude the current field to prevent self-referential conditions
        const isNotCurrentField = attr.name !== currentFieldName;
        return isCorrectType && isNotCurrentField;
    }).map((attr)=>({
            name: attr.name,
            type: attr.type,
            enum: attr.type === 'enumeration' ? attr.enum : undefined
        }));
};

exports.checkDependentRows = checkDependentRows;
exports.getAvailableConditionFields = getAvailableConditionFields;
exports.getFirstVisibleConditionEntry = getFirstVisibleConditionEntry;
exports.getVisibleConditionEntries = getVisibleConditionEntries;
